<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta name="generator" content="Cosma v2.0.4">
    
    <meta name="title" content="">
    
    <meta name="author" content="">
    
    <meta name="description" content="">
    
    <meta name="keywords" content="">
    
    <meta name="date" content="September 24, 2023 at 12:03:47 PM">
    <title>Cosma — </title>

    <link rel="icon" type="image/png" sizes="16x16" href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAMAAAAoLQ9TAAAABGdBTUEAALGPC/xhBQAAAAFzUkdCAK7OHOkAAAIGaVRYdFhNTDpjb20uYWRvYmUueG1wAAAAAAA8eDp4bXBtZXRhIHhtbG5zOng9ImFkb2JlOm5zOm1ldGEvIiB4OnhtcHRrPSJYTVAgQ29yZSA2LjAuMCI+CiAgIDxyZGY6UkRGIHhtbG5zOnJkZj0iaHR0cDovL3d3dy53My5vcmcvMTk5OS8wMi8yMi1yZGYtc3ludGF4LW5zIyI+CiAgICAgIDxyZGY6RGVzY3JpcHRpb24gcmRmOmFib3V0PSIiCiAgICAgICAgICAgIHhtbG5zOnRpZmY9Imh0dHA6Ly9ucy5hZG9iZS5jb20vdGlmZi8xLjAvIgogICAgICAgICAgICB4bWxuczpleGlmPSJodHRwOi8vbnMuYWRvYmUuY29tL2V4aWYvMS4wLyI+CiAgICAgICAgIDx0aWZmOk9yaWVudGF0aW9uPjE8L3RpZmY6T3JpZW50YXRpb24+CiAgICAgICAgIDxleGlmOlBpeGVsWERpbWVuc2lvbj4xMDI0PC9leGlmOlBpeGVsWERpbWVuc2lvbj4KICAgICAgICAgPGV4aWY6UGl4ZWxZRGltZW5zaW9uPjEwMjQ8L2V4aWY6UGl4ZWxZRGltZW5zaW9uPgogICAgICA8L3JkZjpEZXNjcmlwdGlvbj4KICAgPC9yZGY6UkRGPgo8L3g6eG1wbWV0YT4KokyVsAAAAAlwSFlzAAALEwAACxMBAJqcGAAAATtQTFRFR3BMDVJzDFx+GHSUAP//DVByCVd3AAAAAAD/AIKBCXKSBFJ0F1x/GFZ2EGCANmF8D2+QGHaXFG6NAFh8EnaXG3aWLHCOE2iIH3mZAEx4AFV6AG2QAXOUOn+dprG8AHGSGmuLGWmJfZqtUoahprrGD1V4CVV4FHeWFnWWLmeIAFp8FWaGAFh7LmeGF2OGUHWNfZapAF+HUHSNfZaoAF6H/////Pz87O7x4uLio6Oj3t7eFWaJE16DGGCEIWuOAFqAGoCiADxyAFyHRG+MHoGjAGeQE3GUEG6S1tvhAH2iRX+cAG2Y5OjrAGWMiqCxoa+7iqGyobXBB2KGCF6DCGGFCl2DE4CiEX2gGXqdEX2fFX+iJHmcGXqcInmbAEx3aIWa193jAE12Z4Sa5ensZ5CmaZCn1t3j5Ons3t7fp3RRZgAAADV0Uk5TAHRwQQFHRAEBAkJ4+3h0eHBudPpy+3V1cvv7+vpy+/v7+/v7+52dmZl1+nX6dXX7+/v7+/siYXfnAAAA2ElEQVQYGQXBy07CUBRA0X1OD9zCQAloS4Nj40ii/gG/7UdoSYwxjowCgo9EQxrpfdS1wA0AAGDgEHcgO8qshw/xN+IscnG9tp6SfCjqp2hheHXrMhO6EB8Wr43Rz/qX2fSe+TbW0m8MlepMdvBVxCJTDHv/vkNulqs3fnaGIpJmqnWnOksiKKVIAoAkcoLR+fEGMWK3OfYJ5SWsJ2XlfVVOtmGFso+tP53AaOzbuMewqt4+Tj9pP0aH0lrJ/87nG1Elpa5cPufiDvTCUJWUGvM4cDkAAOSOf+M3WO+ZgJOzAAAAAElFTkSuQmCC" />

    <style> 
        :root {
            
            --n_undefined: #858585;
            
            
            --l_undefined: #e1e1e1;
            
            --highlight: #ff6a6a;
        }
        @charset "UTF-8";

/** Global
------------------------------------------------------------*/

:root {
  /* colors */
  --cosma-blue: #147899;
  --cosma-blue-dark: #02385c;
  --border-color: #d8d8d8;
  --border-light-color: #f3f3f3;
  --indicator-background-color: #fafafa;
  --lateral-background-color: white;
  /* typography */
  --sans: system-ui, -apple-system, 'San Francisco', 'Helvetica Neue', 'Segoe UI', 'Arial',
    sans-serif;
  --serif: 'Georgia', 'Garamond', 'Time New Roman', 'Times', serif;
  --mono: 'Consolas', 'Courrier New', 'Courrier', monospace;
  --condensed: 'Avenir Next Condensed', 'Arial Condensed', sans-serif;
  /* global size */
  font-size: 62.5%;
  /* others */
  --shadow: 0 1px 3px rgba(3, 2, 2, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
  --offblack: rgb(51, 51, 51);
  --offwhite: rgb(253, 252, 251);
  --gray: #d6d6d6;
  --background-gray: rgb(246, 246, 246);
}

* {
  box-sizing: border-box;
}

/** Page
------------------------------------------------------------*/

body {
  overflow: hidden;
  width: 100%;
  margin: 0;
  padding: 0;
  font-family: var(--sans);
  font-size: 1.4rem;
}

/* ---
- Containers
--- */

article {
  padding: 0.8rem 0;
}

article + article:not(.record) {
  border-top: 1px solid var(--border-light-color);
}

/** Elements
------------------------------------------------------------*/

a {
  text-decoration: none;
}

/* ---
- Buttons
--- */

button {
  cursor: pointer;
  margin: 0.5rem 0.2rem;
  font-size: 1.3rem;
}

.btn {
  background-color: white;
  border: 1px solid var(--border-color);
  border-radius: 2px;
}

.btn.active,
.btn[data-active='true'] {
  border: 1px solid black;
}

[onclick] {
  cursor: pointer;
}

/* ---
- Inputs
--- */

.outline {
  outline: var(--cosma-blue) solid 1px;
}

label {
  cursor: pointer;
}

input {
  font-size: 1.3rem;
}

input[type='checkbox'] {
  cursor: pointer;
  width: 1.2rem;
  height: 1.2rem;
  margin: 0 0.8rem 0 0;
}

input[type='range'] {
  cursor: pointer;
  width: 100%;
  margin-top: 0.3rem;
}

/* ---
- Lists
--- */

details {
  padding: 0.5rem 0 1rem 0;
}

details[data-sort] .sort-box:not(.active) {
  display: none;
}

details > summary {
  cursor: pointer;
  text-transform: uppercase;
  font-size: 1.3rem;
  margin: 1rem 0 0.5rem 0;
  font-weight: bold;
  outline: none;
}

summary + button {
  margin: 1rem 0;
}

.list-content-container ul {
  list-style: none;
  margin: 0;
  padding: 0;
}

.list-content-container ul li {
  display: flex;
  align-items: baseline;
  padding: 0.3rem 0 0.3rem 1rem;
}

.list-content-container .menu-controls-list {
  margin-top: 1.3rem;
}

.list-content-container ul li span:last-of-type:not(.badge) {
  line-height: 1.5;
}

.record-type-point {
  /* colored dot before links */
  font-size: 0.8em;
  padding-right: 0.8em;
}

.menu-controls-container p:first-child {
  font-weight: bold;
}

.menu-types-container {
  display: flex;
  flex-direction: column;
}

.filter {
  margin: 0.1em 0;
}

.menu-tags-container {
  display: flex;
  flex-wrap: wrap;
  margin-top: 1em;
}

.menu-tags-container label {
  display: flex;
  align-items: center;
  margin: 0.15em 0.15em 0.15em 0;
}

.menu-tags-container .label {
  margin: 0 0.2em;
}

.menu-tags-container .badge {
  margin: 0.15em 0.15em 0.15em 0;
}

.menu-tags-container input[type='checkbox'] {
  margin: 0.2em;
}

.menu-index-list {
  display: flex;
  flex-direction: column;
}

.menu-index-list.reverse {
  flex-direction: column-reverse;
}

/* ---
- Texts
--- */

a {
  color: inherit;
}

code {
  font-family: var(--mono);
}

/* ---
- Tips
--- */

.badge {
  display: inline-block;
  background-color: var(--indicator-background-color);
  border: 1px solid var(--gray);
  border-radius: 1px;
  padding: 0 0.3rem;
  font-size: 1rem;
  margin-left: 0.2rem;
  text-align: center;
  line-height: 1.5rem;
  white-space: nowrap;
  font-family: var(--mono);
  color: black;
}

.tooltip {
  cursor: pointer;
  position: relative;
  display: inline-block;
  background-color: var(--cosma-blue);
  border-radius: 50%;
  color: white;
  width: 16px;
  text-align: center;
  height: 16px;
  font-weight: normal;
}

.tooltip .tooltiptext {
  visibility: hidden;
  width: 120px;
  line-height: 130%;
  background-color: var(--cosma-blue-dark);
  box-shadow: var(--shadow);
  text-align: center;
  padding: 5px;
  border-radius: 2px;
  position: absolute;
  z-index: 1;
  width: 160px;
  bottom: 130%;
  left: 50%;
  margin-left: -80px;
}

.tooltip:hover .tooltiptext {
  visibility: visible;
}

/** Menu (left side board)
------------------------------------------------------------*/

aside.menu {
  position: relative;
  z-index: 100;
  overflow: auto;
  height: 100vh;
  width: 30rem;
  margin: 0;
  padding: 0 20px 50px 20px;
  background-color: var(--background-gray);
  color: var(--offblack);
  box-shadow: var(--shadow);
}

/* ---
- User title
--- */

.title {
  text-transform: uppercase;
  font-size: 1.8rem;
  color: var(--cosma-blue-dark);
  text-align: center;
}

/* ---
- Toolbar
--- */

#menu-electron {
  display: flex;
  flex-direction: row;
  justify-content: space-evenly;
  align-items: center;
  margin-top: 1.5rem;
}

#menu-electron button {
  padding: 0;
  margin: 0;
  background-color: transparent;
  border: none;
}

#menu-electron button + button {
  margin-left: 7px;
}

#menu-electron button svg {
  height: 25px;
  width: auto;
}

#menu-electron button.icon-chevron svg {
  height: 18px;
}

/* ---
- Spatialization load bar
--- */

.load-bar {
  height: 0.3em;
  padding: 0.1rem;
  display: flex;
  margin: 1rem 0;
  background-color: var(--indicator-background-color);
}

.load-bar #load-bar-value {
  background-color: var(--cosma-blue);
  flex-basis: 100%;
}

/* ---
- Search input & list
--- */

.search-bar {
  width: 100%;
  margin: 0.8rem 0;
}

#search-result-list {
  margin-bottom: 0.8rem;
}

#search-result-list li {
  cursor: pointer;
}

/* ---
- Controls from graph params
--- */

.menu-controls input[type='number'] {
  width: 7rem;
  margin-left: 1rem;
}

.menu-controls-title {
  font-weight: bold;
}

/* ---
- Footer
--- */

aside.menu footer {
  position: fixed;
  bottom: 0px;
  left: 0px;
  width: inherit;
  margin-top: 1.8rem;
  background-color: inherit;
  padding: 0.8rem 0 0.8rem 0.8rem;
  text-transform: uppercase;
  border-top: solid 1px var(--border-color);
  font-size: 1.1rem;
  text-align: center;
}

aside.menu footer span:not(:first-of-type)::before {
  content: '—';
}

/** Record container (right side board)
------------------------------------------------------------*/

main.record-container {
  position: absolute;
  z-index: 100;
  top: 0;
  right: 0;
  padding: 1rem 2rem;
  width: 50rem;
  height: 100vh;
  overflow: auto;
  background-color: var(--offwhite);
  color: var(--offblack);
  box-shadow: var(--shadow);
  transition: transform 0.6s;
  transition-timing-function: cubic-bezier(0.02, 1.18, 0, 0.98);
}

main.record-container:not(.active) {
  /* hide on right when class 'active' removed */
  transform: translate(100%);
}

.record:not(.active) {
  display: none;
}

/* ---
- Record header
--- */

.record header {
  padding-bottom: 1.5rem;
  margin-bottom: 3rem;
  border-bottom: 1px solid var(--gray);
}

.record header div + div,
.record-focus-container {
  margin-top: 0.8rem;
}

.record header div span:first-of-type,
.record-focus-container > span {
  font-size: 1.3rem;
  font-weight: bold;
  text-transform: uppercase;
  margin-right: 0.8rem;
}

.record-last-edit {
  display: none;
}

.record-focus-lever {
  display: none;
}

.record-focus-lever input {
  width: 80%;
}

/* ---
- Record content
--- */

.record-content a {
  text-decoration: underline;
}

.record img {
  height: auto;
  max-width: 100%;
}

.record header img {
  height: auto;
  max-width: 100%;
  margin: 1rem 0;
}

/* From Typesafe.css, by Bradley Taunt */

.record main {
  font-size: 1.6rem;
  line-height: 1.5;
  font-family: var(--serif);
}

.record main h1,
h2,
h3,
h4,
h5,
h6 {
  font-family: var(--sans);
  margin: 3rem 0 0 0;
}

.record main h1 {
  font-size: 1.8rem;
}

.record main h2 {
  font-size: 1.6rem;
}

.record main h3 {
  font-size: 1.4rem;
}

.record main a {
  word-wrap: break-word;
}

.record main a.record-link {
  color: #337ab7;
}

.record main ul,
.record main ol {
  padding: 0;
  list-style-position: inside;
}

.record main ul ul,
.record main ol ol {
  padding-left: 2rem;
}

.record main ul li,
.record ol li {
  margin: 0.5rem 0 0 0;
}

.record main pre {
  overflow: auto;
  font-size: 90%;
}

.record main blockquote {
  border-left: 5px solid var(--border-color);
  margin: 0;
  padding-left: 1rem;
}

.record main table {
  border-collapse: collapse;
  margin: 2rem 0;
  text-align: left;
  width: 100%;
}

.record main tr {
  border-bottom: 1px solid var(--border-color);
}

.record main th,
.record main td {
  padding: 0.5rem;
}

.record main hr {
  background: var(--border-color);
  border: 0;
  height: 1px;
  margin: 2rem 0;
}

/* ---
- Record bibliography
--- */

.record .bibliography div + div {
  margin-top: 1rem;
}

.record div.csl-entry {
  width: inherit;
  text-indent: -2rem;
  padding-left: 2rem;
  margin: 0.75rem 0 0 0;
  line-height: 1.3;
  overflow-wrap: break-word;
  word-wrap: break-word;
  -ms-word-break: break-all;
  word-break: break-word;
}

/* ---
- Record footer
--- */

.record footer {
  margin-top: 3rem;
  border-top: 1px solid var(--gray);
}

.record footer li {
  /* links & backlinks */
  position: relative;
}

.record footer li .record-links-context {
  display: none;
  background: white;
  border: 1px black solid;
  padding: 0.8rem;
  border-radius: 2px;
  margin-top: 0.8rem;
  position: absolute;
  bottom: calc(100% + 10px);
  box-shadow: var(--shadow);
  text-indent: 0;
}

.record footer li .record-links-context li {
  display: block;
}

.record footer li:hover .record-links-context {
  display: block;
}

.record footer li .record-links-context li::before {
  content: '-';
  position: relative;
  left: -5px;
}

.record footer li .record-links-context p,
.record footer li .record-links-context blockquote {
  margin: 0px;
}

.record footer li .record-links-context p + p {
  margin-top: 0.8rem;
}

.id-context {
  font-style: normal;
  background-color: rgba(255, 255, 182, 1);
}

/** Graph
------------------------------------------------------------*/

.graph-wrapper {
  position: absolute;
  top: 0;
  right: 0;
  height: 100vh;
  width: calc(100vw - 30rem);
}

#graph-canvas {
  cursor: move;
  width: 100%;
  height: 100%;
}

/* ---
- Graph elements
--- */

/* nodes */
circle,
text {
  cursor: pointer;
}

g.focus circle {
  stroke: var(--highlight);
}

/* labels */
text {
  font-family: var(--condensed);
  fill: gray;
}

/* arrows */
marker {
  /* set into config.yml file by 'undefined' link type */
  stroke: var(--l_undefined);
  fill: var(--l_undefined);
}

/* ---
- Buttons
--- */

.graph-controls {
  position: absolute;
  display: flex;
  align-items: flex-start;
  flex-direction: column;
  left: 30px;
  bottom: 85px;
  flex-wrap: wrap;
  width: 20rem;
}

.graph-controls::before {
  content: '';
  position: absolute;
  backdrop-filter: blur(7px);
  width: 100%;
  height: 100%;
  border-radius: 5px;
}

.graph-timeline {
  position: absolute;
  display: flex;
  align-items: flex-start;
  flex-direction: column;
  left: 30px;
  bottom: 15px;
  width: 20rem;
}

.graph-controls button {
  margin: 2px, 0;
}

.graph-controls div {
  position: relative;
}

.graph-controls div.graph-controls-zoom {
  display: flex;
  justify-content: space-between;
  position: relative;
}

.graph-controls-focus {
  display: flex;
  justify-content: flex-start;
  flex-flow: row wrap;
}

.graph-controls-focus label {
  display: inline-block;
  margin: 0.8rem 0 0.8rem 0.4rem;
}

.graph-controls-focus input[type='range']:not(.active) {
  visibility: hidden;
}

.graph-controls #reset-isolate {
  display: none;
}

.graph-controls #reset-highlight {
  display: none;
}

.graph-timeline {
  width: calc(100% - 60px);
}

#timeline-form {
  width: 100%;
  backdrop-filter: blur(7px);
}

#timeline-form:not(.active) {
  visibility: hidden;
}

#timeline-form input[type='range'] {
  width: 100%;
}

#timeline-range output[name='timeline_output'] {
  color: gray;
  position: relative;
  text-align: center;
}

#chonos-list {
  display: flex;
  color: gray;
  justify-content: space-between;
}

/* thanks to https://stackoverflow.com/a/62983699/13491646 */

input[type='range'] + datalist {
  display: block;
  margin-top: -4px;
}

input[type='range'] + datalist option {
  display: inline-block;
  width: calc((100% - 12px) / (var(--list-length) - 1));
  text-align: center;
  font-size: 11px;
  color: gray;
}

input[type='range'] + datalist option:first-child {
  width: calc((100% - 12px) / ((var(--list-length) - 1) * 2) + 6px);
  text-align: left;
}

input[type='range'] + datalist option:last-child {
  width: calc((100% - 12px) / ((var(--list-length) - 1) * 2) + 6px);
  text-align: right;
}

/** Credits
------------------------------------------------------------*/

article#credits {
  text-align: center;
}

article#credits ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

article#credits dt {
  font-weight: bold;
}

article#credits dd,
article#credits li {
  margin: 0;
  padding: 0 0 0.5rem 0;
}

article#credits h1.metas-title {
  font-size: 2rem;
}

/* ---
- Logo
--- */

.logo {
  cursor: pointer;
}

.logo svg {
  display: block;
  margin: 0 auto;
  max-width: 300px;
}

@keyframes rotate-center {
  0% {
    transform: rotate(0deg);
  }

  50% {
    transform: rotate(270deg);
  }

  100% {
    transform: rotate(360deg);
  }
}

#cosma-roll {
  transform-origin: 96px 42px;
}

#cosma-roll.anim {
  animation: 3s cubic-bezier(0.2, -0.2, 0.8, 1) 1 rotate-center;
}

/** Help
------------------------------------------------------------*/

#help td:first-child {
  font-weight: bold;
  text-align: right;
}

#help td {
  vertical-align: top;
}

#help tr {
  border-top: 1px solid var(--border-color);
  border-bottom: 1px solid var(--border-color);
}

/** Bibliography references
------------------------------------------------------------*/

#citation-references button {
  margin-bottom: 2rem;
}

#citation-references code {
  display: block;
  word-wrap: break-word;
  font-size: 1.2rem;
  max-height: 75vh;
  overflow: scroll;
  border-top: 1px solid var(--border-color);
  border-bottom: 1px solid var(--border-color);
}

        @media print {
  /* ---
	- Global
	--- */

  :root {
    font-size: 52.5%;
  }

  @page {
    margin: 10mm 20mm;
  }

  body {
    overflow: unset;
  }

  /* ---
	- Elements
	--- */

  p,
  tr,
  li {
    orphans: 3;
  }

  a[href]:not([onclick]):after {
    content: ' (' attr(href) ')';
  }

  /* ---
	- Page containers
	--- */

  aside,
  .graph-wrapper {
    display: none;
  }

  /* ---
	- Record header
	--- */

  .record header {
    margin-bottom: 10mm;
  }

  .record-focus-container,
  .record-btn-close {
    display: none;
  }

  /* ---
	- Record main
	--- */

  main.record-container {
    position: unset;
    width: 100%;
    overflow: unset;
    box-shadow: unset;
  }

  /* ---
	- Record footer
	--- */

  main.record-container footer {
    break-before: page;
  }

  main.record-container footer .record-links-context {
    display: block;
    position: unset;
    box-shadow: unset;
  }

  main.record-container footer li {
    display: block;
    margin-top: 2mm;
  }

  .id-context {
    font-style: normal;
    text-decoration: underline;
  }
}


        
    </style>
</head>

<body>

    

    <aside class="menu list-content-container">

        <header role="banner">
            

            

            <div class="load-bar" onclick="updateForces()"
                title="Spatialization (Space)">
                <div id="load-bar-value"></div>
            </div>
        </header>

        <section>
            <input class="search-bar" id="search" placeholder="Search… (S)"
                autocomplete="off" role="search" />
            <ul id="search-result-list"></ul>
        </section>

        <details class="menu-types" open="open">
            <summary class="menu-title">
                Types
                <span id="types-counter" class="badge">0</span>
            </summary>

            <form id="types-form" class="menu-types-container">
                
                <label class="filter">
                    <input type="checkbox" name="undefined">
                    <span class="label" style="color: var(--n_undefined);">undefined</span>
                    <span class="badge">57</span>
                </label>
                
            </form>
        </details>

        

        <details class="menu-index" id="index">
            <summary class="menu-title">
                Index
                <span id="index-counter" class="badge">57</span>
            </summary>

            <select class="sorting-select" name="index_sorting">
                <option value="alphabetical">A-Z</option>
                <option value="alphabetical:reverse">Z-A</option>
                <option value="chronological">Chronologique</option>
                <option value="chronological:reverse">Antéchronologique</option>
            </select>

            <ul class="menu-index-list">
                
                <li data-index="20230924120339" style="order: 1">
                    <a href="#20230924120339">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Actinobacteria</span>
                    </a>
                </li>
                
                <li data-index="20230924120340" style="order: 2">
                    <a href="#20230924120340">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Active Learning</span>
                    </a>
                </li>
                
                <li data-index="20230924120341" style="order: 3">
                    <a href="#20230924120341">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>AlphaFold1, AlphaFold2, and RoseTTAFold</span>
                    </a>
                </li>
                
                <li data-index="20230924120342" style="order: 4">
                    <a href="#20230924120342">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Alphaproteobacteria</span>
                    </a>
                </li>
                
                <li data-index="20230924120343" style="order: 5">
                    <a href="#20230924120343">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Amplicon Target Sequencing</span>
                    </a>
                </li>
                
                <li data-index="20230924120344" style="order: 6">
                    <a href="#20230924120344">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Bayesian statistics</span>
                    </a>
                </li>
                
                <li data-index="20230924120345" style="order: 7">
                    <a href="#20230924120345">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Build Workflow from scratch - HaplotypeCaller</span>
                    </a>
                </li>
                
                <li data-index="20230924120346" style="order: 8">
                    <a href="#20230924120346">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Building Docker in Google Cloud</span>
                    </a>
                </li>
                
                <li data-index="20230924120347" style="order: 9">
                    <a href="#20230924120347">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Central Dogma</span>
                    </a>
                </li>
                
                <li data-index="20230924120348" style="order: 10">
                    <a href="#20230924120348">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Curse of The Dimensionality</span>
                    </a>
                </li>
                
                <li data-index="20230924120349" style="order: 11">
                    <a href="#20230924120349">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Dark microbiome and extremely low organics in Atacama fossil delta unveil Mars life detection limits</span>
                    </a>
                </li>
                
                <li data-index="20230924120350" style="order: 12">
                    <a href="#20230924120350">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Data Integration</span>
                    </a>
                </li>
                
                <li data-index="20230924120351" style="order: 13">
                    <a href="#20230924120351">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Dealing with Dimensionality - The Application of Machine Learning to Multi-omics Data</span>
                    </a>
                </li>
                
                <li data-index="20230924120352" style="order: 14">
                    <a href="#20230924120352">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Deep Learning</span>
                    </a>
                </li>
                
                <li data-index="20230924120353" style="order: 15">
                    <a href="#20230924120353">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>DNA</span>
                    </a>
                </li>
                
                <li data-index="20230924120354" style="order: 16">
                    <a href="#20230924120354">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>edwith_LSnDP_Ch1</span>
                    </a>
                </li>
                
                <li data-index="20230924120355" style="order: 17">
                    <a href="#20230924120355">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>edwith_LSnDP_Ch2</span>
                    </a>
                </li>
                
                <li data-index="20230924120356" style="order: 18">
                    <a href="#20230924120356">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>edwith_LSnDP_Ch3</span>
                    </a>
                </li>
                
                <li data-index="20230924120357" style="order: 19">
                    <a href="#20230924120357">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>edwith_LSnDP_Ch4</span>
                    </a>
                </li>
                
                <li data-index="20230924120358" style="order: 20">
                    <a href="#20230924120358">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>edwith_LSnDP_Ch5</span>
                    </a>
                </li>
                
                <li data-index="20230924120359" style="order: 21">
                    <a href="#20230924120359">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>edwith_LSnDP_Ch6</span>
                    </a>
                </li>
                
                <li data-index="20230924120360" style="order: 22">
                    <a href="#20230924120360">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>edwith_LSnDP_Ch7</span>
                    </a>
                </li>
                
                <li data-index="20230924120361" style="order: 23">
                    <a href="#20230924120361">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Ensemble</span>
                    </a>
                </li>
                
                <li data-index="20230924120362" style="order: 24">
                    <a href="#20230924120362">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Exon</span>
                    </a>
                </li>
                
                <li data-index="20230924120363" style="order: 25">
                    <a href="#20230924120363">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Feature Extraction</span>
                    </a>
                </li>
                
                <li data-index="20230924120364" style="order: 26">
                    <a href="#20230924120364">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Feature Selection</span>
                    </a>
                </li>
                
                <li data-index="20230924120365" style="order: 27">
                    <a href="#20230924120365">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Greatest Common Divisor, Least Common Multiple, and Euclidean algorithm</span>
                    </a>
                </li>
                
                <li data-index="20230924120366" style="order: 28">
                    <a href="#20230924120366">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Integration Strategies of Multi-omics Data for Machine Learning Analysis</span>
                    </a>
                </li>
                
                <li data-index="20230924120367" style="order: 29">
                    <a href="#20230924120367">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>IntGen_1.1-1.3_MendelsFirstLawMeiosis</span>
                    </a>
                </li>
                
                <li data-index="20230924120368" style="order: 30">
                    <a href="#20230924120368">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>IntGen_1.4_Monohybrid</span>
                    </a>
                </li>
                
                <li data-index="20230924120369" style="order: 31">
                    <a href="#20230924120369">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>IntGen_1.5_PunnettSquares</span>
                    </a>
                </li>
                
                <li data-index="20230924120370" style="order: 32">
                    <a href="#20230924120370">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>IntGen_1.6_SingleLocusTestCrosses</span>
                    </a>
                </li>
                
                <li data-index="20230924120371" style="order: 33">
                    <a href="#20230924120371">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>IntGen_6.2_SomaticMutVsGermlineMut</span>
                    </a>
                </li>
                
                <li data-index="20230924120372" style="order: 34">
                    <a href="#20230924120372">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Intron</span>
                    </a>
                </li>
                
                <li data-index="20230924120373" style="order: 35">
                    <a href="#20230924120373">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>lecture-list</span>
                    </a>
                </li>
                
                <li data-index="20230924120374" style="order: 36">
                    <a href="#20230924120374">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Long-read Sequencing</span>
                    </a>
                </li>
                
                <li data-index="20230924120375" style="order: 37">
                    <a href="#20230924120375">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Machine Learning</span>
                    </a>
                </li>
                
                <li data-index="20230924120376" style="order: 38">
                    <a href="#20230924120376">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>mRNA</span>
                    </a>
                </li>
                
                <li data-index="20230924120377" style="order: 39">
                    <a href="#20230924120377">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Multi-omics</span>
                    </a>
                </li>
                
                <li data-index="20230924120378" style="order: 40">
                    <a href="#20230924120378">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Mutation</span>
                    </a>
                </li>
                
                <li data-index="20230924120379" style="order: 41">
                    <a href="#20230924120379">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Nanopore</span>
                    </a>
                </li>
                
                <li data-index="20230924120380" style="order: 42">
                    <a href="#20230924120380">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>NGS</span>
                    </a>
                </li>
                
                <li data-index="20230924120381" style="order: 43">
                    <a href="#20230924120381">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>OTUs</span>
                    </a>
                </li>
                
                <li data-index="20230924120382" style="order: 44">
                    <a href="#20230924120382">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>p-value(유의확률)</span>
                    </a>
                </li>
                
                <li data-index="20230924120383" style="order: 45">
                    <a href="#20230924120383">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Prime number and Sieve of Eratosthenes</span>
                    </a>
                </li>
                
                <li data-index="20230924120384" style="order: 46">
                    <a href="#20230924120384">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Protein</span>
                    </a>
                </li>
                
                <li data-index="20230924120385" style="order: 47">
                    <a href="#20230924120385">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Query Strategy</span>
                    </a>
                </li>
                
                <li data-index="20230924120386" style="order: 48">
                    <a href="#20230924120386">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>README</span>
                    </a>
                </li>
                
                <li data-index="20230924120387" style="order: 49">
                    <a href="#20230924120387">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Real-Time DNA Sequencing in the Antarctic Dry Valleys Using the Oxford Nanopore Sequencer</span>
                    </a>
                </li>
                
                <li data-index="20230924120388" style="order: 50">
                    <a href="#20230924120388">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>RNA</span>
                    </a>
                </li>
                
                <li data-index="20230924120389" style="order: 51">
                    <a href="#20230924120389">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>scRNA-seq</span>
                    </a>
                </li>
                
                <li data-index="20230924120390" style="order: 52">
                    <a href="#20230924120390">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>SingleCellBestPractice-1.Introduction</span>
                    </a>
                </li>
                
                <li data-index="20230924120391" style="order: 53">
                    <a href="#20230924120391">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>SMRT</span>
                    </a>
                </li>
                
                <li data-index="20230924120392" style="order: 54">
                    <a href="#20230924120392">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Spatial Transcriptomics</span>
                    </a>
                </li>
                
                <li data-index="20230924120393" style="order: 55">
                    <a href="#20230924120393">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Underfitting and Overfitting</span>
                    </a>
                </li>
                
                <li data-index="20230924120394" style="order: 56">
                    <a href="#20230924120394">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Untitled</span>
                    </a>
                </li>
                
                <li data-index="20230924120395" style="order: 57">
                    <a href="#20230924120395">
                        <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                        <span>Why Bioinformaticiansuse Nextflow Thesedays</span>
                    </a>
                </li>
                
            </ul>
        </details>

        <details class="menu-view" open="open">
            <summary class="menu-title">
                Views
                <span class="badge">0</span>
            </summary>

            <button onclick="window.location.replace(window.location.pathname)">
                Back to default view
            </button>

            <button id="view-save">
                
                Set the view to URL
                
            </button>

            <span class="tooltip">?
                <span class="tooltiptext">Save the current state of the interface (selected node, types and keywords, focus) and creates a shortcut to this state in the Views section.
</span>
            </span>

            <div class="menu-view-container" id="views-container">
            
                
            
            </div>
        </details>

        <details class="menu-controls">
            <summary class="menu-title">
                Graph parameters
            </summary>

            <ul class="menu-controls-list">
                <li>
                    <label>
                        <input type="checkbox" checked onchange="linksDisplayToggle(this.checked)">
                        Display links
                        <span class="badge">21</span>
                    </label>
                </li>
                <li>
                    <label>
                        <input type="checkbox" checked onchange="labelDisplayToggle(this.checked)">
                        Display labels
                    </label>
                </li>
                <li>
                    <label>
                        <input type="checkbox" checked
                            onchange="graphProperties.graph_highlight_on_hover = this.checked">
                        Highlight on hover
                    </label>
                </li>
            </ul>

            

            <article class="menu-controls-container">
                <p class="menu-controls-title">
                    Forces
                    <span class="tooltip">?
                        <span class="tooltiptext"></span>
                    </span>
                </p>
                
            <form oninput="graphProperties.attraction_force = range.value; updateForces()">
                <label>repulsion</label>
                <input type="number" value="200" name="number" min="50" max="600" oninput="range.value = number.value"/>
                <input type="range" min="50" max="600" value="200" name="range" step="0.1" oninput="number.value = range.value"/>
            </form>
            
                
            <form oninput="graphProperties.attraction_distance_max = range.value; updateForces()">
                <label>max distance</label>
                <input type="number" value="250" name="number" min="200" max="800" oninput="range.value = number.value"/>
                <input type="range" min="200" max="800" value="250" name="range" step="0.1" oninput="number.value = range.value"/>
            </form>
            
                
            <form oninput="graphProperties.attraction_vertical = range.value; updateForces()">
                <label>vertical attraction</label>
                <input type="number" value="0" name="number" min="0" max="1" oninput="range.value = number.value"/>
                <input type="range" min="0" max="1" value="0" name="range" step="0.01" oninput="number.value = range.value"/>
            </form>
            
                
            <form oninput="graphProperties.attraction_horizontal = range.value; updateForces()">
                <label>horizontal attraction</label>
                <input type="number" value="0" name="number" min="0" max="1" oninput="range.value = number.value"/>
                <input type="range" min="0" max="1" value="0" name="range" step="0.01" oninput="number.value = range.value"/>
            </form>
            
            </article>

            <article class="menu-controls-container">
                <p class="menu-controls-title">
                    Labels
                </p>
                
            <form oninput="graphProperties.text_size = range.value; updateFontsize()">
                <label>font size</label>
                <input type="number" value="10" name="number" min="2" max="15" oninput="range.value = number.value"/>
                <input type="range" min="2" max="15" value="10" name="range" step="0.5" oninput="number.value = range.value"/>
            </form>
            
            </article>

            
            <article class="menu-controls-container">
                <p class="menu-controls-title">Mode de focus</p>
                <select name="focus_mode" id="focus-mode-select">
                    <option value="directed">Tous</option>
                    <option value="inbound">Entrant</option>
                    <option value="outbound">Sortant</option>
                </select>
            </article>
            
        </details>

        <footer class="menu-footer">
            

            <button class="btn-help" onclick="openRecord('help', false)">
                Help
            </button>

            <button class="btn-about" onclick="openRecord('credits', false)">
                
                    Made with Cosma – About
                
            </button>
        </footer>
    </aside>

    

    <main id="record-container" class="record-container">
        <button class="record-btn-close" onclick="closeRecord()">
            Close (Escape)
        </button>

        
        <article id="20230924120339" class="record">
            <header>
                <h1 class="record-title">Actinobacteria</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120339</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p>#Archaea</p>


                
            </main>

            <footer class="list-content-container">
                

                
                <details class="record-backlinks" open="open">
                    <summary class="menu-title">
                        backlinks
                        <span class="badge">1</span>
                    </summary>

                    <ul class="record-backlinks-list">
                        
                        <li data-target-id="20230924120349">
                            <a href="#20230924120349">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Dark microbiome and extremely low organics in Atacama fossil delta unveil Mars life detection limits</span>
                                <div class="record-links-context">
                                    <h1>Abstract</h1>
<ul>
<li>칠레의 Atacama Desert의 red stone(암석)에서 채취한 미생물군집 샘플에서 DNA 흔적을 조사했는데, 발견한 DNA의 9%는 과학계에 보고된 적이 없어 분류 불가능한 미생물의 존재였음.</li>
<li>과학자들은 이렇게 sequencing 기술을 통하여 존재를 확인할 수 있지만, 정확하게 분류할 수 없는 미생물을 &quot;dark microbiome&quot;이라고 정의함.</li>
<li>해당 연구는 지구의 매우 건조한 환경에서도 미생물의 흔적도 많이 손상되 보존되기 어려움을 시사함.</li>
<li>따라서 화성에서 암석샘플 회수에 성공한다고 할지라도, 현재 기술로는 과거에 화성에 존재했던 미생물의 흔적을 놓칠 수 있다는 점을 시사함.</li>
</ul>
<h3>Site microbial diversity and dark microbione</h3>
<ul>
<li>암석샘플은 1 그램당 1μg의 DNA를 갖고 있었음.</li>
<li>Operational Taxonomical Units(OTUs)이 제일 다양하게 검출된 부분은 바로 풍화된 상태의 역암(conglomerate)이었다.</li>
<li>가장 많이 검출된 종은 Alphaproteobacteria 와 <em class="id-context" data-target-id="20230924120339">Actinobacteria</em>였다.<br>
<img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41467-023-36172-1/MediaObjects/41467_2023_36172_Fig4_HTML.png?as=webp"></li>
</ul>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                
            </footer>
        </article>
        
        <article id="20230924120340" class="record">
            <header>
                <h1 class="record-title">Active Learning</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120340</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <ul>
<li>Active Learning은 machine learning(ML) 모델을 훈련 시키기 위해 인간이 직접 데이터를 모두 라벨링 시키기 어렵다는 점을 해결하기 위해 만들어지 개념이다.</li>
<li>인간이 일부 데이터에 대하여 라벨링을 하고 제시하면, 모델이 나머지 데이터들을 평가하고, 이 중 라벨링이 필요하다고 판단되는 중요한 데이터는 다시 인간에게 건내준다.<br>
<img src="https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FY3XYy%2FbtqGishAHQX%2FOaVY0ywxxsd3yPwckOB81k%2Fimg.png"></li>
</ul>
<h1>Three scienarios of Active Learning</h1>
<h3>1. Membership Query Synthesis</h3>
<img src="https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FwDUQn%2FbtqGewyBhJH%2FddZ2JWK7n3mNPsK4N7Nr20%2Fimg.png">
<ul>
<li>학습모델이 주어진 데이터를 활용해서 약간 왜곡된 데이터 인스턴스를 생성한 후, 이에 대한 라벨링을 인간에게 요구한다.</li>
</ul>
<h3>2. Stream-Based Selective Sampling</h3>
<img src="https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FuZHQk%2FbtqGh5trbmp%2FShGf5Y6wvONJzRlFBsDC6K%2Fimg.png">
<ul>
<li>학습모델이 주어진 데이터의 정보량을 평가하고, 라벨링이 필요하다고 판단되는 경우에만 인간에게 전달한다.</li>
<li>정보량을 평가할 때는 <a href="#20230924120385" title="Query Strategy" class="record-link">Query Strategy</a>를 사용한다. 라벨링이 필요하다고 판단되는 경우는 query하고 아닌 것은 버린다.</li>
</ul>
<h3>3. Pool-Based Sampling</h3>
<img src="https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FbYge1J%2FbtqGishA6lv%2FUlUodeq23UIJYd1WNNnqWK%2Fimg.png">
<ul>
<li>가장 널리 사용된다.</li>
<li>라벨링 되지 않은 데이터가 매우 많을 때 사용한다.</li>
<li>학습모델이 주어진 데이터를 평가하고, 정보량이 가장 많은 데이터만 선택해서 인간에게 전달한다.</li>
<li>정보량을 평가할 때 버리는 instance가 존재하지 않는다.</li>
</ul>
<p>참고: <a href="https://littlefoxdiary.tistory.com/52">https://littlefoxdiary.tistory.com/52</a></p>


                
            </main>

            <footer class="list-content-container">
                
                <details class="record-links" open="open">
                    <summary class="menu-title">
                        links
                        <span class="badge">1</span>
                    </summary>

                    <ul class="record-links-list">
                        
                        <li data-target-id="20230924120385">
                            <a href="#20230924120385">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Query Strategy</span>
                                <div class="record-links-context">
                                    <ul>
<li>학습모델이 주어진 데이터의 정보량을 평가하고, 라벨링이 필요하다고 판단되는 경우에만 인간에게 전달한다.</li>
<li>정보량을 평가할 때는 <em class="id-context" data-target-id="20230924120385">Query Strategy</em>를 사용한다. 라벨링이 필요하다고 판단되는 경우는 query하고 아닌 것은 버린다.</li>
</ul>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                

                
                <details class="record-backlinks" open="open">
                    <summary class="menu-title">
                        backlinks
                        <span class="badge">1</span>
                    </summary>

                    <ul class="record-backlinks-list">
                        
                        <li data-target-id="20230924120375">
                            <a href="#20230924120375">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Machine Learning</span>
                                <div class="record-links-context">
                                    <h2>Example</h2>
<p>Deep Learning<br>
<em class="id-context" data-target-id="20230924120340">Active Learning</em></p>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                
            </footer>
        </article>
        
        <article id="20230924120341" class="record">
            <header>
                <h1 class="record-title">AlphaFold1, AlphaFold2, and RoseTTAFold</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120341</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <h1>알파폴드1</h1>
<p><a href="https://taehojo.github.io/alphafold/alphafold1.html">알파폴드 리뷰와 해석</a></p>
<ul>
<li><a href="http://www.incodom.kr/Multiple_alignment">Multiple Sequence Alignment(MSA, 다중서열정렬)</a> 데이터를 준비</li>
<li>Deep Learning 알고리즘 중 Convolutional Neural Network(CNN) 알고리즘을 이용
<ul>
<li>데이터에 대한 Distance Map을 이용</li>
<li>Distance Map이란 실제 단백질의 3차원 구조에 기반하여 아미노산 서열들 사이의 거리를 표현한 이미지</li>
<li>실제 이미지를 기반으로 알고리즘을 학습시킴</li>
</ul>
</li>
<li>Gradient descent(경사하강법)을 통하여 실제 이미지와 예측된 이미지의 오차를 조절</li>
</ul>
<h1>알파폴드2</h1>
<p><a href="https://taehojo.github.io/alphafold/alphafold2.html">알파폴드2 모델 분석</a><br>
<a href="https://blog.naver.com/economic_moat/222506288911">알파폴드 2 작동원리 / alphafodl2</a><br>
<a href="https://happyhaelee.tistory.com/98">Alphafold2 논문 리뷰2 - 입력부분</a><br>
<a href="https://happyhaelee.tistory.com/97">Alphafold2 논문리뷰3[Evoformer -  MSA representation]</a></p>
<ul>
<li>Evoformer에 넣기 위한 두 개의 representation 준비
<ul>
<li>알파폴드1과 유사하게 MSA representation을 먼저 준비함</li>
<li>거기에 DB에서 검색된 query sequence와 유사한 sequence를 가진 단백질 template를 활용, template 구조를 pair representation으로 통합하여 준비함</li>
</ul>
</li>
<li>Evoformer
<ul>
<li>Deep transformer-like Network를 적용함</li>
<li>MSA representation과 Pair representation의 정보가 서로 교환되면서 개선</li>
</ul>
</li>
<li>Structure module
<ul>
<li>개선된 두 개의 representation은 아직 3차원 단백질 구조가 아니라 2차원의 데이터임.</li>
<li>단백질의 모든 잔기각 동일한 지점에서 동일한 방향성을 갖고 있다고 설정함(블랙홀 초기화)</li>
<li>Invariant Point Attention(IPA) 연산과 sequence representation을 이용하여 residue gas를 반복적으로 개선함.</li>
<li>Resnet을 추가하여 side chain의 모든 원자에 대한 비틀림 각도를 예측하여 3차원 좌표로 만듦</li>
</ul>
</li>
</ul>
<h1>로제타폴드</h1>
<p><a href="http://www.aitimes.com/news/articleView.html?idxno=140110">구글 인력·컴퓨팅 없이 알파폴드2 재현한 로제타폴드, 어떻게 가능했나?...연구 주도한 백민경 박사 발표 내용</a></p>
<ul>
<li>CNN이 아닌 트랜스포머(Transformer) 일종인 어텐션(Attention)을 도입</li>
<li>&quot;CNN에서는 바로 옆 픽셀만이 중요한 정보를 담기에 여기에 비중을 둔다. 반면 단백질 구조 파악을 위해서는 멀리 떨어진 곳의 정보도 필요하다.&quot;</li>
<li>어텐션을 활용하면 전체 투입정보(input)를 보고 연관도에 따라 정보를 가져오는 양을 조절.</li>
<li>“중간 과정에서부터 3차원 구조를 만들어가면서 서열정보와 2D 이미지 상호작용 정보를 업데이트했다. 단백질 서열과 구조 정보 간 타이트한 커넥션을 학습했다”<br>
즉, 성능 개선을 위한 결단으로 1, 2, 3차원 정보 간 연결을 강화한 것이다. 로제타폴드 이름인 ‘RoseTTA(Three-Track Attention)’도 여기에서 나왔다.</li>
<li>단백질 구조 예측 관련 학계 전체적으로 주목하는 주제는 cryoEM 데이터 활용을 통한 단백질 멀티스테이트 구조 예측이다.</li>
</ul>
<p><a href="https://blog.ksaidev.com/alphafole2%EC%99%80-rosetta-fold-%EB%8B%A8%EB%B0%B1%EC%A7%88-%EA%B5%AC%EC%A1%B0-%EC%98%88%EC%B8%A1/">Alphafold2와 RoseTTA fold – 단백질 구조 예측</a></p>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120342" class="record">
            <header>
                <h1 class="record-title">Alphaproteobacteria</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120342</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p>#Archaea</p>


                
            </main>

            <footer class="list-content-container">
                

                
                <details class="record-backlinks" open="open">
                    <summary class="menu-title">
                        backlinks
                        <span class="badge">1</span>
                    </summary>

                    <ul class="record-backlinks-list">
                        
                        <li data-target-id="20230924120349">
                            <a href="#20230924120349">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Dark microbiome and extremely low organics in Atacama fossil delta unveil Mars life detection limits</span>
                                <div class="record-links-context">
                                    <h1>Abstract</h1>
<ul>
<li>칠레의 Atacama Desert의 red stone(암석)에서 채취한 미생물군집 샘플에서 DNA 흔적을 조사했는데, 발견한 DNA의 9%는 과학계에 보고된 적이 없어 분류 불가능한 미생물의 존재였음.</li>
<li>과학자들은 이렇게 sequencing 기술을 통하여 존재를 확인할 수 있지만, 정확하게 분류할 수 없는 미생물을 &quot;dark microbiome&quot;이라고 정의함.</li>
<li>해당 연구는 지구의 매우 건조한 환경에서도 미생물의 흔적도 많이 손상되 보존되기 어려움을 시사함.</li>
<li>따라서 화성에서 암석샘플 회수에 성공한다고 할지라도, 현재 기술로는 과거에 화성에 존재했던 미생물의 흔적을 놓칠 수 있다는 점을 시사함.</li>
</ul>
<h3>Site microbial diversity and dark microbione</h3>
<ul>
<li>암석샘플은 1 그램당 1μg의 DNA를 갖고 있었음.</li>
<li>Operational Taxonomical Units(OTUs)이 제일 다양하게 검출된 부분은 바로 풍화된 상태의 역암(conglomerate)이었다.</li>
<li>가장 많이 검출된 종은 <em class="id-context" data-target-id="20230924120342">Alphaproteobacteria</em> 와 Actinobacteria였다.<br>
<img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41467-023-36172-1/MediaObjects/41467_2023_36172_Fig4_HTML.png?as=webp"></li>
</ul>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                
            </footer>
        </article>
        
        <article id="20230924120343" class="record">
            <header>
                <h1 class="record-title">Amplicon Target Sequencing</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120343</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                

                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120344" class="record">
            <header>
                <h1 class="record-title">Bayesian statistics</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120344</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                

                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120345" class="record">
            <header>
                <h1 class="record-title">Build Workflow from scratch - HaplotypeCaller</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120345</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <h1>목표</h1>
<ul>
<li>어떠한 준비도 되어있지 않은 방금 설치한 linux system에 대하여</li>
<li>HaplotypeCaller만 돌릴 수 있는 환경을 구성해본다.</li>
</ul>
<h1>준비된 환경</h1>
<ol>
<li>Rocky Linux 9
<ul>
<li>Docker를 활용하여 windows11 환경에서 구축</li>
</ul>
</li>
<li>서버를 다루기 편하도록 Docker와 Visual Studio Code를 서로 연결
<ul>
<li>VSCode의 Docker 확장프로그램을 통하여 연결</li>
<li>Docker 내의 Rocky Linux 9에서 터미널 열기
<ul>
<li>Docker Window의 bash 혹은 VSCode에서 attach to bash로 접속하면 cluster system에 접속할 때 처럼 bash가 뜬다. (<code>sh-5.1#</code>)</li>
<li>Docker에서 Rocky Linux 9이 깔린 컨테이너를 실행한 다음, VSCode에서 해당 컨테이어와 터미널을 연결시키면 새 창에서 접속하게 된다.</li>
</ul>
</li>
</ul>
</li>
</ol>
<h1>1. 자주 사용하는 기본적인 패키지, 라이브러리 설치</h1>
<h2>기본상식: yum? dnf? apt? sudo?</h2>
<ul>
<li><code>yum</code>과 <code>apt</code>는 모두 패키지 설치 관리자다. 다만 같은 리눅스 계열이라 할지라도 어느 리눅스 시스템을 쓰느냐에 따라 다른 패키지 설치 관리자를 사용하게 된다.</li>
<li><code>yum</code>은 Rocky Linux 9와 같은 '레드헷 리눅스(RHEL)' 계열의 리눅스 시스템의 패키지 설치 관리자이다.
<ul>
<li>RHEL에는 대표적으로 Centos 시리즈가 잘 알려져있다.</li>
</ul>
</li>
<li><code>dnf</code>는 RHEL 계열 중에서도 버전 8 이상부터 지원하는 패키지 설치 관리자이다. <code>yum</code>을 사용하는 방식과 똑같이 사용할 수 있다. 다만 상대적으로 RHEL8 이상 버전이 아직 널리 쓰이지 않아서 많은 인터넷의 팁들이 <code>yum</code>으로 명령어를 소개하고 있다. Rocky Linux의 경우 RHEL9 계열이므로 당연히 dnf로도 설치가 가능하다.</li>
<li><code>apt</code> 패키지 설치 관리자는 Ubuntu에서 사용되는 패키지 설치 관리자다.
<ul>
<li>RHEL계열에서 기본적으로는 사용되지 않는다.</li>
</ul>
</li>
<li><code>sudo</code>는 패키지의 한 종류로, 비밀번호만 알고 있다면 일반 사용자도 관리자 권한으로 명령어를 수행할 수 있게 해준다. 리눅스 시스템 계열과 상관없이 사용할 수 있지만, Rocky Linux에는 기본적으로 설치가 되어있지 않 패키지라 직접 설치해줘야한다.</li>
</ul>
<h1>기본적으로 다운받은 패키지들</h1>
<p>모든 명령어는 일반사용자가 아닌 root 사용자로 로그인한 상태에서 진행하였다.</p>
<p><strong>which</strong>  <code>yum install which</code></p>
<p><strong>python</strong> <code>yum install python</code></p>
<p><strong>pip</strong> <code>yum install pip</code></p>
<p><strong>wget</strong> <code>pip install wget</code></p>
<p><strong>sudo</strong> <code>yum install sudo</code></p>
<p><strong>vim</strong> <code>yum install vim</code></p>
<p><strong>unzip</strong> <code>yum install unzip</code></p>
<h1>2. GATK부터 설치</h1>
<h2>GATK를 다운받자</h2>
<ul>
<li>간단하게 파일만 받아서 설치하자.
<ul>
<li>파일을 다운 받는다: <a href="https://github.com/broadinstitute/gatk/releases">GATK github /release</a></li>
<li><code>unzip gatk-4.4.0.0.zip</code></li>
</ul>
</li>
<li>위와 같은 간단한 방법 말고도, Anaconda로 받거나, 공식 github에서 clone으로 받거나, docker image로 받거나 여러 방법이 있다. <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360036194592-Getting-started-with-GATK4">참고: Getting started with GATK4</a>
<ul>
<li>그러나 이런 방법들은 다시 추가적인 환경 구축을 요구한다 -&gt; 귀찮다</li>
</ul>
</li>
<li>대신 수동설치는 최신버전을 받기 위해서 주기적으로 수동설치를 요구한다는 단점 아닌 단점이 있다.
<ul>
<li>파이프라인의 안정성 유지와 최신 버전을 통한 성능향상은 모두 중요하다.</li>
<li>최신 버전이 항상 더 나은 퍼포먼스를 약속하지 않으며, 예상치 못한 버그를 일으킬 수 있다.</li>
</ul>
</li>
</ul>
<h2>GATK를 위한 필수프로그램들을 다운받자</h2>
<p><a href="https://github.com/broadinstitute/gatk#requirements">GATK requirements</a></p>
<ul>
<li>필수 프로그램들은 run/build 목적에 따라 다르게 소개된다. 이번에는 run만을 위한 프로그램만 설치한다.</li>
</ul>
<ol>
<li>특정 버전의 JAVA
<ul>
<li>4.4.0.0 버전의 경우 OpenJDK 17 버전을 요구하고 있다.</li>
<li>향후에 요구사항이 바뀔 수 있으니 설치 전에 항상 requirement 부분을 잘 읽어보자</li>
<li><a href="https://computingforgeeks.com/install-java-openjdk-17-on-centos-rhel-7/?expand_article=1">install openjdk17 on centos RHEL 7</a></li>
<li><code>wget https://download.oracle.com/java/17/latest/jdk-17_linux-x64_bin.rpm</code></li>
<li><code>yum -y install ./jdk-17_linux-x64_bin.rpm</code></li>
<li>버전확인도 수행하자<pre><code class="language-shell"># java --version
java 17.0.8 2023-07-18 LTS
</code></pre>
</li>
</ul>
</li>
<li>Python
<ul>
<li>이미 설치했다</li>
</ul>
</li>
<li>R
<ul>
<li>R은 잘 알려진 통계 및 그래프 그리기에 특화된 프로그램이다.</li>
<li>Centos, Rocky Linux와 같은 RHEL 계열의 패키지 설치 관리자는 기본적으로는 R을 제공하지 않는다.</li>
<li>대신, Powertools라고 하여 개발자 전용 패키지를 추가로 설치하여 주면, R이 포함되어 있다.</li>
<li><code>yum install epel-release</code></li>
<li><code>yum config-manager --set-enabled crb</code></li>
<li><code>yum install R</code></li>
</ul>
</li>
</ol>
<h2>GATK가 정상적으로 실행되는지 확인해보자</h2>
<ul>
<li>gatk가 설치되어 있는 디렉토리에서 gatk를 찾는다.
<ul>
<li>만약 이 과정이 귀찮다면, gatk를 bash enviroment에 추가하는 방법도 있다. 다만 개인적으로 필수적이거나 설치 과정 중에서 자연스럽게 추가되지 않는 경우에는 지양하는 편이다.</li>
</ul>
</li>
<li>나의 경우 <code>/download/gatk-4.4.0.0</code> 디렉토리에 gatk가 들어있었다.</li>
<li><code>--list</code> 를 통하여 GATK를 돌려본다.<pre><code class="language-shell"># /download/gatk-4.4.0.0/gatk --list
Using GATK jar /download/gatk-4.4.0.0/gatk-package-4.4.0.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /download/gatk-4.4.0.0/gatk-package-4.4.0.0-local.jar --help
USAGE:  &lt;program name&gt; [-h]

Available Programs:
...
</code></pre>
</li>
<li>이렇게 프로그램 목록이 쫙 뜨면 성공이다.</li>
</ul>
<h1>3. SAMtools 설치</h1>
<ul>
<li>SAMtools는 BAM/SAM/CRAM 파일을 다루는데 특화되어있다.</li>
<li>특히 간단하게 BAM 파일에서 무언가를 확인할 때 좋다.</li>
</ul>
<h2>SAMtools를 다운받자</h2>
<p><a href="https://github.com/samtools/samtools/releases">https://github.com/samtools/samtools/releases</a></p>
<blockquote>
<p><em>samtools release 1.18<br>
Download the source code here: <a href="https://github.com/samtools/samtools/releases/download/1.18/samtools-1.18.tar.bz2">samtools-1.18.tar.bz2</a>. (The &quot;Source code&quot; downloads are generated by GitHub and are incomplete as they don't bundle HTSlib and are missing some generated files.)</em></p>
</blockquote>
<ul>
<li>어떤 이유인지 1.18 버전의 경우 링크가 따로 걸려있다.</li>
<li><code>wget</code> 명령어를 통해 다운받는다.</li>
</ul>
<h2>tar.bz2 압축파일을 풀자</h2>
<ul>
<li><code>bz2</code> 확장자 압축파일을 풀기위해서는 <code>bzip2</code> 패키지를 먼저 설치해줘야 한다.
<ul>
<li><code>yum install bzip2</code></li>
<li>이제 bzip2를 활용하여 압축파일을 해제한다.
<ul>
<li><code>bzip2 -d samtools-1.18.tar.bz2</code></li>
</ul>
</li>
</ul>
</li>
<li><code>tar</code> 확장자 압축파일을 풀기 위해서는 <code>tar</code> 패키지를 사용한다.
<ul>
<li>옵션이 굉장히 많은 편인데, 보통 <code>-xvf</code> 옵션을 사용한다.<br>
-x : 압축 파일 풀기 (create)<br>
-v : 묶음/해제 과정을 화면에 표시 (verbose)<br>
-f : 파일 이름을 지정 (file)
<ul>
<li><code>tar -xvf samtools-1.18.tar</code></li>
<li>이제 samtools-1.18 폴더가 생성된다.</li>
</ul>
</li>
</ul>
</li>
</ul>
<h2>SAMtools를 설치하자</h2>
<ul>
<li>SAMtools는 GATK와 달리 압축을 풀었어도 바로 사용할 수 없다.</li>
<li>압축이 풀린 SAMtools 폴더에 들어가면 굉장히 파일이 많다는 사실을 알 수 있다.</li>
<li>github에서 clone하는 방식으로 다운로드 받은 경우 <code>autoheader</code> 라던가 <code>autoconf</code> 라던가 추가적인 패키지가 필요된다. 다행히 단순하게 다운받았기 때문에 이러한 과정은 생략한다.</li>
<li>곧바로 SAMtools 폴더 내에서 설치환경 구성을 시도한다.
<ul>
<li><code>./configure</code></li>
<li>그러면 다음과 같은 에러 메시지를 만나게 된다.</li>
</ul>
<pre><code class="language-shell">checking for NcursesW wide-character library... no
checking for Ncurses library... no
checking for Curses library... no
configure: error: curses development files not found

The 'samtools tview' command uses the curses text user interface library.
Building samtools with tview requires curses/ncurses/etc development files
to be installed on the build machine; you may need to ensure a package such
as libncurses5-dev (on Debian or Ubuntu Linux) or ncurses-devel (on RPM-based
Linux distributions) is installed.

FAILED.  Either configure --without-curses or resolve this error to build
samtools successfully.
</code></pre>
<ul>
<li>위의 메시지를 종합하면 SAMtools의 tview 기능을 사용하기 위해서는 ncurse 패키지를 설치하거나 --without-curses 옵션을 사용하라는 것이다.</li>
<li>SAMtools를 설치하는 과정에서 이와 같은 오류 메시지는 4번 정도 발생하게 되므로, 옵션으로 생략하기 보다는 그냥 다 설치하는 편이 낫다.</li>
<li>메시지에서 중요하게 읽어야 하는 부분은 다음의 부분이다:<br>
<em>&quot;you may need to ensure a package such as libncurses5-dev (on Debian or Ubuntu Linux) or ncurses-devel (on RPM-based Linux distributions) is installed.&quot;</em>
<ul>
<li>리눅스 시스템 계열에 따라 다른 패키지를 설치해야한다.</li>
<li>Rocky Linux의 경우 RPM-based Linux distribution이다.</li>
<li>따라서, <code>yum install ncurses-devel</code> 명령어로 ncurse를 설치한 다음, 다시 <code>./configure</code> 명령어로 설치환경 구성을 시도해야한다.</li>
</ul>
</li>
<li><code>ncurse</code>를 포함하여 총 4번의 패키지 설치 요청이 들어온다.
<ul>
<li><code>yum install ncurses-devel</code></li>
<li><code>yum install zlib-devel</code></li>
<li><code>yum install bzip2-devel</code></li>
<li><code>yum install xz-devel</code></li>
</ul>
</li>
<li>마지막으로 <code>make</code> 명령어를 통해 설치환경 구성을 마친다.
<ul>
<li><code>make</code></li>
</ul>
</li>
</ul>
</li>
<li>이제 진짜로 설치를 한다.
<ul>
<li><code>make install</code></li>
</ul>
</li>
<li>SAMtools의 경우 설치가 끝나면 env에 자동으로 명령어가 등록된다. 설치가 잘 되었는지 확인하자.<pre><code class="language-shell"># samtools --help

Program: samtools (Tools for alignments in the SAM format)
Version: 1.18 (using htslib 1.18)

Usage:   samtools &lt;command&gt; [options]

Commands:
...
</code></pre>
</li>
</ul>
<h2>4. 인터넷에서 BAM파일을 받아 variant calling을 시도해보자</h2>
<ul>
<li>Variant calling을 수행하게 하는 caller의 종류는 여러가지가 있다.
<ul>
<li>DeepVariant</li>
<li>GATK의 HaplotypeCaller</li>
<li>Vardict 등등....</li>
</ul>
</li>
<li>오늘은 GATK의 HaplotypeCaller를 통하여 calling을 시도해보자.</li>
</ul>
<h2>NCBI에서 받고자 하는 데이터 검색하기</h2>
<ul>
<li>National Center for Biotechnology Information(NCBI)는 미국 국립보건원에서 운영하고 있는 생물정보학 DB다.</li>
<li>NCBI의 DB는 여러가지 종류가 있는데, 그중 Sequence Read Archive(SRA)에 sequencing 데이터 모음을 받을 수 있다.
<ul>
<li>SRA에서 BRCA와 관련된 용량이 작은 데이터를 검색해봤다.
<ul>
<li>SRA에서 한번 검색한 다음, SRA run selector로 결과를 보내겠냐는 메시지를 클릭하면 좀 더 편하게 정렬하여 검색이 가능하다.</li>
<li>추천하는 방식으로는 '관심있는 질병의 유전자명'으로 검색한 다음, 필터의 access 항목에 'public'으로 필터를 건 다음에 SRA run selector로 결과를 보내서 용량이 작은 순으로 정렬하여 파일을 받는 것이다. Access type에 대한 필터는 run selector에 없는 것으로 보이기 때문이다.</li>
</ul>
</li>
<li><a href="https://trace.ncbi.nlm.nih.gov/Traces/index.html?view=run_browser&amp;page_size=10&amp;acc=SRR9929551&amp;display=alignment">SRR9929551</a> 를 선택</li>
<li>그러나 SRA 웹사이트에서 데이터를 바로 받을 수는 없다.</li>
</ul>
</li>
<li>SRA DB에서 데이터를 받기 위해서는 SRAtoolkit을 설치한 다음, 명령어를 통해 받을 수 있다.</li>
</ul>
<h2>NCBI SRA DB에서 데이터를 받기 위해 SRAToolKit 설치하기</h2>
<p><a href="https://github.com/ncbi/sra-tools/wiki/02.-Installing-SRA-Toolkit">Installing SRA Toolkit</a></p>
<ul>
<li>툴킷을 Rocky Linux 내로 바로 다운받는다.
<ul>
<li><code>wget https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/current/sratoolkit.current-centos_linux64.tar.gz</code></li>
<li><code>gzip -d sratoolkit.current-centos_linux64.tar.gz</code></li>
<li><code>tar -xvf sratoolkit.current-centos_linux64.tar</code></li>
</ul>
</li>
<li>필요한 경우 Github의 설명에 따라 환경변수에 경로를 추가할 수도 있다.</li>
<li>BAM파일만 받을 생각이기 때문에 여러가지 툴 중에서 <code>sam-dump</code> 툴을 활용한다.
<ul>
<li>받으려는 SRA 데이터의 번호는 <code>SRR9929551</code></li>
<li><code>sam-dump</code> 툴은 파일 내용을 그대로 읽어서 출력하기 때문에, 이를 저장해줘야한다.
<ul>
<li><code>/download/sratoolkit.3.0.6-centos_linux64/bin/sam-dump SRR9929551 &gt;&gt; SRR9929551.sam</code></li>
</ul>
</li>
<li><code>sam-dump</code> 이외에 자주 사용되는 명령어로는....<br>
<code>prefetch</code>: BAM파일 뿐 아니라 관련된 전체 데이터를 받는다.<br>
<code>fastq-dump</code>: fastq파일만 받는다</li>
</ul>
</li>
</ul>
<h2>SAMtools를 이용하여 SAM 파일을 BAM 파일로 변환하기</h2>
<ul>
<li>SAMtools의 <code>view</code> 기능을 이용하여 BAM 파일로 저장하여준다.
<ul>
<li><code>samtools view -b -o SRR9929551.bam SRR992551.sam</code></li>
</ul>
</li>
<li>BAM파일의 index파일도 작성하여 준다.
<ul>
<li><code>samtools index SRR9929551.bam -o SRR9929551.bam.bai</code></li>
</ul>
</li>
</ul>
<h2>SAMtools를 이용하여 BAM파일의 reference 파일 확인하고 다운받기</h2>
<ul>
<li>GATK HaployypeCaller의 경우 두 가지의 input 파일이 필요하다.
<ol>
<li>BAM 파일</li>
<li>BAM 파일이 만들어질 때 사용된 reference fasta 파일
<ul>
<li>Reference 파일은 index 파일과 dict 파일과 함께 존재해야한다.</li>
</ul>
</li>
</ol>
</li>
<li><code>samtools view --header-only</code> 옵션을 통하여 reference 파일을 파악하자<pre><code class="language-shell"># samtools view --header-only SRR9929551.bam | grep &quot;reference&quot;
</code></pre>
</li>
</ul>
<p>@PG     ID:tmap CL:mapall -n 12 -f /results/referenceLibrary/tmap-f3/hg19/hg19.fasta -r basecaller_results/IonXpress_016_rawlib.basecaller.bam -v -Y -u --prefix-exclude 5 -o 2 stage1 map4     VN:5.0.7 (517ac65) (201509092132)<br>
```</p>
<ul>
<li>구글링을 통하여 <code>tmap</code>은 torrent mapping의 준말로 iontorrent에서 제공되는 reference 파일로 파악되었다.</li>
<li><a href="https://github.com/domibel/IonTorrent-VariantCaller#get-reference-genome-hg19-from-ion-torrent-size-868170684-828m">구글링</a>을 통하여 reference fasta 파일을 다운로드 받았다.
<ul>
<li><code>wget http://ionupdates.com/reference/hg19.zip</code></li>
<li><code>unzip hg19.zip</code></li>
</ul>
</li>
<li>Reference fasta 파일의 index과 dict 파일을 생성하여 HaplotypeCaller를 위한 준비를 한다.
<ul>
<li><code>samtools faidx hg19.fasta</code></li>
<li><code>samtools dict hg19.fasta -o hg19.dict</code></li>
</ul>
</li>
</ul>
<h2>HaplotypeCaller 실행하기</h2>
<pre><code class="language-shell"># ./gatk-4.4.0.0/gatk HaplotypeCaller -R hg19.fasta -I SRR9929551.bam -O test.vcf.gz
Using GATK jar /download/gatk-4.4.0.0/gatk-package-4.4.0.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /download/gatk-4.4.0.0/gatk-package-4.4.0.0-local.jar HaplotypeCaller -R hg19.fasta -I SRR9929551.bam -O test.vcf.gz
20:39:42.774 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/download/gatk-4.4.0.0/gatk-package-4.4.0.0-local.jar!/com/intel/gkl/native/libgkl_compression.so
20:39:42.949 INFO  HaplotypeCaller - ------------------------------------------------------------
20:39:42.956 INFO  HaplotypeCaller - The Genome Analysis Toolkit (GATK) v4.4.0.0
20:39:42.956 INFO  HaplotypeCaller - For support and documentation go to https://software.broadinstitute.org/gatk/
20:39:42.956 INFO  HaplotypeCaller - Executing as root@ROCKY on Linux v5.15.90.1-microsoft-standard-WSL2 amd64
20:39:42.957 INFO  HaplotypeCaller - Java runtime: Java HotSpot(TM) 64-Bit Server VM v17.0.8+9-LTS-211
20:39:42.957 INFO  HaplotypeCaller - Start Date/Time: August 3, 2023 at 6:39:42 PM UTC
20:39:42.958 INFO  HaplotypeCaller - ------------------------------------------------------------
20:39:42.958 INFO  HaplotypeCaller - ------------------------------------------------------------
20:39:42.960 INFO  HaplotypeCaller - HTSJDK Version: 3.0.5
20:39:42.961 INFO  HaplotypeCaller - Picard Version: 3.0.0
20:39:42.961 INFO  HaplotypeCaller - Built for Spark Version: 3.3.1
20:39:42.963 INFO  HaplotypeCaller - HTSJDK Defaults.COMPRESSION_LEVEL : 2
20:39:42.964 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false
20:39:42.965 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true
20:39:42.968 INFO  HaplotypeCaller - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false
20:39:42.968 INFO  HaplotypeCaller - Deflater: IntelDeflater
20:39:42.969 INFO  HaplotypeCaller - Inflater: IntelInflater
20:39:42.969 INFO  HaplotypeCaller - GCS max retries/reopens: 20
20:39:42.969 INFO  HaplotypeCaller - Requester pays: disabled
20:39:42.971 INFO  HaplotypeCaller - Initializing engine
20:39:43.816 INFO  HaplotypeCaller - Done initializing engine
20:39:43.927 INFO  HaplotypeCallerEngine - Disabling physical phasing, which is supported only for reference-model confidence output
20:39:43.993 INFO  NativeLibraryLoader - Loading libgkl_utils.so from jar:file:/download/gatk-4.4.0.0/gatk-package-4.4.0.0-local.jar!/com/intel/gkl/native/libgkl_utils.so
20:39:44.003 INFO  NativeLibraryLoader - Loading libgkl_pairhmm_omp.so from jar:file:/download/gatk-4.4.0.0/gatk-package-4.4.0.0-local.jar!/com/intel/gkl/native/libgkl_pairhmm_omp.so
20:39:44.096 INFO  IntelPairHmm - Using CPU-supported AVX-512 instructions
20:39:44.097 INFO  IntelPairHmm - Flush-to-zero (FTZ) is enabled when running PairHMM
20:39:44.098 INFO  IntelPairHmm - Available threads: 8
20:39:44.101 INFO  IntelPairHmm - Requested threads: 4
20:39:44.102 INFO  PairHMM - Using the OpenMP multi-threaded AVX-accelerated native PairHMM implementation
20:39:44.264 INFO  ProgressMeter - Starting traversal
20:39:44.266 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Regions Processed   Regions/Minute
20:39:54.299 INFO  ProgressMeter -        chr1:15651665              0.2                 52180         313080.0
</code></pre>
<p>ProgressMeter가 뜨면 variant calling이 진행되기 시작한 것이다.</p>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120346" class="record">
            <header>
                <h1 class="record-title">Building Docker in Google Cloud</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120346</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://brunch.co.kr/@topasvga/705">개발 서버 무료로 구하는 법 9가지</a><br>
<a href="https://brunch.co.kr/@topasvga/168">1.(시작) 구글 Cloud 가입과 서버 1대 받기</a><br>
<a href="https://kibbomi.tistory.com/241">구글 클라우드 플랫폼에 Docker로 웹서버 띄우기</a><br>
<a href="https://hwan-hobby.tistory.com/8">Google Cloud Machine Type에 대해 알아보기 (E2, N2, N2D, N1)</a><br>
<a href="https://cloud.google.com/compute/docs/machine-resource?hl=ko">머신 계열 리소스 및 비교 가이드</a><br>
<a href="https://cloud.google.com/container-optimized-os/docs/how-to/create-configure-instance?hl=ko">Google의 Container-Optimized OS를 실행하는 Compute Engine 인스턴스 만들고 구성하기</a><br>
<a href="https://blog.naver.com/rayhaha/223000083672">구글 클라우드 플랫폼</a><br>
<a href="https://wnsgml972.github.io/setting/2020/07/20/docker/">Docker 설치하고, Docker로 웹 어플리케이션 배포해보기</a><br>
<a href="https://youngjinmo.github.io/2019/10/docker-image/">Docker - 이미지 생성/배포하기</a><br>
<a href="https://herojoon-dev.tistory.com/120">windows 11에 docker 설치하기 (wsl2 이슈 해결)</a><br>
<a href="https://learn.microsoft.com/ko-kr/windows/wsl/install-manual">이전 버전 WSL의 수동 설치 단계</a><br>
<a href="https://docs.rockylinux.org/guides/interoperability/import_rocky_to_wsl/#import-rocky-linux-to-wsl" title="Permanent link">Import Rocky Linux to WSL</a><br>
<a href="https://tech.cloudmt.co.kr/2022/06/29/%EB%8F%84%EC%BB%A4%EC%99%80-%EC%BB%A8%ED%85%8C%EC%9D%B4%EB%84%88%EC%9D%98-%EC%9D%B4%ED%95%B4-1-3-%EC%BB%A8%ED%85%8C%EC%9D%B4%EB%84%88-%EC%82%AC%EC%9A%A9%EB%B2%95/">도커와 컨테이너의 이해</a><br>
<a href="https://velog.io/@dailylifecoding/docker-install-centos-and-run-basic">[Docker] 도커로 Centos 다루기</a></p>
<h1>목표</h1>
<ol>
<li>Docker container에 QIIME, scRNA-seq 분석이 가능한 파이프라인 구축</li>
<li>생성한 docker container를 Google cloud에 적용하고 시험</li>
<li>Google cloud 3개월의 무료 체험이 끝나면 docker container만 다른 클라우드 서비스로 이동</li>
</ol>
<h1>과정</h1>
<h2>1. Windows11에 Docker desktop설치</h2>
<p><a href="https://docs.docker.com/desktop/install/windows-install/">https://docs.docker.com/desktop/install/windows-install/</a></p>
<h2>2. Linux 시스템의 image를 받고 container run까지 한 명령어로 끝내기</h2>
<ul>
<li>Docker에 Centos, Ubuntu, Rocky Linux같은 시스템을 받은 다음, container를 생성하여 virtual machine을 사용하는 것과 비슷하게 환경을 구축하기로 결정하였다.
<ul>
<li>Docker는 프로세스를 격리하는 container 역할을 해주므로 엄밀하게 virtual machine이라고 말할 수는 없다.</li>
<li>다만 OS system으로 container를 처음부터 생성하고 이를 기반으로 개발환경을 구축해간다면 이와 비슷한 효과를 낼 수 있다고 생각된다.</li>
<li>재밌는 점은, WSL에 linux system을 설치하고 그 안에 docker를 실행하는 것도 가능하고, 심지어 docker 안 container에 있는 linux system에서 다시 docker를 중복으로 실행하는 것도 가능하다.</li>
</ul>
</li>
<li>어느 Linux 시스템을 선택할 것인가?
<ul>
<li>RHEL 계열의 Rocky Linux를 사용하기로 결정하였다.</li>
<li>Centos가 유료 stream version 외에는 유지보수를 종료한다는 점과, Centos 개발진이 오픈소스 무료 linux system으로 배포한다는 점이 좋았다.</li>
</ul>
</li>
<li>Docker에 설치는...
<ol>
<li>Powershell에서 명령어를 사용하는 방법, 혹은</li>
<li>Windows에서 UI환경의 Docker desktop을 활용하는 방법이 있다.</li>
</ol>
<ul>
<li>다만, linux system을 설치하는 경우 interactive 옵션(-i)과 TTY 옵션(-t)을 켜지 않으면 command terminal의 충돌로 container가 실행되지 않는다.</li>
<li>위 옵션을 Docker desktop에서 키는 방법을 알 수 없었다.</li>
<li>인터넷에서 조사를 했을 때 대부분의 경우 아직 1) 처럼 shell 명령어로 설치가 이루어지고 있었다.</li>
</ul>
</li>
</ul>
<h3>Powershell 명령어를 사용하여 설치하기</h3>
<pre><code class="language-shell">docker run --name RockyLinux9_test --hostname ROCKY -it rockylinux:9
</code></pre>
<h2>3. Container를 Image로 변환하기.</h2>
<ul>
<li>만들어진 container 내에서 여러가지 프로그램 설치 및 작업을 수행하였다. 해당 container를 새로운 image로 제작하여, Google Cloud에 올릴 수 있게 준비한다.</li>
<li></li>
</ul>
<h2>4. Google Cloud와 VScode 연결하기</h2>
<p><a href="https://cloud.google.com/code/docs/vscode/install?hl=ko">VS Code용 Cloud Code 확장 프로그램 설치</a></p>
<ul>
<li>먼저 VS Code 확장프로그램 중에서 Google Cloud Code를 설치</li>
<li>VS Code를 재시작하면 Google Cloud Code 메뉴가 좌측에 생긴다
<ul>
<li>Cloud Run 항목을 누르면 Google Cloud와 연결할 수 있게 웹페이지를 띄운다</li>
<li>구글로 로그인한다</li>
</ul>
</li>
</ul>
<h1>번외: Docker에 Rocky Linux image 받아오기</h1>
<ul>
<li><code>docker run</code> 명령어의 경우 image가 없으면 자동으로 다운받는 과정이 포함되어 있다.</li>
<li>만약 container를 생성하지 않고 image만 받고 싶은 경우, 다음과 같이 가능하다.</li>
<li>Docker hub에서 자신이 원하는 linux system을 검색한 뒤, image를 받는 것이 가능하다. <a href="https://hub.docker.com/_/rockylinux">#</a></li>
<li>Docker desktop 가장 상단의 검색창에서 받고 싶은 image를 검색한 다음, <code>Pull</code>을 눌러서 다운받 것도 가능하다.</li>
<li>어떤 이름(태그)으로 다운받아야 하는지 확인한 다음, Powershell을 열어서 다음의 명령어를 수행한다</li>
</ul>
<pre><code class="language-shell">docker pull rockylinux:9
</code></pre>
<ul>
<li>Docker window 프로그램에서 Images 메뉴를 보면 Rocky Linux가 추가된 것을 알 수 있다.
<ul>
<li>혹은 Powershell에서 <code>docker images -a</code> 명령어로 확인할 수 있다.</li>
</ul>
<pre><code class="language-shell"></code></pre>
</li>
</ul>
<p>PS C:\WINDOWS\system32&gt; docker images -a<br>
REPOSITORY                 TAG       IMAGE ID       CREATED       SIZE<br>
rockylinux                 9         eeea865f4111   7 weeks ago   176MB<br>
```</p>
<h3>참고로...</h3>
<blockquote>
<p><em>&quot;The <code>rockylinux:latest</code> tag is intentionally missing. Please choose a major version (currently 8 or 9) tag, or a more specific tag to ensure you are pulling the version of Rocky Linux you want: e.g. <code>rockylinux:8</code> or <code>rockylinux:9</code>&quot;</em></p>
</blockquote>
<ul>
<li>
<p>보통은 <code>docker pull rockylinux</code>라는 명령어로 쉽게 최신 버전을 받을 수 있지만, 2023년 7월 기준으로 불가능하다. 따라서 뒤에 태그로 <code>:9</code>를 적어줘야 한다.</p>
</li>
<li>
<p>Rocky Linux가 두 버전이 검색되서 혼란이 가중되었다.</p>
<ul>
<li>Docker Official Image로 배포되는 것  <a href="https://hub.docker.com/_/rockylinux">#</a></li>
<li>Sponser OSS로 배포되는 것 <a href="https://hub.docker.com/r/rockylinux/rockylinux">#</a></li>
</ul>
</li>
<li>
<p>두 배포처의 주소가 조금 다르다.</p>
<ul>
<li>Official image의 경우 다운받기 위해서는 rockylinux 뒤에 tag를 적어주지 않으면 latest 버전이 없다는 에러가 뜨면서 다운이 되지 않는다. 단순하게 9를 tag로 사용할 경우 최신버전인 9.2로 다운되었다. (`rockylinux:9)</li>
<li>Sponser OSS에서도 9.2버전을 받을 수 있지만 따로 tag를 해줘야 한다(<code>rockylinux/rockylinux:9</code>)</li>
<li>태그를 하지 않을 경우 Sponser OSS에서는 8.6버전을 받았다(<code>rockylinux/rockylinux</code>)</li>
</ul>
</li>
</ul>
<h1>번외: WSL2에 Rocky Linux 설치</h1>
<ul>
<li>Rocky linux를 windows11에서 사용하기 위해서는
<ol>
<li>WSL2(Windows Subsystem for Linux) 설치</li>
<li>Rocky linux 사이트에서 최신버전 tar.xz 파일 다운로드</li>
<li>Powershell에서 WSL2를 통해 os import 수행</li>
</ol>
</li>
</ul>
<pre><code class="language-shell">wsl --import RockyLinux9 &lt;Rocky Linux 설치할 장소. 나의 경우 C:\Users\사용자\rockylinux&gt; &lt;다운받은 tar.xz 파일 위치&gt; --version 2
</code></pre>
<ul>
<li>이제 rocky linux를 powershell이나 visual studio code에서 실행한 다음 <code>docker</code>명령어를 실행하면 옵션이 떠야한다.</li>
</ul>
<pre><code class="language-shell">[root@SurfacePro8 system32]# docker

Usage:  docker [OPTIONS] COMMAND

A self-sufficient runtime for containers

Common Commands:
  run         Create and run a new container from an image
  exec        Execute a command in a running container
  ps          List containers
  ...
</code></pre>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120347" class="record">
            <header>
                <h1 class="record-title">Central Dogma</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120347</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="#20230924120353" title="DNA" class="record-link">DNA</a><br>
<a href="#20230924120388" title="RNA" class="record-link">RNA</a><br>
<a href="#20230924120384" title="Protein" class="record-link">Protein</a></p>


                
            </main>

            <footer class="list-content-container">
                
                <details class="record-links" open="open">
                    <summary class="menu-title">
                        links
                        <span class="badge">3</span>
                    </summary>

                    <ul class="record-links-list">
                        
                        <li data-target-id="20230924120353">
                            <a href="#20230924120353">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>DNA</span>
                                <div class="record-links-context">
                                    <p><em class="id-context" data-target-id="20230924120353">DNA</em><br>
RNA<br>
Protein</p>

                                </div>
                            </a>
                        </li>
                        
                        <li data-target-id="20230924120388">
                            <a href="#20230924120388">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>RNA</span>
                                <div class="record-links-context">
                                    <p>DNA<br>
<em class="id-context" data-target-id="20230924120388">RNA</em><br>
Protein</p>

                                </div>
                            </a>
                        </li>
                        
                        <li data-target-id="20230924120384">
                            <a href="#20230924120384">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Protein</span>
                                <div class="record-links-context">
                                    <p>DNA<br>
RNA<br>
<em class="id-context" data-target-id="20230924120384">Protein</em></p>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                

                
            </footer>
        </article>
        
        <article id="20230924120348" class="record">
            <header>
                <h1 class="record-title">Curse of The Dimensionality</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120348</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <h2>How to solve?</h2>
<p><a href="#20230924120364" title="Feature Selection" class="record-link">Feature Selection</a><br>
<a href="#20230924120363" title="Feature Extraction" class="record-link">Feature Extraction</a></p>


                
            </main>

            <footer class="list-content-container">
                
                <details class="record-links" open="open">
                    <summary class="menu-title">
                        links
                        <span class="badge">2</span>
                    </summary>

                    <ul class="record-links-list">
                        
                        <li data-target-id="20230924120364">
                            <a href="#20230924120364">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Feature Selection</span>
                                <div class="record-links-context">
                                    <h2>How to solve?</h2>
<p><em class="id-context" data-target-id="20230924120364">Feature Selection</em><br>
Feature Extraction</p>

                                </div>
                            </a>
                        </li>
                        
                        <li data-target-id="20230924120363">
                            <a href="#20230924120363">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Feature Extraction</span>
                                <div class="record-links-context">
                                    <h2>How to solve?</h2>
<p>Feature Selection<br>
<em class="id-context" data-target-id="20230924120363">Feature Extraction</em></p>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                

                
                <details class="record-backlinks" open="open">
                    <summary class="menu-title">
                        backlinks
                        <span class="badge">1</span>
                    </summary>

                    <ul class="record-backlinks-list">
                        
                        <li data-target-id="20230924120366">
                            <a href="#20230924120366">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Integration Strategies of Multi-omics Data for Machine Learning Analysis</span>
                                <div class="record-links-context">
                                    <h1>1. introduction</h1>
<h3>왜 multi-omics 연구가 필요할까?</h3>
<ul>
<li>맞춤의학, 정밀의료의 부상
<ul>
<li>개인의 임상데이터를 통해 진단하고 적절한 치료법을 제시</li>
<li>임상데이터로 genomics, epigenomics, trasncriptiomics, proteomics, metabolomics 같은 데이터를 수집함</li>
<li>과거에는 이러한 각각의 omics 데이터를 따로 관찰해서 진단에 활용하려고 했음</li>
<li>그러나 대부분의 질병은 omics 데이터들을 서로 관통하면서 연결되어있음</li>
<li>따라서 여러 omics 데이터를 합쳐서 통합적인 관점으로 보는 multi-omics 연구가 필요함</li>
</ul>
</li>
</ul>
<h3>Multi-omics 연구의 장단점은?</h3>
<ul>
<li>한 omics 데이터에서 부족한 부분을 다른 omics 데이터를 참고하여 메꾼다던가, 하나의 큰 시스템에 대한 그림을 얻어낸다던가 하는 장점이 있다.</li>
<li>그러나 적은 샘플에서 가져온 데이터만 있다는 점이나, 개개의 생물학 데이터가 갖고있는 오류를 넘어서서 데이터를 합쳐야 한다는 어려움들이 단점으로 남는다.
<ul>
<li>특히 multiomics 연구에서 데이터를 합치는(integration) 작업이 정확하게 이루어지지 않으면, 오류 투성이로 합쳐진 데이터집합에서 결과를 제대로 얻는 것은 불가능하다.</li>
</ul>
</li>
</ul>
<h3>Data integration</h3>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6242341/">Zitnik et al. (2019)</a>은 integration의 방법을 크게 두 가지로 분류하였다.
<ul>
<li>Horizontal inegration: 동일 omics 여러 샘플</li>
<li>Vertical integration: 여러 omics 동일 샘플</li>
</ul>
</li>
<li>이번 리뷰에서는 Vertical integration을 위주로 리뷰할 것이다.<br>
- Vertical integration의 방법을 수학적인 관점에서 bayesian, network-based, deep learning-based, kernel-based, matrix factorization-based로 나눠서 볼 것이다.
<ul>
<li>또한 <a href="https://pubmed.ncbi.nlm.nih.gov/25582081/">Ritchie et al.(2015)</a>의 분류방법에 따라 vertical integration(논문에서는 meta-dimensinal)을 concatenation-based, transformation-based, model-based로 나눠서 볼 것이다.</li>
</ul>
</li>
</ul>
<h1>2. Challenges</h1>
<ul>
<li>생물학 데이터는 필연적으로 오류를 동반하고 있는데, 이렇게 깨끗하지 못한 여러 omics 데이터를 종합하여 특정한 패턴을 관찰한다는 것은 쉽지 않다.</li>
<li>또한 multi-omics 연구의 특성상 적은 샘플에서 많은 feature를 관찰해야 하는데, 이는 machine learing에서 가장 이상적이지 못한 상황으로 차원의 저주(<em class="id-context" data-target-id="20230924120348">Curse of The Dimensionality</em>)라고 불리운다.
<ul>
<li>이런 상황에서 ML을 통해 데이터를 훈련시키면 overfit된다.</li>
</ul>
</li>
<li>생물학 omics 데이터들이 각각 너무 성격이 다르다는 점도 하나의 문제이다.
<ul>
<li>유전자 발현 데이터의 경우 수만개의 variable이 있을 수 있지만, 대사물질 데이터는 기껏해서 몇천일 수가 있다.</li>
<li>이렇게 서로 다른 데이터들 사이의 수, 종류, 연속성, 등의 차이를 극복하고 데이터들을 잘 integration 해야한다는 어려움이 있다.</li>
</ul>
</li>
<li>데이터 간의 class imbalance 또한 ML 훈려에 악영향을 미친다.
<ul>
<li>데이터에서 특정 class의 데이터만 기하급수적으로 많다면, 이를 바탕으로 훈련시키는 것이 부적절하다.</li>
<li>왜냐하면 ML 훈련은 모든 class의 데이터 량이 엇비슷하다고 가정하고 훈련되기 때문이다.</li>
<li>이를 해결하기 위한 방법으로 sampling, cost-sensitive learning과 같은 방법이 존재한다.</li>
</ul>
</li>
<li>Missing data 문제의 경우 해당 샘플을 지워버린다던가 혹은 Random Forest, K-Nearest Neighbor같은 방법이 있다.</li>
</ul>
<h1>3. Main integration strategies</h1>
<h3>Integration 5가지 소개</h3>
<ul>
<li><strong>Early integration</strong>이란, 데이터들의 복잡도를 낮추지 않고, 단순히 둘 이상의 데이터들을 합친 뒤(wise concatenation) 그 결과물인 matrix를 ML모델에 넣는 것이다.
<ul>
<li>현실적으로는 단순하게 합쳐진 데이터는 여전히 복잡하기 때문에 ML모델이 잘 돌아가지 않는다다</li>
</ul>
</li>
<li><strong>Mixed integration</strong>이란 데이터들의 복잡도를 개별적으로 낮춰서 사용하는 것이다.</li>
<li><strong>Intermediate integration</strong>이란 여러 데이터들의 복잡도를 함께 낮춰서 사용하는 것이다.</li>
<li><strong>Late integration</strong>이란 정반대로 데이터를 합치지 않고 각각의 omics 데이터를 개별적으로 분석하는 것이다.</li>
<li><strong>Hierarchical integration</strong>이란 DNA-RNA-단백질로 이루어지는 시스템을 표현한 central dogma에 입각해서 omics 데이터 간의 조절 관계를 반영하여 합치는 것이다.</li>
</ul>
<h2>3.1 Dimensionality reduction for multi-omic integration</h2>
<p><strong>Dimensionality reduction은 왜 필요할까?</strong></p>
<ul>
<li>Multi-omics 연구에 있어서 샘플들이 너무 많은 variable을 가질 경우 machine learning 기술 적용에 차질을 빚는다. 따라서 variable의 수를 줄여서 차원의 축소와 데이터셋의 noise 감소를 유도한다.</li>
<li>차원축소는 필수적이지는 않고, 위에서 얘기했던 5가지의 integration 방법에 모두 적용이 가능하다.
<ul>
<li>다만 early integration와 intermediate integration의 경우 차원축소를 미리 적용하는 것이 정확도 향상에 기여한다.</li>
</ul>
</li>
<li>차원축소를 위한 방법은 크게 두 가지가 있다:
<ul>
<li>Feature selection</li>
<li>Redundant varaibles and feature extraction</li>
</ul>
</li>
</ul>
<h3>3.1.1. Feature Selection (FS)</h3>
<ul>
<li>Feature를 줄이는 작업</li>
<li>그 결과 데이터의 복잡도, noise가 줄어들어서 데이터가 깔끔해진다.</li>
<li>데이터가 깔끔해지니까, ML 모델이 overfitting될 위험도 적어진다.</li>
<li>데이터가 깔끔해져서 해석하기도 편해진다.</li>
<li>너무 많은 variable이 삭제될 경우?
<ul>
<li>데이터셋 사이에 feautre의 수가 다르면 좋지 않다</li>
<li>Feature Selection은 각 omics block이 동일한 수의 feature를 가지도록 할 수 있다</li>
</ul>
</li>
<li>Feature Selection은 세 가지로 나뉜다:
<ul>
<li>Filter-based Feature Selection</li>
<li>Wrapper-based Feature Selection</li>
<li>Embedded Feature Selection</li>
</ul>
</li>
</ul>
<h3>Filter-based Feature Selection</h3>
<ul>
<li>ML 모델과 상관없이 적용 가능</li>
<li>Feature 수를 줄일 수 있는 방향으로 관련된 variable을 추려내는 통계학적인 분석방법임</li>
<li><strong>CFS, RCA</strong> 같이 correlation 방법을 사용하거나,</li>
<li><strong>ReliefF</strong> 같이 distance 방법을 사용하거나,</li>
<li><strong>mRMR</strong> 같이 information gain 방법을 사용함</li>
</ul>
<h3>Wrapper-based Feature Selection</h3>
<ul>
<li>ML 모델을 활용하여 서로 다른 feature 그룹을 학습시키고 평가한다.</li>
<li>가장 성능이 좋았던 feature 그룹만 선택한다.</li>
<li>Recursive Feature Selection이 한 가지 예시이다.
<ul>
<li>먼저 전체 feauture를 대상으로 모델을 학습시킨다.</li>
<li>그 다음 모델 퍼포먼스에 기여도가 떨어지는 feature를 차례대로 제거해간다.</li>
</ul>
</li>
<li>이 방법은 어느 ML 모델에서든 활용 가능하다는 장점이 있다.</li>
<li>그러나 반복적으로 ML 모델을 훈련시켜야 하기 때문에, 데이터셋이 크다면 컴퓨터 성능에 따라 시간이 오래 걸린다는 단점이 있다.</li>
</ul>
<h3>Embedded Feature Selection</h3>
<ul>
<li>Classifier 모델 안에 feature selection 알고리즘이 넣어져있는 경우</li>
<li>Tree-based feature selection과 regularization 방법이 있다.
<ul>
<li>Regularization 방법은 loss function과 penalization funtion을 결합한 방법이다.
<ul>
<li>Loss funtion은 모델이 주어진 데이터에 얼마나 잘 맞는지 확인하는 함수다.</li>
<li>Penalization funtion은 복잡도(complexity)에 대하여 페널티를 부가하는 방식으로 feature 수를 적은 쪽으로 유도한다.</li>
</ul>
</li>
</ul>
</li>
<li>Embedded Feature Selection 모델은 Wrapper-based Feature Selection과 마찬가지로 ML모델과 함께 사용되지만, 컴퓨터 자원을 덜 소비하기 때문에 컴퓨터 성능에 덜 의존적이라는 장점이 있다.</li>
</ul>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                
            </footer>
        </article>
        
        <article id="20230924120349" class="record">
            <header>
                <h1 class="record-title">Dark microbiome and extremely low organics in Atacama fossil delta unveil Mars life detection limits</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120349</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://www.nature.com/articles/s41467-023-36172-1">Azua-Bustos, A., Fairén, A.G., González-Silva, C. <em>et al.</em> Dark microbiome and extremely low organics in Atacama fossil delta unveil Mars life detection limits. <em>Nat Commun</em> <strong>14</strong>, 808 (2023). https://doi.org/10.1038/s41467-023-36172-1</a></p>
<h1>Abstract</h1>
<ul>
<li>칠레의 Atacama Desert의 red stone(암석)에서 채취한 미생물군집 샘플에서 DNA 흔적을 조사했는데, 발견한 DNA의 9%는 과학계에 보고된 적이 없어 분류 불가능한 미생물의 존재였음.</li>
<li>과학자들은 이렇게 sequencing 기술을 통하여 존재를 확인할 수 있지만, 정확하게 분류할 수 없는 미생물을 &quot;dark microbiome&quot;이라고 정의함.</li>
<li>해당 연구는 지구의 매우 건조한 환경에서도 미생물의 흔적도 많이 손상되 보존되기 어려움을 시사함.</li>
<li>따라서 화성에서 암석샘플 회수에 성공한다고 할지라도, 현재 기술로는 과거에 화성에 존재했던 미생물의 흔적을 놓칠 수 있다는 점을 시사함.</li>
</ul>
<h3>Site microbial diversity and dark microbione</h3>
<ul>
<li>암석샘플은 1 그램당 1μg의 DNA를 갖고 있었음.</li>
<li>Operational Taxonomical Units(<a href="#20230924120381" title="OTUs" class="record-link">OTUs</a>)이 제일 다양하게 검출된 부분은 바로 풍화된 상태의 역암(conglomerate)이었다.</li>
<li>가장 많이 검출된 종은 <a href="#20230924120342" title="Alphaproteobacteria" class="record-link">Alphaproteobacteria</a> 와 <a href="#20230924120339" title="Actinobacteria" class="record-link">Actinobacteria</a>였다.<br>
<img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41467-023-36172-1/MediaObjects/41467_2023_36172_Fig4_HTML.png?as=webp"></li>
</ul>
<p>#Sarah_Stewart_Johnson #Dark_Microbiome #Armando_AzuaBustos</p>


                
            </main>

            <footer class="list-content-container">
                
                <details class="record-links" open="open">
                    <summary class="menu-title">
                        links
                        <span class="badge">3</span>
                    </summary>

                    <ul class="record-links-list">
                        
                        <li data-target-id="20230924120381">
                            <a href="#20230924120381">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>OTUs</span>
                                <div class="record-links-context">
                                    <h1>Abstract</h1>
<ul>
<li>칠레의 Atacama Desert의 red stone(암석)에서 채취한 미생물군집 샘플에서 DNA 흔적을 조사했는데, 발견한 DNA의 9%는 과학계에 보고된 적이 없어 분류 불가능한 미생물의 존재였음.</li>
<li>과학자들은 이렇게 sequencing 기술을 통하여 존재를 확인할 수 있지만, 정확하게 분류할 수 없는 미생물을 &quot;dark microbiome&quot;이라고 정의함.</li>
<li>해당 연구는 지구의 매우 건조한 환경에서도 미생물의 흔적도 많이 손상되 보존되기 어려움을 시사함.</li>
<li>따라서 화성에서 암석샘플 회수에 성공한다고 할지라도, 현재 기술로는 과거에 화성에 존재했던 미생물의 흔적을 놓칠 수 있다는 점을 시사함.</li>
</ul>
<h3>Site microbial diversity and dark microbione</h3>
<ul>
<li>암석샘플은 1 그램당 1μg의 DNA를 갖고 있었음.</li>
<li>Operational Taxonomical Units(<em class="id-context" data-target-id="20230924120381">OTUs</em>)이 제일 다양하게 검출된 부분은 바로 풍화된 상태의 역암(conglomerate)이었다.</li>
<li>가장 많이 검출된 종은 Alphaproteobacteria 와 Actinobacteria였다.<br>
<img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41467-023-36172-1/MediaObjects/41467_2023_36172_Fig4_HTML.png?as=webp"></li>
</ul>

                                </div>
                            </a>
                        </li>
                        
                        <li data-target-id="20230924120342">
                            <a href="#20230924120342">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Alphaproteobacteria</span>
                                <div class="record-links-context">
                                    <h1>Abstract</h1>
<ul>
<li>칠레의 Atacama Desert의 red stone(암석)에서 채취한 미생물군집 샘플에서 DNA 흔적을 조사했는데, 발견한 DNA의 9%는 과학계에 보고된 적이 없어 분류 불가능한 미생물의 존재였음.</li>
<li>과학자들은 이렇게 sequencing 기술을 통하여 존재를 확인할 수 있지만, 정확하게 분류할 수 없는 미생물을 &quot;dark microbiome&quot;이라고 정의함.</li>
<li>해당 연구는 지구의 매우 건조한 환경에서도 미생물의 흔적도 많이 손상되 보존되기 어려움을 시사함.</li>
<li>따라서 화성에서 암석샘플 회수에 성공한다고 할지라도, 현재 기술로는 과거에 화성에 존재했던 미생물의 흔적을 놓칠 수 있다는 점을 시사함.</li>
</ul>
<h3>Site microbial diversity and dark microbione</h3>
<ul>
<li>암석샘플은 1 그램당 1μg의 DNA를 갖고 있었음.</li>
<li>Operational Taxonomical Units(OTUs)이 제일 다양하게 검출된 부분은 바로 풍화된 상태의 역암(conglomerate)이었다.</li>
<li>가장 많이 검출된 종은 <em class="id-context" data-target-id="20230924120342">Alphaproteobacteria</em> 와 Actinobacteria였다.<br>
<img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41467-023-36172-1/MediaObjects/41467_2023_36172_Fig4_HTML.png?as=webp"></li>
</ul>

                                </div>
                            </a>
                        </li>
                        
                        <li data-target-id="20230924120339">
                            <a href="#20230924120339">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Actinobacteria</span>
                                <div class="record-links-context">
                                    <h1>Abstract</h1>
<ul>
<li>칠레의 Atacama Desert의 red stone(암석)에서 채취한 미생물군집 샘플에서 DNA 흔적을 조사했는데, 발견한 DNA의 9%는 과학계에 보고된 적이 없어 분류 불가능한 미생물의 존재였음.</li>
<li>과학자들은 이렇게 sequencing 기술을 통하여 존재를 확인할 수 있지만, 정확하게 분류할 수 없는 미생물을 &quot;dark microbiome&quot;이라고 정의함.</li>
<li>해당 연구는 지구의 매우 건조한 환경에서도 미생물의 흔적도 많이 손상되 보존되기 어려움을 시사함.</li>
<li>따라서 화성에서 암석샘플 회수에 성공한다고 할지라도, 현재 기술로는 과거에 화성에 존재했던 미생물의 흔적을 놓칠 수 있다는 점을 시사함.</li>
</ul>
<h3>Site microbial diversity and dark microbione</h3>
<ul>
<li>암석샘플은 1 그램당 1μg의 DNA를 갖고 있었음.</li>
<li>Operational Taxonomical Units(OTUs)이 제일 다양하게 검출된 부분은 바로 풍화된 상태의 역암(conglomerate)이었다.</li>
<li>가장 많이 검출된 종은 Alphaproteobacteria 와 <em class="id-context" data-target-id="20230924120339">Actinobacteria</em>였다.<br>
<img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41467-023-36172-1/MediaObjects/41467_2023_36172_Fig4_HTML.png?as=webp"></li>
</ul>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                

                
            </footer>
        </article>
        
        <article id="20230924120350" class="record">
            <header>
                <h1 class="record-title">Data Integration</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120350</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                

                
            </main>

            <footer class="list-content-container">
                

                
                <details class="record-backlinks" open="open">
                    <summary class="menu-title">
                        backlinks
                        <span class="badge">1</span>
                    </summary>

                    <ul class="record-backlinks-list">
                        
                        <li data-target-id="20230924120366">
                            <a href="#20230924120366">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Integration Strategies of Multi-omics Data for Machine Learning Analysis</span>
                                <div class="record-links-context">
                                    <p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8258788/">Picard M, Scott-Boyer MP, Bodein A, Périn O, Droit A. Integration strategies of multi-omics data for machine learning analysis. Comput Struct Biotechnol J. 2021 Jun 22;19:3735-3746. doi: 10.1016/j.csbj.2021.06.030. PMID: 34285775; PMCID: PMC8258788.</a></p>
<ul>
<li>2021년 기준으로 Multi-omics <em class="id-context" data-target-id="20230924120350">Data Integration</em> 방법을 5가지로 분류함
<ul>
<li>early</li>
<li>mixed</li>
<li>intermediate</li>
<li>late</li>
<li>hierarchical<br>
<img src="https://ars.els-cdn.com/content/image/1-s2.0-S2001037021002683-ga1_lrg.jpg" width=800></li>
</ul>
</li>
</ul>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                
            </footer>
        </article>
        
        <article id="20230924120351" class="record">
            <header>
                <h1 class="record-title">Dealing with Dimensionality - The Application of Machine Learning to Multi-omics Data</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120351</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://academic.oup.com/bioinformatics/article/39/2/btad021/6986971">https://academic.oup.com/bioinformatics/article/39/2/btad021/6986971</a></p>
<ul>
<li>Review 논문</li>
<li>Omcis data에 maching learning(ML) 기술이 필요한 이유는?
<ul>
<li>Data set의 크기가 크다</li>
<li>이 큰 data set을 전부 인간이 일일이 정리하고 통합적인 관점에서 분석하는 것은 무리다</li>
</ul>
</li>
<li>현재(2023년 초)까지 생물학에서 ML을 이용하여 여러 논문이 발표되었다.
<ul>
<li>Cancer Cell LIne Encyclopedia</li>
<li>Microbiome research</li>
</ul>
</li>
<li>ML이 이상적으로 작동하기 위해서는 대량에 샘플에서 적은 수의 feature들을 뽑아내 모델을 훈련시켜야 한다.
<ul>
<li>Omic data의 문제는 한 샘플에서 feature가 너무 많다는 것이다.
<ul>
<li>예를 들어 종양샘플을 RNAseq한 데이터를 사용하려고 할 때, 20,000개에 달하는 유전자를 모두 feature로 잡아버리게 된. 단 하나의 샘플인데!</li>
</ul>
</li>
<li>물론 이렇게 여러 feature들을 한꺼번에 봐서 통합적인 관점에서 보는 것이 omic 분석의 핵심이다.</li>
<li>하지만 여전히 살펴봐야하는 feature들이 너무 많은데 샘플 수는 이에 미치지 못한다는 점, 그렇게 때문에 효과적인 ML 훈련 및 모델 생성이 어렵다는 점이 난관이다.</li>
</ul>
</li>
<li>해당 리뷰는 scoping review의 방식으로 리뷰를 진행했다.
<ul>
<li>따라서 다른 연구자들의 방식이 더 효과적이었는지, 가장 좋은 방법이 무엇인지 비교하지 않음</li>
<li>다른 연구자들의 방식들을 나열하기만 함</li>
</ul>
</li>
</ul>
<h1>1. What is multi-omic data in practice?</h1>
<h2>1.1. Which types of -omics features made up the ‘data dimensions’?</h2>
<img src="https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bioinformatics/39/2/10.1093_bioinformatics_btad021/1/btad021f2.jpeg?Expires=1696441666&Signature=HTyf33ZsKx-RGcUitMwyt4gakbJa8~Ep6MUmpPWF4R2mExn-WwNEaB8-caE0IgTJVx0LxC1TF7qNnc98qsYSDULj~8-W56jQ-xzsR0X2hjlQJdLq7iBGSfE8HVLKzKBRxOaSD5LshIll0HqK4wEJjbZRBrLTbOrbpUfs9GPjmKT64Px8qA-yAEpR5tnm7lJbWa~p~sx3lZ~MH99dTqnD0MvqR1GlJM7KIQ-o3cfW-B5QYFR5ax~ZtyakaLOltkslnQt7e~h53gF-EW6MEssmjhCeBxDOFB40GPgHnhPxmgHvcbuEcq6JD3-ut6STfCKuEwm~uFOlK2zptSxAAgW01w__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA" width="800">    
<p><em>(a) Number of uses of each -omics category in the reviewed papers.</em><br>
<em>(b) Number of -omics used per paper in the reviewed papers.</em><br>
<em>(c) The number of appearances -omics pairs across papers.</em></p>
<ul>
<li>사람들은 Omics 분석인 만큼 주로 4개 정도의 서로 다른 데이터들을 함께 분석했다.</li>
<li>Transcriptomics를 포함하는 경우가 가장 많았고 그 뒤로 epigenomics, genomics 데이터를 쓰는 경우가 따랐다.</li>
<li>TCGA database의 등장으로 이런 경향이 두드러진다고 생각했으나, TCGA DB를 참조하지 않는 논문들도 비슷한 조합의 경향을 따랐다.</li>
</ul>
<h2>1.2. How is the data structured: how many features versus samples?</h2>
<ul>
<li>위에서 말했듯이 omic data의 경우 샘플마다 살펴봐야하는 특성의 개수가 샘플의 수 보다 훨씬 많다.</li>
<li>각 논문들이 omics 분석엣 사용하는 특성의 수가 샘플의 수 보다 7~8배 정도 많았다.
<ul>
<li>중간값으로 따졌을 때 특성의 수가 33,415이고 샘플의 수가 447</li>
<li>평균값으로 따졌을 때 특성의 수가 73,996이 샘플의 수가 1767</li>
</ul>
</li>
<li>TCGA DB의 등장, omics 분석의 어려움의 특성 상 온라인 토론이 활발하다는 점에서 자연스럽게 데이터 공유가 많아졌다.
<ul>
<li>이를 이점으로 삼아 다양한 ML 기술들이 테스트 되는 환경이 구축될 수 있었다.</li>
</ul>
</li>
</ul>
<h1>2. What analysis has been done on multi-omics data?</h1>
<h2>2.1. Which ML techniques were used?</h2>
<ul>
<li>Data integration은 비단 omics data 뿐만 아니라 여러 상황에서 데이터들을 합쳐야 할 때 필요하다.</li>
<li>Data integration을 위해서 machine learning이 사용되기도 하는데, bioinformatics와 관련없는 그룹들도 data integration을 수행한다.</li>
<li>그렇다면 일반적인 데이터 사이언스 연구자 그룹과 multiomics data를 연구하는 그룹 사이에서 data integration을 목적으로 사용하는 machine learning 기술에 차이가 있을까?<br>
<img src="https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bioinformatics/39/2/10.1093_bioinformatics_btad021/1/btad021f4.jpeg?Expires=1696441617&Signature=zk231bVQ1rG39cc1An3JJlaNOm0iJgRd-0cYURpp7dfFNwRN4UrGoXMeSMkHDC4qBK8urijWQDE1MO1Q32SVs3QXL38QPqRqQybMQCmx6-WpCaeKjqqUxgdq6VMHYPvaFUHgBFNqU2ji21T3OkGYn8k8JFY-Xy0hHWrhUM35TdcAcDyIwwj7z3RM0iFTQyj8tmTC3gbRzn2ZfnMI07r5Dx-WaUx7tcznIuMH-BgsgYqnS~jymyuqu0l3RLyXEVz4AEBN~Sdnnp5k3asOiVAaQGkBihn5s75S6UuFgGar6B2rdz1OqYnrCP~0OeGSeR8D0BIrSGQB6wOUVg~OyEXBUA__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA" width="800"></li>
</ul>
<p><em>(a) Number of ML techniques being used more than once in the reviewed papers.</em><br>
<em>The publications on ‘machine learning AND multi-omics AND integration’ are plotted in green, while the publications on ‘machine learning AND integration’ are plotted in purple.</em><br>
<em>Number in the reviewed multi-omics ML papers of ML goals (b) and labels used for classification (c)</em></p>
<ul>
<li>Multiomics 그룹의 경우 Autoencoder, Cox PH 기술을 많이 이용했다.</li>
<li>반면에 일반 ML 그룹은 Random forest와 SVM 기술 사용에서 큰 차이를 보였다.</li>
</ul>
<h2>2.2. What were the goals of the ML application?</h2>
<ul>
<li>Multiomics 그룹이 ML을 이용한 주 목적은 첫번째는 classification, 두번째는 dimensionality reduction이었다.</li>
<li>Classfication을 통하여 질병을 subtype별로 분리하기, 환자의 생존 예측을 라벨(5년이상, 5년이하) 별로 분리하는 경우에 쓰였다.
<ul>
<li>수명 문제에 관해서 얘기하자면, 정확하게 몇 년 남았다는 식으로 분류하는 경우 이는 회귀문제가 된다. 그러나 이 회귀문제를 기준을 정하고 구간을 나눠서 '이산화'(discretize)하여서, 몇가지 편의를 도모할 수 있다.</li>
</ul>
</li>
</ul>
<h2>2.3 What were the targets/labels of classification tasks?</h2>
<ul>
<li>Classification의 경우 환자의 생존 예측이 제일 많았다.</li>
</ul>
<h1>3. Can we explain trends using an analysis of the citation of papers since 2015?</h1>
<p><img src="https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bioinformatics/39/2/10.1093_bioinformatics_btad021/1/m_btad021f5.jpeg?Expires=1695325471&amp;Signature=Q8ufkAm9rlkrhHe8ZIXi9qTNWHJWP93HEEQU-n0enRc~jJ624LLlODN6P3jkuU2YT4yoou~-migr~noHI9hFtAkXYORhMdQ7feJFTTVxyXcmh5QMVQZPyG-mnTLYngXg-LvNdGueGBD73Vpuxh5356vl9DAlje12M6ZR-pOHyyoxRKNhcIYeSebvJKAX9Swpk9-VXMZvRRkuYjWzwwJB-Tb6x3k~aZG2S~NBPtCUYWSHByZ7OQPloeXS1Cz5u7yu~5UUnTNmSoTcjafCV9xKJ4dbFKpsAcBrxyNjbWrBZIckQo3TtHUg4uIF0OBFWVEs2DE1HzXsDEiVenPVIc17Gg__&amp;Key-Pair-Id=APKAIE5G5CRDK6RD3PGA" alt=""></p>
<p><em>Number of citations per year, of papers published in different years</em></p>
<ul>
<li>2018년 <a href="https://pubmed.ncbi.nlm.nih.gov/28982688/">Deep Learning-Based Multi-Omics Integration Robustly Predicts Survival in Liver Cancer</a> 논문이 발표된 이후, multiomics 연구에서 ML 기술을 이용한 논문이 급격하게 늘어났다.</li>
</ul>
<h1>4. Discussion</h1>
<img src="https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bioinformatics/39/2/10.1093_bioinformatics_btad021/1/btad021f6.jpeg?Expires=1696301496&Signature=4mKcdi8ADjARTc0jzXDH9FQ1VH8VF7URMK4ZzJXdzlAhK-~7oWZ-osXgX0tbdhS9QWEf3~UBhDlu7EN5P8XW56xhFot2-gESjIBqvcC14d8hgs8Fun~sn1S4GkK8Y-KbXevop20uPMG20kzE4Sf3UPtmyM4ZNRzhhKTNgH1wte2Lt8VP89NSfftjallHiwgyhUq79nrUpCrUsyVaKdWa0iIc0jnwPngSHA0rHUqNvTAM~NblAghzEMhOxA0Y6v4-s0vnZwrAIcS69yBDZjaVBQ--w1YXsORw72GDrK5PrHQ5VDOMfN6jtQI8KQnD3DtYzA-x-D42cod2V6d3bpbBEQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA" width="800">    
<p><em>Shapes in datasets. Features (</em>$P$<em>) Samples (</em>$n$<em>)</em><br>
<em>(a) A dataset where</em> $n$ ≫ $P$<em>, the ideal ‘shape’ for many ML techniques.</em><br>
<em>(b) In multi-omics analyses, researchers face very wide datasets, where</em> $n$ ≪ $P$.<br>
<em>(c) Feature selection and extraction are often used to reduce the number of features.</em><br>
<em>In feature selection, a subset of the original features is kept. In feature extraction, features are merged and transformed into a smaller number of new ones</em></p>
<ul>
<li>다시 말하지만 샘플($n$)의 수가 feature($P$)의 수보다 많은 상황이 ML에서 가장 이상적이다. 하지만 Multiomics data는 정반대의 성향을 지닌다.</li>
<li>확인해야 하는 feature($P$)의 수가 엄청 많은 경우를 '높은 차원을 가진다'라고 표현하는데, 이때 발생하는 문제는 바로 희소(sparsity)이다. (혹은 sparse data라고 표현할 수도 있겠다.)</li>
<li>이러한 sparse data를 가지고 실제로 ML 모델을 훈련시키려고 하면 잘 안된다는 단점이 있다. 각 샘플의 feature을 ML모델이 제대로 잡아내지 못한다. 이를 차원의 저주(curse of dimensionality)라고 부른다.</li>
<li>그렇기 때문에 ML 연구에서는 이러한 sprase data에 대하여 여러가지 전략을 시도하여 모델 훈련이 잘 되도록 유도한다.</li>
</ul>
<h2>4.1 Reducing the number of features (P)</h2>
<ul>
<li>Sparse data가 가진 차원의 저주 문제를 풀기 위한 시도 중 하나로는, feature($P$)의 수를 줄이려는 시도가 있다. (feature을 줄이는 것 만이 유일한 전략은 아니다.)</li>
<li>feature를 줄이는 방법에도 여러가지가 있는데...
<ul>
<li>feature들 중에서도 ML 훈련에 사용하기 좋은 경향성이 뚜렷한 feature들만 <strong>취사선택</strong>하는 방법이라던가,</li>
<li>알고리즘을 활용하여 여러 feature들을 <strong>압축</strong>하는 방법이라던가,</li>
<li>혹은 위 두 가지를 아예 한꺼번에 적용하는 방법이 있다.</li>
</ul>
</li>
</ul>
<h2>4.2 Feature selection</h2>
<ul>
<li>자신이 어떤 목적으로, 어떤 방향으로 ML 모델을 생성하고 싶은지 미리 알고 있고, 모델을 구성하는 요소에 대한 사전지식이 충분한 경우, 이러한 지식을 활용하여 모델을 최적화 할 수 있다.</li>
<li>Omics 분석에서 생물학적으로 봤을 때 비교적 확률이 낮은 feature($P$)를 의도적으로 제거하면 ML 모델의 최적화에 큰 이득을 줄 수 있다.
<ul>
<li>예를 들어, SNPs와 대사물질 7백만개로 이루어진 feature들을 65개의 feature로 줄인 <a href="http://ieeexplore.ieee.org/abstract/document/8416980/">논문</a>이 있다. 논문에서는 feature를 줄이기 위하여 다양한 기준을 사용했는데, 그 중에는 대사물질 농도와 특별하게 연관성이 높은 SNPs만 차출하는 기준도 있었다.</li>
<li>일반적으로는 유전자들 중에서도 활성빈도(activity level)이 낮은 유전자를 걸러내는 식으로 feature selection이 이루어진다.</li>
<li>후성유전체학 데이터의 경우에는 관련성 높은 유전자가 인접한 지역이나 여러 메틸화부위를 둘러싼 지역만 따로 고려하는 식으로 feature selection이 이루어진다.</li>
</ul>
</li>
<li>Omics 분석에서 Cox PH 모델(생존율 분석을 위한 모델)과 feature selection을 함께 사용하는 경우가 많이 보였다.
<ul>
<li>Cox PH 모델이란 선형회귀를 활용하여 특정 변수(약물, 유전자 등 수명에 영향을 끼칠 수 있는 것)에 의한 시간에 따른 생존율을 예측하는 모델이다.</li>
<li>Cox PH 모델은 일반적인 ML 연구에서는 인기가 없지만 multiomics 분석에서 전체에서 두 번째로 가장 많이 사용될 정도로 인기가 많았다.
<ul>
<li>이는 TCGA database에서 생존율 데이터에 접근이 가능하다는 점과 multiomics data 분석에 feature selection이 필수불가결해지면서 채택빈도가 늘어난 것으로 보인다.</li>
</ul>
</li>
<li>Feature selection을 통해 나온 결과들을 비교할 수 있기 때문에, 어느 변수(feature)들이 생존율에 더 영향력이 있는지 판가름하기 좋다.</li>
<li>게다가 일부 생존율 데이터는 몇 가지 이유로 불완전한데(예를 들어 환자가 아직까지 생존 중이라 변수와의 관계가 계속해서 갱신 중이라던가, 혹은 임상데이터가 접근 불가능하게 바뀌었다던가...), Cox PH 모델의 경우 불완전한 데이터에 대해서도 적용이 가능하므로, feature selection을 통해 몇 안되는 불완전한 데이터(샘플 사이즈가 작다) 마저도 활용할 수 있다.</li>
</ul>
</li>
</ul>
<h2>4.3 Feature extraction</h2>
<ul>
<li>Feature selection을 통해서 꽤 많은 feature($P$)들이 필터링됐을 것이라고 생각하지만, 대부분의 multiomics data의 경우 여전히 feature의 수가 sample에 비해 너무 많다.</li>
<li>이때 feature의 수를 더 줄이는 방법 중 하나가 feature extraction이다.
<ul>
<li>이는 기존의 feature들을 연구자가 설정한 다른 feature로 압축하는 것을 의미한다.</li>
<li>압축하는 과정에서 일부 feature들과 관련성이 떨어지는 것을 감안해야 한다.</li>
</ul>
</li>
<li>Feature extraction가 사용되는 예시는 여러가지가 있는데, Principal component analysis (PCA)와 Autoencoder가 있다.
<ul>
<li>PCA의 경우 기존의 feature들을 최대한 많이 정확하게 설명할 수 있으면서도 차원이 한 단계 낮춘 plot을 그린다.</li>
<li>Autoencoder는 neural networks다.
<ul>
<li>Autoencoder는 여러 layer로 이루어진 데이터들을 input으로 받는데, 이 중 가장 작은 bottleneck을 가진 layer에서 feature를 뽑아내서 기준으로 잡는다.</li>
<li>Autoencoder는 Cox PH와 마찬가지로 일반적인 ML연구 그룹보다는 multiomics연구 그룹에서 사용하는 빈도가 훨씬 높았다.</li>
</ul>
</li>
</ul>
</li>
</ul>
<h2>4.4 Increasing the number of samples (<em>n</em>)</h2>
<ul>
<li>
<p>위에서는 feature($P$)를 줄이는 방법에 대해서 고민했다.</p>
</li>
<li>
<p>그치만 샘플($n$)의 수를 늘리는 것 또한 omic 데이터가 가진 문제를 해결하는 전략 중 하나일 수 있다.</p>
</li>
<li>
<p>많은 omic 연구 그룹이 TCGA와 같은 public multiomics DB에서 얻은 데이터를 활용(혹은 의지)했다..</p>
<ul>
<li>TCGA는 10,000개가 넘는 샘플과 33개의 암종류에 대한 데이터를 보유하고 있다.(<a href="https://pubmed.ncbi.nlm.nih.gov/25691825/">2015</a>)</li>
<li>TCGA의 데이터는 다루기도 쉽고 연구자의 목적에 따라서 필요한 자료만 가져와서 홀용할 수도 있다.</li>
<li>다만 문제점은 TCGA는 암 데이터에 치중한 데이터들이라는 점이다.
<ul>
<li>암은 물질대사가 제대로 조절되지 못하거나 대사경로의 신호전달이 고장났을 때 발생한다.</li>
<li>이러한 환경에서 얻은 데이터를 암 외의 multiomics 분석에 적용하는 것은 문제가 있을 수 밖에 없다.</li>
</ul>
</li>
</ul>
</li>
<li>
<p>따라서 multiomics 분석에서 TCGA 데이터를 맹목적으로 활용하는 것은 지양해야할 것이다.</p>
<ul>
<li>다만 여전히 TCGA가 다른 DB들의 특징이나 경향성을 나타낼 수 있다는 점도 상기해야할 것이다.<br>
<img src="https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bioinformatics/39/2/10.1093_bioinformatics_btad021/1/m_btad021f3.jpeg?Expires=1695325471&Signature=aNd-OCRNyN3BEK5KCwrPh2Tw0ZyRCc0~hfaIBtpkspZ~yo2A11UqjWm5UFBFfvU3OHckdoQyvZ5nEVmj3OpByWMpT7vt3jL6DeJgvc6dYSKcaH37bNMUX4lox8rKtufzuaL2Zu~yN-RSOftov0zpmaclDRXyPlEFbprSuCNO9sDo7e-W01apuAvTDZHNbeCsbCaTMYYCGm12S6JlVMEaWL~WBem4VsWgFidiZ52hWiCzpMKnaygXRal7KBzO4XKy4ntHi7zE64aP62RF3EWTB2pzqsSNgFxPpHTO~q~FFcyuV-9OtI5y3vGJuQoAfzdtDPb~Uf0nU1tGTBZ6wXuTTQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA"></li>
</ul>
</li>
<li>
<p>multiomics 분석에서 샘플 수에 대해서 더 얘기하자면, 샘플 수가 무조건 많다고 분석에 활용이 가능한 것도 아니다.</p>
<ul>
<li>omic 데이터를 모아놓고 봤더니 overlapping되는 부분이 없어서 다 활용하지 못하는 경우도 있다.</li>
<li>실제로 <a href="https://www.frontiersin.org/articles/10.3389/fgene.2018.00477/full">2018년 논문</a>에서는 copy number alteration과 gene expression 데이터에 대한 407개의 샘플을 모았으나, 두 항목이 모두 조사된 샘플을 190개 밖에 없어서 190개만 활용해서 분석한 경우가 있다.</li>
</ul>
</li>
</ul>
<h2>4.5 After reshaping the data</h2>
<ul>
<li>Autoencoder, Cox PH는 ML을 활용하여 multiomics 분석을 하는 bioinformatics 그룹들이 가장 많이 쓴 기술이었으며, 둘 다 feature extraction을 통한 차원축소를 통해 모델의 정확성을 높이고자 했다.</li>
<li>다른 기술들은 classification, clustering, regression, inferring networks와 같은 용도의 기술들이었다.</li>
</ul>
<h2>4.6 New ML developments of potential use in multi-omics</h2>
<ul>
<li>2017~2018년도 이후로 ML 기술에서 여러 발전이 있었지만 이에 대한 multiomics 분석에 적용은 아직 많이 진척되지 않았다.
<ul>
<li>ML기술의 발전으로 말하자면 1) 인과관계에 대한 연구 2) 그래프 기반의 모델 3) 트랜스포머가 있었다.</li>
</ul>
</li>
<li>인과관계에 대한 연구는 '인과관계에 대한 가설정립'-&gt;'데이터를 통한 가설 확인'과 같은 방식으로 진행된다.
<ul>
<li>생물학의 경우 DNA-RNA-단백질로 구성된 센트럴도그마 시스템이 있는데, 이 요소들의 인과관계를 연구하기 위해 한꺼번에 multiomics로 분석한 사례가 있었다.</li>
<li>시스템 내에 존재하는 인과관계에 대한 가설을 정립할 때는 구조적인 가정 뿐 아니라 분포에 대한 가정으로도 가능하다(LiNGAM, CAM)</li>
<li>그러나 인과관계에 대한 연구는 일반적인 ML모델 훈련보다 훨씬 많은 샘플을 요구하기 때문에 multiomics 연구가 많이 이루어지지는 않는다.</li>
</ul>
</li>
<li>그래프 기반의 모델, 그래프신경망은 딥러닝 접근방식 중 하나이다.
<ul>
<li>특히 그래프 합성곱(convolutional) 네트워크가 유망하다.</li>
<li>합성곱 신경망은 합성곱 레이어를 쌓기 때문에, input으로 주어진 그래프에서 주요한 feature를 잘 배울 수 있고, 그 결과 예측, 분류와 같은 작업을 효과적으로 수행한다.</li>
<li>실제로 multiomics연구에서도 합성곱 신경망은 활용되고 있다.</li>
</ul>
</li>
<li>트랜스포머는 딥러닝 모델 중 하나다.
<ul>
<li>트랜스포머 딥러닝 모델은 합성곱이나 fully connected layer 대신 attention mechanism을 활용한다.</li>
<li>이 attention mechanism의 강력한 성능 때문에 ML 연구에서도 자연어(NLP)분야에서 많이 사용되고 있다.</li>
<li>트랜스포머 모델은 생물학에서도 잘 사용되는데 AlphaFold에서 사용된 것으로 유명하다.</li>
<li>생물학에서 이용된 또 다른 예시로 sequence 분석을 위해 개발 트랜스포머 모델인 <a href="https://proceedings.neurips.cc/paper/2020/hash/c8512d142a2d849725f31a9a7a361ab9-Abstract.html">Big Bird 모델</a>이 있다.</li>
<li>트랜스포머 모델을 사전에 잘 훈련 시킨다면, multiomics 연구에서 마주하게 되는 샘플 수가 적은 문제를 해결하는데 도움을 줄 수도 있다.</li>
</ul>
</li>
</ul>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120352" class="record">
            <header>
                <h1 class="record-title">Deep Learning</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120352</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                

                
            </main>

            <footer class="list-content-container">
                

                
                <details class="record-backlinks" open="open">
                    <summary class="menu-title">
                        backlinks
                        <span class="badge">1</span>
                    </summary>

                    <ul class="record-backlinks-list">
                        
                        <li data-target-id="20230924120375">
                            <a href="#20230924120375">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Machine Learning</span>
                                <div class="record-links-context">
                                    <h2>Example</h2>
<p><em class="id-context" data-target-id="20230924120352">Deep Learning</em><br>
Active Learning</p>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                
            </footer>
        </article>
        
        <article id="20230924120353" class="record">
            <header>
                <h1 class="record-title">DNA</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120353</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                

                
            </main>

            <footer class="list-content-container">
                

                
                <details class="record-backlinks" open="open">
                    <summary class="menu-title">
                        backlinks
                        <span class="badge">1</span>
                    </summary>

                    <ul class="record-backlinks-list">
                        
                        <li data-target-id="20230924120347">
                            <a href="#20230924120347">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Central Dogma</span>
                                <div class="record-links-context">
                                    <p><em class="id-context" data-target-id="20230924120353">DNA</em><br>
RNA<br>
Protein</p>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                
            </footer>
        </article>
        
        <article id="20230924120354" class="record">
            <header>
                <h1 class="record-title">edwith_LSnDP_Ch1</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120354</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://www.edwith.org/datastructure-2019s">EDWITH: Linear Structure and Dynamic Programming</a><br>
Chapter 1</p>
<h1>What Makes You As a Good Engineer?</h1>
<ul>
<li>Programming is an implementation tool.</li>
</ul>
<h2>Good engineer needs...</h2>
<ul>
<li>Conceptual thinking and design</li>
<li>Practical design and implementation</li>
</ul>
<h2>To programmer, the design means 'algorithm design'</h2>
<ul>
<li>We will learn UML to understand 'data structure' and 'algorithm design' -&gt; first priority</li>
<li>We will learn Python to implement the design -&gt; second priority
<ul>
<li>If the design is bad, implementation is meaningless</li>
</ul>
</li>
<li>Of course, good programmers should be good at BOTH</li>
</ul>
<h1>What is Python?</h1>
<ul>
<li>Python is Interpreter language
<ul>
<li>Opposite to Compiled language</li>
<li>Compiled language cannot run an imperfect code</li>
<li>However, the Interpreter language code(Python) can run an imperfect code</li>
</ul>
</li>
<li>Python is Object-oriented language</li>
<li>Python can dynamic type of variables.
<ul>
<li>The types of variables can be changed anytime</li>
<li>This is opposite to Compiled language</li>
</ul>
</li>
<li>Python is fasy to develop but slow to execute
<ul>
<li>Meanwhile, Compiled language programming is slow to develop but fast to execute</li>
</ul>
</li>
<li>Python is specialized in data analyses</li>
</ul>
<h1>Procedure-oriented vs Object-oriented</h1>
<h2>What is Procedure-oriented Programming?</h2>
<ul>
<li>Opposite to Object-oriented Programming</li>
<li>Procedure-oriented Programming does not use Class</li>
<li>Largely in two parts: 1) Definition part 2) Execution part</li>
</ul>
<pre><code class="language-Python">'''
This is 'Procedure-oriented' Hello World code.
Meaning, code has only functions.

'''

def main():    #function
    print (&quot;Hello, world&quot;)
    print (&quot;This program computes the average of two exam scores.&quot;)

    score1, score2 = input(&quot;Enter two scores separated by a comma: &quot;).split(&quot;,&quot;)
    average = (int(score1) + int(score2)) / 2.0

    print(&quot;The average of the scores is: &quot;, average)
main()
</code></pre>
<h2>What is Object-oriented Programming</h2>
<ul>
<li>Object-oriented Programming use Class</li>
<li>Largely in two parts: 1) Definition part 2) Execution part</li>
</ul>
<pre><code class="language-Python">'''
This is 'Object-oriented' Hello World.
Meaning, using class to make an instance later.

'''
class HelloWorld: #class name -&gt; noun, camel casing
    def __init__(self): #constructor
        print (&quot;Hello, world&quot;)

    def __del__(self):
        print (&quot;Good Bye!&quot;)

    def performAverage(self, val1, val2): #method name -&gt; verb, start with lower case
        average = (val1 + val2) / 2.0 #variable name -&gt; noun, start with lower case
        print(&quot;The average of the scores is: &quot;, average)

def main(): 
    world = HelloWorld()
    score1, score2 = input(&quot;Enter two scores separated by a comma: &quot;).split(&quot;,&quot;)
    world.performAverage(int(score1), int(score2))

main()

'''
Also, check the naming and styling recommendations

For class,
    - Noun for the concept to eb represented by the class
    - Capitalize the first letter of each word

For variable,
    - Noun for the contents to be stored
    - Start with lower case
    - Not recommend to note the type of variable
      (variable's type can be changed easily from Python)

For method,
    - Verb for the method action
    - Start with lower case
'''
</code></pre>
<h1>Naming and Styling</h1>
<p><strong>For Class</strong></p>
<ul>
<li>Capitalize the first letter of each word</li>
<li>e.g. MyFirstClass<br>
<strong>For Variable</strong></li>
<li>Start with lower case</li>
<li>Do not recommend to put the name of type front (Hungarian Notation style?)</li>
<li>e.g. numberOfStudents = 100<br>
<strong>For Function, Method</strong></li>
<li>Start with lower case</li>
<li>Start with verb</li>
<li>e.g. getNumber</li>
</ul>
<h1>Operator</h1>
<ul>
<li>Result of operation is displaced which has more information</li>
<li>e.g. int(10)/int(3.0) = 3.333335</li>
</ul>
<h1>Function Statement</h1>
<ul>
<li>You can return multiple variables
<ul>
<li>But keep them in order</li>
</ul>
</li>
<li>You do not have to specify return types</li>
</ul>
<pre><code class="language-Python">def name(parameters):
	statements
	return variable1 variable2...
</code></pre>
<ul>
<li>One line function is called <em>lambda</em> function</li>
</ul>
<pre><code class="language-Python">lambda &lt;parameters&gt; : &lt;return&gt;
</code></pre>
<pre><code class="language-Python">numA = 1
numB = 2

lambdaAdd = lambda numParam1, numParam2 : num Param1 + numParam2

numL = lambdaAdd(numA, numB)
print(numL)
</code></pre>
<h1>What is different between 'A is B' and 'A == B'?</h1>
<ul>
<li>'A is B' checks their objects' IDs</li>
<li>'A == B' checks their values</li>
</ul>
<pre><code class="language-Python">k=2
A=[1, 2, 3]
B=[1, k, 3]

# A is B -&gt; False
# A == B -&gt; True
</code></pre>
<h1>Class and Instance</h1>
<ul>
<li>Class is like a blue print</li>
<li>Instance is like a house, which is built based on the blue print, the Class</li>
<li>One blue print -&gt; Several Houses, but with slightly different features</li>
</ul>
<pre><code class="language-Python">class MyHome:
	colorRoof = 'red'
	stateDoor = 'closed'
	def paintRoof(self, color):
		self.colorRoof = color
	def openDoor(self):
		self.stateDoor = 'open'
	def closeDoor(self):
		self.stateDoor = 'close'
	def printStatus(self):
		print(&quot;Roof color is&quot;, self.colorRoof, &quot;, and door is&quot;, self.stateDoor)


homeAtTokyo = MyHome() #instance
homeAtSeoul = MyHome() #instance

homeAtSeoul.openDoor()
homeAtTokyo.paintRoof('blue')

homeAtTokyo.printStatus()
homeAtSeoul.printStatus()
</code></pre>
<h1>Constructor, Destructor in Class</h1>
<h2>Constructor</h2>
<ul>
<li>Called when instantiated</li>
</ul>
<h2>Destructor</h2>
<ul>
<li>Called when the instance is removed from the value table</li>
<li>Clearly removing the class</li>
<li>Rarely used<br>
Let's add the constructor and destructor on the above code.</li>
</ul>
<pre><code class="language-Python">from time import ctime   #Let's add time info

class MyHome:
	colorRoof = 'red'
	stateDoor = 'closed'
	def paintRoof(self, color):
		self.colorRoof = color
	def openDoor(self):
		self.stateDoor = 'open'
	def closeDoor(self):
		self.stateDoor = 'close'
	def printStatus(self):
		print(&quot;Roof color is&quot;, self.colorRoof, &quot;, and door is&quot;, self.stateDoor)
	def __init__(self, strAddress):   #Constructor
		print(&quot;Built on&quot;, strAddress)
		print(&quot;Built at&quot;, ctime())
	def __del__(self):   #Destructor
		print(&quot;Destroyed at&quot;, ctime())


homeAtNewyork = MyHome('NYC')
homeAtNewyork.printStatus()
del homeAtNewyork
</code></pre>
<h1>Importing Module</h1>
<ul>
<li>Write down Class codes in seperated files</li>
<li>You can import them to anohter single code file</li>
<li>Use from to specify the director, or the folder, path</li>
<li>Let's say...
<ol>
<li>Make <em><a href="http://home.py">home.py</a></em> file which has Class, <em>HOME</em></li>
<li>Make <em><a href="http://UsingMyHome.py">UsingMyHome.py</a></em> file, write down 'from <em>path</em> import HOME'</li>
<li>Now you can use the class <em>HOME</em> in <a href="http://UsingMyHome.py">UsingMyHome.py</a> file,</li>
</ol>
</li>
</ul>
<h1>Managing Modules by Package</h1>
<ul>
<li>Several module files can be placed in one directory or folder</li>
<li>We call these directories as <strong>Package</strong></li>
<li>Package folder should have <em><em>init</em>.py</em> in the directory<pre><code class="language-Python">from &lt;directory=package&gt; import module
</code></pre>
</li>
<li>When the module is placed in the same directory with the file which uses the module, we can skip to write <em>from</em><pre><code class="language-Python">import module
</code></pre>
</li>
</ul>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120355" class="record">
            <header>
                <h1 class="record-title">edwith_LSnDP_Ch2</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120355</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://www.edwith.org/datastructure-2019s">EDWITH: Linear Structure and Dynamic Programming</a><br>
Chapter 2: Object Oriented Paradigm</p>
<h1>Object-oriented paradigm(OOP)</h1>
<h2>What is <em>object</em> from programming</h2>
<ul>
<li>Class, Instance, inheritance, encapsulation, polymorphism ...</li>
</ul>
<h2>What is Object-Oriented Design</h2>
<ul>
<li>You have to program three things of object
<ol>
<li>object's name</li>
<li>object's features</li>
<li>object's action</li>
</ol>
</li>
<li>Think about concept, not code.
<ul>
<li>Let's abstract the real world concepts into object design.</li>
<li>How can we abstract the information of customer from online banking?
<ol>
<li>Customer's name</li>
<li>Customer's featuer: bank ID, account number, ...</li>
<li>Customer's action: login, request withdrawl, confirm security code, ...</li>
</ol>
</li>
</ul>
</li>
</ul>
<h2>Class and Instance in Object-Orieted Design</h2>
<ul>
<li>Right above, we designed 'Customer's basic concept'</li>
<li>If we register foure different people as custoemers, each of them would register different names, bank IDs, requests, etc...</li>
<li>Default customer information which is required -&gt; <em>Class</em></li>
<li>Four customers' information following default form -&gt; <em>Instance</em></li>
</ul>
<h3>Class</h3>
<ul>
<li>Result of design and implementation</li>
</ul>
<h3>Instance</h3>
<ul>
<li>Result of execution</li>
</ul>
<h1>The importance fo UML</h1>
<ul>
<li>UML(Unified Modeling Language)</li>
<li>Standarad design for software engineering</li>
<li>Learning standard from field is important to communicate with your team</li>
</ul>
<h1>Encapsulation</h1>
<ul>
<li>Utilizing the visibility of your object</li>
<li>Object = Data + Behavior
<ul>
<li>Behavior: method, member function, operation</li>
</ul>
</li>
<li>Types of visibility
<ul>
<li>Private: seen only within the class</li>
<li>Protected: seen only within the class and its descendants</li>
<li>Public: seen everywhere</li>
</ul>
</li>
<li>Recommended policy is like...
<ul>
<li>Data -&gt; private</li>
<li>Behavior -&gt; public</li>
</ul>
</li>
<li>Python cannot restrict t he visibility
<ul>
<li>One promise between developers is that if the file name has two underscores before the name( _ _ <a href="http://example.py">example.py</a>), <strong>do not touch</strong></li>
</ul>
</li>
</ul>
<h1>Inheritance between <em>Class</em></h1>
<ul>
<li>Giving attributes to descendants
<ul>
<li>Attributes are including member variables, methods</li>
</ul>
</li>
<li>Descendants will get attributes but...
<ul>
<li>Descendants can have their own attributes</li>
<li>Descendants can overlap the attributes they got</li>
</ul>
</li>
<li>JAVA's class can have only one parent</li>
<li>Python's class can have several parents</li>
</ul>
<h2>Superclass</h2>
<ul>
<li>Ancestor, parents</li>
<li>In design, generalized class</li>
</ul>
<h2>Subclass</h2>
<ul>
<li>Descendants</li>
<li>In design, specialized class</li>
</ul>
<h2>Practice</h2>
<ul>
<li>In Python, to be inherited...
<ul>
<li>Write down the Superclass's name as a parameter</li>
</ul>
</li>
</ul>
<pre><code class="language-Python">'''
inheritance: Giving my attribute(member variables, methods) to my descendants

superclass: my ancestors, specifically my father
            Python let the class has more than two one parent class(&lt;-&gt; Java)

subclass: my descendatns, specifically my son
'''

class Father(object): #the superest class is inherited from 'object' class
    strHomeTown = &quot;Jeju&quot;
    def __init__(self): #constructor
        print(&quot;Father is created&quot;)
    def doFatherThing(self): #method
        print(&quot;Father's action&quot;)
    def doRunning(self):
        print(&quot;Slow&quot;)

class Mother(object):
    strHomeTown = &quot;Seoul&quot;
    def __init__(self):
        print(&quot;Mother is created&quot;)
    def doMotherThing(self):
        print(&quot;Mother's action&quot;)

class Child(Father, Mother): #the class 'Child' has two super classes
    strName = &quot;Moon&quot;
    def __init__(self):
        super(Child, self).__init__() #call the super class 'Father'
        print(&quot;Child is created&quot;)
    def doRunning(self): #method override
        print(&quot;Fast&quot;)

me = Child() #instance created
me.doFatherThing()
me.doMotherThing()
me.doRunning()
print(me.strHomeTown)
print(me.strName)
</code></pre>
<ul>
<li>SuperClass1: strHometown, doFatherthing(), doRunning()</li>
<li>SuperClass2: strHometown, doMotherthing()</li>
<li>SubClass: strName, doRunning()
<ul>
<li>SubClass's doRunning() method will be ovelapped</li>
</ul>
</li>
</ul>
<h1>Reference Variables <em>self</em> and <em>super</em></h1>
<ul>
<li>From Python, <em>self</em> is...
<ul>
<li>Reference variable pointing the instance itself</li>
</ul>
</li>
<li>From Python, <em>super</em> is...
<ul>
<li>Reference variable pointing the base class instance</li>
<li><em>super</em> is used to ca;l the base class methods</li>
</ul>
</li>
</ul>
<h2>Practice</h2>
<pre><code class="language-Python">'''
self: reference variable pointing the instance itself

super: reference variable pointing the base class instance
       super is used to call the base class methods.
'''
class Father(object):
    strHometown = &quot;Jeju&quot;
    def __init__(self, paramHome):
        self.strHometown = paramHome
        print(&quot;Father is created&quot;)
    def doFatherThing(self):
        print(&quot;Father's action&quot;)
    def doRunning(self):
        print(&quot;Slow&quot;)

class Mother(object):
    strHomeTown = &quot;Seoul&quot;
    def __init__(self):
        print(&quot;Mother is created&quot;)
    def doMotherThing(self):
        print(&quot;Mother's action&quot;)

class Child(Father, Mother):
    strName = &quot;Moon&quot;
    def __init__(self, paramName, paramHome): #when an instance is created, two arguments will be 'pramName', 'paramHome'
        super(Child, self).__init__(paramHome) #bring the 'paramHome' from the superclass 'Father'
        self.strName = paramName
        print(&quot;Child is created&quot;)
    def doRunning(self):
        print(&quot;Fast&quot;)

me = Child(&quot;Sun&quot;, &quot;Universe&quot;) #Father is created, Child is created
me.doFatherThing() #Father's action
me.doMotherThing() #Mother's action
me.doRunning() #Fast
print(me.strHomeTown) #Universe
print(me.strName) #Sun
</code></pre>
<h1>Polymorphism</h1>
<ul>
<li>Method Overriding &amp; Overloading</li>
<li>After get method from superclass, subclass can modify the method in two ways: overriding, overloading</li>
<li>We can also set the default parameter value</li>
</ul>
<h2>Method Overriding</h2>
<ul>
<li>Base class has a method A(num),<br>
and its derived class has a method A(num)</li>
</ul>
<h2>Method Overloading</h2>
<ul>
<li>A class has a method A(num),<br>
A(num, name),<br>
and A(num, name, home)</li>
</ul>
<h2>Practice</h2>
<pre><code class="language-python">'''
method overriding: change the behavior of existing methods

method overloading: more than one method of the same class shares the same method name having different signatures
'''
class Building:
    strAddress = &quot;Daejeon&quot;
    def openDoor(self):
        print(&quot;Door Opened&quot;)

class Hotel(Building):
    def openDoor(self): #overriding of the method 'openDoor' from the super class 'Building'
        print(&quot;Bellboy opens a door&quot;)
    def checkIn(self, days = 1): #overloading by putting default value of the parameter
        print(&quot;Someone checks in for&quot;, days, &quot;days&quot;)

lotteHotel = Hotel()
lotteHotel.openDoor()
lotteHotel.checkIn() #default value of 'days' is 1. So in here, the method 'checkIn' looks like doesn't have any parameter.
lotteHotel.checkIn(2) #the method 'checkIn' gets parameter but still works with overloading.
</code></pre>
<h1>Abstract class (=Abstract Base Class)</h1>
<ul>
<li>A class with an abstract method
<ul>
<li>The abstract method is a method with signature, but with no implementation</li>
<li>The abstract method is not a complete</li>
<li>You can't make an instance out of it</li>
</ul>
</li>
<li>By abstract class, we can design more carefully
<ul>
<li>The main purpose is inheriting the abstract method to subclass and let that override it</li>
</ul>
</li>
</ul>
<h2>Practice</h2>
<pre><code class="language-python">'''
abstract class: A class with an abstract method

abstract method: A method with signature, but with no implementation.
                 Abstract method is not a complete.
                 You can't make an instance out of it.
                 The concrete class with 'full' implementations and inheriting the abstract class will be a basic for instances.
'''
from abc import ABC, abstractmethod

class Room(ABC):
    @abstractmethod #indicator of abstract base method and class
    def openDoor(self):
        pass
    @abstractmethod
    def openWindow(self):
        pass

class BedRoom(Room):
    def openDoor(self):
        print(&quot;Open bedroom door&quot;)
    def openWindow(self):
        print(&quot;Open bedroom winodw&quot;)

class Lobby(Room): #this class has a lack of method, 'openWindow' which is the abstract base method of the super calss.
    def openDoor(self):
        print(&quot;Open lobby door&quot;)

room1 = BedRoom()
print(issubclass(BedRoom, Room), isinstance(room1, Room))

lobby1 = Lobby()
print(issubclass(Lobby, Room), isinstance(lobby1, Room))
</code></pre>
<h1><em>Object</em> class</h1>
<ul>
<li>The top superclass of all of classes
<ul>
<li>Or, think it as like default type of class</li>
</ul>
</li>
<li>Every Python classes are the descendeants of <em>Object</em>
<ul>
<li>When we code like <code>class name:</code> it's actually like <code>class name(object):</code></li>
</ul>
</li>
<li><em>Object</em> class has several methods
<ul>
<li><em>init</em></li>
<li><em>del</em></li>
<li><em>eq</em></li>
<li><em>add</em><br>
...</li>
</ul>
</li>
<li>We are actually override them to make the methods behave as we please</li>
</ul>
<h2>Practice</h2>
<pre><code class="language-python">'''
All of Python classes are the descendants of Object.

Object class has many hidden methods: __init__  constructor
                                      __del__
                                      __eq__    T/F examination
                                      __cmp__   compare
                                      __add__   

We actually override to those hidden methods.
'''
class Room:
    numWidth = 100
    numHeight = 100
    numDepth = 100

    def __init__(self, parWidth, parHeight, parDepth): #actually overriding from the object class
        self.numDepth = parDepth
        self.numWidth = parWidth
        self.numHeight = parHeight
    
    def getVolume(self):
        return self.numDepth*self.numHeight*self.numWidth

    def __eq__(self, other):
        if isinstance(other, Room):
            if self.getVolume() == other.getVolume(): #Duck Typing: the variable is not defined yet
                return True
        return False
</code></pre>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120356" class="record">
            <header>
                <h1 class="record-title">edwith_LSnDP_Ch3</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120356</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://www.edwith.org/datastructure-2019s">EDWITH: Linear Structure and Dynamic Programming</a><br>
Chapter 3: Abstract Data Types</p>
<h1>Abstract Data Types(ADT)</h1>
<ul>
<li>An abstraction of a data structure</li>
<li>Include...
<ul>
<li>Stack</li>
<li>Queue</li>
<li>List<br>
...</li>
</ul>
</li>
<li>ADT specifies...
<ul>
<li>Data stored</li>
<li>Operations on the data</li>
<li>Error conditions associated with operations
<ul>
<li>Error condition: expect future error and plan the solution</li>
</ul>
</li>
</ul>
</li>
</ul>
<h1>What is <em>array</em> in general programming</h1>
<ul>
<li>Array has elements</li>
<li>The elements is accessible by index</li>
<li>The index is typically zero or a positive integer</li>
</ul>
<h2>Search procedure in <em>array</em> and  retrieval time</h2>
<ul>
<li>Should use indexes only for searching</li>
</ul>
<ol>
<li>Navigate from the first</li>
<li>Keep navigate until hit
<ul>
<li>If there is no hit, reach to the end</li>
</ul>
</li>
</ol>
<ul>
<li>The maximum retrievals are <strong>N</strong></li>
</ul>
<h2>Insert procedure in <em>array</em> and  retrieval time</h2>
<ul>
<li>We have list 'X'</li>
<li>Want to insert 'c' in third place</li>
</ul>
<ol>
<li>Create a new empty list 'Y'</li>
<li>Put first and second elements of 'X' to 'Y'</li>
<li>Put 'c' to 'Y'</li>
<li>Put rest elements of 'X' to 'Y'</li>
<li>Claim, 'X = Y'</li>
</ol>
<ul>
<li>Have to reach to the end of 'X', therefore, the retrievals are N</li>
</ul>
<h2>Delete procedire in <em>array</em> and retrieval time</h2>
<ul>
<li>We have list 'X'</li>
<li>Want to delete 'd' in fourth place</li>
</ul>
<ol>
<li>Create a new empty list 'Y'</li>
<li>Put first to thrid elements of 'X' to 'Y'</li>
<li>Do not input 'd' from 'X' to 'Y'</li>
<li>Put rest elements of 'X' to 'Y'</li>
<li>Claim, 'X = Y'</li>
</ol>
<ul>
<li>Have to reach to the end of 'X', except one element</li>
<li>The retrievals are N-1, almost same as N</li>
</ul>
<h2>The problems in <em>array</em></h2>
<ul>
<li>Have to perform line-wise retrievals, N retrievals</li>
<li>So, the procedure are bounded to the one-dimension</li>
<li><strong>What if we have a magic to create a space in the middel of the line?</strong>
<ul>
<li><em><strong>Linket list</strong></em> -&gt; the procedure is not bounded anymore -&gt; solve!</li>
</ul>
</li>
</ul>
<h1>Singly Linked List</h1>
<ul>
<li>A singly linked list is organized with nodes and references</li>
</ul>
<h2>Node</h2>
<ul>
<li>A node consists of
<ol>
<li>A variable to hold a reference to its next node</li>
<li>A variable to hold a reference to its value object</li>
</ol>
</li>
<li>There are two special nodes: <em>Head, Tail</em>
<ul>
<li><em>Head</em> is the first node of the list, <em>Tail</em> is the last node of the list</li>
<li>Special nodes don't have references, no need to connect to the next node
<ul>
<li><em>Head</em> doesn't have value</li>
<li><em>Tail</em> doesn't have value and next node linkage</li>
</ul>
</li>
<li>Special nodes are necessary to a linked list.<br>
However, they make procedures more convenient such as search, insert, and delete</li>
</ul>
</li>
</ul>
<h2>Reference</h2>
<ul>
<li>The node's variables are linked with references</li>
<li>Reference can be a value or next node</li>
</ul>
<h2>Practice</h2>
<pre><code class="language-python">'''
Singly linked list: a type of data structure.
                    It's built with several nodes.
                    Each nodes are linked with their own reference values. (except head and tail nodes)

Before we implement a singly linked list,
we should define the Node class.
'''
class Node:
    nodeNext = None
    nodePrev = ''
    objValue = ''
    blnHead = False #head and tail nodes are optional.
    blnTail = False #having head and tail makes implementation very convenient.
    
    def __init__(self, objValue='', nodeNext = None, blnHead = False, blnTail = False):
        self.nodeNext = nodeNext
        self.objValue = objValue
        self.blnHead = blnHead
        self.blnTail =blnTail
    
    def getValue(self): #node has a variable to it's reference value
        return self.objValue
    def setValue(self, objValue):
        self.objValue = objValue

    def getNext(self): #node has a variable to it's next node
        return self.nodeNext
    def setNext(self, nodeNext):
        self.nodeNext = nodeNext

    def isHead(self):
        return self.blnHead
    def isTail(self):
        return self.blnTail

node1 = Node(objValue = 'a')
nodeTail = Node(blnTail = True)
nodeHead = Node(blnHead = True, nodeNext = node1)
</code></pre>
<h1>Search Procedure in Singly Linked List</h1>
<ul>
<li>Actually, searching procedure in singly linked list takes time as much as array</li>
<li>Check every values</li>
<li>The procedure is like....
<ol>
<li>Find Head node</li>
<li>Loop to search next node
<ul>
<li>if Tail node -&gt; exit</li>
<li>if not Tail node -&gt; value check</li>
</ul>
</li>
</ol>
</li>
</ul>
<h1>Insert&amp;Delete Procedure in Singly Linked List</h1>
<ul>
<li>Insert and Delete procedures takes much less time than array</li>
</ul>
<h2>Insert Procedure</h2>
<ul>
<li>Let's say, we already have a new node which we want to insert</li>
<li>Let's say, put node 'c', between 'a'-'b'-'d'-'e'</li>
</ul>
<ol>
<li>Find previous node(node 'b')</li>
<li>Link the next variable of the previous node, to the new node(node 'b' -&gt; node 'c')</li>
<li>Modify the next variable of the new node, to the next node(node 'c' -&gt; node 'd')</li>
</ol>
<h2>Delete Procedure</h2>
<ol>
<li>Retrieve a new next node, which is two nodes far from a previous node</li>
<li>Change a reference of the next variable in a previous node, to the new next node</li>
</ol>
<h2>Practice</h2>
<pre><code class="language-python">'''
Implement singly linked list using the node class we definied.
We can do insertion or deletion of nodes from singly link list.

1) Insert procedure: store the new node
                     update the previous node to point the new node
                     update the new node to point the next node

2) Delete procedure: retrieve the next node
                     update the previous node to point the next node
'''
from ImportOnly.NodeClass import Node #please check https://github.com/jhlee0637/CodeExercising 

class SinglyLinkedList:
    nodeHead = ''
    nodeTail = ''
    size = 0
    def __init__(self):
        self.nodeTail = Node(blnTail=True)
        self.nodeHead = Node(blnHead=True, nodeNext = self.nodeTail)

    def insertAt(self, objInsert, idxInsert): 
        nodeNew = Node(objValue = objInsert)
        nodePrev = self.get(idxInsert - 1) #get the previous node through the method 'get' which is defined from below
        nodeNext = nodePrev.getNext()
        nodePrev.setNext(nodeNew)
        nodeNew.setNext(nodeNext)
        self.size = self.size + 1

    def removeAt(self, idxRemove):
        nodePrev = self.get(idxRemove - 1)
        nodeRemove = nodePrev.getNext()
        nodeNext = nodeRemove.getNext()
        nodePrev.setNext(nodeNext)
        self.size = self.size - 1
        return nodeRemove.getValue()

    def get(self, idxRetrieve): #method 'get' to get which index's node
        nodeReturn = self.nodeHead
        for itr in range(idxRetrieve + 1):
            nodeReturn = nodeReturn.getNext()
        return nodeReturn

    def printStatus(self):
        nodeCurrent = self.nodeHead
        while nodeCurrent.getNext().isTail() == False:
            nodeCurrent = nodeCurrent.getNext()
            print(nodeCurrent.getValue(), end=&quot; &quot;)
        print(&quot;&quot;)

    def getSize(self):
        return self.size

'''
Let's try:  a - b - d - e - f

            a - b -&quot;c&quot;- d - e - f

            a - b - c ----- e - f
'''
list1 = SinglyLinkedList()
list1.insertAt('a',0)
list1.insertAt('b',1)
list1.insertAt('d',2)
list1.insertAt('e',3)
list1.insertAt('f',4)
list1.printStatus()

list1.insertAt('c',2) #insert the node at the middle of the array
list1.printStatus()

list1.removeAt(3) #delete the node at the middle of the array 
list1.printStatus()
</code></pre>
<h1>Stack</h1>
<h2>Structure of <em>Stack</em></h2>
<ul>
<li>Stacks are linear like linked lists
<ul>
<li>A variation of a singly linked list</li>
</ul>
</li>
<li>You can only access to <strong>the first instance</strong> in the list
<ul>
<li>This mechanism is called Last-In-First-Out(LIFO)</li>
<li>You cannot modify other instances in the list</li>
</ul>
</li>
<li>Imagine the snack, Pringles
<ul>
<li>You can only access to the top of the container</li>
</ul>
</li>
</ul>
<h2><em>Push</em> operation of <em>Stack</em></h2>
<ul>
<li>Insert an instance at the first in the linked list</li>
<li>Put an instance at the top in the stack</li>
</ul>
<h2>Pop operation of Stack</h2>
<ul>
<li>Remove and return an instance at the first in the linked list</li>
<li>Remove and return an instance at the top in the stack</li>
</ul>
<h2>Practice</h2>
<pre><code class="language-python">'''
Stacks are the name of data structure.
Stacks are linear like linked lists.
You can only access to the end of stack.
You can't insert or delete the data from middle. (Last In First Out)

Stack has two operations.
1) Push: insert an instance at the first in the linked list.
2) Pop: remove and return an instance at the first in the linked list.
'''
from ImportOnly.SinglyLinkedList import SinglyLinkedList #please check https://github.com/jhlee0637/CodeExercising

class Stack(object):
    lstInstance = SinglyLinkedList()

    def pop(self):
        return self.lstInstance.removeAt(0) #remove from first

    def push(self, value):
        self.lstInstance.insertAt(value, 0) #insert to first

stack = Stack()
stack.push(&quot;a&quot;)
stack.push(&quot;b&quot;)
stack.push(&quot;c&quot;)

print(stack.pop()) #c
print(stack.pop()) #b
print(stack.pop()) #a
</code></pre>
<h2>Where to use <em>Stack</em> structure?</h2>
<ul>
<li>For example, check balancing symbols
<ul>
<li>Balancing symbols: ( ), { }, [ ]</li>
<li><code>[2+(1+2)]-3</code> -&gt; balanced symbols</li>
<li><code>{2+(1}+2)-3</code> -&gt; not balanced symbols</li>
</ul>
</li>
<li>It's not about counting symbols</li>
<li>It's about opening and closing at right order</li>
<li>Algorithm for the balanced symbol checking
<ol>
<li>Make an empty stack</li>
<li>Read symbols until end of formula
<ul>
<li>If the symbol is an opening symbol, push it onto the stack</li>
<li>If the symbol is a closing symbol
<ul>
<li>If the stack is empty -&gt; error</li>
<li>If the stack has matched symbol -&gt; pop</li>
<li>If the stack has not-matched symbol -&gt; error</li>
</ul>
</li>
</ul>
</li>
<li>At the end of the formula, if the stack is not empty -&gt; error</li>
</ol>
</li>
</ul>
<h1>Queue</h1>
<h2>Structure of Queue</h2>
<ul>
<li>Queues are linear like linked lists
<ul>
<li>A variation of a singly linked list</li>
</ul>
</li>
<li>You can only access to <strong>the first and last instance</strong> in the list
<ul>
<li>This mechanism is called First-In-First-Out(LIFO)</li>
<li>You cannot modify other instances in the list</li>
</ul>
</li>
<li>Items are inserted at the last, removed at the front</li>
</ul>
<h2>Enqueue operation from Queue</h2>
<ul>
<li>Insert an instance at the last in the linked list</li>
<li>Put an instance at the rear in the queue</li>
</ul>
<h2>Dequeue operation from Queue</h2>
<ul>
<li>Remove and return an instance at the first in the linked list</li>
<li>Remove and return an instance at the front in the queue</li>
</ul>
<h2>Practice</h2>
<pre><code class="language-python">'''
Queues are the name of data structure.
Queues are linear like linked lists.
You can insert to the first of queue.
You can remove from the end of queue.
You can't insert or delete the data from middle. (First In First Out)

Queue has two operations.
1) Enqueue: insert an instance at the last in the linked list.
2) Dequeue: remove and return an instance at the first in the linked list.
'''
from ImportOnly.SinglyLinkedList import SinglyLinkedList #please check https://github.com/jhlee0637/CodeExercising

class Queue(object):
    lstInstance = SinglyLinkedList()

    def dequeue(self):
        return self.lstInstance.removeAt(0) #remove from first

    def enqueue(self, value):
        self.lstInstance.insertAt(value, self.lstInstance.getSize()) #insert to end

queue = Queue()
queue.enqueue(&quot;a&quot;)
queue.enqueue(&quot;b&quot;)
queue.enqueue(&quot;c&quot;)

print (queue.dequeue()) #a
print (queue.dequeue()) #b
print (queue.dequeue()) #c
</code></pre>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120357" class="record">
            <header>
                <h1 class="record-title">edwith_LSnDP_Ch4</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120357</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://www.edwith.org/datastructure-2019s">EDWITH: Linear Structure and Dynamic Programming</a><br>
Chapter 4: Recursions and Dynamic Programming</p>
<h1>Recursion</h1>
<h2>Repeating Problem from Programming</h2>
<ol>
<li>You use a method to solve problem, get a value</li>
<li>Now, you have to use the value to put it in the same method, and get another value</li>
<li>repeating...</li>
</ol>
<ul>
<li>The size of problem decreases</li>
<li>Examples
<ul>
<li>Factorial</li>
<li>Great Common Divisor
<ul>
<li>GCD(32, 24) = 8</li>
<li>Using Euclid's algorithm<br>
GCD(A, B) = GCD(B, A mod B)<br>
GCD(A, 0) = A</li>
</ul>
</li>
</ul>
</li>
<li>Commonality
<ul>
<li>Repeating function calls</li>
<li>Reducing parameters</li>
<li>Just like the mathematical induction</li>
</ul>
</li>
</ul>
<h2>Using Recursion to handle the Repeating Items</h2>
<ul>
<li>Often in a form of calling a function within the function</li>
</ul>
<h2>Practice: Fibonacci</h2>
<pre><code class="language-python">'''
Recursion: A programming method to handle the repeating items in a 'self-similar' way.
           often calling the function again within the function.
'''

def Fibonacci(n):
    if n == 0: #escape when it's not for recursion
        return 0
    if n == 1: #escape when it's not for recursion
        return 1
    intRet = Fibonacci(n-1) + Fibonacci(n-2) #calling the function again within the function
    return intRet

for itr in range(0, 10):
    print(Fibonacci(itr), end=&quot; &quot;) #0 1 1 2 3 5 8 13 21 34
</code></pre>
<h1>Stackframe</h1>
<ul>
<li><em>Recursion</em> is a repeat of function call</li>
<li><em>Stackframe</em> is a stack storing your function call history</li>
<li>Two types of <em>Stackframe</em>'s actions
<ul>
<li><em>Push</em>: When a function is invoked, <strong>record the function call</strong></li>
<li><em>Pop</em>: When a function hits <em>return</em> or ends, <strong>remove the function call</strong></li>
</ul>
</li>
</ul>
<h1>Merge Sort</h1>
<ul>
<li><em>Merge Sort</em> is one example of recursive programming
<ul>
<li>Sorting randomly aligned numbers in ascending order</li>
<li>Convert this problem into repeating problem -&gt; decompose and aggregate</li>
</ul>
</li>
</ul>
<h2>Merge Sort Procedures</h2>
<ul>
<li>Decompose into two smaller lists</li>
<li>Compare two smallest lists</li>
<li>Aggregate to one larger and sorted list</li>
<li>Repeat Comparision and Aggregattion</li>
</ul>
<h2>Practice</h2>
<pre><code class="language-python">'''
Merge sort is an example of recursive programming.
It's a way of sort of the list.
Repeatedly decompose the list into two smaller lists.
When you can't divide the list anymore, aggregate to one sorted list.
'''

import random

def perfomrMergeSort(lstElementToSort):
    if len(lstElementToSort) == 1: #escape when the list not for recursion
        return lstElementToSort

    #decompose of the given list
    lstSubElementToSort1 = [] 
    lstSubElementToSort2 = []
    for itr in range(len(lstElementToSort)):
        if len(lstElementToSort)/2 &gt; itr:
            lstSubElementToSort1.append(lstElementToSort[itr])
        else:
            lstSubElementToSort2.append(lstElementToSort[itr])

    #recursion - repeatedly decompose the list until it can't be divded anymore
    lstSubElementToSort1 = perfomrMergeSort(lstSubElementToSort1)
    lstSubElementToSort2 = perfomrMergeSort(lstSubElementToSort2)

    #aggregate
    idxCount1 = 0
    idxCount2 = 0
    for itr in range(len(lstElementToSort)):
        if idxCount1 == len(lstSubElementToSort1):
            lstElementToSort[itr] = lstSubElementToSort2[idxCount2]
            idxCount2 = idxCount2 + 1
        elif idxCount2 == len(lstSubElementToSort2):
            lstElementToSort[itr] = lstSubElementToSort1[idxCount1]
            idxCount1 = idxCount1 + 1
        elif lstSubElementToSort1[idxCount1] &gt; lstSubElementToSort2[idxCount2]:
            lstElementToSort[itr] = lstSubElementToSort2[idxCount2]
            idxCount2 = idxCount2 + 1
        else:
            lstElementToSort[itr] = lstSubElementToSort1[idxCount1]
            idxCount1 = idxCount1 + 1
    
    return lstElementToSort

#Test
lstRandom = []
for itr in range(0, 10):
    lstRandom.append(random.randrange(0, 100))
print(lstRandom)

lstRandom = perfomrMergeSort(lstRandom)
print(lstRandom)
</code></pre>
<h1>Question: Is Recursion really good?</h1>
<ul>
<li>Cons: longer list needs more time and space
<ul>
<li>Excessive function calls</li>
<li>Same funcions can be overlapped
<ul>
<li>For, instance, Fibonacci(4)<br>
F(4) = F(3) + F(2)<br>
= F(2) + F(1) + F(1) + F(0)<br>
= F(1) + F(0) + F(1) + F(1) + F(0)</li>
<li>Fibonacci(4) has three repeated calls of F(1), two repeated F(0), two repeated F(2)</li>
</ul>
</li>
</ul>
</li>
<li>Solution -&gt; <em><strong>Dynamic Programming</strong></em></li>
</ul>
<h1>Dynamic Programming</h1>
<ul>
<li><em>Dynamic Programming</em> is a general algorithm <strong>design technique</strong><br>
(It's not an algorithm as like <em>Recursion</em>)
<ul>
<li><em>Dynamic Programming</em> is defined by or fomulated as recurrences with overlapping sub-instances</li>
</ul>
</li>
<li>Using a table, writing down the result of each instances in advance</li>
<li>Setting up a recurrence
<ol>
<li>Solve small instances once</li>
<li>Record solutions in a table</li>
<li>Extract a solution of a larger instance from the table</li>
</ol>
</li>
</ul>
<h2>Memoization</h2>
<ul>
<li>One of optimization techniques for <em>Dynamic Programming</em></li>
<li>Sotring the results of previous function calls to reuse the results again in the future</li>
</ul>
<h3>Differences between Recursion's StackFrame and Dynamic Programming's Memoization</h3>
<ul>
<li>StackFram is top-down
<ul>
<li>F(1)</li>
<li>F(2) -&gt; F(1)</li>
<li>F(3) -&gt; F(2) -&gt; F(1)</li>
<li>F(4) -&gt; F(3) -&gt; F(2) -&gt; F(1)</li>
</ul>
</li>
<li>Memoizaiton is bottom-up
<ul>
<li>F(1)</li>
<li>F(2) &lt;- F(1)</li>
<li>F(3) &lt;- F(2)</li>
<li>F(4) &lt;- F(3)</li>
</ul>
</li>
</ul>
<h2>Practice</h2>
<pre><code class="language-python">'''
Dynamic programming(DP): DP is an algorithm 'design' technique.
                         It has recurrences with overlapping sub-instances.

Memoization: Memoization is one of the optimization techniques for DP.
             Simply storing the results of previouis function calls to reuse it later.

We can solve the Fibonacci problem again while using DP's memoization.
'''
def FibonacciDP(n):
    dicFibonacci = {} #using a dictionary collection variable type for memoization.
    dicFibonacci[0] = 0
    dicFibonacci[1] = 1

    for itr in range(2, n+1): #building up a bigger solutions
        dicFibonacci[itr] = dicFibonacci[itr-1] + dicFibonacci[itr-2]
    return dicFibonacci[n]


for itr in range(0, 10):
    print(FibonacciDP(itr), end=&quot; &quot;)

'''
compare with the recursion's Fibonacci solution in big-O,
    recursion spends O(2^n)
    DP spends O(N)
'''
</code></pre>
<h1>Solving Assembly Line Sceduling problem in Recursion and DynamicProgramming</h1>
<p><a href="https://www.geeksforgeeks.org/assembly-line-scheduling-dp-34/">To understand concept -&gt; Geeks for Geeks: Assembly Line Scheduling</a><br>
<img src="https://media.geeksforgeeks.org/wp-content/uploads/AssembleScheduling1.png"></p>
<h2>1. Using Recursion</h2>
<pre><code class="language-python">'''
There are two assembly lines.
Each assemply line has six stations.
Each 12 stations have their own consuming time.
Also, each line has enter and exit belt, which also needs time to pass.
When you pass the one station, you have two options: 1) stay in the same assembly line and go to the next station
                                                     2) cross to another assembly line through the belt and go to the next station of the line
                                                        (when you cross, it needs another additional time)

'''
class AssemblyLines:
   timeStation = [[7, 9, 3, 4, 8, 4],
                  [8, 5, 6, 4, 5, 7]]
   timeBelt = [[2, 2, 3, 1, 3, 4, 3],
               [4, 2, 1, 2, 2, 1, 2]]
   intCount = 0

   def Scheduling(self, idxLine, idxStation):
      print(&quot;Calculate scheduling: line, station: &quot;, idxLine, idxStation, &quot;(&quot;, self.intCount, &quot;recursion calls )&quot;)
      self.intCount = self.intCount + 1

      #escape when it's not for the recursion
      if idxStation == 0: 
         if idxLine == 1:
            return self.timeBelt[0][0] + self.timeStation[0][0]
         elif idxLine == 2:
            return self.timeBelt[1][0] + self.timeStation[1][0]

      #recursion
      if idxLine == 1:  
         costLine1 = self.Scheduling(1, idxStation-1) + self.timeStation[0][idxStation]
         costLine2 = self.Scheduling(2, idxStation-1) + self.timeStation[0][idxStation] + self.timeBelt[1][idxStation]
      if idxLine == 2:
         costLine1 = self.Scheduling(1, idxStation-1) + self.timeStation[1][idxStation] + self.timeBelt[0][idxStation]
         costLine2 = self.Scheduling(2, idxStation-1) + self.timeStation[1][idxStation]

      if costLine1 &gt; costLine2:
         return costLine2
      else:
         return costLine1
      
   def startScheduling(self):
      numStation = len(self.timeStation[0])
      costLine1 = self.Scheduling(1, numStation - 1) + self.timeBelt[0][numStation]
      costLine2 = self.Scheduling(2, numStation - 1) + self.timeBelt[1][numStation]
      if costLine1 &gt; costLine2:
         return costLine2
      else:
         return costLine1


lines = AssemblyLines()
time = lines.startScheduling() # 125 times calling
print(&quot;Fastest production time:&quot;, time) #38
</code></pre>
<h2>2. Using Dynamic Programming</h2>
<pre><code class="language-python">'''
There are two assembly lines.
Each assemply line has six stations.
Each 12 stations have their own consuming time.
Also, each line has enter and exit belt, which also needs time to pass.
When you pass the one station, you have two options: 1) stay in the same assembly line and go to the next station
                                                     2) cross to another assembly line through the belt and go to the next station of the line
                                                        (when you cross, it needs another additional time)

'''
class AssemblyLines:
   timeStation = [[7, 9, 3, 4, 8, 4],
                  [8, 5, 6, 4, 5, 7]]
   timeBelt = [[2, 2, 3, 1, 3, 4, 3],
               [4, 2, 1, 2, 2, 1, 2]]
   
   #setting up a memoization table
   timeScheduling = [list(range(6)), list(range(6))] #consider it as an empty list
   stationTracing = [list(range(6)), list(range(6))] 

   def startSchedulingDP(self):
      #table initiation
      numStation = len(self.timeStation[0])
      self.timeScheduling[0][0] = self.timeStation[0][0] + self.timeBelt[0][0] 
      self.timeScheduling[1][0] = self.timeStation[1][0] + self.timeBelt[1][0] #Scheduling becomes [[9, _, _, _, _, _], [12, _, _, _, _, _]]

      #dynamic programming
      for itr in range(1, numStation):
         if self.timeScheduling[0][itr-1] &gt; self.timeScheduling[1][itr-1] + self.timeBelt[1][itr]:
            self.timeScheduling[0][itr] = self.timeScheduling[0][itr] + self.timeScheduling[1][itr-1] + self.timeBelt[1][itr]
            self.stationTracing[0][itr] = 1
         else:
            self.timeScheduling[0][itr] = self.timeStation[0][itr] +  self.timeScheduling[0][itr-1]
            self.stationTracing[0][itr] = 0

         if self.timeScheduling[1][itr-1] &gt; self.timeScheduling[0][itr-1] + self.timeBelt[0][itr]:
            self.timeScheduling[1][itr] = self.timeStation[1][itr] + self.timeScheduling[0][itr-1] + self.timeBelt[0][itr]
            self.stationTracing[1][itr] = 0
         else:
            self.timeScheduling[1][itr] = self.timeStation[1][itr] + self.timeScheduling[1][itr-1]
            self.stationTracing[1][itr] = 1
      
      #caculate the time cost overall
      costLine1 = self.timeScheduling[0][numStation-1] + self.timeBelt[0][numStation] #adding last belt's time
      costLine2 = self.timeScheduling[1][numStation-1] + self.timeBelt[1][numStation] #adding last belt's time
      
      #compare between two paths
      if costLine1 &gt; costLine2:
         return costLine2, 1
      else:
         return costLine1, 0

   def printTracing(self, lineTracing):
      numStation = len(self.timeStation[0])
      print(&quot;Line: &quot;, lineTracing,&quot;, Station:&quot;, numStation)
      for itr in range(numStation-1, 0, -1):
         lineTracing = self.stationTracing[lineTracing][itr]
         print(&quot;Line:&quot;, lineTracing, &quot;, Station:&quot;, itr)


lines = AssemblyLines()
time, lineTracing = lines.startSchedulingDP() # 6 times calling
print(&quot;Fastest produiction time:&quot;, time) #38
lines.printTracing(lineTracing)
</code></pre>
<ul>
<li>Dynamic Programming calls only 6 times while Recursion calls 125 times</li>
</ul>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120358" class="record">
            <header>
                <h1 class="record-title">edwith_LSnDP_Ch5</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120358</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://www.edwith.org/datastructure-2019s">EDWITH: Linear Structure and Dynamic Programming</a><br>
Chapter 5: The Application of Computer Simulations</p>
<h1>Real World's Problems</h1>
<ul>
<li>Developing a specialized solution -&gt; theorical</li>
<li>Developing a managerial solution -&gt; sophisticated</li>
<li>To solve problems, both views to problems are required</li>
<li>To solve porblems, we need to comprehend the problem first</li>
</ul>
<h2>How to comprehend a problem?</h2>
<ul>
<li>Estimate the future of a given status-quo</li>
<li>Estimate the progress of problems factors in the system</li>
<li>However, the real world's problem is not a well-defined progrmming
<ul>
<li>It's hard to model the real world's problem correctly</li>
</ul>
</li>
</ul>
<h1>Why Modeling?</h1>
<ul>
<li>It is difficult to trace the real-world problems</li>
<li>To simplify the problem, we create a model
<ul>
<li>The model is an essence of the real world problems</li>
</ul>
</li>
</ul>
<h2>What to remove from the real-world problem?</h2>
<ul>
<li>Think what to reamin in the model</li>
<li>Think what to remove</li>
<li>Determined by objectives</li>
</ul>
<h1>Two Types of Models</h1>
<h2>Models to solve: numerical modeling</h2>
<ul>
<li>Goal: Converting the real-world problems into formula. Then, finding optimal solutions through solvers</li>
<li>Example:
<ul>
<li>Linear programming</li>
<li>Integer programming</li>
</ul>
</li>
</ul>
<h2>Models to simulate: system dynamics/discrete event models/agent based models</h2>
<ul>
<li>Goal: Approximating to the real world problems. Then, finding the optimal solutions through <strong>repetitive simulations</strong></li>
<li>Example:
<ul>
<li>Event graphs</li>
<li>petri-net</li>
<li>agent modeling</li>
</ul>
</li>
</ul>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120359" class="record">
            <header>
                <h1 class="record-title">edwith_LSnDP_Ch6</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120359</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://www.edwith.org/datastructure-2019s">EDWITH: Linear Structure and Dynamic Programming</a><br>
Chapter 6: Tree</p>
<h1><em>Tree</em> Structure is an abstract data type</h1>
<h2><em>Abstract Data Type</em></h2>
<ul>
<li><em>ADT</em> is specified in
<ul>
<li>Data stored</li>
<li>Operations on the data</li>
<li>Error conditions associated with operations</li>
</ul>
</li>
</ul>
<h2><em>Tree</em> as an abstract data type</h2>
<ul>
<li>Data is stored as a <em>Tree</em> structure</li>
<li>Operations on the data are...
<ul>
<li>including ordinary data structure's operations such as<br>
*- Insert
<ul>
<li>Delete</li>
<li>Search*</li>
</ul>
</li>
<li>In addition, <em>Tree</em> structure has a special operation called
<ul>
<li><em>Traverse</em>: Special searching approaches for trees and networks</li>
</ul>
</li>
</ul>
</li>
</ul>
<h2>Why do we use <em>Tree</em>?</h2>
<ul>
<li><em>Tree</em> structure is good analogy to the various real world structures
<ul>
<li>Corporate structrues</li>
<li>Group bank accounts</li>
<li>Command and control structures</li>
</ul>
</li>
<li><em>Tree</em> structure is easy to <em>Divide</em> and <em>Conquer</em> the data</li>
</ul>
<h2>Compare with <em>Linked List</em> and <em>Tree</em></h2>
<ul>
<li>Linked List is linear structure, one Linked List node has single 'next' variable</li>
<li>One Tree node can have multiple 'next' variables</li>
<li>Particularly, the structure maintains the next nodes as an array or variables</li>
</ul>
<h1>Some terms of <em>Tree</em> structure</h1>
<ul>
<li>Root: the toppest node</li>
<li>Edge: the line from a node to the next node</li>
<li>Parent-Child: When two nodes are linked through the edge, they are in Parent-Child relationship</li>
<li>Siblings: a group of Child nodes</li>
<li>Leave nodes/Terminal nodes: the nodes at bottom</li>
<li>Internal nodes: nodes between Leave and Root</li>
</ul>
<h2><em>Ancestors, Descendants, Path</em></h2>
<ul>
<li>Ancestors of 'x': Every ancestor nodes of the node 'x'</li>
<li>Descendants of 'x': Every descendants nodes of the node 'x'</li>
<li>Path to 'x': the shortest way from Root to the node 'x'</li>
</ul>
<h2><em>Depth, Size, Height, Degree</em></h2>
<ul>
<li>Depth and level of 'x': same as 'Path'</li>
<li>Size of tree: same as many as nodes</li>
<li>Height: the maximum depth of the Tree structure</li>
<li>Degree of 'x': the maximum nodes which the node 'x' could acquire</li>
</ul>
<h2><em>Full Tree</em> and <em>Complete Tree</em></h2>
<ul>
<li>Ful Tree: utilize every next nodes.
<ul>
<li>Tree height = Tree depth</li>
</ul>
</li>
<li>Complete Tree: the next nodes are not fully utilized yet
<ul>
<li>The nodes are filled from the left side of the graphic of the structure</li>
</ul>
</li>
</ul>
<h1>Characteristics of Tree</h1>
<ul>
<li>{Number of edges} = {Number of nodes } - 1</li>
<li>{Depth of Root} = 0</li>
<li>{Maximum number of nodes at level <em>i</em> with degree <em>d</em>} = <em>d</em>$^i$</li>
<li>{Maximum number of leaves with height <em>h</em> and degree <em>d</em>} = <em>d</em>$^ h$</li>
<li>{Maximum size of a tree with height <em>h</em> and degree <em>d</em>} = 1 + <em>d</em> + <em>d</em>$^ 2$ + <em>d</em>$^ 3$ + ... + <em>d</em>$^ h$</li>
<li>{Height of a complete tree with size <em>s</em> and degree <em>d</em>} = [log$ d$(<em>s</em>(<em>d</em> - 1) + 1] - 1</li>
</ul>
<h1><em>Binary Search Tree</em>(<em>BST</em>)</h1>
<h2><em>Binary Tree</em> vs <em>Binary 'search' Tree</em></h2>
<ul>
<li><em>Binary Tree</em> is a tree structure with degree 2</li>
<li><em>Binary Search Tree</em> is a tree structure with degree 2, and designed for a <strong>fast search</strong> of stored data</li>
</ul>
<h2>How to perform a faster search from Binary Search Tree?</h2>
<ul>
<li>Single tree node can be linked to several different tree nodes</li>
<li>It does not need to search as long as the length of data</li>
</ul>
<h2>Let's code Tree node</h2>
<ul>
<li>This time, we will design Tree node has two next variables
<ul>
<li>Left hand side (LHS): usually to lower value</li>
<li>Right hand side (RHS): usually to higher value</li>
</ul>
</li>
<li>Tree node have its own value</li>
<li>Tree node have its parent node</li>
</ul>
<pre><code class="language-python">'''
Tree structure is a type of an adstract data type. (Same as like stack, queue, ...)
Tree structure has operations, 'insert', 'delete', 'research'.
    Tree structure has special search approach called 'Traverse'

Compare with the Singly linked list, Tree structure's node involves several 'next nodes' pointing the next node.
Let's build a node for tree structure within two next nodes.
'''
class TreeNode:
    nodeLHS = None #next node1: Left hand side - when values have &quot;lower&quot; than it's own value
    nodeRHS = None #next node2: Right hand side - when values have &quot;higher&quot; than it's own value
    nodeParent = None #previous node
    value = None

    def __init__(self, value, nodeParent):
        self.value = value
        self.nodeParent = nodeParent

    def getLHS(self):
        return self.nodeLHS
    def getRHS(self):
        return self.nodeRHS
    def getValue(self):
        return self.value
    def getParent(self):
        return self.nodeParent

    def setLHS(self, LHS):
        self.nodeLHS = LHS
    def setRHS(self, RHS):
        self.nodeRHS = RHS
    def setValue(self, value):
        self.value = value
    def setParent(self, nodeParent):
        self.nodeParent = nodeParent
</code></pre>
<h1><em>BST</em>'s operations</h1>
<ul>
<li><em>BST</em> handles thedata sotred throught its root
<ul>
<li><em>Root</em> has its own value</li>
<li><em>Tree</em> instance access to the <em>Root</em></li>
<li>Only through the <em>Root</em>, the <em>Tree</em> instances access to the descendant nodes of the <em>Root</em></li>
</ul>
</li>
</ul>
<h2>Insert operation</h2>
<ul>
<li>Retrieve the current node value</li>
<li>If the value is...
<ul>
<li>...equal to the value to insert
<ul>
<li>Return</li>
</ul>
</li>
<li>...smaller to the value to insert
<ul>
<li>If there is a node in the RHS -&gt; move to RHS (Recursion)</li>
<li>If there is no node in the RHS -&gt; create RHS node -&gt; insert the value</li>
</ul>
</li>
<li>...larger to the value to insert
<ul>
<li>If there is a node in the LHS -&gt; move to LHS (Recursion)</li>
<li>If there is no node in the LHS -&gt; create LHS node -&gt; insert the value</li>
</ul>
</li>
</ul>
</li>
</ul>
<h2>Search operation</h2>
<ul>
<li>Retrieve the current node value</li>
<li>IF the value is...
<ul>
<li>...equal t the value to search
<ul>
<li>Return <strong>True</strong></li>
</ul>
</li>
<li>...smaller to the value to insert
<ul>
<li>If there is a node in the RHS -&gt; move to RHS (Recursion)</li>
<li>If there is no node in the RHS -&gt; return <strong>False</strong></li>
</ul>
</li>
<li>...larger to the value to insert
<ul>
<li>If there is a node in the LHS -&gt; move to LHS (Recursion)</li>
<li>If there is no node in the LHS -&gt; return <strong>False</strong></li>
</ul>
</li>
</ul>
</li>
</ul>
<h2>Delete operation</h2>
<ul>
<li>Delete operation can be under three cases
<ol>
<li>Case1. deleting a node with no children</li>
<li>Case2. deleting a node with 1 child</li>
<li>Case3. deleting a node with 2 children</li>
</ol>
</li>
</ul>
<h3>Case1. deleting a node with no children</h3>
<ul>
<li>Just remove the node by modifying its parent</li>
</ul>
<h3>Case2. deleting a node with 1 child</h3>
<ul>
<li>Replace the node with the child</li>
</ul>
<h3>Case3. deleting a node with 2 children</h3>
<ul>
<li>Find either
<ul>
<li>A maximum in the LHS or RHS</li>
<li>Substitute the node to delete with the found value</li>
<li>Delete the found node in the LHS or RHS</li>
</ul>
</li>
<li>The Tree structure can be looked different after the delete operation which you choosed between LHS or RHS</li>
</ul>
<h2>Escape</h2>
<ul>
<li>In a recursive function, there must be a termination condition that allows the function to exit without calling itself.</li>
</ul>
<h2>Practice(BST, Insert, Search, Delete, find)</h2>
<pre><code class="language-python">'''
Binary Search Tree(BST) is one of the simplest tree structure.
Each node from BST is built within two next nodes, as like the tree node we just built from '6_1_TreeStructure_TreeNode.py'

Let's implement BST with the several operations(insert, delete, search, travse, ...)
'''
from ImportOnly.TreeNode import TreeNode #please check https://github.com/jhlee0637/CodeExercising
from queue import Queue

class BinarySearchTree:
    root = None

    def __init__(self):
        pass

    def insert(self, value, node = None):
        if node is None:
            node = self.root
        #Before insertion, the operation to look for a place to insert is run by recursions. -&gt; so we write down 'escape route' from the recursions on here.
        if self.root is None:
            self.root = TreeNode(value, None) #if there is no root, you are the root now
            return
        if value == node.getValue(): #if the value is equal to the vaule to insert
            return #return already there
        #try to insert to RHS
        if value &gt; node.getValue(): #if the value is smaller than the value to insert, try to insert to RHS
            if node.getRHS() is None:
                node.setRHS(TreeNode(value, node)) #insert to RHS
            else:
                self.insert(value, node.getRHS()) #if RHS is not empty, recursion(go down to the next level of that RHS)
        #try to insert to LHS
        if value &lt; node.getValue():
            if node.getLHS() is None:
                node.setLHS(TreeNode(value, node))
            else:
                self.insert(value, node.hetHLS)

    def search(self, value, node = None):
        #initiating from a root
        if node is None: 
            node = self.root
        #Searching uses recursion too -&gt; we need 'escape' in here too.
        if value == node.getValue():
            return True
        #search from the RHS
        if value &gt; node.getValue(): #if the value is smaller than the value to search, look at the RHS
            if node.getRHS() is None:
                return False #if there is no node in RHS, FALSE
            else:
                return self.search(value, node.getRHS()) #if there is node, recursion
        #search from the LHS
        if value &lt; node.getValue():
            if node.getLHS() is None:
                return False
            else:
                return self.search(value, node.getLHS()) #recursion
        
    def delete(self, value, node = None):
        #initiating from a root
        if node is None:
            node = self.root
        #find a node to delete through recursions
        if node.getValue() &lt; value:
            return self.delete(value, node.getRHS())
        if node.getValue() &gt; value:
            return self.delete(value, node.getLHS())
        #there are three cases to delete node from the BST
        if node.getValue() == value:
            if node.getLHS() is not None and node.getRHS() is not None: #case1) the node has both LHS and RHS -&gt; we can choose one of them to replace the deleted node. -&gt; Let's choose RHS -&gt; the minium value from RHS's childs will replace the current value.
                nodeMin = self.findMin(node.getRHS) #function 'findMin' will be defined later.
                node.setValue(nodeMin.getValue())
                self.delete(nodeMin.getValue(), node.getRHS())
                return
            parent = node.getParent()
            if node.getLHS() is not None: #case2) the node has one child(LHS)
                if node == self.root:
                    self.root = node.getLHS()
                elif parent.getLHS() == node:
                    parent.setLHS(node.getLHS())
                    node.getLHS().setParent(parent)
                else:
                    parent.setRHS(node.getLHS())
                    node.getLHS().setParent(parent)
                return
            if node.getRHS() is not None: #case2) the node has one child(RHS)
                if node == self.root:
                    self.root = node.getRHS()
                elif parent.getLHS() == node:
                    parent.setLHS(node.getRHS())
                    node.getRHS().setParent(parent)
                else:
                    parent.setRHS(node.getRHS())
                    node.getRHS().setParent(parent)
                return
            if node == self.root: #case3) if there are no childs
                self.root = None
            elif parent.getLHS() == node:
                parent.setLHS(None)
            else:
                parent.setRHS(None)
            return

    def findMax(self, node = None): #the end of RHS is the maximum - keep recursion
        if node is None:
            node - self.root
        if node.getRHS() is None:
            return node
        return self.findMax(node.getRHS())

    def findMin(self, node = None): #the end of LHS is the minimum - keep recursion
        if node is None:
            node = self.root
        if node.getLHS() is None:
            return node
        return self.findMin(node.getLHS())
    
    '''
    Tree traversing: Tree structure has several ways of traversing
        Breadth first traverse: 1) Queue-based lavel-order traverse
        Depth first traverse: 1) Pre-order traverse, 2) In-order traverse, 3) Post-order traverse
    '''
    def traverseLevelOrder(self): #Breaadth first traverse
        ret = []
        Q = Queue()
        Q.enqueue(self.root)
        while not Q.isEmpty(): #enqueue the root while until queue is empty
            node = Q.dequeue()
            if node is None:
                continue
            ret.append(node.getValue())
            if node.getLHS() is not None: #if current's LHS exist -&gt; enqueue current LHS
                Q.enqueue(node.getLHS()) 
            if node.getRHS() is not None: #if current's RHS exist -&gt; enqueue current RHS
                Q.enqueue(node.getRHS())
        return ret
    
    def traverseInOrder(self, node = None): #In-order traverse
        if node is None: #Recursion -&gt; need exigt
            node = self.root
        ret = []
        if node.getLHS() is not None: #Put LHS first
            ret = ret + self.traverseInOrder(node.getLHS())
        ret.append(node.getValue()) #Put current second
        if node.getRHS() is not None: #Put RHS last
            ret = ret + self.traverseInOrder(node.getRHS())
        return ret
    
    def traversePreOrder(self, node = None): #Pre-order traverse
        if node is None: #Recursion -&gt; need exigt
            node = self.root
        ret = []
        ret.append(node.getValue()) #Put current first
        if node.getLHS() is not None: #Put LHS second
            ret = ret + self.traversePreOrder(node.getLHS())
        if node.getRHS() is not None: #Put RHS last
            ret = ret + self.traversePreOrder(node.getRHS())
        return ret
    
    def traversePostOrder(self, node = None): #Post-order traverse
        if node is None: #Recursion -&gt; need exigt
            node = self.root
        ret = []
        if node.getLHS() is not None: #Put LHS first
            ret = ret + self.traversePostOrder(node.getLHS())
        if node.getRHS() is not None: #Put RHS second
            ret = ret + self.traversePostOrder(node.getRHS())
        ret.append(node.getValue()) #Put current last
        return ret

</code></pre>
<h1>Performance of <em>Binary Search Tree</em></h1>
<table>
<thead>
<tr>
<th style="text-align:center"></th>
<th style="text-align:center">Linked List</th>
<th style="text-align:center">BST in Average</th>
<th style="text-align:center">BST in Worst Case</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center">Search</td>
<td style="text-align:center">O(n)</td>
<td style="text-align:center">O(log n)</td>
<td style="text-align:center">O(n)</td>
</tr>
<tr>
<td style="text-align:center">Insert after search</td>
<td style="text-align:center">O(1)</td>
<td style="text-align:center">O(1)</td>
<td style="text-align:center">O(1)</td>
</tr>
<tr>
<td style="text-align:center">Delete after search</td>
<td style="text-align:center">O(1)</td>
<td style="text-align:center">O(1)</td>
<td style="text-align:center">O(1)</td>
</tr>
<tr>
<td style="text-align:center">Traverse</td>
<td style="text-align:center">O(n)</td>
<td style="text-align:center">O(n)</td>
<td style="text-align:center">O(n)</td>
</tr>
</tbody>
</table>
<ul>
<li>What is the 'BST in Worst Case'?
<ul>
<li>Imagine the tree structure which every nodes have only single next node</li>
<li>The structure would be looked like a list</li>
</ul>
</li>
<li>At 'Search', 'BST in Average' has O(log n) while 'BST in Worst case' has O(n)
<ul>
<li>This is coming from divide and conquer</li>
<li>Heigh confirms the performance</li>
</ul>
</li>
</ul>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120360" class="record">
            <header>
                <h1 class="record-title">edwith_LSnDP_Ch7</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120360</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://www.edwith.org/datastructure-2019s">EDWITH: Linear Structure and Dynamic Programming</a><br>
Chapter 7: Algorithm Analysis</p>
<h1>What is the Algorithm?</h1>
<ul>
<li>A clearly specified set of simple instructions to be followed to solve a problem</li>
<li>Take input -&gt; give output</li>
<li>Program is a combination of algorithm and data structures</li>
<li>Examples of algorithms...
<ul>
<li>insertion, deletion, search, stack, queue, <strong>sorting</strong>, ...</li>
<li>Examples of <strong>sorting</strong> algorithm
<ul>
<li>Bubble sort, Quick sort, Merge sort, ...</li>
</ul>
</li>
</ul>
</li>
</ul>
<h1>Example of algorithm: Bubble Sort Algorithm</h1>
<ul>
<li>Bubble Sort Algorithm is a simple and basic sorting algorithm</li>
<li>The algorithm to sort the list in descending order</li>
<li>Let's say we have a list, $[3, 2, 1, 5, 4]$
<ul>
<li>compare the first element($3$) with rest of them, one by one, following the list</li>
<li>when you find the bigger element, exchange the location of them</li>
<li>If you reach to the end, start from the second element</li>
</ul>
</li>
</ul>
<table>
<thead>
<tr>
<th style="text-align:center"></th>
<th style="text-align:center"></th>
<th style="text-align:center"></th>
<th style="text-align:center"></th>
<th style="text-align:center"></th>
<th style="text-align:left"></th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center">3*</td>
<td style="text-align:center">2</td>
<td style="text-align:center">1</td>
<td style="text-align:center">5*</td>
<td style="text-align:center">4</td>
<td style="text-align:left">3 is smaller than 5. Exchange the location.</td>
</tr>
<tr>
<td style="text-align:center">5*</td>
<td style="text-align:center">2</td>
<td style="text-align:center">1</td>
<td style="text-align:center">3</td>
<td style="text-align:center">4</td>
<td style="text-align:left">nothing is smller than 5.</td>
</tr>
<tr>
<td style="text-align:center">5</td>
<td style="text-align:center">2*</td>
<td style="text-align:center">1</td>
<td style="text-align:center">3*</td>
<td style="text-align:center">4</td>
<td style="text-align:left">2 is smaller than 3. Exchange the location.</td>
</tr>
<tr>
<td style="text-align:center">5</td>
<td style="text-align:center">3*</td>
<td style="text-align:center">1</td>
<td style="text-align:center">2</td>
<td style="text-align:center">4*</td>
<td style="text-align:left">3 is smaller than 4. Exchange the location.</td>
</tr>
<tr>
<td style="text-align:center">5</td>
<td style="text-align:center">4*</td>
<td style="text-align:center">1</td>
<td style="text-align:center">2</td>
<td style="text-align:center">3</td>
<td style="text-align:left">nothing is smller than 4.</td>
</tr>
<tr>
<td style="text-align:center">5</td>
<td style="text-align:center">4</td>
<td style="text-align:center">1*</td>
<td style="text-align:center">2*</td>
<td style="text-align:center">3</td>
<td style="text-align:left">1 is smaller than 2. Exchange the location.</td>
</tr>
<tr>
<td style="text-align:center">5</td>
<td style="text-align:center">4</td>
<td style="text-align:center">2*</td>
<td style="text-align:center">1</td>
<td style="text-align:center">3*</td>
<td style="text-align:left">2 is smaller than 3. Exchange the location.</td>
</tr>
<tr>
<td style="text-align:center">5</td>
<td style="text-align:center">4</td>
<td style="text-align:center">3*</td>
<td style="text-align:center">1</td>
<td style="text-align:center">2</td>
<td style="text-align:left">nothing is smller than 3.</td>
</tr>
<tr>
<td style="text-align:center">5</td>
<td style="text-align:center">4</td>
<td style="text-align:center">3</td>
<td style="text-align:center">2*</td>
<td style="text-align:center">1</td>
<td style="text-align:left">nothing is smller than 2.</td>
</tr>
</tbody>
</table>
<ul>
<li>We can describe the algorithm in pseudo code</li>
</ul>
<pre><code>for itr1 = 0 to length(list)
	for itr2 = itr + 1 to length(list)
		if list[itr1] &lt; list[itr2]
			swap list[itr1], list[itr2]
return list
</code></pre>
<ul>
<li>Now, we can pratice this pseudo code into practical python code</li>
</ul>
<h2>Python practice</h2>
<pre><code class="language-python">import random

def performSelectionSort(lst):
    for itr1 in range(0, len(lst)):
        for itr2 in range(itr1+1, len(lst)):
            if lst[itr1] &lt; lst[itr2]:
                lst[itr1], lst[itr2] = lst[itr2], lst[itr1]
    return lst


N = 10
lstNumbers = list(range(N))
random.shuffle(lstNumbers)
print(lstNumbers)
print(performSelectionSort(lstNumbers))


lstNumbers2 = [2, 5, 0, 3, 3, 3, 1, 5, 4, 2]
print(lstNumbers2) #[[2, 5, 0, 3, 3, 3, 1, 5, 4, 2]
print(performSelectionSort(lstNumbers2)) #[5, 5, 4, 3, 3, 3, 2, 2, 1, 0]
</code></pre>
<h1>Algorithm Efficiency</h1>
<ul>
<li>Program could be <strong>inefficient</strong>
<ul>
<li>Large data -&gt; running time becomes a big issue</li>
</ul>
</li>
<li>We need a gurantee of the worst-case scenario
<ul>
<li>Worst-case running time of a single transaction</li>
<li>Worst-case transaction request numbers of a single day</li>
</ul>
</li>
</ul>
<h1>Algorithm Analysis</h1>
<ul>
<li>Estimate the resources like...
<ul>
<li>Memory</li>
<li>Communication bandwidth</li>
<li><strong>Computational time</strong></li>
</ul>
</li>
<li>Factors affecting the running time
<ul>
<li>Computer used for executions</li>
<li>Algorithms</li>
<li>Data structures</li>
<li>Input data size</li>
</ul>
</li>
<li>After analyzing the algorithms
<ul>
<li>Estimate the worst-case of the costs by the factors</li>
</ul>
</li>
</ul>
<h2>Simple algorithm analysis</h2>
<ul>
<li>Caculate the iteration of each line</li>
</ul>
<pre><code class="language-python">def calculateIntegerRangeSum(intFrom, intTo):
	intSum=0 # iteration: 1

	for itr in range(intFrom, intTo): # iteration: N
		intSum = intSUm + itr # iteration: N

	return intSum # iteration: 1

print(calculateInegerRangeSum(0,10))
</code></pre>
<ul>
<li>Total iteration<br>
= 1 + N + N + 1<br>
= 2N + 2</li>
</ul>
<h1>Big O notation</h1>
<ul>
<li>
<p>One example of asymptotic notation .</p>
</li>
<li>
<p>Show the worst-case running time<br>
$$\huge f(N) = O(g(N))$$</p>
</li>
<li>
<p>There are positive constants $c$, and $n_0$.<br>
when $N ≥ n_0$,<br>
$$\huge f(N) ≤ c＊g(N)$$</p>
</li>
<li>
<p>Big O notation ignores the lower-order terms, and the constants(coefficients) of highest-order term</p>
<ul>
<li>When we have $N^3$, $N^2$ and $N$ means nothing in terms of Big O</li>
<li>From the growth rate order<br>
$\huge C^N &gt; N^K &gt; N^2 &gt; NlogN &gt; N &gt; logN &gt; C$<br>
$C≥2$ and $K&gt;2$</li>
<li>Therefore, given example of simple algorithm analysis' total iteration $2N+2$ turns into $O(N)$. The constant term $2N$ are disregarded, leaving only the $N$ term in the Big O notation.</li>
</ul>
</li>
<li>
<p>The growth rate of $f(N)$ is less than or equal to the growth rate of  $g(N)$</p>
</li>
<li>
<p>$g(N)$ is an uppder bound on $f(N)$</p>
</li>
</ul>
<h2>Analysis the Bio-O of Bubble Sort Algorithm</h2>
<pre><code class="language-python">def performSelectionSort(lst):
    for itr1 in range(0, len(lst)): # iteration: N
        for itr2 in range(itr1+1, len(lst)): # iteration: N-i
            if lst[itr1] &lt; lst[itr2]: 
                lst[itr1], lst[itr2] = lst[itr2], lst[itr1]
                
    return lst # iteration: 1
</code></pre>
<ul>
<li>
<p>$N$ is the length of the list.</p>
</li>
<li>
<p>About the second $for$ loop, <code>for itr2 in range(itr1+1, len(lst))</code></p>
<ul>
<li>To calculate the efficiency, worst case should be considered.
<ul>
<li>Assume that $for$ loops will reach to the end without problems, and $if$ always results in true</li>
</ul>
</li>
<li>How many items will be given from the second $for$ loop?
<ul>
<li>Consider the frst $for$ loop together</li>
<li><code>for itr1 in range(0, len(lst)</code><br>
$\huge \sum_{i=0}^{N-1}$</li>
<li><code>for itr2 in range(itr1+1, len(lst)</code><br>
$\huge \sum_{j=i+1}^{N-1}$<br>
$$\huge \sum_{i=0}^{N-1} \sum_{j=i+1}^{N-1} 1= {1 \over 2}n^2 - {1 \over 2}n$$</li>
</ul>
</li>
<li>for single item from <code>for itr2 in range(itr1+1, len(lst)</code>,<br>
3 iterations are needed
<ol>
<li>for itr2 in range(itr1+1, len(lst)</li>
<li>if lst[itr1] &lt; lst[itr2] :</li>
<li>lst[itr1], lst[itr2] = lst[itr2], lst[itr1]</li>
</ol>
</li>
<li>Therefore, the iteration for the second $for$ loop multiple three times</li>
</ul>
</li>
<li>
<p>Total iteration<br>
= $\huge n + ({1 \over 2}n^2-{1 \over 2}n)3 + 1$</p>
<blockquote>
<p>Big O notation ignores constants and lower-order terms.</p>
</blockquote>
<p>= $\huge O(N^2)$</p>
</li>
</ul>
<h2>Examples of Big O notation</h2>
<p>$$ O({N^2 \over 2-3N}) = O(N^2)$$</p>
<p>$$O(1+4N) = O(N)$$</p>
<p>$$O(7N^2 + 1-N + 3) =O(N^2)$$</p>
<p>$$\begin{align}<br>
\huge O(logN)=O(log_{2}N) \<br>
(log_{10}N = {{log_{2}N} \over {log_{2}10}}) \<br>
\end{align}$$</p>
<p>$$O(sinN) = O(1)$$</p>
<p>$$ O(10) = O(1)$$</p>
<p>$$O(10^10)=O(1)$$</p>
<p>$$O(log{N}+N)=O(N)$$</p>
<h2>Big O notation of list, stack, and queue</h2>
<table>
<thead>
<tr>
<th style="text-align:center"></th>
<th style="text-align:center">List</th>
<th style="text-align:center">Stack</th>
<th style="text-align:center">Queue</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center">pop</td>
<td style="text-align:center">X</td>
<td style="text-align:center">1 retrieval<br>$O(1)$</td>
<td style="text-align:center">X</td>
</tr>
<tr>
<td style="text-align:center">Push</td>
<td style="text-align:center">X</td>
<td style="text-align:center">1 retrieval<br>$O(1)$</td>
<td style="text-align:center">X</td>
</tr>
<tr>
<td style="text-align:center">Enqueue</td>
<td style="text-align:center">X</td>
<td style="text-align:center">X</td>
<td style="text-align:center">1 retrieval<br>$O(1)$</td>
</tr>
<tr>
<td style="text-align:center">Dequeue</td>
<td style="text-align:center">X</td>
<td style="text-align:center">X</td>
<td style="text-align:center">1 retrieval<br>$O(1)$</td>
</tr>
<tr>
<td style="text-align:center">Search</td>
<td style="text-align:center">$i$ retrieval<br>(if the target instance<br>at $i^th$ in the list)<br>$O(N)$</td>
<td style="text-align:center">X<br>(Does not allow search<br>in the stack)</td>
<td style="text-align:center">X<br>(Does not allow search<br>in the queue)</td>
</tr>
</tbody>
</table>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120361" class="record">
            <header>
                <h1 class="record-title">Ensemble</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120361</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                

                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120362" class="record">
            <header>
                <h1 class="record-title">Exon</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120362</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                

                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120363" class="record">
            <header>
                <h1 class="record-title">Feature Extraction</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120363</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                

                
            </main>

            <footer class="list-content-container">
                

                
                <details class="record-backlinks" open="open">
                    <summary class="menu-title">
                        backlinks
                        <span class="badge">1</span>
                    </summary>

                    <ul class="record-backlinks-list">
                        
                        <li data-target-id="20230924120348">
                            <a href="#20230924120348">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Curse of The Dimensionality</span>
                                <div class="record-links-context">
                                    <h2>How to solve?</h2>
<p>Feature Selection<br>
<em class="id-context" data-target-id="20230924120363">Feature Extraction</em></p>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                
            </footer>
        </article>
        
        <article id="20230924120364" class="record">
            <header>
                <h1 class="record-title">Feature Selection</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120364</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                

                
            </main>

            <footer class="list-content-container">
                

                
                <details class="record-backlinks" open="open">
                    <summary class="menu-title">
                        backlinks
                        <span class="badge">2</span>
                    </summary>

                    <ul class="record-backlinks-list">
                        
                        <li data-target-id="20230924120348">
                            <a href="#20230924120348">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Curse of The Dimensionality</span>
                                <div class="record-links-context">
                                    <h2>How to solve?</h2>
<p><em class="id-context" data-target-id="20230924120364">Feature Selection</em><br>
Feature Extraction</p>

                                </div>
                            </a>
                        </li>
                        
                        <li data-target-id="20230924120366">
                            <a href="#20230924120366">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Integration Strategies of Multi-omics Data for Machine Learning Analysis</span>
                                <div class="record-links-context">
                                    <h1>1. introduction</h1>
<h3>왜 multi-omics 연구가 필요할까?</h3>
<ul>
<li>맞춤의학, 정밀의료의 부상
<ul>
<li>개인의 임상데이터를 통해 진단하고 적절한 치료법을 제시</li>
<li>임상데이터로 genomics, epigenomics, trasncriptiomics, proteomics, metabolomics 같은 데이터를 수집함</li>
<li>과거에는 이러한 각각의 omics 데이터를 따로 관찰해서 진단에 활용하려고 했음</li>
<li>그러나 대부분의 질병은 omics 데이터들을 서로 관통하면서 연결되어있음</li>
<li>따라서 여러 omics 데이터를 합쳐서 통합적인 관점으로 보는 multi-omics 연구가 필요함</li>
</ul>
</li>
</ul>
<h3>Multi-omics 연구의 장단점은?</h3>
<ul>
<li>한 omics 데이터에서 부족한 부분을 다른 omics 데이터를 참고하여 메꾼다던가, 하나의 큰 시스템에 대한 그림을 얻어낸다던가 하는 장점이 있다.</li>
<li>그러나 적은 샘플에서 가져온 데이터만 있다는 점이나, 개개의 생물학 데이터가 갖고있는 오류를 넘어서서 데이터를 합쳐야 한다는 어려움들이 단점으로 남는다.
<ul>
<li>특히 multiomics 연구에서 데이터를 합치는(integration) 작업이 정확하게 이루어지지 않으면, 오류 투성이로 합쳐진 데이터집합에서 결과를 제대로 얻는 것은 불가능하다.</li>
</ul>
</li>
</ul>
<h3>Data integration</h3>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6242341/">Zitnik et al. (2019)</a>은 integration의 방법을 크게 두 가지로 분류하였다.
<ul>
<li>Horizontal inegration: 동일 omics 여러 샘플</li>
<li>Vertical integration: 여러 omics 동일 샘플</li>
</ul>
</li>
<li>이번 리뷰에서는 Vertical integration을 위주로 리뷰할 것이다.<br>
- Vertical integration의 방법을 수학적인 관점에서 bayesian, network-based, deep learning-based, kernel-based, matrix factorization-based로 나눠서 볼 것이다.
<ul>
<li>또한 <a href="https://pubmed.ncbi.nlm.nih.gov/25582081/">Ritchie et al.(2015)</a>의 분류방법에 따라 vertical integration(논문에서는 meta-dimensinal)을 concatenation-based, transformation-based, model-based로 나눠서 볼 것이다.</li>
</ul>
</li>
</ul>
<h1>2. Challenges</h1>
<ul>
<li>생물학 데이터는 필연적으로 오류를 동반하고 있는데, 이렇게 깨끗하지 못한 여러 omics 데이터를 종합하여 특정한 패턴을 관찰한다는 것은 쉽지 않다.</li>
<li>또한 multi-omics 연구의 특성상 적은 샘플에서 많은 feature를 관찰해야 하는데, 이는 machine learing에서 가장 이상적이지 못한 상황으로 차원의 저주(Curse of The Dimensionality)라고 불리운다.
<ul>
<li>이런 상황에서 ML을 통해 데이터를 훈련시키면 overfit된다.</li>
</ul>
</li>
<li>생물학 omics 데이터들이 각각 너무 성격이 다르다는 점도 하나의 문제이다.
<ul>
<li>유전자 발현 데이터의 경우 수만개의 variable이 있을 수 있지만, 대사물질 데이터는 기껏해서 몇천일 수가 있다.</li>
<li>이렇게 서로 다른 데이터들 사이의 수, 종류, 연속성, 등의 차이를 극복하고 데이터들을 잘 integration 해야한다는 어려움이 있다.</li>
</ul>
</li>
<li>데이터 간의 class imbalance 또한 ML 훈려에 악영향을 미친다.
<ul>
<li>데이터에서 특정 class의 데이터만 기하급수적으로 많다면, 이를 바탕으로 훈련시키는 것이 부적절하다.</li>
<li>왜냐하면 ML 훈련은 모든 class의 데이터 량이 엇비슷하다고 가정하고 훈련되기 때문이다.</li>
<li>이를 해결하기 위한 방법으로 sampling, cost-sensitive learning과 같은 방법이 존재한다.</li>
</ul>
</li>
<li>Missing data 문제의 경우 해당 샘플을 지워버린다던가 혹은 Random Forest, K-Nearest Neighbor같은 방법이 있다.</li>
</ul>
<h1>3. Main integration strategies</h1>
<h3>Integration 5가지 소개</h3>
<ul>
<li><strong>Early integration</strong>이란, 데이터들의 복잡도를 낮추지 않고, 단순히 둘 이상의 데이터들을 합친 뒤(wise concatenation) 그 결과물인 matrix를 ML모델에 넣는 것이다.
<ul>
<li>현실적으로는 단순하게 합쳐진 데이터는 여전히 복잡하기 때문에 ML모델이 잘 돌아가지 않는다다</li>
</ul>
</li>
<li><strong>Mixed integration</strong>이란 데이터들의 복잡도를 개별적으로 낮춰서 사용하는 것이다.</li>
<li><strong>Intermediate integration</strong>이란 여러 데이터들의 복잡도를 함께 낮춰서 사용하는 것이다.</li>
<li><strong>Late integration</strong>이란 정반대로 데이터를 합치지 않고 각각의 omics 데이터를 개별적으로 분석하는 것이다.</li>
<li><strong>Hierarchical integration</strong>이란 DNA-RNA-단백질로 이루어지는 시스템을 표현한 central dogma에 입각해서 omics 데이터 간의 조절 관계를 반영하여 합치는 것이다.</li>
</ul>
<h2>3.1 Dimensionality reduction for multi-omic integration</h2>
<p><strong>Dimensionality reduction은 왜 필요할까?</strong></p>
<ul>
<li>Multi-omics 연구에 있어서 샘플들이 너무 많은 variable을 가질 경우 machine learning 기술 적용에 차질을 빚는다. 따라서 variable의 수를 줄여서 차원의 축소와 데이터셋의 noise 감소를 유도한다.</li>
<li>차원축소는 필수적이지는 않고, 위에서 얘기했던 5가지의 integration 방법에 모두 적용이 가능하다.
<ul>
<li>다만 early integration와 intermediate integration의 경우 차원축소를 미리 적용하는 것이 정확도 향상에 기여한다.</li>
</ul>
</li>
<li>차원축소를 위한 방법은 크게 두 가지가 있다:
<ul>
<li>Feature selection</li>
<li>Redundant varaibles and feature extraction</li>
</ul>
</li>
</ul>
<h3>3.1.1. <em class="id-context" data-target-id="20230924120364">Feature Selection</em> (FS)</h3>
<ul>
<li>Feature를 줄이는 작업</li>
<li>그 결과 데이터의 복잡도, noise가 줄어들어서 데이터가 깔끔해진다.</li>
<li>데이터가 깔끔해지니까, ML 모델이 overfitting될 위험도 적어진다.</li>
<li>데이터가 깔끔해져서 해석하기도 편해진다.</li>
<li>너무 많은 variable이 삭제될 경우?
<ul>
<li>데이터셋 사이에 feautre의 수가 다르면 좋지 않다</li>
<li>Feature Selection은 각 omics block이 동일한 수의 feature를 가지도록 할 수 있다</li>
</ul>
</li>
<li>Feature Selection은 세 가지로 나뉜다:
<ul>
<li>Filter-based Feature Selection</li>
<li>Wrapper-based Feature Selection</li>
<li>Embedded Feature Selection</li>
</ul>
</li>
</ul>
<h3>Filter-based Feature Selection</h3>
<ul>
<li>ML 모델과 상관없이 적용 가능</li>
<li>Feature 수를 줄일 수 있는 방향으로 관련된 variable을 추려내는 통계학적인 분석방법임</li>
<li><strong>CFS, RCA</strong> 같이 correlation 방법을 사용하거나,</li>
<li><strong>ReliefF</strong> 같이 distance 방법을 사용하거나,</li>
<li><strong>mRMR</strong> 같이 information gain 방법을 사용함</li>
</ul>
<h3>Wrapper-based Feature Selection</h3>
<ul>
<li>ML 모델을 활용하여 서로 다른 feature 그룹을 학습시키고 평가한다.</li>
<li>가장 성능이 좋았던 feature 그룹만 선택한다.</li>
<li>Recursive Feature Selection이 한 가지 예시이다.
<ul>
<li>먼저 전체 feauture를 대상으로 모델을 학습시킨다.</li>
<li>그 다음 모델 퍼포먼스에 기여도가 떨어지는 feature를 차례대로 제거해간다.</li>
</ul>
</li>
<li>이 방법은 어느 ML 모델에서든 활용 가능하다는 장점이 있다.</li>
<li>그러나 반복적으로 ML 모델을 훈련시켜야 하기 때문에, 데이터셋이 크다면 컴퓨터 성능에 따라 시간이 오래 걸린다는 단점이 있다.</li>
</ul>
<h3>Embedded Feature Selection</h3>
<ul>
<li>Classifier 모델 안에 feature selection 알고리즘이 넣어져있는 경우</li>
<li>Tree-based feature selection과 regularization 방법이 있다.
<ul>
<li>Regularization 방법은 loss function과 penalization funtion을 결합한 방법이다.
<ul>
<li>Loss funtion은 모델이 주어진 데이터에 얼마나 잘 맞는지 확인하는 함수다.</li>
<li>Penalization funtion은 복잡도(complexity)에 대하여 페널티를 부가하는 방식으로 feature 수를 적은 쪽으로 유도한다.</li>
</ul>
</li>
</ul>
</li>
<li>Embedded Feature Selection 모델은 Wrapper-based Feature Selection과 마찬가지로 ML모델과 함께 사용되지만, 컴퓨터 자원을 덜 소비하기 때문에 컴퓨터 성능에 덜 의존적이라는 장점이 있다.</li>
</ul>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                
            </footer>
        </article>
        
        <article id="20230924120365" class="record">
            <header>
                <h1 class="record-title">Greatest Common Divisor, Least Common Multiple, and Euclidean algorithm</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120365</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://codingpractices.tistory.com/34">[Python] 최소공배수, 최대공약수란? 파이썬 알고리즘으로 쉽게 구현하기</a><br>
<a href="https://en.wikipedia.org/wiki/Greatest_common_divisor">Greatest common divisor</a><br>
<a href="https://en.wikipedia.org/wiki/Least_common_multiple">Least common multiple</a><br>
<a href="https://en.wikipedia.org/wiki/Euclidean_algorithm">Euclidean algorithm</a></p>
<h1>Greatest Common Divisor(GCD), 최대공약수</h1>
<h2>What is GCD</h2>
<ul>
<li>In mathematics, the greatest common divisor (GCD) of two or more integers, which are not all zero, is the largest positive integer that divides each of the integers.
<ul>
<li>Divisors of 8
<ul>
<li>1, 2, 4, 8</li>
</ul>
</li>
<li>Divisors of 10
<ul>
<li>1, 2, 5, 10</li>
</ul>
</li>
<li>Common divisors of 8 and 10
<ul>
<li>1, 2</li>
</ul>
</li>
<li>Greatest Common Divisor
<ul>
<li>2</li>
</ul>
</li>
</ul>
</li>
</ul>
<h2>What is the feature of GCD</h2>
<ul>
<li>When you divide two numbers by their GCD, there is no remainder.</li>
</ul>
<h2>Python practice</h2>
<pre><code class="language-python">A = 8
B = 10

for number in range(1, min(A, B)):
	if number!=1 and A%number == 0 and B%number == 0:
		print(number)
		break
</code></pre>
<h1>Least Common Multiple(LCM), 최소공배수</h1>
<h2>What is LCM</h2>
<ul>
<li>In arithmetic and number theory, the least common multiple, lowest common multiple, or smallest common multiple of two integers $\huge a$ and $\huge b$, usually denoted by $\huge lcm(a, b)$, is the smallest positive integer that is divisible by both $\huge a$ and $\huge b$.
<ul>
<li>Multiples of 4
<ul>
<li>4 , 8 , 12 , 16 , 20 , 24 , 28, 32, 36, 40, 44, 48, ...</li>
</ul>
</li>
<li>Multiples of 6
<ul>
<li>6 , 12 , 18 , 24 , 30 , 36 , 42, 48, ...</li>
</ul>
</li>
<li>Common multiples of 4 and 6 are the numbers that are in both lists
<ul>
<li>12, 24, 36, 48, 60, 72, ...</li>
</ul>
</li>
<li>Least Common Multiple
<ul>
<li>12</li>
</ul>
</li>
</ul>
</li>
</ul>
<h2>What is the feature of LCM</h2>
<ul>
<li>When LCM is divided by two numbers, there is no remainder.</li>
<li>LCM can be computed from the GCD with the formula, $$\Huge lcm(a,b) = \frac{|ab|}{gcd(a,b)}$$</li>
</ul>
<h2>Python practice</h2>
<pre><code class="language-python">A = 4
B = 6

for number in range(max(A, B), (A*B)+1):
	if number%A == 0 and number%B ==0:
		print(number)
		break
</code></pre>
<h1>Euclidean Algorithm, 유클리드 호제법</h1>
<ul>
<li>In mathematics, the Euclidean algorithm, or Euclid's algorithm, is an efficient method for computing the greatest common divisor (GCD) of two integers (numbers), the largest number that divides them both without a remainder.
<ul>
<li>Remainder of 10/12 = 10</li>
<li>Remainder of 12/10 = 2</li>
<li>Remainder of 10/2 = 0
<ul>
<li>2 is the GCD of 10, 12</li>
</ul>
</li>
</ul>
</li>
</ul>
<p>|10|%|12|=|10|<br>
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|<br>
| | |12|%|10|=|2|<br>
| | |  | |10|%|2|=|0|</p>
<h2>Python practice</h2>
<pre><code class="language-python">A = 10
B = 12

while(B):
	A, B = B, A%B

print (&quot;GCD:&quot;, A)
</code></pre>
<ul>
<li>To get LCD with Euclidean Algorithm, using the feature of LCD with the formula, $\Huge lcm(a,b) = \frac{|ab|}{gcd(a,b)}$</li>
</ul>
<pre><code class="language-python">#GCD
A = 10
B = 12

while(B):
	A, B = B, A%B

GCD=A


#LCM
A = 10
B = 12

LCM=(A*B)//GCD

print(&quot;LCM:&quot;, LCM)
</code></pre>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120366" class="record">
            <header>
                <h1 class="record-title">Integration Strategies of Multi-omics Data for Machine Learning Analysis</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120366</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8258788/">Picard M, Scott-Boyer MP, Bodein A, Périn O, Droit A. Integration strategies of multi-omics data for machine learning analysis. Comput Struct Biotechnol J. 2021 Jun 22;19:3735-3746. doi: 10.1016/j.csbj.2021.06.030. PMID: 34285775; PMCID: PMC8258788.</a></p>
<ul>
<li>2021년 기준으로 <a href="#20230924120377" title="Multi-omics" class="record-link">Multi-omics</a> <a href="#20230924120350" title="Data Integration" class="record-link">Data Integration</a> 방법을 5가지로 분류함
<ul>
<li>early</li>
<li>mixed</li>
<li>intermediate</li>
<li>late</li>
<li>hierarchical<br>
<img src="https://ars.els-cdn.com/content/image/1-s2.0-S2001037021002683-ga1_lrg.jpg" width=800></li>
</ul>
</li>
</ul>
<h1>1. introduction</h1>
<h3>왜 multi-omics 연구가 필요할까?</h3>
<ul>
<li>맞춤의학, 정밀의료의 부상
<ul>
<li>개인의 임상데이터를 통해 진단하고 적절한 치료법을 제시</li>
<li>임상데이터로 genomics, epigenomics, trasncriptiomics, proteomics, metabolomics 같은 데이터를 수집함</li>
<li>과거에는 이러한 각각의 omics 데이터를 따로 관찰해서 진단에 활용하려고 했음</li>
<li>그러나 대부분의 질병은 omics 데이터들을 서로 관통하면서 연결되어있음</li>
<li>따라서 여러 omics 데이터를 합쳐서 통합적인 관점으로 보는 multi-omics 연구가 필요함</li>
</ul>
</li>
</ul>
<h3>Multi-omics 연구의 장단점은?</h3>
<ul>
<li>한 omics 데이터에서 부족한 부분을 다른 omics 데이터를 참고하여 메꾼다던가, 하나의 큰 시스템에 대한 그림을 얻어낸다던가 하는 장점이 있다.</li>
<li>그러나 적은 샘플에서 가져온 데이터만 있다는 점이나, 개개의 생물학 데이터가 갖고있는 오류를 넘어서서 데이터를 합쳐야 한다는 어려움들이 단점으로 남는다.
<ul>
<li>특히 multiomics 연구에서 데이터를 합치는(integration) 작업이 정확하게 이루어지지 않으면, 오류 투성이로 합쳐진 데이터집합에서 결과를 제대로 얻는 것은 불가능하다.</li>
</ul>
</li>
</ul>
<h3>Data integration</h3>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6242341/">Zitnik et al. (2019)</a>은 integration의 방법을 크게 두 가지로 분류하였다.
<ul>
<li>Horizontal inegration: 동일 omics 여러 샘플</li>
<li>Vertical integration: 여러 omics 동일 샘플</li>
</ul>
</li>
<li>이번 리뷰에서는 Vertical integration을 위주로 리뷰할 것이다.<br>
- Vertical integration의 방법을 수학적인 관점에서 bayesian, network-based, deep learning-based, kernel-based, matrix factorization-based로 나눠서 볼 것이다.
<ul>
<li>또한 <a href="https://pubmed.ncbi.nlm.nih.gov/25582081/">Ritchie et al.(2015)</a>의 분류방법에 따라 vertical integration(논문에서는 meta-dimensinal)을 concatenation-based, transformation-based, model-based로 나눠서 볼 것이다.</li>
</ul>
</li>
</ul>
<h1>2. Challenges</h1>
<ul>
<li>생물학 데이터는 필연적으로 오류를 동반하고 있는데, 이렇게 깨끗하지 못한 여러 omics 데이터를 종합하여 특정한 패턴을 관찰한다는 것은 쉽지 않다.</li>
<li>또한 multi-omics 연구의 특성상 적은 샘플에서 많은 feature를 관찰해야 하는데, 이는 machine learing에서 가장 이상적이지 못한 상황으로 차원의 저주(<a href="#20230924120348" title="Curse of The Dimensionality" class="record-link">Curse of The Dimensionality</a>)라고 불리운다.
<ul>
<li>이런 상황에서 ML을 통해 데이터를 훈련시키면 overfit된다.</li>
</ul>
</li>
<li>생물학 omics 데이터들이 각각 너무 성격이 다르다는 점도 하나의 문제이다.
<ul>
<li>유전자 발현 데이터의 경우 수만개의 variable이 있을 수 있지만, 대사물질 데이터는 기껏해서 몇천일 수가 있다.</li>
<li>이렇게 서로 다른 데이터들 사이의 수, 종류, 연속성, 등의 차이를 극복하고 데이터들을 잘 integration 해야한다는 어려움이 있다.</li>
</ul>
</li>
<li>데이터 간의 class imbalance 또한 ML 훈려에 악영향을 미친다.
<ul>
<li>데이터에서 특정 class의 데이터만 기하급수적으로 많다면, 이를 바탕으로 훈련시키는 것이 부적절하다.</li>
<li>왜냐하면 ML 훈련은 모든 class의 데이터 량이 엇비슷하다고 가정하고 훈련되기 때문이다.</li>
<li>이를 해결하기 위한 방법으로 sampling, cost-sensitive learning과 같은 방법이 존재한다.</li>
</ul>
</li>
<li>Missing data 문제의 경우 해당 샘플을 지워버린다던가 혹은 Random Forest, K-Nearest Neighbor같은 방법이 있다.</li>
</ul>
<h1>3. Main integration strategies</h1>
<h3>Integration 5가지 소개</h3>
<ul>
<li><strong>Early integration</strong>이란, 데이터들의 복잡도를 낮추지 않고, 단순히 둘 이상의 데이터들을 합친 뒤(wise concatenation) 그 결과물인 matrix를 ML모델에 넣는 것이다.
<ul>
<li>현실적으로는 단순하게 합쳐진 데이터는 여전히 복잡하기 때문에 ML모델이 잘 돌아가지 않는다다</li>
</ul>
</li>
<li><strong>Mixed integration</strong>이란 데이터들의 복잡도를 개별적으로 낮춰서 사용하는 것이다.</li>
<li><strong>Intermediate integration</strong>이란 여러 데이터들의 복잡도를 함께 낮춰서 사용하는 것이다.</li>
<li><strong>Late integration</strong>이란 정반대로 데이터를 합치지 않고 각각의 omics 데이터를 개별적으로 분석하는 것이다.</li>
<li><strong>Hierarchical integration</strong>이란 DNA-RNA-단백질로 이루어지는 시스템을 표현한 central dogma에 입각해서 omics 데이터 간의 조절 관계를 반영하여 합치는 것이다.</li>
</ul>
<h2>3.1 Dimensionality reduction for multi-omic integration</h2>
<p><strong>Dimensionality reduction은 왜 필요할까?</strong></p>
<ul>
<li>Multi-omics 연구에 있어서 샘플들이 너무 많은 variable을 가질 경우 machine learning 기술 적용에 차질을 빚는다. 따라서 variable의 수를 줄여서 차원의 축소와 데이터셋의 noise 감소를 유도한다.</li>
<li>차원축소는 필수적이지는 않고, 위에서 얘기했던 5가지의 integration 방법에 모두 적용이 가능하다.
<ul>
<li>다만 early integration와 intermediate integration의 경우 차원축소를 미리 적용하는 것이 정확도 향상에 기여한다.</li>
</ul>
</li>
<li>차원축소를 위한 방법은 크게 두 가지가 있다:
<ul>
<li>Feature selection</li>
<li>Redundant varaibles and feature extraction</li>
</ul>
</li>
</ul>
<h3>3.1.1. <a href="#20230924120364" title="Feature Selection" class="record-link">Feature Selection</a> (FS)</h3>
<ul>
<li>Feature를 줄이는 작업</li>
<li>그 결과 데이터의 복잡도, noise가 줄어들어서 데이터가 깔끔해진다.</li>
<li>데이터가 깔끔해지니까, ML 모델이 overfitting될 위험도 적어진다.</li>
<li>데이터가 깔끔해져서 해석하기도 편해진다.</li>
<li>너무 많은 variable이 삭제될 경우?
<ul>
<li>데이터셋 사이에 feautre의 수가 다르면 좋지 않다</li>
<li>Feature Selection은 각 omics block이 동일한 수의 feature를 가지도록 할 수 있다</li>
</ul>
</li>
<li>Feature Selection은 세 가지로 나뉜다:
<ul>
<li>Filter-based Feature Selection</li>
<li>Wrapper-based Feature Selection</li>
<li>Embedded Feature Selection</li>
</ul>
</li>
</ul>
<h3>Filter-based Feature Selection</h3>
<ul>
<li>ML 모델과 상관없이 적용 가능</li>
<li>Feature 수를 줄일 수 있는 방향으로 관련된 variable을 추려내는 통계학적인 분석방법임</li>
<li><strong>CFS, RCA</strong> 같이 correlation 방법을 사용하거나,</li>
<li><strong>ReliefF</strong> 같이 distance 방법을 사용하거나,</li>
<li><strong>mRMR</strong> 같이 information gain 방법을 사용함</li>
</ul>
<h3>Wrapper-based Feature Selection</h3>
<ul>
<li>ML 모델을 활용하여 서로 다른 feature 그룹을 학습시키고 평가한다.</li>
<li>가장 성능이 좋았던 feature 그룹만 선택한다.</li>
<li>Recursive Feature Selection이 한 가지 예시이다.
<ul>
<li>먼저 전체 feauture를 대상으로 모델을 학습시킨다.</li>
<li>그 다음 모델 퍼포먼스에 기여도가 떨어지는 feature를 차례대로 제거해간다.</li>
</ul>
</li>
<li>이 방법은 어느 ML 모델에서든 활용 가능하다는 장점이 있다.</li>
<li>그러나 반복적으로 ML 모델을 훈련시켜야 하기 때문에, 데이터셋이 크다면 컴퓨터 성능에 따라 시간이 오래 걸린다는 단점이 있다.</li>
</ul>
<h3>Embedded Feature Selection</h3>
<ul>
<li>Classifier 모델 안에 feature selection 알고리즘이 넣어져있는 경우</li>
<li>Tree-based feature selection과 regularization 방법이 있다.
<ul>
<li>Regularization 방법은 loss function과 penalization funtion을 결합한 방법이다.
<ul>
<li>Loss funtion은 모델이 주어진 데이터에 얼마나 잘 맞는지 확인하는 함수다.</li>
<li>Penalization funtion은 복잡도(complexity)에 대하여 페널티를 부가하는 방식으로 feature 수를 적은 쪽으로 유도한다.</li>
</ul>
</li>
</ul>
</li>
<li>Embedded Feature Selection 모델은 Wrapper-based Feature Selection과 마찬가지로 ML모델과 함께 사용되지만, 컴퓨터 자원을 덜 소비하기 때문에 컴퓨터 성능에 덜 의존적이라는 장점이 있다.</li>
</ul>


                
            </main>

            <footer class="list-content-container">
                
                <details class="record-links" open="open">
                    <summary class="menu-title">
                        links
                        <span class="badge">4</span>
                    </summary>

                    <ul class="record-links-list">
                        
                        <li data-target-id="20230924120377">
                            <a href="#20230924120377">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Multi-omics</span>
                                <div class="record-links-context">
                                    <p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8258788/">Picard M, Scott-Boyer MP, Bodein A, Périn O, Droit A. Integration strategies of multi-omics data for machine learning analysis. Comput Struct Biotechnol J. 2021 Jun 22;19:3735-3746. doi: 10.1016/j.csbj.2021.06.030. PMID: 34285775; PMCID: PMC8258788.</a></p>
<ul>
<li>2021년 기준으로 <em class="id-context" data-target-id="20230924120377">Multi-omics</em> Data Integration 방법을 5가지로 분류함
<ul>
<li>early</li>
<li>mixed</li>
<li>intermediate</li>
<li>late</li>
<li>hierarchical<br>
<img src="https://ars.els-cdn.com/content/image/1-s2.0-S2001037021002683-ga1_lrg.jpg" width=800></li>
</ul>
</li>
</ul>

                                </div>
                            </a>
                        </li>
                        
                        <li data-target-id="20230924120350">
                            <a href="#20230924120350">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Data Integration</span>
                                <div class="record-links-context">
                                    <p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8258788/">Picard M, Scott-Boyer MP, Bodein A, Périn O, Droit A. Integration strategies of multi-omics data for machine learning analysis. Comput Struct Biotechnol J. 2021 Jun 22;19:3735-3746. doi: 10.1016/j.csbj.2021.06.030. PMID: 34285775; PMCID: PMC8258788.</a></p>
<ul>
<li>2021년 기준으로 Multi-omics <em class="id-context" data-target-id="20230924120350">Data Integration</em> 방법을 5가지로 분류함
<ul>
<li>early</li>
<li>mixed</li>
<li>intermediate</li>
<li>late</li>
<li>hierarchical<br>
<img src="https://ars.els-cdn.com/content/image/1-s2.0-S2001037021002683-ga1_lrg.jpg" width=800></li>
</ul>
</li>
</ul>

                                </div>
                            </a>
                        </li>
                        
                        <li data-target-id="20230924120348">
                            <a href="#20230924120348">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Curse of The Dimensionality</span>
                                <div class="record-links-context">
                                    <h1>1. introduction</h1>
<h3>왜 multi-omics 연구가 필요할까?</h3>
<ul>
<li>맞춤의학, 정밀의료의 부상
<ul>
<li>개인의 임상데이터를 통해 진단하고 적절한 치료법을 제시</li>
<li>임상데이터로 genomics, epigenomics, trasncriptiomics, proteomics, metabolomics 같은 데이터를 수집함</li>
<li>과거에는 이러한 각각의 omics 데이터를 따로 관찰해서 진단에 활용하려고 했음</li>
<li>그러나 대부분의 질병은 omics 데이터들을 서로 관통하면서 연결되어있음</li>
<li>따라서 여러 omics 데이터를 합쳐서 통합적인 관점으로 보는 multi-omics 연구가 필요함</li>
</ul>
</li>
</ul>
<h3>Multi-omics 연구의 장단점은?</h3>
<ul>
<li>한 omics 데이터에서 부족한 부분을 다른 omics 데이터를 참고하여 메꾼다던가, 하나의 큰 시스템에 대한 그림을 얻어낸다던가 하는 장점이 있다.</li>
<li>그러나 적은 샘플에서 가져온 데이터만 있다는 점이나, 개개의 생물학 데이터가 갖고있는 오류를 넘어서서 데이터를 합쳐야 한다는 어려움들이 단점으로 남는다.
<ul>
<li>특히 multiomics 연구에서 데이터를 합치는(integration) 작업이 정확하게 이루어지지 않으면, 오류 투성이로 합쳐진 데이터집합에서 결과를 제대로 얻는 것은 불가능하다.</li>
</ul>
</li>
</ul>
<h3>Data integration</h3>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6242341/">Zitnik et al. (2019)</a>은 integration의 방법을 크게 두 가지로 분류하였다.
<ul>
<li>Horizontal inegration: 동일 omics 여러 샘플</li>
<li>Vertical integration: 여러 omics 동일 샘플</li>
</ul>
</li>
<li>이번 리뷰에서는 Vertical integration을 위주로 리뷰할 것이다.<br>
- Vertical integration의 방법을 수학적인 관점에서 bayesian, network-based, deep learning-based, kernel-based, matrix factorization-based로 나눠서 볼 것이다.
<ul>
<li>또한 <a href="https://pubmed.ncbi.nlm.nih.gov/25582081/">Ritchie et al.(2015)</a>의 분류방법에 따라 vertical integration(논문에서는 meta-dimensinal)을 concatenation-based, transformation-based, model-based로 나눠서 볼 것이다.</li>
</ul>
</li>
</ul>
<h1>2. Challenges</h1>
<ul>
<li>생물학 데이터는 필연적으로 오류를 동반하고 있는데, 이렇게 깨끗하지 못한 여러 omics 데이터를 종합하여 특정한 패턴을 관찰한다는 것은 쉽지 않다.</li>
<li>또한 multi-omics 연구의 특성상 적은 샘플에서 많은 feature를 관찰해야 하는데, 이는 machine learing에서 가장 이상적이지 못한 상황으로 차원의 저주(<em class="id-context" data-target-id="20230924120348">Curse of The Dimensionality</em>)라고 불리운다.
<ul>
<li>이런 상황에서 ML을 통해 데이터를 훈련시키면 overfit된다.</li>
</ul>
</li>
<li>생물학 omics 데이터들이 각각 너무 성격이 다르다는 점도 하나의 문제이다.
<ul>
<li>유전자 발현 데이터의 경우 수만개의 variable이 있을 수 있지만, 대사물질 데이터는 기껏해서 몇천일 수가 있다.</li>
<li>이렇게 서로 다른 데이터들 사이의 수, 종류, 연속성, 등의 차이를 극복하고 데이터들을 잘 integration 해야한다는 어려움이 있다.</li>
</ul>
</li>
<li>데이터 간의 class imbalance 또한 ML 훈려에 악영향을 미친다.
<ul>
<li>데이터에서 특정 class의 데이터만 기하급수적으로 많다면, 이를 바탕으로 훈련시키는 것이 부적절하다.</li>
<li>왜냐하면 ML 훈련은 모든 class의 데이터 량이 엇비슷하다고 가정하고 훈련되기 때문이다.</li>
<li>이를 해결하기 위한 방법으로 sampling, cost-sensitive learning과 같은 방법이 존재한다.</li>
</ul>
</li>
<li>Missing data 문제의 경우 해당 샘플을 지워버린다던가 혹은 Random Forest, K-Nearest Neighbor같은 방법이 있다.</li>
</ul>
<h1>3. Main integration strategies</h1>
<h3>Integration 5가지 소개</h3>
<ul>
<li><strong>Early integration</strong>이란, 데이터들의 복잡도를 낮추지 않고, 단순히 둘 이상의 데이터들을 합친 뒤(wise concatenation) 그 결과물인 matrix를 ML모델에 넣는 것이다.
<ul>
<li>현실적으로는 단순하게 합쳐진 데이터는 여전히 복잡하기 때문에 ML모델이 잘 돌아가지 않는다다</li>
</ul>
</li>
<li><strong>Mixed integration</strong>이란 데이터들의 복잡도를 개별적으로 낮춰서 사용하는 것이다.</li>
<li><strong>Intermediate integration</strong>이란 여러 데이터들의 복잡도를 함께 낮춰서 사용하는 것이다.</li>
<li><strong>Late integration</strong>이란 정반대로 데이터를 합치지 않고 각각의 omics 데이터를 개별적으로 분석하는 것이다.</li>
<li><strong>Hierarchical integration</strong>이란 DNA-RNA-단백질로 이루어지는 시스템을 표현한 central dogma에 입각해서 omics 데이터 간의 조절 관계를 반영하여 합치는 것이다.</li>
</ul>
<h2>3.1 Dimensionality reduction for multi-omic integration</h2>
<p><strong>Dimensionality reduction은 왜 필요할까?</strong></p>
<ul>
<li>Multi-omics 연구에 있어서 샘플들이 너무 많은 variable을 가질 경우 machine learning 기술 적용에 차질을 빚는다. 따라서 variable의 수를 줄여서 차원의 축소와 데이터셋의 noise 감소를 유도한다.</li>
<li>차원축소는 필수적이지는 않고, 위에서 얘기했던 5가지의 integration 방법에 모두 적용이 가능하다.
<ul>
<li>다만 early integration와 intermediate integration의 경우 차원축소를 미리 적용하는 것이 정확도 향상에 기여한다.</li>
</ul>
</li>
<li>차원축소를 위한 방법은 크게 두 가지가 있다:
<ul>
<li>Feature selection</li>
<li>Redundant varaibles and feature extraction</li>
</ul>
</li>
</ul>
<h3>3.1.1. Feature Selection (FS)</h3>
<ul>
<li>Feature를 줄이는 작업</li>
<li>그 결과 데이터의 복잡도, noise가 줄어들어서 데이터가 깔끔해진다.</li>
<li>데이터가 깔끔해지니까, ML 모델이 overfitting될 위험도 적어진다.</li>
<li>데이터가 깔끔해져서 해석하기도 편해진다.</li>
<li>너무 많은 variable이 삭제될 경우?
<ul>
<li>데이터셋 사이에 feautre의 수가 다르면 좋지 않다</li>
<li>Feature Selection은 각 omics block이 동일한 수의 feature를 가지도록 할 수 있다</li>
</ul>
</li>
<li>Feature Selection은 세 가지로 나뉜다:
<ul>
<li>Filter-based Feature Selection</li>
<li>Wrapper-based Feature Selection</li>
<li>Embedded Feature Selection</li>
</ul>
</li>
</ul>
<h3>Filter-based Feature Selection</h3>
<ul>
<li>ML 모델과 상관없이 적용 가능</li>
<li>Feature 수를 줄일 수 있는 방향으로 관련된 variable을 추려내는 통계학적인 분석방법임</li>
<li><strong>CFS, RCA</strong> 같이 correlation 방법을 사용하거나,</li>
<li><strong>ReliefF</strong> 같이 distance 방법을 사용하거나,</li>
<li><strong>mRMR</strong> 같이 information gain 방법을 사용함</li>
</ul>
<h3>Wrapper-based Feature Selection</h3>
<ul>
<li>ML 모델을 활용하여 서로 다른 feature 그룹을 학습시키고 평가한다.</li>
<li>가장 성능이 좋았던 feature 그룹만 선택한다.</li>
<li>Recursive Feature Selection이 한 가지 예시이다.
<ul>
<li>먼저 전체 feauture를 대상으로 모델을 학습시킨다.</li>
<li>그 다음 모델 퍼포먼스에 기여도가 떨어지는 feature를 차례대로 제거해간다.</li>
</ul>
</li>
<li>이 방법은 어느 ML 모델에서든 활용 가능하다는 장점이 있다.</li>
<li>그러나 반복적으로 ML 모델을 훈련시켜야 하기 때문에, 데이터셋이 크다면 컴퓨터 성능에 따라 시간이 오래 걸린다는 단점이 있다.</li>
</ul>
<h3>Embedded Feature Selection</h3>
<ul>
<li>Classifier 모델 안에 feature selection 알고리즘이 넣어져있는 경우</li>
<li>Tree-based feature selection과 regularization 방법이 있다.
<ul>
<li>Regularization 방법은 loss function과 penalization funtion을 결합한 방법이다.
<ul>
<li>Loss funtion은 모델이 주어진 데이터에 얼마나 잘 맞는지 확인하는 함수다.</li>
<li>Penalization funtion은 복잡도(complexity)에 대하여 페널티를 부가하는 방식으로 feature 수를 적은 쪽으로 유도한다.</li>
</ul>
</li>
</ul>
</li>
<li>Embedded Feature Selection 모델은 Wrapper-based Feature Selection과 마찬가지로 ML모델과 함께 사용되지만, 컴퓨터 자원을 덜 소비하기 때문에 컴퓨터 성능에 덜 의존적이라는 장점이 있다.</li>
</ul>

                                </div>
                            </a>
                        </li>
                        
                        <li data-target-id="20230924120364">
                            <a href="#20230924120364">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Feature Selection</span>
                                <div class="record-links-context">
                                    <h1>1. introduction</h1>
<h3>왜 multi-omics 연구가 필요할까?</h3>
<ul>
<li>맞춤의학, 정밀의료의 부상
<ul>
<li>개인의 임상데이터를 통해 진단하고 적절한 치료법을 제시</li>
<li>임상데이터로 genomics, epigenomics, trasncriptiomics, proteomics, metabolomics 같은 데이터를 수집함</li>
<li>과거에는 이러한 각각의 omics 데이터를 따로 관찰해서 진단에 활용하려고 했음</li>
<li>그러나 대부분의 질병은 omics 데이터들을 서로 관통하면서 연결되어있음</li>
<li>따라서 여러 omics 데이터를 합쳐서 통합적인 관점으로 보는 multi-omics 연구가 필요함</li>
</ul>
</li>
</ul>
<h3>Multi-omics 연구의 장단점은?</h3>
<ul>
<li>한 omics 데이터에서 부족한 부분을 다른 omics 데이터를 참고하여 메꾼다던가, 하나의 큰 시스템에 대한 그림을 얻어낸다던가 하는 장점이 있다.</li>
<li>그러나 적은 샘플에서 가져온 데이터만 있다는 점이나, 개개의 생물학 데이터가 갖고있는 오류를 넘어서서 데이터를 합쳐야 한다는 어려움들이 단점으로 남는다.
<ul>
<li>특히 multiomics 연구에서 데이터를 합치는(integration) 작업이 정확하게 이루어지지 않으면, 오류 투성이로 합쳐진 데이터집합에서 결과를 제대로 얻는 것은 불가능하다.</li>
</ul>
</li>
</ul>
<h3>Data integration</h3>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6242341/">Zitnik et al. (2019)</a>은 integration의 방법을 크게 두 가지로 분류하였다.
<ul>
<li>Horizontal inegration: 동일 omics 여러 샘플</li>
<li>Vertical integration: 여러 omics 동일 샘플</li>
</ul>
</li>
<li>이번 리뷰에서는 Vertical integration을 위주로 리뷰할 것이다.<br>
- Vertical integration의 방법을 수학적인 관점에서 bayesian, network-based, deep learning-based, kernel-based, matrix factorization-based로 나눠서 볼 것이다.
<ul>
<li>또한 <a href="https://pubmed.ncbi.nlm.nih.gov/25582081/">Ritchie et al.(2015)</a>의 분류방법에 따라 vertical integration(논문에서는 meta-dimensinal)을 concatenation-based, transformation-based, model-based로 나눠서 볼 것이다.</li>
</ul>
</li>
</ul>
<h1>2. Challenges</h1>
<ul>
<li>생물학 데이터는 필연적으로 오류를 동반하고 있는데, 이렇게 깨끗하지 못한 여러 omics 데이터를 종합하여 특정한 패턴을 관찰한다는 것은 쉽지 않다.</li>
<li>또한 multi-omics 연구의 특성상 적은 샘플에서 많은 feature를 관찰해야 하는데, 이는 machine learing에서 가장 이상적이지 못한 상황으로 차원의 저주(Curse of The Dimensionality)라고 불리운다.
<ul>
<li>이런 상황에서 ML을 통해 데이터를 훈련시키면 overfit된다.</li>
</ul>
</li>
<li>생물학 omics 데이터들이 각각 너무 성격이 다르다는 점도 하나의 문제이다.
<ul>
<li>유전자 발현 데이터의 경우 수만개의 variable이 있을 수 있지만, 대사물질 데이터는 기껏해서 몇천일 수가 있다.</li>
<li>이렇게 서로 다른 데이터들 사이의 수, 종류, 연속성, 등의 차이를 극복하고 데이터들을 잘 integration 해야한다는 어려움이 있다.</li>
</ul>
</li>
<li>데이터 간의 class imbalance 또한 ML 훈려에 악영향을 미친다.
<ul>
<li>데이터에서 특정 class의 데이터만 기하급수적으로 많다면, 이를 바탕으로 훈련시키는 것이 부적절하다.</li>
<li>왜냐하면 ML 훈련은 모든 class의 데이터 량이 엇비슷하다고 가정하고 훈련되기 때문이다.</li>
<li>이를 해결하기 위한 방법으로 sampling, cost-sensitive learning과 같은 방법이 존재한다.</li>
</ul>
</li>
<li>Missing data 문제의 경우 해당 샘플을 지워버린다던가 혹은 Random Forest, K-Nearest Neighbor같은 방법이 있다.</li>
</ul>
<h1>3. Main integration strategies</h1>
<h3>Integration 5가지 소개</h3>
<ul>
<li><strong>Early integration</strong>이란, 데이터들의 복잡도를 낮추지 않고, 단순히 둘 이상의 데이터들을 합친 뒤(wise concatenation) 그 결과물인 matrix를 ML모델에 넣는 것이다.
<ul>
<li>현실적으로는 단순하게 합쳐진 데이터는 여전히 복잡하기 때문에 ML모델이 잘 돌아가지 않는다다</li>
</ul>
</li>
<li><strong>Mixed integration</strong>이란 데이터들의 복잡도를 개별적으로 낮춰서 사용하는 것이다.</li>
<li><strong>Intermediate integration</strong>이란 여러 데이터들의 복잡도를 함께 낮춰서 사용하는 것이다.</li>
<li><strong>Late integration</strong>이란 정반대로 데이터를 합치지 않고 각각의 omics 데이터를 개별적으로 분석하는 것이다.</li>
<li><strong>Hierarchical integration</strong>이란 DNA-RNA-단백질로 이루어지는 시스템을 표현한 central dogma에 입각해서 omics 데이터 간의 조절 관계를 반영하여 합치는 것이다.</li>
</ul>
<h2>3.1 Dimensionality reduction for multi-omic integration</h2>
<p><strong>Dimensionality reduction은 왜 필요할까?</strong></p>
<ul>
<li>Multi-omics 연구에 있어서 샘플들이 너무 많은 variable을 가질 경우 machine learning 기술 적용에 차질을 빚는다. 따라서 variable의 수를 줄여서 차원의 축소와 데이터셋의 noise 감소를 유도한다.</li>
<li>차원축소는 필수적이지는 않고, 위에서 얘기했던 5가지의 integration 방법에 모두 적용이 가능하다.
<ul>
<li>다만 early integration와 intermediate integration의 경우 차원축소를 미리 적용하는 것이 정확도 향상에 기여한다.</li>
</ul>
</li>
<li>차원축소를 위한 방법은 크게 두 가지가 있다:
<ul>
<li>Feature selection</li>
<li>Redundant varaibles and feature extraction</li>
</ul>
</li>
</ul>
<h3>3.1.1. <em class="id-context" data-target-id="20230924120364">Feature Selection</em> (FS)</h3>
<ul>
<li>Feature를 줄이는 작업</li>
<li>그 결과 데이터의 복잡도, noise가 줄어들어서 데이터가 깔끔해진다.</li>
<li>데이터가 깔끔해지니까, ML 모델이 overfitting될 위험도 적어진다.</li>
<li>데이터가 깔끔해져서 해석하기도 편해진다.</li>
<li>너무 많은 variable이 삭제될 경우?
<ul>
<li>데이터셋 사이에 feautre의 수가 다르면 좋지 않다</li>
<li>Feature Selection은 각 omics block이 동일한 수의 feature를 가지도록 할 수 있다</li>
</ul>
</li>
<li>Feature Selection은 세 가지로 나뉜다:
<ul>
<li>Filter-based Feature Selection</li>
<li>Wrapper-based Feature Selection</li>
<li>Embedded Feature Selection</li>
</ul>
</li>
</ul>
<h3>Filter-based Feature Selection</h3>
<ul>
<li>ML 모델과 상관없이 적용 가능</li>
<li>Feature 수를 줄일 수 있는 방향으로 관련된 variable을 추려내는 통계학적인 분석방법임</li>
<li><strong>CFS, RCA</strong> 같이 correlation 방법을 사용하거나,</li>
<li><strong>ReliefF</strong> 같이 distance 방법을 사용하거나,</li>
<li><strong>mRMR</strong> 같이 information gain 방법을 사용함</li>
</ul>
<h3>Wrapper-based Feature Selection</h3>
<ul>
<li>ML 모델을 활용하여 서로 다른 feature 그룹을 학습시키고 평가한다.</li>
<li>가장 성능이 좋았던 feature 그룹만 선택한다.</li>
<li>Recursive Feature Selection이 한 가지 예시이다.
<ul>
<li>먼저 전체 feauture를 대상으로 모델을 학습시킨다.</li>
<li>그 다음 모델 퍼포먼스에 기여도가 떨어지는 feature를 차례대로 제거해간다.</li>
</ul>
</li>
<li>이 방법은 어느 ML 모델에서든 활용 가능하다는 장점이 있다.</li>
<li>그러나 반복적으로 ML 모델을 훈련시켜야 하기 때문에, 데이터셋이 크다면 컴퓨터 성능에 따라 시간이 오래 걸린다는 단점이 있다.</li>
</ul>
<h3>Embedded Feature Selection</h3>
<ul>
<li>Classifier 모델 안에 feature selection 알고리즘이 넣어져있는 경우</li>
<li>Tree-based feature selection과 regularization 방법이 있다.
<ul>
<li>Regularization 방법은 loss function과 penalization funtion을 결합한 방법이다.
<ul>
<li>Loss funtion은 모델이 주어진 데이터에 얼마나 잘 맞는지 확인하는 함수다.</li>
<li>Penalization funtion은 복잡도(complexity)에 대하여 페널티를 부가하는 방식으로 feature 수를 적은 쪽으로 유도한다.</li>
</ul>
</li>
</ul>
</li>
<li>Embedded Feature Selection 모델은 Wrapper-based Feature Selection과 마찬가지로 ML모델과 함께 사용되지만, 컴퓨터 자원을 덜 소비하기 때문에 컴퓨터 성능에 덜 의존적이라는 장점이 있다.</li>
</ul>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                

                
            </footer>
        </article>
        
        <article id="20230924120367" class="record">
            <header>
                <h1 class="record-title">IntGen_1.1-1.3_MendelsFirstLawMeiosis</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120367</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://opengenetics.pressbooks.tru.ca/chapter/introduction-2/">https://opengenetics.pressbooks.tru.ca/chapter/introduction-2/</a></p>
<h1>Mendle's study of garden pea plant</h1>
<ul>
<li>After eight years of tedious experiments with garden pea plants, Mendel proposed his foundational principles of inheritance.<br>
8년간의 지루한 배나무 실험 끝에, 멘델의 유전(inheritance)에 관한 기초적인 원리를 제안한다.</li>
<li>Mendel showed that white-flowered plants could be produced by crossing two purple-flowered plants, but only if the purple-flowered plants themselves had at least one white-flowered parent.<br>
멘데은 보라색 꽃의 부모에 하얀 꽃이 있다면, 두 개의 보라색 꽃에서 하얀 꽃이 만들어 질 수 있다는 점을 보여줬다.</li>
<li>Mendel’s observations disproved blending inheritance and favoured an alternative concept, called <strong>particulate inheritance</strong>, in which heredity is the product of discrete factors that control independent traits.<br>
멘델의 관찰결과는 혼합유전(blending inheritance)를 반증하고, 별개의 요소가 개체의 특성(trait)을 조절한다는 입자유전(particulate inheritance)을 새롭게 주장한다.</li>
</ul>
<h1>Mendel's First law</h1>
<ul>
<li>Through careful study of patterns of inheritance, Mendel recognized that a single trait could exist in different versions, or <strong>alleles</strong>, even within an individual plant or animal.<br>
유전 패턴을 주의깊게 연구한 끝에, 멘델은 하나의 식물 혹은 동물 개체 내에서도 개체의 특성(trait)은 서로 다른 버전 혹은 형질(allele)들로 존재할 수 있다는 점을 발견했다.</li>
<li>This is the basis of <strong>Mendel’s First Law</strong>, also called <strong>The Law of Equal Segregation</strong>, which states that, during gamete formation the two alleles at a gene locus segregate from each other, and each gamete has an equal probability of containing either allele.<br>
이것이 <strong>멘델의 첫번째 법칙</strong>, 혹은 <strong>평등분리의 법칙(The Law of Equal Segregation)</strong> 의 기본이며, 이는 생식세포(gamete)의 형성과정 중에서 유전자 자리(locus) 위의 두 개의 형질(allele)이 서로 분리되며, 각 생식세포는 동일한 확률로 형질 중 하나를 가진다는 뜻이다.</li>
</ul>
<h1>Dominant and Recessive Alleles</h1>
<ul>
<li>The dominant/recessive character is a relationship between two alleles and must be determined by observation of the heterozygous phenotype.<br>
우성/열성 특징은 두 개의 형질(allele)의 관계이며 이형접합체(heterozygous)의 표현형(phenotype)에 의해서 결정되어야만 한다.</li>
<li>Remember, alleles are different versions of a gene. The relationship of different alleles of a gene can be described as complete dominance, incomplete dominance, or co-dominance.<br>
형질(allele)는 유전자의 다른 버전이라는 점을 기억하자. 한 유전자에서 유래한 서로 다른 형질의 관계는 완전우성(complete dominance), 불완전우성(incomplete dominance), 혹은 공동우성(co-dominnace)의 형태로 설명될 수 있다.</li>
<li>The traits Mendel studied with his peas were all completely dominant.<br>
멘델이 연구했던 배나무의 특성들은 모두 완전우성이었다.</li>
<li>In a diploid organism, if an allele is <strong>dominant,</strong> only one copy of that allele is necessary to express the dominant phenotype.<br>
이배체(diploid) 생물의 경우, 만약 형질(allele)이 <strong>우성(dominant)</strong> 이라면, 우성 표현형(phenotype)을 발현하기 위해서는 형질 하나만 필요하다.</li>
<li>If an allele is <strong>recessive</strong>, then the gene needs to have two copies (or be homozygous) to express the recessive phenotype.<br>
만약 형질(allele)이 <strong>열성(recessive)</strong> 이라면, 열성 표현형(phenotype)을 발현하기 위해서는 형질 두 개(혹은 동형접합(homozygous))를 가져야 한다.</li>
</ul>
<h1>Meiosis(감수분열)</h1>
<ul>
<li>Most eukaryotes reproduce sexually — a cell from one individual joins with a cell from another to create offspring.<br>
대부분의 진핵생물(eukaryotes)는 생식을 통해 번식한다 — 한 개체에서의 세포가 다른 개체의 세포와 만나서 후손을 만든다.</li>
<li>To be successful, the cells that fuse must contain half the number of chromosomes as in the adult organism.<br>
성공하기 위해서는, 융합될 세포는 반드시 성인개체가 가진 염색체(chromosome) 수의 절반을 가지고 있어야만 한다.</li>
<li>The chromosome number is reduced through the process of <strong>meiosis</strong>.<br>
염색체(chromosome) 수는 <strong>감수분열(meiosis)</strong> 을 통해 줄어든다.</li>
<li>Most multicellular organisms use meiosis to produce <strong>gametes</strong>, the cells that fuse to make offspring.<br>
대부분의 다세포 생물들은 감수분열(meiosis)를 통해 생식세포(gamete)를 생성하고, 이 생식세포가 후손을 만들기 위해 융합된다.</li>
<li>You will hear of cells that have not yet undergone meiosis to become egg or sperm cells called oocytes or spermatocytes respectively.<br>
감수분열(meiosis)를 아직 겪지 않아 난자(egg)나 정자(sperm) 세포가 되지 않은 것들은 각각 난모세포(oocyte)와 정모세포(spermatocyte)라고 불린다.</li>
<li>Meiosis begins similarly to mitosis in that a cell has grown large enough to divide and has replicated its chromosomes. However, Meiosis requires two rounds of division.<br>
감수분열(meiosis)는 체세포분열(mitosis)과 마찬가지로 세포가 거대해지고 염색체가 복제되는 과정에서부터 시작한다. 그러나 감수분열의 경우 체세포분열과 달리 분열 과정을 두 번 거친다.</li>
<li>Meiosis has two main stages, designated by the roman numerals I and II.<br>
감수분열(meiosis)는 두 개의 단계로 나누어지는데, 로마숫자에 따라 I과 II로 나뉜다.
<ul>
<li>In <strong>Meiosis I</strong> homologous chromosomes segregate<br>
감수분열(meiosis) I에서는 상동염색체(homologous chromosome)가 분리된다</li>
<li>In <strong>Meiosis II</strong> sister chromatids segregate<br>
감수분열(meiosis) II에서는 자매 염색분체(sister chromatid)가 분리된다</li>
</ul>
</li>
</ul>
<h2>Meiosis I</h2>
<ul>
<li>Cells that will undergo meiosis are called <strong>meiocytes</strong> and are diploid (2N)<br>
감수분열(meiosis)를 거치는 세포들은 <strong>감수모세포(meiocyte)</strong> 라고 불리며 이배체(diploid)(2N) 상태이다.</li>
<li>Meiosis I is called a <strong>reductional</strong> division, because it reduces the number of chromosomes inherited in each of the daughter cells – the parent cell is 2N while the two daughter cells are each 1N.<br>
감수분열(meiosis) I 단계는 염색체(chromosome)의 수가 줄어들어 딸세포들로 유전되기 때문에 분열(a reductional divison)이라고 불린다 - 부모세포는 2N인 반면에 두 개의 딸세포들은 각각 1N 상태다.</li>
<li>Meiosis I is further divided into...<br>
감수분열 I은 다음과 같이 세분화된다...
<ul>
<li>Prophase I<br>
전기</li>
<li>Metaphase I<br>
중기</li>
<li>Anaphase I<br>
후기</li>
<li>Telophase I<br>
말</li>
</ul>
</li>
<li>During synapsis, proteins bind to both homologous chromosomes along their entire length and form the <strong>synaptonemal complex</strong> (synapse means junction). These proteins hold the chromosomes in the transient structure of a bivalent. The proteins are released when the cell enters Anaphase I.<br>
접합(synapsis)과정 동안에, 단백질이 두 개의 상동염색체(homologous chromosome)에 전체에 달라붙어 <strong>이음실복합체(synaptonemal complex)</strong> 를 형성한다. 이 단백질은 염색체(chromosome)들을 이가(bivalent)의 임시형태로 유지 시킨다. 단백질들은 세포가 후기(Anaphase) I에 진입할 때 풀어준다.<br>
<img src="http://opengenetics.pressbooks.tru.ca/wp-content/uploads/sites/42/2020/11/ch1-pic7.png" alt="image"></li>
</ul>
<h3>Meiosis I - Prophase I</h3>
<ul>
<li>Initially, chromosomes condense and become visible and centrosomes begin to migrate to opposite poles of the cell<br>
초기에, 염색체(chromosome)들은 응축하여 관찰 가능한 형태가 되고 중심체(centrosome)들은 세포의 반대방향의 극으로 움직이기 시작한다.</li>
<li>Homologous chromosomes enter synapsis and the synaptonemal complex forms<br>
상동염색체(homologous chromosome)은 접합(synapsis) 과정에 들어가고 이음실복합체(synaptonemal complex)를 형성한다.</li>
<li><strong>Crossing over</strong> occurs resulting in an exchange of genetic material between non-sister chromatids of a homologous chromosome pair<br>
<strong>교차(crossing over)</strong> 는 상동염색체(a homologous chromosome pair) 사이에서 서로 자매가 아닌 염색분체(chromatid) 사이의 유전물질 교환을 의미한다.
<ul>
<li><strong>Crossing over</strong> occurs within the synaptonemal complex.<br>
교차(crossing over)는 이음실복합체(synaptonemal complex)에서 일어난다.</li>
<li>A crossover is a place where DNA repair enzymes break the DNA of two non-sister chromatids in similar locations and then covalently reattach non-sister chromatids together to create a crossover between non-sister chromatids.<br>
교차점(a crossover)은 DNA 복구 효소들이 서로 자매가 아닌 두 개의 염색분체(chromatid)의 DNA를 부순 다음에 공유결합을 통해 재결합 시키는 지점이다.</li>
<li>Crossovers function to hold homologous chromosomes together during meiosis I so they orient correctly and segregate successfully and reshuffles the allele combinations along a chromosome.<br>
교차점은 상동염색체(homologous chromosome)들을 감수분열(meiosis) I 동안 유지시키는 기능을 함으로서 상동염색체들의 방향과 분리가 잘 되도록 하고 형질(allele) 조합이 염색체(chromosome)를 따라서 섞이도록 한다.</li>
</ul>
</li>
<li>Following this, the synaptonemal complex disappears and tetrads are visible<br>
이에 따라, 이음실복합체(synaptonemal complex)는 사라지고 2가염색체(tetrad, bivalent chromosome)로 불린다</li>
<li>Crossover points appear as <strong>chiasmata</strong> which hold non-sister chromatids together<br>
교차점은 서로 자매가 아닌 염색분체(chromatid)를 함께 지닌 <strong>염색체교차점(chiasmata)</strong> 의 형태로 보인다</li>
<li>Finally, chromatids thicken and shorten, the nuclear membrane dissolves and spindle fibers begin forming.<br>
마지막으로, 염색분체(chromatid)는 두껍고 짧아지며, 핵막(nuclear membrane)이 분해되고 방추사(spindle fiber)가 형성된다.</li>
</ul>
<h3>Meiosis I - Metaphase I</h3>
<ul>
<li>Tetrads line up on the equator or the metaphase plate and each chromosome of a homologous pair attaches to spindle fibers from opposite ends of the poles – sister chromatids attach to fibers from the same pole.<br>
2가염색체(tetrad)들이 세포의 적도판(equator) 혹은 중기판(metaphase plate)에 나열하고, 상동염색체(homologous pair)의 각 염색체(chromosome)는 세포의 양극에서 나온 방추사(spinde fiber)에 연결된다 - 자매염색분체(sister chromatid)들은 동일한 극에서 나온 방추사에 연결된다.</li>
</ul>
<h3>Meiosis I - Anaphase I</h3>
<ul>
<li>Chiasmata dissolve<br>
염색체교차점이 사라진다</li>
<li>Homologous chromosomes move to opposite poles<br>
상동염색체(homologous chromosome)가 양 극단으로 이동한다
<ul>
<li>one member of each pair migrates to each daughter cell (1N)<br>
2가염색체(pair)를 이루는 자매염색분체(one member)가 각각 딸세포(1N)로 이동한다<br>
(note: centromeres do not separate here.)<br>
(노트: 자매염색분체를 구성하는 동원체(cetromere)는 사라지지 않는다.)</li>
</ul>
</li>
</ul>
<h3>Meiosis I - Telophase I</h3>
<ul>
<li>The nuclear envelope reforms and the resulting cells have half the number of chromosomes, each consisting of two sister chromatids.<br>
핵막(nuclear envelope)이 재구성되며 세포의 염색체의 수는 반으로 줄어들고 각 세포는 두 개의 자매염색분체(sister chromatids)를 지니고 있다.</li>
</ul>
<h3>Interkinesis/Cytokinesis</h3>
<ul>
<li>Similar to interphase except no chromosome duplication occurs. Daughter nuclei become enclosed into separate daughter cells.<br>
감수분열 이전의 휴지기(interphase)와 비슷하나 염색체 복제 과정은 일어나지 않는다. 딸핵이 닫히며 딸세포들이 분리된다.</li>
</ul>
<h2>Meiosis II</h2>
<ul>
<li>At the completion of Meiosis I, there are two cells, each with one, replicated copy of each chromosome (1N).<br>
감수분열(meiosis) I이 완료되면, 두 개의 세포가 생기며, 각 세포들은 복제되었던 염색체(1N)을 지니고 있다.</li>
<li>Meiosis II resembles mitosis, with one sister chromatid from each chromosome separating to produce two daughter cells.<br>
감수분열(meiosis) II는 체세포분열(mitosis)와 비슷하며, 하나의 자매염색분체(sister chromatid)가 염색체에서 분리되어 두 개의 딸세포들을 생성하게 된다.</li>
<li>Because Meiosis II, like mitosis, results in the segregation of sister chromatids, Meiosis II is called an <strong>equational</strong> division<br>
왜냐하면 감수분열(meiosis) II는 체세포분열(mitosis)처럼, 자매염색분체(sister chromatid)의 분리가 목적이며, 감수분열 II는 <strong>유사분열(an equational division)</strong> 이라고 불린다.</li>
</ul>
<h3>Meiosis II - Prophase II</h3>
<ul>
<li>Chromosomes condense<br>
염색체 응축</li>
<li>Centrioles move towards the poles<br>
중심립이 극으로 이동</li>
<li>The nuclear envelope disintegrates<br>
핵막의 분해해</li>
</ul>
<h3>Meiosis II - Metaphase II</h3>
<ul>
<li>Chromosomes align at the equator or the metaphase plate<br>
염색체가 적도판 혹은 중기판에 나열열</li>
<li>Sister chromatids attach to spindle fibres from opposite poles<br>
자매염색분체가 양 극에서 나온 방추사와 연결</li>
</ul>
<h3>Meiosis II - Anaphase II</h3>
<ul>
<li>Centromeres divide<br>
중심립 분리</li>
<li>Sister chromatids move to opposite poles<br>
자매염색분체가 양 극으로 이동</li>
</ul>
<h3>Meiosis II - Telophase II</h3>
<ul>
<li>Chromosomes begin to uncoil<br>
염색체가 풀어짐</li>
<li>Nuclear envelope and nucleoli begin to reform<br>
핵막과 핵소체들(nucleoli)이 형성되기 시작</li>
</ul>
<h3>Cytokinesis 세포질분열</h3>
<ul>
<li>
<p>Division of the cytoplasm occurs, resulting in four new daughter cells<br>
세포질(cytoplasm)의 분리가 발생하여 4개의 딸세포가 생성됨</p>
</li>
<li>
<p>Each containing haploid number of chromosomes<br>
각 딸세포들은 반수체(haploid) 수 만큼의 염색체를 가지고 있음</p>
</li>
<li>
<p>Note how Meiosis I and II are both divided into prophase, metaphase, anaphase, and telophase, since those stages have similar features to mitosis. After two rounds of cytokinesis, four cells will be produced, each with a single copy of each chromosome in the set.<br>
감수분열(meiosis) I과 II는 모두 전기, 중기, 후기, 말기를 거친다는 점에서 체세포분열(mitosis)와 비슷하다는 점을 기억하자. 세포질분열(cytokinesis)를 두 번 거치고 나서, 4개의 세포들이 만들어질 것이며, 각 딸세포들은 각 염색체의 한 copy를 지니고 있다.</p>
</li>
</ul>
<h1>Gamete Maturation 생식세포 성숙</h1>
<ul>
<li>In animals and plants, the cells produced by meiosis need to mature before they become functional gametes.<br>
동물과 식물에서, 감수분열(meiosis)을 통해 만들어진 세포들은 작동하는 생식세포가 되기 이전에 성숙단계를 거칠 필요가 있다.</li>
<li>In male animals, the four products of meiosis are called <strong>spermatids</strong>.<br>
수컷동물에서, 감수분열을 통해 만들어진 4개의 결과물을 <strong>정아세포(spermatid)</strong> 라고 부른다.
<ul>
<li>They grow structures, like tails, and become functional sperm cells.<br>
꼬리와 같은 구조가 자라고 나서야 작동하는 정자세포(sperm cell)이 된다</li>
</ul>
</li>
<li>In female animals, the gametes are <strong>eggs</strong>.<br>
암컷동물에서, 생식세포는 <strong>난자(egg)</strong> 라고 부른다
<ul>
<li>For each egg to contain the maximum amount of nutrients.<br>
각 난자(egg)는 최대량의 영양분을 지니고 있다.</li>
<li>Typically only one of the four products of meiosis becomes <strong>an egg.</strong><br>
일반적으로 감수분열을 통해 만들어진 4개의 결과물 중 단 하나만이 난자(egg)가 된다.</li>
<li>The other three cells end up as tiny disposable cells called <strong>polar bodies.</strong><br>
나머지 3개의 세포들은 소모되는 세포들로서 <strong>극세포(polar bodies)</strong> 라고 불린다</li>
</ul>
</li>
<li>In plants, the products of meiosis reproduce a few times using mitosis as they develop into functional male or female gametes.<br>
식물에서, 감수분열(meiosis)의 생산물들은 작동하는 수컷 혹은 암컷 생식세포로 성장하기 위해 몇번의 체세포분열(mitosis)를 거친다</li>
</ul>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120368" class="record">
            <header>
                <h1 class="record-title">IntGen_1.4_Monohybrid</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120368</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://opengenetics.pressbooks.tru.ca/chapter/one-locus-on-a-chromosome-segregation-monohybrid/">https://opengenetics.pressbooks.tru.ca/chapter/one-locus-on-a-chromosome-segregation-monohybrid/</a></p>
<ul>
<li>Mendel’s first law, the law of segregation, states that every individual contains a pair of alleles for each gene<br>
멘델의 첫번째 법칙, 분리의 법칙은 모든 개체는 유전자에 대한 한 쌍의 형질(allele)들을 갖고있다</li>
<li>Alleles segregate during the formation of gametes, and so for every gene pair, each parent passes on a random allele to its offspring.<br>
형질(allele)들은 생식세포(gamete)의 생성과정 중에서 분리되며, 따라서 모든 유전자 쌍에 대해, 부모는 각각 무작위 형질을 그 자손에게 전달하게 된다.</li>
<li>A specific position, region, or segment along a chromosome is called a <strong>locus</strong>.<br>
염색체 위의 특정한 위치, 지역, 혹은 부분은 <strong>유전자좌(locus)</strong> 라고 불린다
<ul>
<li>Each gene occupies a specific locus (so the terms locus and gene are often used interchangeably).<br>
각 유전자는 특정한 유전자좌(locus)에 존재한다(따라서 유전자좌와 유전자라는 단어는 혼용되고는 한다)</li>
<li>Each locus will have an allelic form (allele).<br>
각 유전자좌(locus)는 대립형질을 띈다 (형질(allele))</li>
</ul>
</li>
<li>The complete set of alleles (at all loci of interest) in an individual is its <strong>genotype</strong>.<br>
(관심있는 영역의 모든 위치에서의) 형질들(alleles)의 전체 모음을 유전자형(genotype)이라고 부른다.
<ul>
<li>Typically, when writing out a genotype, only the alleles at the locus (loci) of interest are considered – all the others are present and assumed to be wild type but are normally not written in the genotype.<br>
전형적으로, 유전자형(genotype)을 적을 때는, 오직 관심있는 영역의 유전자좌(locus) (위치)에 있는 형질(allele)만을 고려한다 - 다른 형질들은 변이가 없는 야생형(wild type)이라고 간주하지만 유전자형으로 적지는 않는다.</li>
</ul>
</li>
<li>The observable or detectable effect of these alleles on the structure or function of that individual is called its <strong>phenotype</strong>.<br>
형질(alleles)에 의하여 개체에 대하여 구조적으로나 기능적으로 영향을 미치는 것이 눈으로 관찰 가능하거나 확인 가능한 영향을 표현형(phenotype)이라고 부른다.
<ul>
<li>The phenotype studied in any particular genetic experiment may range from simple, visible traits such as hair color, to more complex phenotypes including disease susceptibility or behavior.<br>
표현형(phenotype)에 관한 유전실험은 그 범주가 머리카락 색깔과 같은 간단한 것부터 질병민감도 혹은 행동과 같은 복잡한 것까지 아우른다.</li>
<li>If two alleles are present in an individual, then various interactions between them may influence their expression in the phenotype.<br>
만약 두 형질들(alleles)이 한 개체에 존재한다면, 두 형질들 사이의 다양한 상호작용이 표현형(phenotype)의 발현에 영향을 끼칠 수 있다.</li>
</ul>
</li>
</ul>
<h2>True-Bredding Lines 순종계보</h2>
<ul>
<li>True-Bredding Lines are in-bred populations of plants or animals in which all parents and their offspring have the same phenotypes with respect to a particular trait.<br>
순종계보(True-Bredding Lines)란 모든 부모와 그 후손들이 특정한 특성으로 대표되는 동일한 표현형(phenotype)을 가지는 식물 혹은 동물의 동계집단(in-bred population)을 말한다.</li>
<li>They are typically assumed to be homozygous for the alleles that affect the trait of interest.<br>
순종계보의 개체들은 우리가 관심있어 하는 특성들에 대하여 동형의(homozygous) 형질(allele)을 가진다고 가정한다.</li>
<li>When two individuals that are homozygous for the same alleles are crossed, all of their offspring will all also be homozygous.<br>
만약 동형의 형질(allele)을 가지는 두 개체가 서로 교배(cross)한다면, 두 개체의 후손들은 모두 동형일것이다.</li>
<li>The continuation of such crosses constitutes a true breeding line or strain.<br>
이러한 교접의 연속이 순종계보를 구성한다</li>
</ul>
<p><img src="http://opengenetics.pressbooks.tru.ca/wp-content/uploads/sites/42/2020/11/ch1-pic8.png" alt=""><br>
<em>(a) A True-Breeding Line<br>
(b) A Monohybrid Cross Produced by Mating Two Different Pure-Breeding Lines</em></p>
<h2>Monohybrid Crosses 단성잡종교배</h2>
<ul>
<li>A <strong>monohybrid cross</strong> is one in which both parents are heterozygous (or a hybrid) for a single (mono) trait.<br>
<strong>단성잡종교배(monohybrid)</strong> 란 두 부모 모두 한 가지의 특징에 대하여 이형접합체(heterozygous) (혹은 잡종(hybrid))인 경우를 말한다.</li>
<li>By using monohybrid crosses, Mendel discovered that genes were discrete units that separated in the creation of offspring.<br>
단성잡종교배(monohybrid cross)을 활용하여, 멘델은 유전자가 후손의 생성과는 별개의 존재라는 것을 발견했다.</li>
<li>Through the monohybrid cross Mendel was able to discern the <strong>dominant</strong> and <strong>recessive</strong> alleles of each gene he studied in the pea plants.<br>
단성잡종교배(monohybrid cross)을 통하여 멘델은 배나무 연구에서 각 유전자의 <strong>우성(dominant)</strong> 와 <strong>열성(recessive)</strong> 형질(allele)을 파악했다</li>
<li>In further crosses, these traits were continuously transmitted and not lost.<br>
이후의 교배에서, 이러한 특성들은 계속해서 전달되고 사라지지 않았다.</li>
</ul>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120369" class="record">
            <header>
                <h1 class="record-title">IntGen_1.5_PunnettSquares</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120369</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://opengenetics.pressbooks.tru.ca/chapter/punnett-squares-31-ratio/">https://opengenetics.pressbooks.tru.ca/chapter/punnett-squares-31-ratio/</a></p>
<ul>
<li>The specific ratios seen in the monohybrid cross can be described using a <strong>Punnett square</strong>, named after R.C. Punnett who devised this approach.<br>
단성잡종교배(monohybrid cross)는 R.C.퍼넷의 이름을 딴 <strong>퍼넷 사각형(a Punnett square)</strong> 으로 표현될 수 있다.<br>
<img src="https://opengenetics.pressbooks.tru.ca/wp-content/uploads/sites/42/2020/11/ch1-pic9.png" alt="image"></li>
<li>A Punnett square is a matrix in which all of the possible gametes produced by one parent are listed along one axis, and the gametes from the other parent are listed along the other axis.<br>
퍼넷 사각형(a Punnett square)는 한 부모에서 유래한 모든 이용가능한 생식세포들을 한 축에 배열하고, 다른 한 축에는 다른 부모에서 온 생식세포를 배열한 행렬이다</li>
<li>Each possible combination of gametes is listed at the intersection of each row and column, since we know through the process of meiosis that the alleles on each chromosome separate to form the gametes.<br>
우리가 감수분열(meiosis)을 통해 각 염색체의 형질(allele)가 분리되어 생식세포를 구성한다는 점을 알기 때문에, 각각 가능한 생식세포의 조합들은 행과 열의 교차점에 나열된다.</li>
<li>The offspring ratios will be 3:1 of dominant phenotype (purple): recessive phenotype (white).<br>
후손의 비율은 우성(dominant) 표현형(보라색)과 열성(recessive) 표현형(하얀색)이 각각 3:1의 비율이다</li>
<li>Punnett squares can also be used to calculate the frequency of offspring.<br>
퍼넷 사각형(Punnett square)은 후손의 빈도를 계산할 때도 사용될 수 있다.</li>
<li>The frequency of each offspring is the frequency of the male gametes multiplied by the frequency of the female gamete.<br>
각 후손의 빈도는 남성생식세포의 빈도와 여성생식세포의 빈도의 곱이다.</li>
</ul>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120370" class="record">
            <header>
                <h1 class="record-title">IntGen_1.6_SingleLocusTestCrosses</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120370</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://opengenetics.pressbooks.tru.ca/chapter/single-locus-test-crosses/">https://opengenetics.pressbooks.tru.ca/chapter/single-locus-test-crosses/</a></p>
<ul>
<li>Genotypes must be inferred based on phenotypes.<br>
유전형(genotype)은 표현형(phenotype)을 기본으로 한다</li>
<li>Because of dominance, it is often not possible to distinguish between a heterozygote and a homozygote based on phenotype alone (e.g., see the purple-flowered F2 plants).<br>
우성인자(dominance) 때문에, 이형접합(heterozygote)과 동형접합(homozygote)을 표현형(phenotype)만 보고서는 판단할 수 없다</li>
<li>To determine the genotype of a specific individual, a <strong>test cross</strong> can be performed in which the individual with an unknown genotype is crossed with an individual that is homozygous recessive for all of the loci being tested.
<ul>
<li>For example, if you were given a pea plant with purple flowers it might be a homozygote (<em>AA</em>) or a heterozygote (<em>Aa</em>).</li>
<li>You could cross this purple-flowered plant to a white-flowered plant as a <strong>tester</strong>, since you know the genotype of the tester is <em>aa</em>.</li>
<li>Depending on the genotype of the purple-flowered parent (<strong>Figure 1.6.1</strong>), you will observe one of two phenotypic ratios in the F1 generation.<br>
<img src="https://opengenetics.pressbooks.tru.ca/wp-content/uploads/sites/42/2020/11/1.18.png" alt=""><br>
<em><strong>Figure 1.6.1</strong> Punnett Squares Showing the Two Possible Outcomes of a Single Locus Test Cross</em></li>
<li>If the purple-flowered parent was a homozygote <em>AA</em>, all of the F1 progeny will be purple.</li>
<li>If the purple-flowered parent was a heterozygote <em>Aa</em>, the F1 progeny should segregate purple-flowered and white-flowered plants in a 1:1 ratio.</li>
</ul>
</li>
</ul>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120371" class="record">
            <header>
                <h1 class="record-title">IntGen_6.2_SomaticMutVsGermlineMut</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120371</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://opengenetics.pressbooks.tru.ca/chapter/somatic-vs-germline-mutations/">Opengenetics: Somatic vs. Germline Mutations</a></p>
<h1>Somatic cells 체세포 vs. Germline cells 생식세포계열 세포</h1>
<ul>
<li>In multicellular animals, there is a partitioning early in development into <strong>somatic cells</strong>, which form the body cells, and <strong>germline cells</strong>, which form the gametes for the next generation.<br>
다세포 생물에서, 발생의 이른 단계에서 몸을 구성하는 체세포(somatic cell)와 다음 세대를 위한 생식세포(gamete)를 구성하는 생식세포계열세포(germline cell)로 나뉜다.</li>
<li>Somatic cells form the tissues of the organism and are not passed on as gametes.<br>
체세포(somatic cell)는 개체의 조직(tissue)을 구성하며 생식세포처럼 전달되지 않는다.</li>
<li>Germline cells are those that form the eggs or sperm cells (ovum or pollen in plants), and are passed on to form the next generation.<br>
생식세포계열 세포(germline cell)는 난자(egg) 혹은 정자(sperm)로구성되며(식물에서는 난세포(ovum)과 화분(pollen)), 이들은 다음 세대로 전달된다.</li>
</ul>
<h1>Somatic mutation 체세포 돌연변이</h1>
<ul>
<li>Any mutations in somatic cells will only affect the individual in which they occur, not its progeny.<br>
체세포(somatic cell)에서 발생하는 어떠한 돌연변이도 그 개체에서만 발생할 뿐, 후손(progeny)로 전달되지 않는다.</li>
<li>If mutations occur in somatic cells, its mutant descendants will exist alongside other non-mutant (wild type) cells.<br>
만약 체세포(somatic cell)에서 돌연변이가 일어났다면, 그 돌연변이 세포의 후손세포들은 다른 비-돌연변이(야생형)세포들과 체내에 함께 존재할 것이다</li>
<li>If the mutation occurs at a very early stage of development, the mutation will be present in more cells.<br>
만약 돌연변이가 발생의 매우 이른 단계에서 발생했다면, 돌연변이는 훨씬 많은 세포들에서 발견될 수 있다.</li>
<li>This gives rise to an individual composed of two or more types of cells that differ in their genetic composition. Such an individual is said to be a <strong>mosaic</strong>.<br>
이러한 이른 단계의 돌연변이는 개체로 하여금 두 개 혹은 그 이상의 종류의 유전자 구성의 세포들을 지니게 한다, 이러한 개체들을 모자이크 돌연변이를 가졌다고 부른다.</li>
</ul>
<h1>Germline mutation 생식세포 돌연변이</h1>
<ul>
<li>Mutations in germline cells will be passed on to the next generation but won’t affect the individual in which they occur.<br>
생식세포에서의 돌연변이는 다음 세대로 전달되지만, 정작 돌연변이가 발생한 개체에는 영향을 끼치지 않는다.</li>
<li>In animals, somatic cells are segregated from germ line cells.<br>
동물에서, 체세포는 생식세포로부터 분리된다.
<ul>
<li>In plants, somatic cells become germline cells; so somatic mutations can become germline mutations.<br>
식물에서, 체세포가 생식세포로 바뀐다; 따라서 식물에서의 체세포 돌연변이는 생식세포 돌연변이가 될 수 있다.</li>
</ul>
</li>
</ul>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120372" class="record">
            <header>
                <h1 class="record-title">Intron</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120372</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                

                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120373" class="record">
            <header>
                <h1 class="record-title">lecture-list</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120373</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <table>
<thead>
<tr>
<th style="text-align:left">name of lecture</th>
<th style="text-align:center">study status</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left"><a href="https://www.edwith.org/datastructure-2019s/home">KAIST: 데이터 구조 및 분석: Linear Structure and Dynamic Programming 2018</a></td>
<td style="text-align:center">complete</td>
</tr>
<tr>
<td style="text-align:left"><a href="https://www.edwith.org/datastructure-2019s2/home">KAIST: 데이터 구조 및 분석: Non-Linear Structure, Optimization, and Algorithms 2018</a></td>
<td style="text-align:center">ongoing</td>
</tr>
<tr>
<td style="text-align:left"><a href="https://youtube.com/playlist?list=PLUl4u3cNGP63gFHB6xb-kVBiQHYe_4hSi&amp;si=dQ8sMZqmWMudHkEh">MIT: 6.034 Artificial Intelligence, Fall 2010</a><a href="https://www.edwith.org/mitai/joinLectures/21462">인공지능 개론</a></td>
<td style="text-align:center">-</td>
</tr>
<tr>
<td style="text-align:left"><a href="https://atcold.github.io/NYU-DLSP20/ko/">New York University: Deep Learning Spring 2020</a></td>
<td style="text-align:center">-</td>
</tr>
<tr>
<td style="text-align:left"><a href="https://youtube.com/playlist?list=PLypiXJdtIca6dEYlNoZJwBaz__CdsaoKJ&amp;si=Fmb5rexc6L9Ra376">MIT: 6.047/6.878 Machine Learning for Genomics Fall 2020</a></td>
<td style="text-align:center">-</td>
</tr>
<tr>
<td style="text-align:left"><a href="https://mit6874.github.io/">MIT: 6.874 Computational Systems Biology: Deep Learning in the Life Sciences Spring 2021</a></td>
<td style="text-align:center">-</td>
</tr>
</tbody>
</table>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120374" class="record">
            <header>
                <h1 class="record-title">Long-read Sequencing</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120374</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p>Also known as third-generation sequencing<br>
<a href="#20230924120379" title="Nanopore" class="record-link">Nanopore</a><br>
<a href="#20230924120391" title="SMRT" class="record-link">SMRT</a></p>


                
            </main>

            <footer class="list-content-container">
                
                <details class="record-links" open="open">
                    <summary class="menu-title">
                        links
                        <span class="badge">2</span>
                    </summary>

                    <ul class="record-links-list">
                        
                        <li data-target-id="20230924120379">
                            <a href="#20230924120379">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Nanopore</span>
                                <div class="record-links-context">
                                    <p>Also known as third-generation sequencing<br>
<em class="id-context" data-target-id="20230924120379">Nanopore</em><br>
SMRT</p>

                                </div>
                            </a>
                        </li>
                        
                        <li data-target-id="20230924120391">
                            <a href="#20230924120391">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>SMRT</span>
                                <div class="record-links-context">
                                    <p>Also known as third-generation sequencing<br>
Nanopore<br>
<em class="id-context" data-target-id="20230924120391">SMRT</em></p>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                

                
            </footer>
        </article>
        
        <article id="20230924120375" class="record">
            <header>
                <h1 class="record-title">Machine Learning</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120375</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <h2>Example</h2>
<p><a href="#20230924120352" title="Deep Learning" class="record-link">Deep Learning</a><br>
<a href="#20230924120340" title="Active Learning" class="record-link">Active Learning</a></p>


                
            </main>

            <footer class="list-content-container">
                
                <details class="record-links" open="open">
                    <summary class="menu-title">
                        links
                        <span class="badge">2</span>
                    </summary>

                    <ul class="record-links-list">
                        
                        <li data-target-id="20230924120352">
                            <a href="#20230924120352">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Deep Learning</span>
                                <div class="record-links-context">
                                    <h2>Example</h2>
<p><em class="id-context" data-target-id="20230924120352">Deep Learning</em><br>
Active Learning</p>

                                </div>
                            </a>
                        </li>
                        
                        <li data-target-id="20230924120340">
                            <a href="#20230924120340">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Active Learning</span>
                                <div class="record-links-context">
                                    <h2>Example</h2>
<p>Deep Learning<br>
<em class="id-context" data-target-id="20230924120340">Active Learning</em></p>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                

                
            </footer>
        </article>
        
        <article id="20230924120376" class="record">
            <header>
                <h1 class="record-title">mRNA</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120376</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                

                
            </main>

            <footer class="list-content-container">
                

                
                <details class="record-backlinks" open="open">
                    <summary class="menu-title">
                        backlinks
                        <span class="badge">1</span>
                    </summary>

                    <ul class="record-backlinks-list">
                        
                        <li data-target-id="20230924120388">
                            <a href="#20230924120388">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>RNA</span>
                                <div class="record-links-context">
                                    <p><em class="id-context" data-target-id="20230924120376">mRNA</em></p>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                
            </footer>
        </article>
        
        <article id="20230924120377" class="record">
            <header>
                <h1 class="record-title">Multi-omics</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120377</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                

                
            </main>

            <footer class="list-content-container">
                

                
                <details class="record-backlinks" open="open">
                    <summary class="menu-title">
                        backlinks
                        <span class="badge">2</span>
                    </summary>

                    <ul class="record-backlinks-list">
                        
                        <li data-target-id="20230924120366">
                            <a href="#20230924120366">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Integration Strategies of Multi-omics Data for Machine Learning Analysis</span>
                                <div class="record-links-context">
                                    <p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8258788/">Picard M, Scott-Boyer MP, Bodein A, Périn O, Droit A. Integration strategies of multi-omics data for machine learning analysis. Comput Struct Biotechnol J. 2021 Jun 22;19:3735-3746. doi: 10.1016/j.csbj.2021.06.030. PMID: 34285775; PMCID: PMC8258788.</a></p>
<ul>
<li>2021년 기준으로 <em class="id-context" data-target-id="20230924120377">Multi-omics</em> Data Integration 방법을 5가지로 분류함
<ul>
<li>early</li>
<li>mixed</li>
<li>intermediate</li>
<li>late</li>
<li>hierarchical<br>
<img src="https://ars.els-cdn.com/content/image/1-s2.0-S2001037021002683-ga1_lrg.jpg" width=800></li>
</ul>
</li>
</ul>

                                </div>
                            </a>
                        </li>
                        
                        <li data-target-id="20230924120389">
                            <a href="#20230924120389">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>scRNA-seq</span>
                                <div class="record-links-context">
                                    <p>Single Cell RNA Sequencing<br>
<em class="id-context" data-target-id="20230924120377">Multi-omics</em><br>
Spatial Transcriptomics</p>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                
            </footer>
        </article>
        
        <article id="20230924120378" class="record">
            <header>
                <h1 class="record-title">Mutation</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120378</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <h1>Loss of Heterozygosity(LOH)</h1>
<p>이형상실/이형접합성 소실<br>
<img src="https://preview.redd.it/can-this-be-considered-a-somatic-mutation-v0-3d2xmyo37dpb1.png?width=1819&format=png&auto=webp&s=89f8595717b435ea42b57ebd78cb78396f3e708b" alt="example of LOH allele proportion" width="100%"><br>
<em>Reference allele는 G, alternative allele는 A가 확인된 샘플.</em></p>
<h1>Substitution</h1>
<h1>Point Mutation</h1>
<h1>Missense Mutation</h1>
<h1>Frameshift Mutation</h1>
<p><a href="https://www.genome.gov/genetics-glossary/Frameshift-Mutation">https://www.genome.gov/genetics-glossary/Frameshift-Mutation</a><br>
<img src="https://www.genome.gov/sites/default/files/media/images/tg/Frameshift-mutation.jpg"></p>
<h2>Deletion</h2>
<h2>Insertion</h2>
<h1>Nonsense Mutation</h1>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120379" class="record">
            <header>
                <h1 class="record-title">Nanopore</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120379</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <h2>2D sequencing</h2>
<ul>
<li>가장 처음으로 nanopore에서 활용된 기술</li>
<li>헤어핀구조가 이중나선 구조의 DNA에 붙음</li>
<li>두 DNA 가닥을 모두 sequencing 함</li>
</ul>
<h2>1D sequencing</h2>
<ul>
<li>One direction library kit를 활용함</li>
<li>하나의 DNA strand만 sequencing 함</li>
</ul>
<h2>1D$^2$ sequencing</h2>
<ul>
<li>헤어핀구조를 사용하지 않고 이중나선 구조를 sequencing</li>
<li><a href="https://nanoporetech.com/about-us/news/1d-squared-kit-available-store-boost-accuracy-simple-prep">1D 기술보다 더 높은 정확도를 가짐</a></li>
</ul>
<p>참고: <a href="https://www.mdpi.com/2079-6374/11/7/214">https://www.mdpi.com/2079-6374/11/7/214</a></p>


                
            </main>

            <footer class="list-content-container">
                

                
                <details class="record-backlinks" open="open">
                    <summary class="menu-title">
                        backlinks
                        <span class="badge">2</span>
                    </summary>

                    <ul class="record-backlinks-list">
                        
                        <li data-target-id="20230924120374">
                            <a href="#20230924120374">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Long-read Sequencing</span>
                                <div class="record-links-context">
                                    <p>Also known as third-generation sequencing<br>
<em class="id-context" data-target-id="20230924120379">Nanopore</em><br>
SMRT</p>

                                </div>
                            </a>
                        </li>
                        
                        <li data-target-id="20230924120387">
                            <a href="#20230924120387">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Real-Time DNA Sequencing in the Antarctic Dry Valleys Using the Oxford Nanopore Sequencer</span>
                                <div class="record-links-context">
                                    <p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5362188/">Johnson SS, Zaikova E, Goerlitz DS, Bai Y, Tighe SW. Real-Time DNA Sequencing in the Antarctic Dry Valleys Using the Oxford Nanopore Sequencer. J Biomol Tech. 2017 Apr;28(1):2-7. doi: 10.7171/jbt.17-2801-009. Epub 2017 Mar 22. PMID: 28337073; PMCID: PMC5362188.</a></p>
<h1>Abstract</h1>
<ul>
<li>Oxford Nanopore Technologies에서 개발한 소형 <em class="id-context" data-target-id="20230924120379">Nanopore</em> sequencer를 활용하여, 연구실이 아닌 야외에서 DNA sequencing 작업을 수행하고, 인터넷에 연결되지 않은 환경에서 소프트웨어를 활용해 분석까지 수행했다.</li>
<li>장소는 남극의 McMurdo Dry Valleys였다.</li>
<li>사용된 sequencer는 MK1B MinION sequencer였다.</li>
<li>2시간 30분만에 야외환경에서 sequencing protocol을 완수할 수 있었다.</li>
</ul>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                
            </footer>
        </article>
        
        <article id="20230924120380" class="record">
            <header>
                <h1 class="record-title">NGS</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120380</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p>Next Generation Sequencing</p>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120381" class="record">
            <header>
                <h1 class="record-title">OTUs</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120381</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p>Operational Taxonomical Units</p>


                
            </main>

            <footer class="list-content-container">
                

                
                <details class="record-backlinks" open="open">
                    <summary class="menu-title">
                        backlinks
                        <span class="badge">1</span>
                    </summary>

                    <ul class="record-backlinks-list">
                        
                        <li data-target-id="20230924120349">
                            <a href="#20230924120349">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Dark microbiome and extremely low organics in Atacama fossil delta unveil Mars life detection limits</span>
                                <div class="record-links-context">
                                    <h1>Abstract</h1>
<ul>
<li>칠레의 Atacama Desert의 red stone(암석)에서 채취한 미생물군집 샘플에서 DNA 흔적을 조사했는데, 발견한 DNA의 9%는 과학계에 보고된 적이 없어 분류 불가능한 미생물의 존재였음.</li>
<li>과학자들은 이렇게 sequencing 기술을 통하여 존재를 확인할 수 있지만, 정확하게 분류할 수 없는 미생물을 &quot;dark microbiome&quot;이라고 정의함.</li>
<li>해당 연구는 지구의 매우 건조한 환경에서도 미생물의 흔적도 많이 손상되 보존되기 어려움을 시사함.</li>
<li>따라서 화성에서 암석샘플 회수에 성공한다고 할지라도, 현재 기술로는 과거에 화성에 존재했던 미생물의 흔적을 놓칠 수 있다는 점을 시사함.</li>
</ul>
<h3>Site microbial diversity and dark microbione</h3>
<ul>
<li>암석샘플은 1 그램당 1μg의 DNA를 갖고 있었음.</li>
<li>Operational Taxonomical Units(<em class="id-context" data-target-id="20230924120381">OTUs</em>)이 제일 다양하게 검출된 부분은 바로 풍화된 상태의 역암(conglomerate)이었다.</li>
<li>가장 많이 검출된 종은 Alphaproteobacteria 와 Actinobacteria였다.<br>
<img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41467-023-36172-1/MediaObjects/41467_2023_36172_Fig4_HTML.png?as=webp"></li>
</ul>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                
            </footer>
        </article>
        
        <article id="20230924120382" class="record">
            <header>
                <h1 class="record-title">p-value(유의확률)</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120382</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://adnoctum.tistory.com/332">p-value란 무엇인가</a><br>
<a href="https://velog.io/@cualquier/%EA%B7%80%EB%AC%B4%EA%B0%80%EC%84%A4-%EB%8C%80%EB%A6%BD%EA%B0%80%EC%84%A4%EC%9D%80-%EB%AD%98%EA%B9%8C">귀무가설, 대립가설의 의미와 예시</a><br>
<a href="https://angeloyeo.github.io/2020/03/29/p_value.html">p-value의 의미</a></p>
<hr>
<p>귀무가설(null hypothesis)이 맞다고 전제할 때, 통계값(statistics)이 실제 관측된 값 이상일 확률<br>
The probability that the test statistic is equal to or greater than the observed value, assuming the null hypothesis is true, is called the p-value.</p>
<p>p-value는 귀무가설이 맞다고 가정했을 때, 관측된 데이터나 그보다 극단적인 데이터가 나타날 확률을 의미한다.<br>
The p-value represents the probability of observing the data or more extreme data, given that the null hypothesis is true.</p>
<p>즉, 가설검정의 과정에서 전체 데이터가 아니라 sampling된 데이터를 가지고 검정을 할 때, 그 통계값이 얼마나 믿을 수 있는가에 대한 것이다.<br>
In other words, when conducting hypothesis testing using a sample of data instead of the entire population, the p-value indicates the level of confidence we can have in the test statistic.</p>
<hr>
<p>교수님의 연구실에 10,000개의 샘플로 이루어진 데이터가 있다고 하자. 이 모집단 데이터는 정규분포의 모습을 하고 있고, 평균 값은 100이다. 이때, 교수님이 우리에게 모집단 데이터와 평균값에 대한 내용을 숨기고 무작위로 고른 샘플 100개를 주고는 모집단의 평균값을 추측해보라고 한다.<br>
Let's say you have a dataset consisting of 10,000 samples in your professor's research lab. This population data follows a 'normal distribution', with a mean value of 100. Now, your professor has randomly selected 100 samples from the population and provided them to you, without revealing any information about the population data or the mean value. Your task is to estimate the population mean based on this sample.</p>
<p>10,000개의 샘플이 정규분포를 그리고 있기 때문에, 데이터의 샘플 값은 0부터 200까지 다양하다. 만약 교수님이 주신 샘플 100개의 값이 전부 150 이상이었다면, 우리가 구한 평균값은 150 이상이 될 것이다. 이는 모집단의 평균값 100을 상회하므로, 우리가 받은 샘플들은 모집단의 모습을 반영하지 못한다고 할 수 있다.<br>
Since the 10,000 samples follow a normal distribution, the sample values can range from 0 to 200, exhibiting various values in between. If all the 100 values provided by your professor were 150 or higher, the mean value you calculate would be greater than 150. This would imply that the estimated mean exceeds the population mean of 100, suggesting that the sample you received does not accurately reflect the characteristics of the population.</p>
<p>어떤 방식으로도 10,000개의 샘플에서 무작위로 뽑은 100개의 평균값이 모집단과 동일하게 100일 가능성은 적다. 그렇다면 98이라면? 95라면? 93이라면? 우리는 그 평균값은 믿을 수 있을까?<br>
Using any method, it is unlikely that the average of 100 randomly selected samples out of 10,000 will be identical to the population. So, what about 98? What about 95? What about 93? Can we trust those average values?</p>
<p>이번에는 다르게 생각해보자. 10,000개로 이루어진 모집단에서 무작위하게 뽑은 100개의 샘플의 평균값이 95라는 사실을 알았다. 우리는 모집단에 대하여 아는 사실이 없다. 이번에는 교수님도 모른다. 다만 우리는 95라는 값이 모집단의 평균값과 정확하게 일치하지 않는다는 사실을 알고있다. 그렇다면 모집단의 평균값은 100일까 90일까?<br>
Let's consider a different scenario this time. We know the average of 100 randomly selected samples from a population of 10,000 is 95. We have no information about the population. Even the professor is unaware of it. However, we do know that the value of 95 does not precisely match the population mean. So, would the population mean be 100 or 90?</p>
<hr>
<p>가설검증이라는 것은 전체 데이터의 일부만을 추출하여 평균을 내고, 그 평균이 전체 데이터의 평균을 잘 반영한다는 가정 하에 전체 데이터의 평균을 구하는 작업인데, 아무리 무작위 추출을 잘 한다 하더라도 추출된 데이터의 평균은 전체 데이터의 평균에서 멀어질 수 있게 된다. 따라서, 내가 추출한 이 데이터의 평균이 원래의 전체 데이터의 평균과 얼마나 다른 값인지를 알 수 있는 방법이 필요하게 된다. 이와 같은 문제 때문에 나온 값이 p-value 이다.<br>
Hypothesis testing involves taking a subset of the entire data, calculating the average under the assumption that it reflects the mean of the entire dataset. However, no matter how well we perform random sampling, the average of the sampled data can deviate from the mean of the entire dataset. Therefore, we need a method to determine how different the average of the extracted data is from the original mean of the entire dataset. This is where the concept of p-value comes into play.</p>
<hr>
<p>&quot;모집단 정규분포 그래프의 평균값이 100 이다&quot;라는 귀무가설이 참이라는 가정 할 때, 10,000개의 데이터에서 100개의 샘플을 계속해서 뽑고 평균값을 구하고 있다고 하자. 100개의 샘플을 sampling 할 때 이론적으로 나올 수 있는 평균값의 분포가 있다. 이때, 내가 방금 뽑은 95라는 평균값 보다 더 큰 평균값이 나올 수 있는 확률이 p-value이다.<br>
When assuming that the null hypothesis states &quot;the mean of the population normal distribution is 100,&quot; and continuously drawing 100 samples from a dataset of 10,000 to calculate the mean, there is a theoretical distribution of possible mean values that can occur when sampling the 100 values. In this case, the p-value represents the probability of obtaining a mean greater than the observed average of 95 that you just drew.</p>
<hr>
<p>위의 예시로 다시 설명해보자면</p>
<ol>
<li>모집단의 평균값이 100이라는 '귀무가설'을 세웠다.</li>
<li>무작위하게 뽑은 샘플들의 평균값이 95였다.</li>
<li>95라는 평균값이 앞으로도 수행할 무작위 sampling에서 나올 확률이 p-value다.<br>
Let's explain again using the example you provided:</li>
</ol>
<ol>
<li>We set the null hypothesis as &quot;the mean of the population is 100.&quot;</li>
<li>The average of the randomly selected samples was 95.</li>
<li>The p-value represents the probability of obtaining a value as extreme as 95 or higher in future random sampling.</li>
</ol>
<p>이때, 95라는 평균값보다 더 높은 값, 96~100이 앞으로도 수행할 무작위 sampling에서 나올 확률이 97% 라고하자. (p-value=0.97)<br>
그렇다면 우리가 1) 에서 말한 '모집단의 평균값이 100'이라는 '귀무가설'은 맞는 귀무가설일까 틀린 가설일까?<br>
97% 확률로 일어날 사건이기 때문에 맞는 귀무가설이라고 할 수 있을 것이다.<br>
In this case, let's say the probability of obtaining values higher than 95, specifically in the range of 96 to 100, in future random sampling is 97% (p-value = 0.97).<br>
So, is the null hypothesis we stated in step 1, &quot;the mean of the population is 100,&quot; a valid or invalid hypothesis?<br>
Considering that the event has a 97% chance of occurring, we can conclude that it is a valid null hypothesis.</p>
<hr>
<p>이때, 귀무가설의 의미를 다시 상기해야 한다. 귀무가설은 &quot;새로울 게 없다&quot;인 가설이다. 다른 말로 귀무가설은 부정적, 소극적, 보수적, 전통적인 입장이며, 과학자들이 제발 틀리기를 기도하는 가설이다. '기존에 알려진 사실'로 생각해도 된다.<br>
It is important to revisit the meaning of the null hypothesis. The null hypothesis states that &quot;there is no new effect&quot; or &quot;there is no difference.&quot; In other words, the null hypothesis takes a negative, conservative, and traditional stance. It is a hypothesis that scientists hope to reject, as it represents the status quo or the &quot;already known fact.&quot;</p>
<p>따라서, 귀무가설이 0.05 이하의 p-value를 가질 때 우리는 일반적으로 이 귀무가설이 틀렸다고 말한다. 귀무가설이 틀리고 나서야 과학자들이 그토록 원하는 대립가설이 참이라고 말할 수 있는 것이다.<br>
Indeed, when the null hypothesis has a p-value of less than 0.05, it is generally concluded that the null hypothesis is rejected. Scientists can then state that the alternative hypothesis, which they had been hoping to support, is likely to be true.</p>
<hr>
<p>위에서는 주로 평균값에 대해서 얘기하였으나, 평균과 분산을 동시에 알아야 가설검증을 할 수 있다. 또한, 모분포의 평균과 분산값에 대한 조건이 있으면, 거기서 n 개를 sampling 했을 때 그 n 개의 평균이 얼마나 잘 나올 수 있는 값인지 중심극한정리와 표준정규분포의 성질을 이용하여 정확하게 계산이 가능하다. 따라서 그 경우마다 p-value를 구할 수 있게 된다.<br>
In the previous discussion, we mainly focused on the average value, but hypothesis testing requires knowledge of both the mean and variance. Moreover, if there are conditions regarding the mean and variance of the population distribution, it is possible to calculate precisely how well the sample mean of n values will perform using the properties of the central limit theorem and the standard normal distribution. Therefore, it becomes possible to calculate the p-value for each specific case.</p>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120383" class="record">
            <header>
                <h1 class="record-title">Prime number and Sieve of Eratosthenes</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120383</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://www.geeksforgeeks.org/python-program-to-check-whether-a-number-is-prime-or-not/">Python Program to Check Prime Number</a><br>
<a href="https://this-programmer.tistory.com/409">에라토스테네스의 체 (소수구하기. 파이썬)</a><br>
<a href="https://aiday.tistory.com/68">[Python] programmers, 프로그래머스 소수 찾기 Lv.1 (feat.에라토스테네스의 체)</a><br>
<a href="https://ye333.tistory.com/24">[파이썬/python] 에라토스테네스의 체</a><br>
<a href="https://en.wikipedia.org/wiki/Prime_number">Prime number</a><br>
<a href="https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes">Sieve of Eratosthenes</a></p>
<h1>Prime number</h1>
<ul>
<li>A prime number (or a prime) is a natural number greater than 1 that is not a product of two smaller natural numbers.</li>
<li>For example,
<ul>
<li>2, 3, 5, ,7, 11, 13, 17, 19, ...</li>
</ul>
</li>
</ul>
<h1>Sieve of Eratosthenes</h1>
<ul>
<li>Sieve of Eratosthenes is an ancient algorithm for finding all prime numbers.</li>
<li>The multiples of a given prime are generated as a sequence of numbers starting from that prime, with constant gap between them that is equal to that prime.</li>
<li>To find all the prime numbers less than or equal to a given integer $\large n$ by Eratosthenes' method:
<ol>
<li>Create a list of consecutive integers from 2 through $\large n$:<br>
(2, 3, 4, ..., $\large n$).</li>
<li>Initially, let $\large p$ equal 2, the smallest prime number.</li>
<li>Enumerate the multiples of $\large p$ by counting in increments of $\large p$ from 2$\large p$ to $\large n$,<br>
and mark them in the list.<br>
(these will be 2$\large p$, 3$\large p$, 4$\large p$, ...; the $\large p$ itself should not be marked)</li>
<li>Find the smallest number in the list greater than $\large p$ that is not marked.<br>
If there was no such number, stop.<br>
Otherwise, let $\large p$ now equal this new number (which is the next prime),<br>
and repeat from step 3.</li>
<li>When the algorithm terminates, the numbers remaining not marked in the list are all the primes below $\large n$.</li>
</ol>
</li>
<li>Note that some of the numbers may be marked more than once<br>
(e.g., 15 will be marked both for 3 and 5).
<h2>Python practice</h2>
</li>
</ul>
<pre><code class="language-python">def prime_list(n):
    sieve = [True] * n #list of 'True'

    m = int(n ** 0.5) # numbers larger than the square root of 'n' will surely be removed by the sieve.
    for i in range(2, m + 1):
        if sieve[i] == True:           # if 'i' is prime 
            for j in range(i+i, n, i): # All multiples of 'i' are 'False'
                sieve[j] = False

    #return numbers based on the list of 'True'
    return [i for i in range(2, n) if sieve[i] == True]
</code></pre>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120384" class="record">
            <header>
                <h1 class="record-title">Protein</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120384</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                

                
            </main>

            <footer class="list-content-container">
                

                
                <details class="record-backlinks" open="open">
                    <summary class="menu-title">
                        backlinks
                        <span class="badge">1</span>
                    </summary>

                    <ul class="record-backlinks-list">
                        
                        <li data-target-id="20230924120347">
                            <a href="#20230924120347">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Central Dogma</span>
                                <div class="record-links-context">
                                    <p>DNA<br>
RNA<br>
<em class="id-context" data-target-id="20230924120384">Protein</em></p>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                
            </footer>
        </article>
        
        <article id="20230924120385" class="record">
            <header>
                <h1 class="record-title">Query Strategy</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120385</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <ul>
<li>Active Learning에서</li>
</ul>
<p>참고: <a href="https://littlefoxdiary.tistory.com/52">https://littlefoxdiary.tistory.com/52</a></p>


                
            </main>

            <footer class="list-content-container">
                

                
                <details class="record-backlinks" open="open">
                    <summary class="menu-title">
                        backlinks
                        <span class="badge">1</span>
                    </summary>

                    <ul class="record-backlinks-list">
                        
                        <li data-target-id="20230924120340">
                            <a href="#20230924120340">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Active Learning</span>
                                <div class="record-links-context">
                                    <ul>
<li>학습모델이 주어진 데이터의 정보량을 평가하고, 라벨링이 필요하다고 판단되는 경우에만 인간에게 전달한다.</li>
<li>정보량을 평가할 때는 <em class="id-context" data-target-id="20230924120385">Query Strategy</em>를 사용한다. 라벨링이 필요하다고 판단되는 경우는 query하고 아닌 것은 버린다.</li>
</ul>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                
            </footer>
        </article>
        
        <article id="20230924120386" class="record">
            <header>
                <h1 class="record-title">README</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120386</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <h1>2023</h1>
<h2>September</h2>
<p>21/September/2023 paper: Real-Time DNA Sequencing in the Antarctic Dry Valleys Using the Oxford Nanopore Sequencer (ongoing)<br>
20/September/2023 paper: Dark microbiome and extremely low organics in Atacama fossil delta unveil Mars life detection limits (ongoing)<br>
03/September/2023 paper: Integration Strategies of Multi-omics Data for Machine Learning Analysis (ongoing)<br>
02/September/2023 paper: Integration Strategies of Multi-omics Data for Machine Learning Analysis (ongoing)</p>
<h2>August</h2>
<p>31/August/2023 paper: Dealing with dimensionality: the application of machine learning to multi-omics data<br>
30/August/2023 paper: Dealing with dimensionality: the application of machine learning to multi-omics data (ongoing)<br>
28/August/2023 paper: Dealing with dimensionality: the application of machine learning to multi-omics data (ongoing)<br>
27/August/2023 paper: Dealing with dimensionality: the application of machine learning to multi-omics data (ongoing)<br>
22/August/2023 Single Cell Best Practice 1.Introduction (ongoing)<br>
18/August/2023 Punnett Squares<br>
16/August/2023 Monohybrid<br>
16/August/2023 Monohybrid (ongoing)<br>
09/August/2023 Mendel's First Law and Meiosis<br>
08/August/2023 Somatic mutation vs. Germline mutation<br>
07/August/2023 Single Cell Best Practice  1.Introduction (ongoing)<br>
05/August/2023 AlphaFold1, AlphaFold2, and RoseTTAFold<br>
04/August/2023 Build Workflow from scratch - HaplotypeCaller<br>
03/August/2023 Build Workflow from scratch - HaplotypeCaller (ongoing)<br>
02/August/2023 Build Workflow from scratch - HaplotypeCaller (ongoing)<br>
01/August/2023 Build Workflow from scratch - HaplotypeCaller (ongoing)</p>
<h2>July</h2>
<p>31/July/2023 Build Workflow from scratch - HaplotypeCaller (ongoing)<br>
30/July/2023 Edwith - LinearStructureandDynamicProgramming - Chapter 7<br>
21/July/2023 Edwith - LinearStructureandDynamicProgramming - Chapter 7 (ongoing)<br>
19/July/2023 Prime number and Sieve of Eratosthenes<br>
18/July/2023 Greatest Common Divisor, Least Common Multiple, and Euclidean algorithm<br>
17/July/2023 Building Docer &amp; Google Cloud (ongoing)<br>
15/July/2023 Edwith - LinearStructureandDynamicProgramming - Chapter 6, Why Bioinformaticians use Nextflow thesedays?<br>
14/July/2023 Code modify from Edwith classes<br>
12/JUly/2023 Building Docer &amp; Google Cloud(ongoing)<br>
11/July/2023 Edwith - LinearStructureandDynamicProgramming - Chapter 6(ongoing)<br>
10/July/2023 Edwith - LinearStructureandDynamicProgramming - Chapter 5, Edwith - LinearStructureandDynamicProgramming - Chapter 6 (ongoing)<br>
08/July/2023 Edwith - LinearStructureandDynamicProgramming - Chapter 4<br>
07/July/2023 Edwith - LinearStructureandDynamicProgramming - Chapter 4 (ongoing)<br>
06/July/2023 p-value(유의확률)<br>
05/July/2023 p-value(유의확률) (ongoing)<br>
04/July/2023 Reviewing Edwith - LinearStructureandDynamicProgramming - Chapter 3<br>
03/July/2023 Reviewing Edwith - LinearStructureandDynamicProgramming - Chapter 3 (ongoing)<br>
02/July/2023 Reviewing Edwith - LinearStructureandDynamicProgramming - Chapter 2<br>
01/July/2023 Code modify from Edwith classes</p>
<h2>June</h2>
<p>30/June/2023 Reviewing Edwith - LinearStructureandDynamicProgrammng - Chapter 2 (ongoing)<br>
29/June/2023 Underfitting and Overfitting<br>
28/June/2023 Reviewing Edwith - LinearStructureandDynamicProgrammng - Chapter 1</p>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120387" class="record">
            <header>
                <h1 class="record-title">Real-Time DNA Sequencing in the Antarctic Dry Valleys Using the Oxford Nanopore Sequencer</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120387</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5362188/">Johnson SS, Zaikova E, Goerlitz DS, Bai Y, Tighe SW. Real-Time DNA Sequencing in the Antarctic Dry Valleys Using the Oxford Nanopore Sequencer. J Biomol Tech. 2017 Apr;28(1):2-7. doi: 10.7171/jbt.17-2801-009. Epub 2017 Mar 22. PMID: 28337073; PMCID: PMC5362188.</a></p>
<h1>Abstract</h1>
<ul>
<li>Oxford Nanopore Technologies에서 개발한 소형 <a href="#20230924120379" title="Nanopore" class="record-link">Nanopore</a> sequencer를 활용하여, 연구실이 아닌 야외에서 DNA sequencing 작업을 수행하고, 인터넷에 연결되지 않은 환경에서 소프트웨어를 활용해 분석까지 수행했다.</li>
<li>장소는 남극의 McMurdo Dry Valleys였다.</li>
<li>사용된 sequencer는 MK1B MinION sequencer였다.</li>
<li>2시간 30분만에 야외환경에서 sequencing protocol을 완수할 수 있었다.</li>
</ul>
<p>#Sarah_Stewart_Johnson</p>


                
            </main>

            <footer class="list-content-container">
                
                <details class="record-links" open="open">
                    <summary class="menu-title">
                        links
                        <span class="badge">1</span>
                    </summary>

                    <ul class="record-links-list">
                        
                        <li data-target-id="20230924120379">
                            <a href="#20230924120379">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Nanopore</span>
                                <div class="record-links-context">
                                    <p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5362188/">Johnson SS, Zaikova E, Goerlitz DS, Bai Y, Tighe SW. Real-Time DNA Sequencing in the Antarctic Dry Valleys Using the Oxford Nanopore Sequencer. J Biomol Tech. 2017 Apr;28(1):2-7. doi: 10.7171/jbt.17-2801-009. Epub 2017 Mar 22. PMID: 28337073; PMCID: PMC5362188.</a></p>
<h1>Abstract</h1>
<ul>
<li>Oxford Nanopore Technologies에서 개발한 소형 <em class="id-context" data-target-id="20230924120379">Nanopore</em> sequencer를 활용하여, 연구실이 아닌 야외에서 DNA sequencing 작업을 수행하고, 인터넷에 연결되지 않은 환경에서 소프트웨어를 활용해 분석까지 수행했다.</li>
<li>장소는 남극의 McMurdo Dry Valleys였다.</li>
<li>사용된 sequencer는 MK1B MinION sequencer였다.</li>
<li>2시간 30분만에 야외환경에서 sequencing protocol을 완수할 수 있었다.</li>
</ul>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                

                
            </footer>
        </article>
        
        <article id="20230924120388" class="record">
            <header>
                <h1 class="record-title">RNA</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120388</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="#20230924120376" title="mRNA" class="record-link">mRNA</a></p>


                
            </main>

            <footer class="list-content-container">
                
                <details class="record-links" open="open">
                    <summary class="menu-title">
                        links
                        <span class="badge">1</span>
                    </summary>

                    <ul class="record-links-list">
                        
                        <li data-target-id="20230924120376">
                            <a href="#20230924120376">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>mRNA</span>
                                <div class="record-links-context">
                                    <p><em class="id-context" data-target-id="20230924120376">mRNA</em></p>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                

                
                <details class="record-backlinks" open="open">
                    <summary class="menu-title">
                        backlinks
                        <span class="badge">1</span>
                    </summary>

                    <ul class="record-backlinks-list">
                        
                        <li data-target-id="20230924120347">
                            <a href="#20230924120347">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Central Dogma</span>
                                <div class="record-links-context">
                                    <p>DNA<br>
<em class="id-context" data-target-id="20230924120388">RNA</em><br>
Protein</p>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                
            </footer>
        </article>
        
        <article id="20230924120389" class="record">
            <header>
                <h1 class="record-title">scRNA-seq</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120389</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p>Single Cell RNA Sequencing<br>
<a href="#20230924120377" title="Multi-omics" class="record-link">Multi-omics</a><br>
<a href="#20230924120392" title="Spatial Transcriptomics" class="record-link">Spatial Transcriptomics</a></p>


                
            </main>

            <footer class="list-content-container">
                
                <details class="record-links" open="open">
                    <summary class="menu-title">
                        links
                        <span class="badge">2</span>
                    </summary>

                    <ul class="record-links-list">
                        
                        <li data-target-id="20230924120377">
                            <a href="#20230924120377">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Multi-omics</span>
                                <div class="record-links-context">
                                    <p>Single Cell RNA Sequencing<br>
<em class="id-context" data-target-id="20230924120377">Multi-omics</em><br>
Spatial Transcriptomics</p>

                                </div>
                            </a>
                        </li>
                        
                        <li data-target-id="20230924120392">
                            <a href="#20230924120392">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Spatial Transcriptomics</span>
                                <div class="record-links-context">
                                    <p>Single Cell RNA Sequencing<br>
Multi-omics<br>
<em class="id-context" data-target-id="20230924120392">Spatial Transcriptomics</em></p>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                

                
            </footer>
        </article>
        
        <article id="20230924120390" class="record">
            <header>
                <h1 class="record-title">SingleCellBestPractice-1.Introduction</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120390</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://www.sc-best-practices.org/preamble.html">https://www.sc-best-practices.org/preamble.html</a><br>
*2023년 8월 기준으로 작성된 문서 내용을 복사한 것이며, 해당 문서는 꾸준히 수정이 되어지고 있음<br>
*개인적으로 읽을만했던 내용만 복사해옴<br>
*문장 전체를 직역하지 않음</p>
<h1>INTRODUCTION</h1>
<h2><a href="https://www.sc-best-practices.org/introduction/prior_art.html">1. Prior art</a></h2>
<h2><a href="https://www.sc-best-practices.org/introduction/scrna_seq.html#">2. Single-cell RNA sequencing</a></h2>
<h2>2.4 RNA sequencing</h2>
<ul>
<li>RNA-Seq largely follows the DNA sequencing protocols, but includes a reverse transcription step where complementary DNA (cDNA) is synthesized from the RNA template.<br>
<em>RNA-Seq은 DNA sequencing protocol을 많이 따르지만, 차이점으로 RNA template에서 complementary DNA를 합성하기 위한 reverse transcription 과정이 포함된다.</em></li>
</ul>
<h2>2.5 Single-cell RNA sequencing</h2>
<p><strong>2.5.1. Overview</strong></p>
<ul>
<li>
<p>Sequencing of RNA can be mainly conducted in two ways:<br>
<em>RNA sequencing 주요 방법 두 가지:</em></p>
<ol>
<li>Either by sequencing the mixed RNA from the source of interest across cells (bulk sequencing)<br>
<em>여러 세포에서 유래한 RNA들을 sequencing하기</em></li>
<li>by sequencing the transcriptomes of the cells individually (single-cell sequencing).<br>
<em>각 세포에서 유래한 transcriptiomes를 sequencing하기</em></li>
</ol>
</li>
<li>
<p>Some drugs or perturbations may affect only specific cell types or interactions between cell types.<br>
<em>특정 약물이나 교란법(perturbation)은 특정한 세포종류들 혹은 상호작용 영향일 미칠 수 있다.</em></p>
</li>
<li>
<p>To uncover such relationships, it is vital to examine gene expression on a single-cell level.<br>
<em>이러한 관계를 밝혀내기 위해 single-cell 수준의 유전자 발현 조사가 중요.</em></p>
</li>
<li>
<p>Single-cell RNA-Seq (scRNA-Seq) does, however, come with several caveats.<br>
<em>sRNA-seq의 단점(caveat)</em></p>
<ul>
<li>First, single-cell experiments are generally more expensive and more difficult to properly conduct.<br>
<em>비싸고 어렵다</em></li>
<li>Second, the downstream analysis becomes more complex due to the increased resolution, and it is easier to draw false conclusions.<br>
<em>Resolution이 증가 -&gt; downstream 분석과정 복잡해짐 &amp; false 결과빈도 증가</em></li>
</ul>
</li>
<li>
<p>Just like bulk sequencing, single-cell sequencing requires:</p>
<ul>
<li>lysis</li>
<li>reverse transcription</li>
<li>amplification</li>
<li>the eventual sequencing.</li>
</ul>
</li>
<li>
<p>In addition, single-cell sequencing requires cell isolation and a physical separation into smaller reaction chambers or another form of cell labeling to be able to map the obtained transcriptomes back to the cells of origin later on.<br>
<em>Single-cell sequencing은  transcriptome이 어느 세포에서 유래했는지 추적하기 위해 특유의 cell isolation 과정과 모든 세포들을 물리적으로 분리하는 과정이 필수적으로 필요하다.(작은 chamber든 labelling이든)</em></p>
</li>
<li>
<p>Hence, these are also the steps where most single-cell assays differ:<br>
<em>따라서, 여러 single-cell sequencing 방법들이 다음의 과정들을 서로 다르게 수행하는 경우가 많다</em></p>
<ul>
<li>single-cell isolation</li>
<li>transcript amplification</li>
<li>sequencing(depending on the sequencing machine)</li>
</ul>
</li>
</ul>
<p><strong>2.5.2 Transcript quantification</strong></p>
<ul>
<li>
<p>Transcript quantification is the process of counting the hits of the sequenced transcripts against the gene sequences.<br>
<em>Transcript quantification이란 유전자 서열과 대응하는 sequenced transcripts 수를 세는 과정이다.</em></p>
</li>
<li>
<p>There are two major approaches to transcript quantification:</p>
<ol>
<li>Full-length protocols
<ul>
<li>try to cover the whole transcript uniformly with sequencing reads<br>
<em>transcript 전체를 sequencing reads와 비교</em></li>
</ul>
</li>
</ol>
<ul>
<li>Tag-based protocols
<ul>
<li>only capture the 5’ or 3’ ends.<br>
<em>5' end 혹은 3' end만 비교</em></li>
</ul>
</li>
</ul>
</li>
<li>
<p>The usage of UMIs is a common solution to quantify the original, non-duplicated molecules.<br>
<em>UMIs는 증폭과정을 거치지 않은 본래의 (유전)분자물질을 정량하는 일반적인 방법이다.</em></p>
<ul>
<li>The UMIs serve as molecular barcodes and are also sometimes referred to as random barcodes.<br>
<em>UMIs는 분자바코드로 때로는 랜덤바코드로 불린다.</em></li>
<li>These ‘barcodes’ consist of short random nucleotide sequences that are added to every molecule in the sample as a unique tag.<br>
<em>이 '바코드'들은 짧은 뉴클레오타이드 서열들이 랜덤하게 이루어진 형태이다. 이 서열들은 하나의 unique tag로 붙여지게 된다.</em></li>
<li>UMIs must be added during library generation before the amplification step.<br>
<em>UMIs는 반드시 amplication step 이전인 library generation 과정에서 붙여져야한다.</em></li>
</ul>
</li>
</ul>
<p><strong>2.5.3 Single-cell sequencing protocols</strong></p>
<ul>
<li>Currently, three types of single-cell sequencing protocols exist, which are grouped primarily by their cell isolation protocols:<br>
<em>현재는 cell isolation protocol에 따라 3개의 방식으로 분류한다.</em>
<ol>
<li>Microfluidic device-based strategies
<ul>
<li>cells are encapsulated into hydrogel droplets<br>
<em>세포를 hydrogel droplet을 이용해서 분리</em></li>
</ul>
</li>
<li>Well plate based protocols
<ul>
<li>cells are physically separated into wells<br>
<em>세포를 well을 이용해서 분리</em></li>
</ul>
</li>
<li>The commercial Fluidigm C1 microfluidic chip based solution
<ul>
<li>loads and separates cells into small reaction chambers.<br>
<em>세포를 small reaction chamber를 이용해서 분리</em></li>
</ul>
</li>
</ol>
</li>
</ul>
<p><strong>2.5.3.4. Nanopore single-cell transcriptome sequencing</strong></p>
<ul>
<li>
<p><a href="https://www.nature.com/articles/s41467-020-17800-6">Lebrigand <em>et al.</em>, 2020</a>introduced ScNaUmi-seq (Single-cell Nanopore sequencing with UMIs) which combines Nanopore sequencing with cell barcode and UMI assignment.<br>
Lebrigand는 나노포어 시퀀싱과 바코드, UMI 식별법을 결합한 ScNaUmi-seq을 공개했다.</p>
<ul>
<li>The barcode assignment is guided with Illumina data by comparing the cell bar code sequences found in the Nanopore reads with those recovered from the Illumina reads for the same region or gene.<br>
바코드 식별법은</li>
<li>However, this effectively requires two single-cell libraries.</li>
</ul>
</li>
<li>
<p>scCOLOR-seq(<a href="https://www.nature.com/articles/s41587-021-00965-w">Philpott <em>et al.</em>, 2021</a>) computationally identifies barcodes without errors using nucleotide pair complementary across the full length of the barcode.</p>
<ul>
<li>These barcodes are then used as guides to correct the remaining erroneous barcodes.</li>
<li>A modified UMI-tools directional network based method corrects for UMI sequence duplication.</li>
</ul>
</li>
<li>
<p>Strengths:</p>
<ul>
<li>Recovers splicing and sequence heterogeneity information</li>
</ul>
</li>
<li>
<p>Weaknesses:</p>
<ul>
<li>Nanopore reagents are expensive.</li>
<li>High cell barcode recovery error rates.</li>
<li>Depending on the protocol, barcode assignment is guided with Illumina data requiring two sequencing assays.</li>
</ul>
</li>
</ul>
<h4>2.5.3.5 Summary</h4>
<ul>
<li>
<p>In summary, we strongly recommend that wet lab and dry lab scientists select the sequencing protocol based on the aim of the study.</p>
<ul>
<li>Is a deep characterization of a specific cell type population desired?
<ul>
<li>In this case one of the plate-based methods may be more suitable.</li>
</ul>
</li>
<li>On the contrary, droplet based assays will capture heterogeneous mixtures better, allowing for a more broad characterization of the sequenced cells.</li>
<li>Moreover, if the budget is a limiting factor, the protocol of choice should be more cost-effective and robust.</li>
</ul>
</li>
<li>
<p>When analyzing the data, be aware of the sequencing assay specific biases.</p>
<ul>
<li>For an extensive comparison of all single-cell sequencing protocols, we recommend the “Benchmarking single-cell RNA-sequencing protocols for cell atlas projects” paper by <a href="https://www.nature.com/articles/s41587-020-0469-4">Mereu et al., 2020</a>.</li>
</ul>
</li>
</ul>
<h3>2.5.4 single-cell vs single-nuclei</h3>
<ul>
<li>
<p>So far we have only been discussing single-cell assays, but it is also possible to only sequence the nuclei of the cells.</p>
</li>
<li>
<p>Single-cell profiling does not always provide an unbiased view on cell types for specific tissues or organs, such as, for example, the brain.</p>
<ul>
<li>During the tissue dissociation process, some cell types are more vulnerable and therefore difficult to capture.</li>
<li>Moreover, single-cell sequencing highly relies on fresh tissue, making it difficult to make use of tissue biobanks.</li>
</ul>
</li>
<li>
<p>On the other hand, the nuclei are more resistant to mechanical force, and can be safely isolated from frozen tissue without the use of tissue dissociation enzymes[Krishnaswami et al., 2016].</p>
</li>
<li>
<p>Both options have varying applicability across tissues and sample types, and the resulting biases and uncertainties are still not fully uncovered.</p>
<ul>
<li>It has been shown already that nuclei accurately reflect all transcriptional patterns of cells[Ding et al., 2020].</li>
<li>The choice of single-cell versus single-nuclei in the experimental design is mostly driven by the type of tissue sample.</li>
</ul>
</li>
<li>
<p>Data analysis however should be aware of the fact that dissociation ability will have a strong effect on the potentially observable cell types.</p>
</li>
<li>
<p>Therefore, we strongly encourage discussions between wet lab and dry lab scientists concerning the experimental design.</p>
</li>
</ul>
<h1><a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html">3. Raw data processing</a></h1>
<ul>
<li>Here, we will primarily refer to this phase of processing as “raw data processing”, and our focus will be on the phase of data analysis that begins with lane-demultiplexed FASTQ files, and ends with a count matrix representing the estimated number of distinct molecules arising from each gene within each quantified cell, potentially stratified by the inferred splicing status of each molecule (Fig. 3.1).</li>
</ul>
<p><img src="https://www.sc-best-practices.org/_images/overview_raw_data_processing.jpg" alt="Fig. 3.1."><br>
Fig. 3.1 An overview of the topics discussed in this chapter.<br>
In the plot, “txome” stands for transcriptome.</p>
<ul>
<li>This count matrix then serves as the input for the myriad methods that have been developed for various analyses carried out with scRNA-seq data [Zappia and Theis, 2021], from methods for normalization, integration, and filtering through methods for inferring cell types, developmental trajectories, and expression dynamics.</li>
<li>In particular, we will cover the fundamental steps in raw data processing, including read alignment/mapping, cell barcode (CB) identification and correction, and the estimation of molecule counts through the resolution of unique molecular identifiers (UMIs).</li>
</ul>
<blockquote>
<p>A note on what precedes raw data processing</p>
<p>The decision of where to begin discussing raw data processing is somewhat arbitrary. We note that while, here, we consider starting with lane-demultiplexed FASTQ files as the raw input, even this already represents data that has been processed and transformed from raw measurements. Further, some of the processing that precedes the generation of the FASTQ files is relevant to challenges that may arise in subsequent processing.</p>
</blockquote>
<h2>3.1 Raw data quality control</h2>
<ul>
<li>Once raw FASTQ files have been obtained, the quality of the reads themselves can be quickly diagnosed by running a quality control (QC) tool, such as <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/"><code>FastQC</code></a>, to assess read quality, sequence content, etc.
<ul>
<li>If run successfully, <code>FastQC</code> generates a QC report for each input FASTQ file.</li>
<li>Overall, this report summarizes the quality score, base content, and certain relevant statistics to spot potential problems originating from library preparation or sequencing.</li>
</ul>
</li>
<li>Although nowadays, single-cell raw data processing tools internally evaluate some quality checks that are important for single-cell data, such as the N content of sequences and the fraction of mapped reads, it is still a good habit to run a quick quality check before processing the data, as it evaluates other useful QC metrics.</li>
</ul>
<h2>3.2 Alignment and mapping</h2>
<ul>
<li>Mapping or alignment is a fundamental step in single-cell raw data processing. It refers to the process of determining the potential loci of origin of each sequenced fragment (e.g., the set of genomic or transcriptomic loci that are similar to the read).</li>
<li>Depending on the sequencing protocol, the resulting raw sequence file contains:
<ul>
<li>the cell-level information</li>
<li>commonly known as cell barcodes (CB)</li>
<li>the unique molecule identifier (UMI)</li>
<li>the raw cDNA sequence (read sequence)<br>
generated from the molecule.</li>
</ul>
</li>
<li>Tools such as:
<ul>
<li>Cell Ranger (commercial software from 10x Genomics) [Zheng et al., 2017]</li>
<li>zUMIs [Parekh et al., 2018]</li>
<li>alevin [Srivastava et al., 2019]</li>
<li>RainDrop [Niebler et al., 2020]</li>
<li>kallisto|bustools [Melsted et al., 2021]</li>
<li>STARsolo [Kaminow et al., 2021]</li>
<li>alevin-fry [He et al., 2022]<br>
provide dedicated treatment for aligning scRNA-seq reads along with parsing of technical read content (CBs and UMIs), as well as methods for demultiplexing and UMI resolution.</li>
</ul>
</li>
<li>While the specific algorithms, data structures, and time and space trade-offs made by different alignment and mapping approaches can vary greatly, we can roughly categorize existing approaches along two axes:
<ul>
<li>The type of mapping they perform</li>
<li>The type of reference sequence against which they map reads.</li>
</ul>
</li>
</ul>
<h3>3.2.1. Types of mapping</h3>
<ul>
<li>
<p>Here we consider three main types of mapping algorithms that are most commonly applied in the context of mapping sc/snRNA-seq data:</p>
<ul>
<li>spliced alignment</li>
<li>contiguous alignment</li>
<li>variations of lightweight mapping</li>
</ul>
</li>
<li>
<p>First, let us draw a distinction here between alignment-based approaches and lightweight mapping-based approaches (<a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#raw-proc-fig-alignment-mapping">Fig. 3.14</a>).</p>
</li>
</ul>
<p><img src="https://www.sc-best-practices.org/_images/alignment_vs_mapping.png" alt="image"><br>
Fig. 3.14 An abstract overview of the alignment-based method and lightweight mapping-based method.</p>
<ul>
<li>
<p>Alignment approaches apply a range of different heuristics to determine the potential loci from which reads may arise and then subsequently score, at each putative locus, the best nucleotide-level alignment between the read and reference, typically using dynamic programming-based approaches.</p>
</li>
<li>
<p>The exact type of dynamic programming algorithm used depends on the type of alignment being sought.</p>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Needleman%E2%80%93Wunsch_algorithm">Global alignment</a> is applicable for the case where the entirety of the query and reference sequence are to be aligned.</li>
<li><a href="https://en.wikipedia.org/wiki/Smith%E2%80%93Waterman_algorithm">local alignment</a> is applicable when, possibly, only a subsequence of the query is expected to match a subsequence of the reference.</li>
</ul>
</li>
<li>
<p>Typically, the models most applicable for short-read alignment are neither fully global nor fully local, but fall into a category of semi-global alignment where the majority of the query is expected to align to some substring of the reference (this is often called “fitting” alignment).</p>
</li>
<li>
<p>Moreover, it is sometimes common to allow soft-clipping of the alignment so that the penalties for mismatches, insertions, or deletions appearing at the very start or end of the read are ignored or down-weighted. This is commonly done using <a href="https://github.com/smarco/WFA2-lib#-33-alignment-span">“extension” alignment</a>.</p>
</li>
<li>
<p>Though these modifications change the specific rules used in the dynamic programming recurrence and traceback, they do not fundamentally change its overall complexity.</p>
</li>
<li>
<p>Apart from these general alignment techniques, a number of more sophisticated modifications and heuristics have been designed to make the alignment process more practical and efficient in the context of aligning genomic sequencing reads.</p>
<ul>
<li>For example, <code>banded alignment</code> <a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id96" title="Kun-Mao Chao, William R. Pearson, and Webb Miller. Aligning two sequences within a specified diagonal band. Bioinformatics, 8(5):481–487, 1992. URL: https://doi.org/10.1093/bioinformatics/8.5.481, doi:10.1093/bioinformatics/8.5.481.">Chao <em>et al.</em>, 1992</a> is a popular heuristic included in the alignment implementation of many popular tools, which avoids computing large parts of the dynamic programming table if the user is uninterested in alignment scores below a certain threshold.</li>
<li>Likewise, other heuristics like X-drop <a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id108" title="Zheng Zhang, Scott Schwartz, Lukas Wagner, and Webb Miller. A greedy algorithm for aligning term`DNA` sequences. Journal of Computational Biology, 7(1-2):203–214, February 2000. URL: https://doi.org/10.1089/10665270050081478, doi:10.1089/10665270050081478.">Zhang <em>et al.</em>, 2000</a> and Z-drop <a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id97" title="Heng Li. Minimap2: pairwise alignment for nucleotide sequences. Bioinformatics, 34(18):3094–3100, May 2018. URL: https://doi.org/10.1093/bioinformatics/bty191, doi:10.1093/bioinformatics/bty191.">Li, 2018</a> are popular for pruning un-promising alignments early.</li>
<li>Recent algorithmic progress, such as wavefront alignment <a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id107" title="Santiago Marco-Sola, Jordan M Eizenga, Andrea Guarracino, Benedict Paten, Erik Garrison, and Miquel Moreto. Optimal gap-affine alignment in o(s) space. bioRxiv, April 2022. URL: https://doi.org/10.1101/2022.04.14.488380, doi:10.1101/2022.04.14.488380.">Marco-Sola <em>et al.</em>, 2022</a>, <a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id105" title="Santiago Marco-Sola, Juan Carlos Moure, Miquel Moreto, and Antonio Espinosa. Fast gap-affine pairwise alignment using the wavefront algorithm. Bioinformatics, September 2020. URL: https://doi.org/10.1093/bioinformatics/btaa777, doi:10.1093/bioinformatics/btaa777.">Marco-Sola <em>et al.</em>, 2020</a>], allows for determining optimal alignments in asymptotically (and practically) shorter time and smaller space if high-scoring (low divergence) alignments exist.</li>
<li>Finally, considerable effort has also been devoted to optimizing data layout and computation in a way that takes advantage of instruction-level parallelism <a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id100" title="Michael Farrar. Striped Smith–Waterman speeds database searches six times over other SIMD implementations. Bioinformatics, 23(2):156–161, 2007.">Farrar, 2007</a>, <a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id101" title="Torbjørn Rognes and Erling Seeberg. Six-fold speed-up of Smith–Waterman sequence database searches using parallel processing on common microprocessors. Bioinformatics, 16(8):699–706, 2000.">Rognes and Seeberg, 2000</a>, <a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id99" title="Andrzej Wozniak. Using video-oriented instructions to speed up sequence comparison. Bioinformatics, 13(2):145–150, 1997.">Wozniak, 1997</a>],and to expressing the alignment dynamic programming recurrences in a manner that is highly amenable to data parallelism and vectorization (e.g., as in the difference encoding of <a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id106" title="Hajime Suzuki and Masahiro Kasahara. Introducing difference recurrence relations for faster semi-global alignment of long sequences. BMC Bioinformatics, February 2018. URL: https://doi.org/10.1186/s12859-018-2014-8, doi:10.1186/s12859-018-2014-8.">Suzuki and Kasahara 2018</a>.</li>
</ul>
</li>
<li>
<p>Most widely-used alignment tools make use of highly-optimized and vectorized alignment implementations.</p>
</li>
<li>
<p>In addition to the alignment score, a backtrace of the actual alignment that yields this score may be obtained. Such information is typically encoded as a <code>CIGAR</code> string (short for “Concise Idiosyncratic Gapped Alignment Report”), a compressed alphanumeric representation of the alignment, within the SAM or BAM file that is output.</p>
<ul>
<li>An example of a <code>CIGAR</code> string is <code>3M2D4M</code>, which represents that the alignment has three matches or mismatches, followed by a deletion of length two from the read (bases present in the reference but not the read), followed by four more matches or mismatches.</li>
<li>Other variants of the <code>CIGAR</code> string can also delineate between matches or mismatches.
<ul>
<li>For example, <code>3=2D2=2X</code> is a valid extended <code>CIGAR</code> string encoding the same alignment as just described, except that it makes clear that the three bases before the deletion constitute matches and that after the deletion, there are two matched bases followed by two mismatched bases.</li>
</ul>
</li>
<li>A detailed description of the <code>CIGAR</code> string format can be found in <a href="https://samtools.github.io/hts-specs/SAMv1.pdf">the SAMtools manual</a> or <a href="https://genome.sph.umich.edu/wiki/SAM#What_is_a_CIGAR.3F">the SAM wiki page of UMICH</a>.</li>
</ul>
</li>
<li>
<p>At the cost of computing such scores, alignment-based approaches know the quality of each potential mapping of a read, which they can then use to filter reads that align well to the reference from mappings that arise as the result of low complexity or “spurious” matches between the read and reference.</p>
</li>
<li>
<p>Alignment-based approaches include both:</p>
<ul>
<li>traditional “full-alignment” approaches, as implemented in tools such as:
<ul>
<li><code>STAR</code><a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id149" title="Alexander Dobin, Carrie A Davis, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha, Philippe Batut, Mark Chaisson, and Thomas R Gingeras. Star: ultrafast universal rna-seq aligner. Bioinformatics, 29(1):15–21, 2013.">Dobin <em>et al.</em>, 2013</a></li>
<li><code>STARsolo</code><a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id115" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a></li>
</ul>
</li>
<li><em>selective-alignment</em> implemented in:
<ul>
<li><code>salmon</code><a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id139" title="Avi Srivastava, Laraib Malik, Hirak Sarkar, Mohsen Zakeri, Fatemeh Almodaresi, Charlotte Soneson, Michael I Love, Carl Kingsford, and Rob Patro. Alignment and mapping methodology influence transcript abundance estimation. Genome Biology, 21(1):1–29, 2020.">Srivastava <em>et al.</em>, 2020</a></li>
<li><code>alevin</code><a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id111" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a><br>
which score mappings but omit the computation of the optimal alignment’s backtrace.</li>
</ul>
</li>
</ul>
</li>
<li>
<p>Alignment-based approaches can be further categorized into:</p>
<ol>
<li>spliced-alignment</li>
<li>contiguous-alignment approaches<br>
(currently, there are no lightweight-mapping approaches that perform spliced mapping).</li>
</ol>
<ul>
<li>Spliced-alignment approaches are capable of aligning a sequence read over several distinct segments of a reference, allowing potentially large gaps between the regions of the reference that align well with the corresponding parts of the read.
<ul>
<li>These alignment approaches are typically applied when aligning RNA-seq reads to the genome, since the alignment procedure must be able to align reads that span across one or more splice junctions of the transcript, where a sequence that is contiguous in the read may be separated by intron and exon subsequences (potentially spanning kilobases of sequence) in the reference.</li>
<li>Spliced alignment is a challenging problem, particularly in cases where only a small region of the read spans a splicing junction, since very little informative sequence may be available to place the segment of the read that overhangs the splice junction by only a small amount.</li>
</ul>
</li>
<li>On the other hand, contiguous alignment seeks a contiguous substring of the reference that aligns well against the read.</li>
<li>In such alignment problems, though small insertions and deletions may be allowed, large gaps such as those observed in spliced alignments are typically not tolerated.</li>
</ul>
</li>
<li>
<p>Finally, we can draw a distinction between alignment-based methods such as those described above and lightweight-mapping methods, which include approaches such as:</p>
<ul>
<li>pseudoalignment <a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id135" title="Nicolas L Bray, Harold Pimentel, Páll Melsted, and Lior Pachter. Near-optimal probabilistic term`RNA`-seq quantification. Nature biotechnology, 34(5):525–527, 2016.">Bray <em>et al.</em>, 2016</a></li>
<li>quasi-mapping <a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id134" title="Avi Srivastava, Hirak Sarkar, Nitish Gupta, and Rob Patro. Rapmap: a rapid, sensitive and accurate tool for mapping term`rna`-seq reads to transcriptomes. Bioinformatics, 32(12):i192–i200, 2016.">Srivastava <em>et al.</em>, 2016</a></li>
<li>pseudoalignment with structural constraints <a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id118" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>.</li>
</ul>
</li>
<li>
<p>Such approaches generally achieve superior speed by avoiding nucleotide-level alignment between the read and reference sequences.</p>
</li>
<li>
<p>Instead, they base the determination of the reported mapping loci of a read on a separate set of rules and heuristics that may look only at the set of matching k-mers or other types of exact matches (e.g., via a consensus rule) and, potentially, their orientations and relative positions with respect to each other on both the read and reference (e.g., chaining).</p>
<ul>
<li>This can lead to substantially increased speed and mapping throughput, but also precludes easily-interpretable score-based assessments of whether or not the matches between the read and reference constitute a good match capable of supporting a high-quality alignment.</li>
</ul>
</li>
</ul>
<h3>3.2.2. Mapping against different reference sequences</h3>
<ul>
<li>While different choices can be made among the mapping algorithms themselves, different choices can <em>also</em> be made about the reference against which the read is mapped.</li>
<li>We consider three main categories of reference sequence against which a method might map:
<ul>
<li>The full (likely annotated) reference genome</li>
<li>The annotated transcriptome</li>
<li>An augmented transcriptome</li>
</ul>
</li>
<li>It is also worth noting that, currently, not all combinations of mapping algorithms and reference sequences are possible.
<ul>
<li>For example, lightweight mapping-based algorithms do not currently exist that can perform spliced mapping of reads against a reference genome.</li>
</ul>
</li>
</ul>
<h4>3.2.2.1 Mapping to the full genome</h4>
<ul>
<li>
<p>The first type of reference, against which a method may map, consists of the entire genome of the target organism, usually with the annotated transcripts considered during mapping.</p>
<ul>
<li>Tools like:
<ul>
<li><code>zUMIs</code> <a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id121" title="Swati Parekh, Christoph Ziegenhain, Beate Vieth, Wolfgang Enard, and Ines Hellmann. zUMIs - a fast and flexible pipeline to process term`RNA` sequencing data with UMIs. GigaScience, May 2018. URL: https://doi.org/10.1093/gigascience/giy059, doi:10.1093/gigascience/giy059.">Parekh <em>et al.</em>, 2018</a></li>
<li><code>Cell Ranger</code> <a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id151" title="Grace X. Y. Zheng, Jessica M. Terry, Phillip Belgrader, Paul Ryvkin, Zachary W. Bent, Ryan Wilson, Solongo B. Ziraldo, Tobias D. Wheeler, Geoff P. McDermott, Junjie Zhu, Mark T. Gregory, Joe Shuga, Luz Montesclaros, Jason G. Underwood, Donald A. Masquelier, Stefanie Y. Nishimura, Michael Schnall-Levin, Paul W. Wyatt, Christopher M. Hindson, Rajiv Bharadwaj, Alexander Wong, Kevin D. Ness, Lan W. Beppu, H. Joachim Deeg, Christopher McFarland, Keith R. Loeb, William J. Valente, Nolan G. Ericson, Emily A. Stevens, Jerald P. Radich, Tarjei S. Mikkelsen, Benjamin J. Hindson, and Jason H. Bielas. Massively parallel digital transcriptional profiling of single cells. Nature Communications, 8(1):14049, Jan 2017. URL: https://doi.org/10.1038/ncomms14049, doi:10.1038/ncomms14049.">Zheng <em>et al.</em>, 2017</a></li>
<li><code>STARsolo</code> <a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id115" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a><br>
take this approach.</li>
</ul>
</li>
<li>Since many of the reads arise from spliced transcripts, such an approach also requires the use of a splice-aware alignment algorithm where the alignment for a read can be split across one or more splicing junctions.</li>
</ul>
</li>
<li>
<p>The main benefits of this approach are that reads arising from any location in the genome, not just from annotated transcripts, can be accounted for.</p>
<ul>
<li>Since these approaches require constructing a genome-wide index, there is little marginal cost for reporting not only the reads that map to known spliced transcripts but also reads that overlap or align within introns, making the alignment cost when using such approaches very similar for both single-cell and single-nucleus data.</li>
<li>A final benefit is that even reads residing outside of the annotated transcripts, exons, and introns can be accounted for by such methods, which can enable <em>post hoc</em> augmentation of the set of quantified loci (e.g., as is done by Pool <em>et al.</em> <a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id142" title="Allan-Hermann Pool, Helen Poldsam, Sisi Chen, Matt Thomson, and Yuki Oka. Enhanced recovery of single-cell term`RNA`-sequencing reads for missing gene expression data. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449, arXiv:https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449.full.pdf, doi:10.1101/2022.04.26.489449.">2022</a>] by adding expressed UTR extensions to transcript annotations in a sample-specific and data-driven manner) and potentially increase gene detection and quantification sensitivity.</li>
</ul>
</li>
<li>
<p>On the other hand, the versatility of the strategy of performing spliced alignment against the full genome comes with certain trade-offs.</p>
<ul>
<li>First, the most commonly-used alignment tools that adopt this strategy in the single-cell space have substantial memory requirements.
<ul>
<li>Due to its speed and versatility, most of these tools are based upon the STAR <a href="https://www.sc-best-practices.org/introduction/raw_data_processing.html#id149" title="Alexander Dobin, Carrie A Davis, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha, Philippe Batut, Mark Chaisson, and Thomas R Gingeras. Star: ultrafast universal rna-seq aligner. Bioinformatics, 29(1):15–21, 2013.">Dobin <em>et al.</em>, 2013</a> aligner.</li>
<li>Yet, for a human-scale genome, constructing and storing the index can require upwards of 32 GB of memory.</li>
<li>The use of a sparse <a href="https://en.wikipedia.org/wiki/Suffix_array">suffix array</a> can reduce the final index size by close to a factor of 2, but this comes at a reduction in alignment speed, and it still requires larger memory for the initial construction.</li>
</ul>
</li>
<li>Second, the increased difficulty of spliced alignment compared to contiguous alignment and the fact that current spliced-alignment tools must explicitly compute a score for each read alignment, means that this approach comes at an increased computational cost compared to the alternatives.</li>
<li>Finally, such an approach requires the genome of the organism under study is available.
<ul>
<li>While this is not problematic for the most commonly-studied reference organisms and is rarely an issue, it can make using such tools difficult for non-model organisms where only a transcriptome assembly may be available.</li>
</ul>
</li>
</ul>
</li>
</ul>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120391" class="record">
            <header>
                <h1 class="record-title">SMRT</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120391</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p>Single molecule real time sequencing from PacBio</p>


                
            </main>

            <footer class="list-content-container">
                

                
                <details class="record-backlinks" open="open">
                    <summary class="menu-title">
                        backlinks
                        <span class="badge">1</span>
                    </summary>

                    <ul class="record-backlinks-list">
                        
                        <li data-target-id="20230924120374">
                            <a href="#20230924120374">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>Long-read Sequencing</span>
                                <div class="record-links-context">
                                    <p>Also known as third-generation sequencing<br>
Nanopore<br>
<em class="id-context" data-target-id="20230924120391">SMRT</em></p>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                
            </footer>
        </article>
        
        <article id="20230924120392" class="record">
            <header>
                <h1 class="record-title">Spatial Transcriptomics</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120392</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                

                
            </main>

            <footer class="list-content-container">
                

                
                <details class="record-backlinks" open="open">
                    <summary class="menu-title">
                        backlinks
                        <span class="badge">1</span>
                    </summary>

                    <ul class="record-backlinks-list">
                        
                        <li data-target-id="20230924120389">
                            <a href="#20230924120389">
                                <span class="record-type-point" style="color: var(--n_undefined);">⬤</span>
                                <span>scRNA-seq</span>
                                <div class="record-links-context">
                                    <p>Single Cell RNA Sequencing<br>
Multi-omics<br>
<em class="id-context" data-target-id="20230924120392">Spatial Transcriptomics</em></p>

                                </div>
                            </a>
                        </li>
                        
                    </ul>
                </details>
                
            </footer>
        </article>
        
        <article id="20230924120393" class="record">
            <header>
                <h1 class="record-title">Underfitting and Overfitting</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120393</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://www.geeksforgeeks.org/underfitting-and-overfitting-in-machine-learning/">https://www.geeksforgeeks.org/underfitting-and-overfitting-in-machine-learning/</a></p>
<h1>First, let's talk about Bias and Variance</h1>
<h2>Bias</h2>
<ul>
<li>Error rate of the training data</li>
<li>High value of error rate = High Bias</li>
<li>Low value of error rate = Low Bias</li>
</ul>
<h2>Variance</h2>
<ul>
<li>The gap of bias between training and testing data</li>
<li>High gap = High Variance</li>
<li>Low gap = Low Variance</li>
<li>Usually, low varinace is preferred to generalize a model</li>
</ul>
<h1>Underfitting</h1>
<ul>
<li>When model cannot capture the underlying trend of data
<ul>
<li>i.e. It only performs well on training data but performs poorly on testing data</li>
</ul>
</li>
<li>Meaning, the model is not fit the data</li>
</ul>
<h2><strong>Reasons for</strong> Underfitting</h2>
<ol>
<li>High bias and low variance.</li>
<li>The size of the training dataset used is not enough.</li>
<li>The model is too simple.</li>
<li>Training data is not cleaned and also contains noise in it.</li>
</ol>
<h2>Techniques to Reduce Underfitting</h2>
<ol>
<li>Increase model complexity.</li>
<li>Increase the number of features, performing feature engineering</li>
<li>Remove noise from the data.</li>
<li>Increase the number of epochs or increase the duration of training to get better results.</li>
</ol>
<h1>Overfitting</h1>
<ul>
<li>When the model does not make accurate predictions on testing data.
<ul>
<li>When a model gets trained with so much data, it starts learning from the noise and inaccurate data entries in our data set.</li>
</ul>
</li>
</ul>
<h2>Reasons for Overfitting</h2>
<ol>
<li>High variance and low bias.</li>
<li>The model is too complex.</li>
<li>The size of the training data.</li>
</ol>
<h2>Techniques to Reduce Overfitting</h2>
<ol>
<li>Increase training data.</li>
<li>Reduce model complexity.</li>
<li>Early stopping during the training phase</li>
<li>Ridge Regularization and Lasso Regularization.</li>
<li>Use dropout for neural networks to tackle overfitting.</li>
</ol>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120394" class="record">
            <header>
                <h1 class="record-title">Untitled</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120394</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://www.blopig.com/blog/2023/09/current-strategies-to-predict-structures-of-multiple-protein-conformational-states/">Current strategies to predict structures of multiple protein conformational states</a></p>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        <article id="20230924120395" class="record">
            <header>
                <h1 class="record-title">Why Bioinformaticiansuse Nextflow Thesedays</h1>

                

                <div class="record-id">
                    <span>id :</span>
                    <span>20230924120395</span>
                </div>

                <div class="record-type">
                    <span>types :</span>
                    <span>undefined</span>
                </div>

                
                <div class="record-tags">
                    <span>keywords :</span>
                    
                </div>
                

                
            </header>

            <main class="record-content">
                <p><a href="https://en.wikipedia.org/wiki/Nextflow">Wikipedia: Nextflow</a><br>
<a href="https://blog.naver.com/jinp7/222996328340">Nextflow &amp; Clara Parabricks</a><br>
<a href="https://www.reddit.com/r/bioinformatics/comments/sc8e0i/does_everyone_use_nextflow_thesedays/">Reddit: Does everyone use Nextflow thesedays?</a><br>
<a href="https://www.reddit.com/r/bioinformatics/comments/wazatv/nextflow_vs_snakemake/">Reddit: Nextflow vs Snakemake</a><br>
<a href="https://nf-co.re/rnaseq/3.12.0">nf-core/rnaseq</a><br>
<a href="https://medium.com/@MarinaBioinfoStarter/bioinformatics-tools-nextflow-snakemake-how-they-are-integrated-with-ci-cd-pipeline-c9f02781ccf7">Bioinformatics Tools: Nextflow &amp; SnakeMake</a><br>
<a href="https://labs.epi2me.io/snakemake-vs-nextflow/">Snakemake vs Nextflow</a><br>
<a href="https://biomadeira.github.io/2022-10-25-workflow-management">Workflow management systems in Bioinformatics</a></p>
<h1>What is Nextflow</h1>
<ul>
<li>Nextflow was originally developed at the Centre for Genomic Regulation in Spain and released as an open-source project on GitHub in July 2013.</li>
<li>Workflow Management System</li>
<li>Choose workflows(pipelines) on your purpose <a href="https://nf-co.re/pipelines">#</a>
<ul>
<li>RNAseq</li>
<li>Amplicon Target Sequencing(microbial)</li>
<li>CRISP data</li>
<li>Nanostring<br>
...</li>
</ul>
</li>
<li>Language: Groovy programming language
<ul>
<li>Java's extension</li>
<li>Similar with Python, Ruby, Smalltalk</li>
</ul>
</li>
<li>Cloud familiar
<ul>
<li>AWS<a href="https://www.nextflow.io/docs/latest/aws.html">#</a></li>
<li>Google Cloud<a href="https://www.nextflow.io/docs/latest/google.html">#</a><a href="https://cloud.google.com/life-sciences/docs/tutorials/nextflow?hl=ko">#</a></li>
<li>Azure Cloud<a href="https://www.nextflow.io/docs/latest/azure.html">#</a><br>
...</li>
</ul>
</li>
<li>Container familiar<a href="https://www.nextflow.io/docs/latest/container.html#">#</a>
<ul>
<li>Docker<a href="https://www.nextflow.io/docs/latest/container.html#docker">#</a></li>
<li>Sarus<a href="https://www.nextflow.io/docs/latest/container.html#sarus">#</a><br>
...</li>
</ul>
</li>
</ul>
<h1>Workflow example: nf-core/rnaseq <a href="https://nf-co.re/rnaseq/3.12.0">#</a></h1>
<ul>
<li>Pre-designed pipeline for RNA-seq</li>
<li>You can choose methods on you desires (change some tools in some stages)</li>
<li>Input...
<ul>
<li>a samplesheet</li>
<li>FASTQ files</li>
</ul>
</li>
<li>Perform...
<ul>
<li>quality control (QC)</li>
<li>trimming</li>
<li>(pseudo-)alignment</li>
</ul>
</li>
<li>Produce....
<ul>
<li>a gene expression matrix</li>
<li>extensive QC report</li>
</ul>
</li>
</ul>
<h2>Stage</h2>
<ol>
<li>Pre-processing</li>
<li>Genome alignment &amp; quantification</li>
<li>Pseudo-alignment $ quantification</li>
<li>Final QC</li>
</ol>
<h1>&quot;Is Nextflow better than Snakemake?&quot;</h1>
<blockquote>
<p><em>&quot;...The popularity of <strong>Python</strong> among the Bioinformatics community would suggest <strong>Snakemake</strong> would be the most adopted option, but it turns out that <strong>Nextflow</strong> <strong>is very popular</strong> and the choice of many researchers and groups.&quot;</em> <a href="https://biomadeira.github.io/2022-10-25-workflow-management">#</a></p>
</blockquote>
<h2>What is Snakemake</h2>
<ul>
<li>Another Workflow Managerment System</li>
<li>Pythonic</li>
</ul>
<h2>Snakemake vs Nextflow?</h2>
<ul>
<li>
<p>Snakemake is Python-like</p>
</li>
<li>
<p>Snakemake needs user to pre-understand of Python</p>
</li>
<li>
<p>Snakemake may need to install other tools first(conda, mambaforge)</p>
</li>
<li>
<p>Snakemake is integrated with other bioinformatics tools such as BioConda, Maba</p>
</li>
<li>
<p>Nextflow is UNIX-like</p>
</li>
<li>
<p>Nextflow is easy to install, easy to learn</p>
</li>
<li>
<p>Nextflow is well documentized</p>
</li>
<li>
<p>Nextflow has an active community</p>
</li>
<li>
<p>Both systems enable the automatic parallelization of jobs and automatic retrying of failed jobs.</p>
</li>
<li>
<p>Both workflow management systems support cloud execution via Kubernetes.</p>
</li>
<li>
<p>Snakemake is faster than Nextflow(Dec 2022)</p>
</li>
<li>
<p>Snakemake supports dry-run execution testing</p>
</li>
<li>
<p>Nextflow onyl supports on selected commands</p>
</li>
</ul>


                
            </main>

            <footer class="list-content-container">
                

                
            </footer>
        </article>
        
        
        
        
        <article id="help" class="record">
            <main>
                <h1>Help</h1>
                
                <p>
                    <a href="https://cosma.graphlab.fr/en/docs/" target="_blank" class="link-documentation">
                        Click here to access Cosma&#39;s documentation
                    </a>
                </p>
                
                <h2>Shortcuts</h2>
                
                    <table>
                        <tbody>
                            <tr>
                                <td>Space</td>
                                <td>Re-run the force-layout algorithm</td>
                            </tr>
                            <tr>
                                <td>S</td>
                                <td>Move the cursor to Search</td>
                            </tr>
                            <tr>
                                <td>Alt + click</td>
                                <td>(on a record type) Deselect other types</td>
                            </tr>
                            <tr>
                                <td>R</td>
                                <td>Reset zoom</td>
                            </tr>
                            <tr>
                                <td>Alt + R</td>
                                <td>Reset the display</td>
                            </tr>
                            <tr>
                                <td>C</td>
                                <td>Zoom in on the selected node</td>
                            </tr>
                            <tr>
                                <td>F</td>
                                <td>Switch to Focus mode</td>
                            </tr>
                            <tr>
                                <td>Escape</td>
                                <td>Close the active record</td>
                            </tr>
                        </tbody>
                    </table>
            </main>
        </article>

        

        <article id="credits" class="record">
            <main>
                
                    

                    

                    
                    
                    <hr>
                

                
                <div class="logo">
                    <?xml version="1.0" encoding="UTF-8"?>
<svg enable-background="new 0 0 283.46 68" version="1.1" viewBox="0 0 283.46 68" xml:space="preserve" xmlns="http://www.w3.org/2000/svg">
<style type="text/css">
	.st0{fill:#147899;}
	.st1{fill:#02385C;}
	.st2{fill:#001A37;}
</style>
	<path class="st0" d="m227.63 41.56v0.68-0.35-0.33m-15.62-15.3c-6.32 0-12.02 3.81-14.44 9.65s-1.08 12.56 3.39 17.03c2.99 2.99 6.98 4.58 11.05 4.58 2.01 0 4.04-0.39 5.98-1.19 0.26-0.11 0.52-0.22 0.77-0.34v-12.1c-0.86 2.91-3.56 5.03-6.75 5.03-2.85 0-5.41-1.71-6.5-4.34s-0.49-5.66 1.53-7.67c1.35-1.35 3.15-2.06 4.98-2.06 0.91 0 1.82 0.18 2.69 0.54 1.99 0.83 3.46 2.5 4.05 4.5v-12.09c-2.05-0.98-4.33-1.53-6.75-1.54"/>
	<path class="st0" d="m67.32 26.18c-4 0-8 1.53-11.05 4.58-6.1 6.1-6.1 16 0 22.1 3.05 3.05 7.05 4.58 11.05 4.58s8-1.53 11.05-4.58l-6.07-6.07c-1.31 1.32-3.1 2.06-4.96 2.06h-0.02c-3.19 0-5.97-2.14-6.8-5.22-0.82-3.08 0.52-6.32 3.28-7.92 1.1-0.63 2.31-0.94 3.52-0.94 1.82 0 3.62 0.71 4.98 2.06l6.07-6.07c-3.05-3.05-7.05-4.58-11.05-4.58"/>
	<path class="st0" d="m137.46 26.26h-12.18c-5.5 0-9.96 4.46-9.96 9.96s4.46 9.96 9.96 9.96h5.13c0.76 0 1.37 0.61 1.37 1.37s-0.61 1.37-1.37 1.37h-12.18v8.59h12.32c4-0.05 7.58-2.5 9.09-6.21s0.65-7.96-2.18-10.79c-1.86-1.87-4.38-2.92-7.01-2.92h-0.03-5.13c-0.76 0-1.37-0.62-1.37-1.38s0.62-1.37 1.37-1.37h12.18v-8.58"/>
	<path class="st0" d="m227.63 42.24c-0.13 5.9-3.58 11.21-8.88 13.74v1.49h8.88v-15.23m-8.87-2.36v4.01c0.19-0.64 0.29-1.31 0.29-2s-0.1-1.37-0.29-2.01m8.87-13.65h-8.88v1.57c5.15 2.48 8.74 7.69 8.88 13.76v-15.33"/>
	<path class="st1" d="m218.76 27.8v12.08c0.19 0.64 0.29 1.31 0.29 2 0 0.7-0.1 1.37-0.29 2v12.09c5.3-2.53 8.74-7.84 8.88-13.74v-0.68c-0.14-6.06-3.73-11.28-8.88-13.75"/>
	<path class="st0" d="m153.79 26.23h-8.52v31.24h8.52-8.52v-17.57-0.18c0.07-5.65 3.57-10.48 8.52-12.49v-1"/>
	<path class="st0" d="m177.75 26.23c-3.65 0-6.96 1.43-9.41 3.75 2.58 2.45 4.21 5.91 4.26 9.74v0.18c0-2.84 2.3-5.15 5.15-5.15 2.84 0 5.15 2.3 5.15 5.15v17.57h8.52v-17.57-0.18c-0.1-7.47-6.18-13.49-13.67-13.49"/>
	<path class="st0" d="m158.93 26.23c-1.82 0-3.56 0.36-5.15 1v30.24-17.57c0-2.84 2.3-5.15 5.15-5.15s5.15 2.3 5.15 5.15v-0.18c0.05-3.83 1.67-7.28 4.26-9.74-2.45-2.32-5.76-3.75-9.41-3.75"/>
	<path class="st1" d="m153.79 27.23c-4.94 2.01-8.45 6.84-8.52 12.49v0.18 17.57h8.52v-30.24"/>
	<path class="st1" d="m168.34 29.98c-2.58 2.45-4.21 5.91-4.26 9.74v0.18 17.57h8.52v-17.57-0.18c-0.05-3.83-1.68-7.28-4.26-9.74"/>
	<g id="cosma-roll">
		<path class="st0" d="m95.49 46.43-4.08 2.36v8.55h8.17v-8.55l-4.09-2.36m4.08-20.34h-8.17v8.55l4.09 2.36 4.08-2.36v-8.55"/>
		<path class="st0" d="m103.65 41.71-4.08 2.36v4.71l7.4 4.27 4.08-7.07-7.4-4.27m-19.65-11.34-4.08 7.07 7.4 4.27 4.08-2.36v-4.71l-7.4-4.27"/>
		<path class="st1" d="m99.57 44.07-4.08 2.36 4.08 2.36v-4.72m-8.17-9.43v4.71l4.09-2.35-4.09-2.36"/>
		<path class="st0" d="m87.32 41.71-7.4 4.27 4.08 7.08 7.4-4.27v-4.71l-4.08-2.37m19.65-11.34-7.4 4.27v4.71l4.08 2.36 7.4-4.27-4.08-7.07"/>
		<path class="st1" d="m91.4 44.07v4.71l4.08-2.36-4.08-2.35m8.17-9.43-4.08 2.36 4.08 2.36v-4.72"/>
		<path class="st1" d="m91.4 39.36-4.08 2.36 4.08 2.36v-4.72m8.17 0v4.71l4.08-2.36-4.08-2.35"/>
		<polyline class="st2" points="95.49 37 91.4 39.36 91.4 44.07 95.49 46.43 99.57 44.07 99.57 39.36 95.49 37"/>
	</g>
	<path class="st0" d="m50.43 16.17h-15c-0.69 0-1.25 0.56-1.25 1.25v48.56c0 0.69 0.56 1.25 1.25 1.25h15v-9.9h-5.09c-0.69 0-1.25-0.56-1.25-1.25v-28.74c0-0.69 0.56-1.25 1.25-1.25h5.09v-9.92"/>
	<path class="st0" d="m32.53 16.17h-15c-0.69 0-1.25 0.56-1.25 1.25v48.56c0 0.69 0.56 1.25 1.25 1.25h15v-9.9h-5.09c-0.69 0-1.25-0.56-1.25-1.25v-28.73c0-0.69 0.56-1.25 1.25-1.25h5.09v-9.93"/>
	<path class="st0" d="m247.89 16.17h-15v9.92h5.09c0.69 0 1.25 0.56 1.25 1.25v28.74c0 0.69-0.56 1.25-1.25 1.25h-5.09v9.9h15c0.69 0 1.25-0.56 1.25-1.25v-48.56c0.01-0.68-0.56-1.25-1.25-1.25"/>
	<path class="st0" d="m265.82 16.17h-15v9.92h5.09c0.69 0 1.25 0.56 1.25 1.25v28.74c0 0.69-0.56 1.25-1.25 1.25h-5.09v9.9h15c0.69 0 1.25-0.56 1.25-1.25v-48.55c0-0.69-0.56-1.26-1.25-1.26"/>
</svg>
                </div>
                

                <p>
                    Version 2.0.4
                    •
                    License GPL-3.0-or-later</p>

                <ul>
                    
                    <li>Arthur Perret</li>
                    
                    <li>Guillaume Brioudes</li>
                    
                    <li>Clément Borel</li>
                    
                    <li>Olivier Le Deuff</li>
                    

                    <li>ANR research programme <a href="https://hyperotlet.hypotheses.org/">HyperOtlet</a></li>
                </ul>

                <dl>
                    <dt><a href="https://d3js.org/" target="_blank">D3 <span class="badge">v4.13.0</span></a></dt>
                    <dd>Mike Bostock (BSD 3-Clause)</dd>
                    <dt><a href="https://mozilla.github.io/nunjucks/" target="_blank">Nunjucks <span class="badge">v3.2.3</span></a></dt>
                    <dd>James Long (BSD 2-Clause)</dd>
                    <dt><a href="https://github.com/nodeca/js-yaml" target="_blank">Js-yaml <span class="badge">v4.1.0</span></a></dt>
                    <dd>Vitaly Puzrin (MIT License)</dd>
                    <dt><a href="https://github.com/markdown-it/markdown-it" target="_blank">Markdown-it <span class="badge">v12.3.0</span></a></dt>
                    <dd>Vitaly Puzrin, Alex Kocharin (MIT License)</dd>
                    <dt><a href="https://github.com/juris-m/citeproc-js" target="_blank">Citeproc <span class="badge">v2.4.62</span></a></dt>
                    <dd>Frank Bennett (CPAL, AGPL)</dd>
                    <dt><a href="https://fusejs.io/" target="_blank">Fuse-js <span class="badge">v6.4.6</span></a></dt>
                    <dd>Kiro Risk (Apache License 2.0)</dd>
                </dl>
            </main>
        </article>

        

    </main>

    

    <div class="graph-wrapper" style="background-color: #ffffff">
        <svg id="graph-canvas">
            <defs>
                <marker id="arrow" viewBox="0 0 10 10" refX="22" refY="5" markerUnits="strokeWidth" markerWidth="5" markerHeight="5" orient="auto">
                    <path d="M 0 0 L 10 5 L 0 10 z"></path>
                </marker>
                <filter id="double">
                    <feMorphology in="SourceGraphic" result="a" operator="dilate" radius="0.2"></feMorphology>
                    <feComposite in="SourceGraphic" in2="a" result="xx" operator="xor"></feComposite>
                </filter>
                
            </defs>
        </svg>

        <div class="graph-controls">
            <div class="graph-controls-zoom">
                <button onclick="zoomLess()">-</button>
                <button onclick="zoomReset()">Reset (R)</button>
                <button onclick="zoomMore()">+</button>
            </div>

            
            <div class="graph-controls-focus">
                <label>
                    <input id="focus-check" type="checkbox" name="checkbox">
                    Focus (F)
                </label>

                <input id="focus-input" type="range" min="1" max="2" value="1" name="focus_range" step="1" />
            </div>
            
        </div>

        
        <div class="graph-timeline">
            <label>
                <input id="timeline-checkbox" type="checkbox" name="checkbox">
                Mode chronologique
            </label>

            <form id="timeline-form">
                <output name="timeline_output">9/24/2023</output>
                <input
                    type="range"
                    name="timeline_range"
                    min="1695571425.298"
                    max="1695571425.361"
                    value="1695571425.298"
                    list="timeline-ticks"
                >
                <datalist id="timeline-ticks"></datalist>
            </form>
        </div>
        
    </div>

    

    <script>
        const data = {"nodes":[{"id":"20230924120339","label":"Actinobacteria","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":5,"strokeWidth":2,"begin":1695571425.298},{"id":"20230924120340","label":"Active Learning","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":8,"strokeWidth":2,"begin":1695571425.298},{"id":"20230924120341","label":"AlphaFold1, AlphaFold2, and RoseTTAFold","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.314},{"id":"20230924120342","label":"Alphaproteobacteria","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":5,"strokeWidth":2,"begin":1695571425.314},{"id":"20230924120343","label":"Amplicon Target Sequencing","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.314},{"id":"20230924120344","label":"Bayesian statistics","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.314},{"id":"20230924120345","label":"Build Workflow from scratch - HaplotypeCaller","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.314},{"id":"20230924120346","label":"Building Docker in Google Cloud","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.314},{"id":"20230924120347","label":"Central Dogma","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":11,"strokeWidth":2,"begin":1695571425.314},{"id":"20230924120348","label":"Curse of The Dimensionality","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":11,"strokeWidth":2,"begin":1695571425.314},{"id":"20230924120349","label":"Dark microbiome and extremely low organics in Atacama fossil delta unveil Mars life detection limits","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":11,"strokeWidth":2,"begin":1695571425.314},{"id":"20230924120350","label":"Data Integration","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":5,"strokeWidth":2,"begin":1695571425.314},{"id":"20230924120351","label":"Dealing with Dimensionality - The Application of Machine Learning to Multi-omics Data","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.314},{"id":"20230924120352","label":"Deep Learning","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":5,"strokeWidth":2,"begin":1695571425.314},{"id":"20230924120353","label":"DNA","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":5,"strokeWidth":2,"begin":1695571425.314},{"id":"20230924120354","label":"edwith_LSnDP_Ch1","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.314},{"id":"20230924120355","label":"edwith_LSnDP_Ch2","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.314},{"id":"20230924120356","label":"edwith_LSnDP_Ch3","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.33},{"id":"20230924120357","label":"edwith_LSnDP_Ch4","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.33},{"id":"20230924120358","label":"edwith_LSnDP_Ch5","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.33},{"id":"20230924120359","label":"edwith_LSnDP_Ch6","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.33},{"id":"20230924120360","label":"edwith_LSnDP_Ch7","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.33},{"id":"20230924120361","label":"Ensemble","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.33},{"id":"20230924120362","label":"Exon","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.33},{"id":"20230924120363","label":"Feature Extraction","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":5,"strokeWidth":2,"begin":1695571425.33},{"id":"20230924120364","label":"Feature Selection","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":8,"strokeWidth":2,"begin":1695571425.33},{"id":"20230924120365","label":"Greatest Common Divisor, Least Common Multiple, and Euclidean algorithm","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.33},{"id":"20230924120366","label":"Integration Strategies of Multi-omics Data for Machine Learning Analysis","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":13.999999999999998,"strokeWidth":2,"begin":1695571425.33},{"id":"20230924120367","label":"IntGen_1.1-1.3_MendelsFirstLawMeiosis","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.33},{"id":"20230924120368","label":"IntGen_1.4_Monohybrid","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.33},{"id":"20230924120369","label":"IntGen_1.5_PunnettSquares","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.33},{"id":"20230924120370","label":"IntGen_1.6_SingleLocusTestCrosses","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.33},{"id":"20230924120371","label":"IntGen_6.2_SomaticMutVsGermlineMut","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.345},{"id":"20230924120372","label":"Intron","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.345},{"id":"20230924120373","label":"lecture-list","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.345},{"id":"20230924120374","label":"Long-read Sequencing","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":8,"strokeWidth":2,"begin":1695571425.345},{"id":"20230924120375","label":"Machine Learning","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":8,"strokeWidth":2,"begin":1695571425.345},{"id":"20230924120376","label":"mRNA","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":5,"strokeWidth":2,"begin":1695571425.345},{"id":"20230924120377","label":"Multi-omics","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":8,"strokeWidth":2,"begin":1695571425.345},{"id":"20230924120378","label":"Mutation","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.345},{"id":"20230924120379","label":"Nanopore","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":8,"strokeWidth":2,"begin":1695571425.345},{"id":"20230924120380","label":"NGS","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.345},{"id":"20230924120381","label":"OTUs","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":5,"strokeWidth":2,"begin":1695571425.345},{"id":"20230924120382","label":"p-value(유의확률)","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.345},{"id":"20230924120383","label":"Prime number and Sieve of Eratosthenes","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.345},{"id":"20230924120384","label":"Protein","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":5,"strokeWidth":2,"begin":1695571425.345},{"id":"20230924120385","label":"Query Strategy","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":5,"strokeWidth":2,"begin":1695571425.361},{"id":"20230924120386","label":"README","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.361},{"id":"20230924120387","label":"Real-Time DNA Sequencing in the Antarctic Dry Valleys Using the Oxford Nanopore Sequencer","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":5,"strokeWidth":2,"begin":1695571425.361},{"id":"20230924120388","label":"RNA","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":8,"strokeWidth":2,"begin":1695571425.361},{"id":"20230924120389","label":"scRNA-seq","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":8,"strokeWidth":2,"begin":1695571425.361},{"id":"20230924120390","label":"SingleCellBestPractice-1.Introduction","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.361},{"id":"20230924120391","label":"SMRT","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":5,"strokeWidth":2,"begin":1695571425.361},{"id":"20230924120392","label":"Spatial Transcriptomics","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":5,"strokeWidth":2,"begin":1695571425.361},{"id":"20230924120393","label":"Underfitting and Overfitting","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.361},{"id":"20230924120394","label":"Untitled","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.361},{"id":"20230924120395","label":"Why Bioinformaticiansuse Nextflow Thesedays","type":"undefined","fill":"#858585","colorStroke":"#858585","highlight":"#ff6a6a","size":2,"strokeWidth":2,"begin":1695571425.361}],"links":[{"id":0,"context":["- 학습모델이 주어진 데이터의 정보량을 평가하고, 라벨링이 필요하다고 판단되는 경우에만 인간에게 전달한다.\n- 정보량을 평가할 때는 [[20230924120385|Query Strategy]]를 사용한다. 라벨링이 필요하다고 판단되는 경우는 query하고 아닌 것은 버린다."],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120340","target":"20230924120385","report":[]},{"id":1,"context":["[[20230924120353|DNA]]\n[[20230924120388|RNA]]\n[[20230924120384|Protein]]"],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120347","target":"20230924120353","report":[]},{"id":2,"context":["[[20230924120353|DNA]]\n[[20230924120388|RNA]]\n[[20230924120384|Protein]]"],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120347","target":"20230924120388","report":[]},{"id":3,"context":["[[20230924120353|DNA]]\n[[20230924120388|RNA]]\n[[20230924120384|Protein]]"],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120347","target":"20230924120384","report":[]},{"id":4,"context":["## How to solve?\n[[20230924120364|Feature Selection]]\n[[20230924120363|Feature Extraction]]"],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120348","target":"20230924120364","report":[]},{"id":5,"context":["## How to solve?\n[[20230924120364|Feature Selection]]\n[[20230924120363|Feature Extraction]]"],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120348","target":"20230924120363","report":[]},{"id":6,"context":["# Abstract\n- 칠레의 Atacama Desert의 red stone(암석)에서 채취한 미생물군집 샘플에서 DNA 흔적을 조사했는데, 발견한 DNA의 9%는 과학계에 보고된 적이 없어 분류 불가능한 미생물의 존재였음.\n- 과학자들은 이렇게 sequencing 기술을 통하여 존재를 확인할 수 있지만, 정확하게 분류할 수 없는 미생물을 \"dark microbiome\"이라고 정의함.\n- 해당 연구는 지구의 매우 건조한 환경에서도 미생물의 흔적도 많이 손상되 보존되기 어려움을 시사함.\n- 따라서 화성에서 암석샘플 회수에 성공한다고 할지라도, 현재 기술로는 과거에 화성에 존재했던 미생물의 흔적을 놓칠 수 있다는 점을 시사함.\n### Site microbial diversity and dark microbione\n- 암석샘플은 1 그램당 1μg의 DNA를 갖고 있었음.\n- Operational Taxonomical Units([[20230924120381|OTUs]])이 제일 다양하게 검출된 부분은 바로 풍화된 상태의 역암(conglomerate)이었다.\n- 가장 많이 검출된 종은 [[20230924120342|Alphaproteobacteria]] 와 [[20230924120339|Actinobacteria]]였다.\n<img src=\"https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41467-023-36172-1/MediaObjects/41467_2023_36172_Fig4_HTML.png?as=webp\">"],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120349","target":"20230924120381","report":[]},{"id":7,"context":["# Abstract\n- 칠레의 Atacama Desert의 red stone(암석)에서 채취한 미생물군집 샘플에서 DNA 흔적을 조사했는데, 발견한 DNA의 9%는 과학계에 보고된 적이 없어 분류 불가능한 미생물의 존재였음.\n- 과학자들은 이렇게 sequencing 기술을 통하여 존재를 확인할 수 있지만, 정확하게 분류할 수 없는 미생물을 \"dark microbiome\"이라고 정의함.\n- 해당 연구는 지구의 매우 건조한 환경에서도 미생물의 흔적도 많이 손상되 보존되기 어려움을 시사함.\n- 따라서 화성에서 암석샘플 회수에 성공한다고 할지라도, 현재 기술로는 과거에 화성에 존재했던 미생물의 흔적을 놓칠 수 있다는 점을 시사함.\n### Site microbial diversity and dark microbione\n- 암석샘플은 1 그램당 1μg의 DNA를 갖고 있었음.\n- Operational Taxonomical Units([[20230924120381|OTUs]])이 제일 다양하게 검출된 부분은 바로 풍화된 상태의 역암(conglomerate)이었다.\n- 가장 많이 검출된 종은 [[20230924120342|Alphaproteobacteria]] 와 [[20230924120339|Actinobacteria]]였다.\n<img src=\"https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41467-023-36172-1/MediaObjects/41467_2023_36172_Fig4_HTML.png?as=webp\">"],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120349","target":"20230924120342","report":[]},{"id":8,"context":["# Abstract\n- 칠레의 Atacama Desert의 red stone(암석)에서 채취한 미생물군집 샘플에서 DNA 흔적을 조사했는데, 발견한 DNA의 9%는 과학계에 보고된 적이 없어 분류 불가능한 미생물의 존재였음.\n- 과학자들은 이렇게 sequencing 기술을 통하여 존재를 확인할 수 있지만, 정확하게 분류할 수 없는 미생물을 \"dark microbiome\"이라고 정의함.\n- 해당 연구는 지구의 매우 건조한 환경에서도 미생물의 흔적도 많이 손상되 보존되기 어려움을 시사함.\n- 따라서 화성에서 암석샘플 회수에 성공한다고 할지라도, 현재 기술로는 과거에 화성에 존재했던 미생물의 흔적을 놓칠 수 있다는 점을 시사함.\n### Site microbial diversity and dark microbione\n- 암석샘플은 1 그램당 1μg의 DNA를 갖고 있었음.\n- Operational Taxonomical Units([[20230924120381|OTUs]])이 제일 다양하게 검출된 부분은 바로 풍화된 상태의 역암(conglomerate)이었다.\n- 가장 많이 검출된 종은 [[20230924120342|Alphaproteobacteria]] 와 [[20230924120339|Actinobacteria]]였다.\n<img src=\"https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41467-023-36172-1/MediaObjects/41467_2023_36172_Fig4_HTML.png?as=webp\">"],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120349","target":"20230924120339","report":[]},{"id":9,"context":["[Picard M, Scott-Boyer MP, Bodein A, Périn O, Droit A. Integration strategies of multi-omics data for machine learning analysis. Comput Struct Biotechnol J. 2021 Jun 22;19:3735-3746. doi: 10.1016/j.csbj.2021.06.030. PMID: 34285775; PMCID: PMC8258788.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8258788/)\n- 2021년 기준으로 [[20230924120377|Multi-omics]] [[20230924120350|Data Integration]] 방법을 5가지로 분류함\n\t- early\n\t- mixed\n\t- intermediate\n\t- late\n\t- hierarchical\n<img src=\"https://ars.els-cdn.com/content/image/1-s2.0-S2001037021002683-ga1_lrg.jpg\" width=800>"],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120366","target":"20230924120377","report":[]},{"id":10,"context":["[Picard M, Scott-Boyer MP, Bodein A, Périn O, Droit A. Integration strategies of multi-omics data for machine learning analysis. Comput Struct Biotechnol J. 2021 Jun 22;19:3735-3746. doi: 10.1016/j.csbj.2021.06.030. PMID: 34285775; PMCID: PMC8258788.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8258788/)\n- 2021년 기준으로 [[20230924120377|Multi-omics]] [[20230924120350|Data Integration]] 방법을 5가지로 분류함\n\t- early\n\t- mixed\n\t- intermediate\n\t- late\n\t- hierarchical\n<img src=\"https://ars.els-cdn.com/content/image/1-s2.0-S2001037021002683-ga1_lrg.jpg\" width=800>"],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120366","target":"20230924120350","report":[]},{"id":11,"context":["# 1. introduction\n### 왜 multi-omics 연구가 필요할까?\n- 맞춤의학, 정밀의료의 부상\n\t- 개인의 임상데이터를 통해 진단하고 적절한 치료법을 제시\n\t- 임상데이터로 genomics, epigenomics, trasncriptiomics, proteomics, metabolomics 같은 데이터를 수집함\n\t- 과거에는 이러한 각각의 omics 데이터를 따로 관찰해서 진단에 활용하려고 했음\n\t- 그러나 대부분의 질병은 omics 데이터들을 서로 관통하면서 연결되어있음\n\t- 따라서 여러 omics 데이터를 합쳐서 통합적인 관점으로 보는 multi-omics 연구가 필요함\n### Multi-omics 연구의 장단점은?\n- 한 omics 데이터에서 부족한 부분을 다른 omics 데이터를 참고하여 메꾼다던가, 하나의 큰 시스템에 대한 그림을 얻어낸다던가 하는 장점이 있다.\n- 그러나 적은 샘플에서 가져온 데이터만 있다는 점이나, 개개의 생물학 데이터가 갖고있는 오류를 넘어서서 데이터를 합쳐야 한다는 어려움들이 단점으로 남는다.\n\t- 특히 multiomics 연구에서 데이터를 합치는(integration) 작업이 정확하게 이루어지지 않으면, 오류 투성이로 합쳐진 데이터집합에서 결과를 제대로 얻는 것은 불가능하다.\n### Data integration\n- [Zitnik et al. (2019)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6242341/)은 integration의 방법을 크게 두 가지로 분류하였다.\n\t- Horizontal inegration: 동일 omics 여러 샘플\n\t- Vertical integration: 여러 omics 동일 샘플\n- 이번 리뷰에서는 Vertical integration을 위주로 리뷰할 것이다.\n\t\t- Vertical integration의 방법을 수학적인 관점에서 bayesian, network-based, deep learning-based, kernel-based, matrix factorization-based로 나눠서 볼 것이다.\n\t- 또한 [Ritchie et al.(2015)](https://pubmed.ncbi.nlm.nih.gov/25582081/)의 분류방법에 따라 vertical integration(논문에서는 meta-dimensinal)을 concatenation-based, transformation-based, model-based로 나눠서 볼 것이다.\n# 2. Challenges\n- 생물학 데이터는 필연적으로 오류를 동반하고 있는데, 이렇게 깨끗하지 못한 여러 omics 데이터를 종합하여 특정한 패턴을 관찰한다는 것은 쉽지 않다.\n- 또한 multi-omics 연구의 특성상 적은 샘플에서 많은 feature를 관찰해야 하는데, 이는 machine learing에서 가장 이상적이지 못한 상황으로 차원의 저주([[20230924120348|Curse of The Dimensionality]])라고 불리운다.\n\t- 이런 상황에서 ML을 통해 데이터를 훈련시키면 overfit된다.\n- 생물학 omics 데이터들이 각각 너무 성격이 다르다는 점도 하나의 문제이다.\n\t- 유전자 발현 데이터의 경우 수만개의 variable이 있을 수 있지만, 대사물질 데이터는 기껏해서 몇천일 수가 있다.\n\t- 이렇게 서로 다른 데이터들 사이의 수, 종류, 연속성, 등의 차이를 극복하고 데이터들을 잘 integration 해야한다는 어려움이 있다.\n- 데이터 간의 class imbalance 또한 ML 훈려에 악영향을 미친다.\n\t- 데이터에서 특정 class의 데이터만 기하급수적으로 많다면, 이를 바탕으로 훈련시키는 것이 부적절하다.\n\t- 왜냐하면 ML 훈련은 모든 class의 데이터 량이 엇비슷하다고 가정하고 훈련되기 때문이다.\n\t- 이를 해결하기 위한 방법으로 sampling, cost-sensitive learning과 같은 방법이 존재한다.\n- Missing data 문제의 경우 해당 샘플을 지워버린다던가 혹은 Random Forest, K-Nearest Neighbor같은 방법이 있다.\n# 3. Main integration strategies\n### Integration 5가지 소개\n- **Early integration**이란, 데이터들의 복잡도를 낮추지 않고, 단순히 둘 이상의 데이터들을 합친 뒤(wise concatenation) 그 결과물인 matrix를 ML모델에 넣는 것이다.\n\t- 현실적으로는 단순하게 합쳐진 데이터는 여전히 복잡하기 때문에 ML모델이 잘 돌아가지 않는다다\n- **Mixed integration**이란 데이터들의 복잡도를 개별적으로 낮춰서 사용하는 것이다.\n- **Intermediate integration**이란 여러 데이터들의 복잡도를 함께 낮춰서 사용하는 것이다.\n- **Late integration**이란 정반대로 데이터를 합치지 않고 각각의 omics 데이터를 개별적으로 분석하는 것이다.\n- **Hierarchical integration**이란 DNA-RNA-단백질로 이루어지는 시스템을 표현한 central dogma에 입각해서 omics 데이터 간의 조절 관계를 반영하여 합치는 것이다.\n## 3.1 Dimensionality reduction for multi-omic integration\n**Dimensionality reduction은 왜 필요할까?**\n- Multi-omics 연구에 있어서 샘플들이 너무 많은 variable을 가질 경우 machine learning 기술 적용에 차질을 빚는다. 따라서 variable의 수를 줄여서 차원의 축소와 데이터셋의 noise 감소를 유도한다.\n- 차원축소는 필수적이지는 않고, 위에서 얘기했던 5가지의 integration 방법에 모두 적용이 가능하다.\n\t- 다만 early integration와 intermediate integration의 경우 차원축소를 미리 적용하는 것이 정확도 향상에 기여한다.\n- 차원축소를 위한 방법은 크게 두 가지가 있다:\n\t- Feature selection\n\t- Redundant varaibles and feature extraction\n### 3.1.1. [[20230924120364|Feature Selection]] (FS) \n- Feature를 줄이는 작업\n- 그 결과 데이터의 복잡도, noise가 줄어들어서 데이터가 깔끔해진다.\n- 데이터가 깔끔해지니까, ML 모델이 overfitting될 위험도 적어진다.\n- 데이터가 깔끔해져서 해석하기도 편해진다.\n- 너무 많은 variable이 삭제될 경우?\n\t- 데이터셋 사이에 feautre의 수가 다르면 좋지 않다\n\t- Feature Selection은 각 omics block이 동일한 수의 feature를 가지도록 할 수 있다\n- Feature Selection은 세 가지로 나뉜다:\n\t- Filter-based Feature Selection\n\t- Wrapper-based Feature Selection\n\t- Embedded Feature Selection\n### Filter-based Feature Selection\n- ML 모델과 상관없이 적용 가능\n- Feature 수를 줄일 수 있는 방향으로 관련된 variable을 추려내는 통계학적인 분석방법임\n- **CFS, RCA** 같이 correlation 방법을 사용하거나,\n- **ReliefF** 같이 distance 방법을 사용하거나,\n- **mRMR** 같이 information gain 방법을 사용함\n### Wrapper-based Feature Selection\n- ML 모델을 활용하여 서로 다른 feature 그룹을 학습시키고 평가한다.\n- 가장 성능이 좋았던 feature 그룹만 선택한다.\n- Recursive Feature Selection이 한 가지 예시이다.\n\t- 먼저 전체 feauture를 대상으로 모델을 학습시킨다.\n\t- 그 다음 모델 퍼포먼스에 기여도가 떨어지는 feature를 차례대로 제거해간다.\n- 이 방법은 어느 ML 모델에서든 활용 가능하다는 장점이 있다.\n- 그러나 반복적으로 ML 모델을 훈련시켜야 하기 때문에, 데이터셋이 크다면 컴퓨터 성능에 따라 시간이 오래 걸린다는 단점이 있다.\n### Embedded Feature Selection\n- Classifier 모델 안에 feature selection 알고리즘이 넣어져있는 경우\n- Tree-based feature selection과 regularization 방법이 있다.\n\t- Regularization 방법은 loss function과 penalization funtion을 결합한 방법이다.\n\t\t- Loss funtion은 모델이 주어진 데이터에 얼마나 잘 맞는지 확인하는 함수다.\n\t\t- Penalization funtion은 복잡도(complexity)에 대하여 페널티를 부가하는 방식으로 feature 수를 적은 쪽으로 유도한다.\n- Embedded Feature Selection 모델은 Wrapper-based Feature Selection과 마찬가지로 ML모델과 함께 사용되지만, 컴퓨터 자원을 덜 소비하기 때문에 컴퓨터 성능에 덜 의존적이라는 장점이 있다."],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120366","target":"20230924120348","report":[]},{"id":12,"context":["# 1. introduction\n### 왜 multi-omics 연구가 필요할까?\n- 맞춤의학, 정밀의료의 부상\n\t- 개인의 임상데이터를 통해 진단하고 적절한 치료법을 제시\n\t- 임상데이터로 genomics, epigenomics, trasncriptiomics, proteomics, metabolomics 같은 데이터를 수집함\n\t- 과거에는 이러한 각각의 omics 데이터를 따로 관찰해서 진단에 활용하려고 했음\n\t- 그러나 대부분의 질병은 omics 데이터들을 서로 관통하면서 연결되어있음\n\t- 따라서 여러 omics 데이터를 합쳐서 통합적인 관점으로 보는 multi-omics 연구가 필요함\n### Multi-omics 연구의 장단점은?\n- 한 omics 데이터에서 부족한 부분을 다른 omics 데이터를 참고하여 메꾼다던가, 하나의 큰 시스템에 대한 그림을 얻어낸다던가 하는 장점이 있다.\n- 그러나 적은 샘플에서 가져온 데이터만 있다는 점이나, 개개의 생물학 데이터가 갖고있는 오류를 넘어서서 데이터를 합쳐야 한다는 어려움들이 단점으로 남는다.\n\t- 특히 multiomics 연구에서 데이터를 합치는(integration) 작업이 정확하게 이루어지지 않으면, 오류 투성이로 합쳐진 데이터집합에서 결과를 제대로 얻는 것은 불가능하다.\n### Data integration\n- [Zitnik et al. (2019)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6242341/)은 integration의 방법을 크게 두 가지로 분류하였다.\n\t- Horizontal inegration: 동일 omics 여러 샘플\n\t- Vertical integration: 여러 omics 동일 샘플\n- 이번 리뷰에서는 Vertical integration을 위주로 리뷰할 것이다.\n\t\t- Vertical integration의 방법을 수학적인 관점에서 bayesian, network-based, deep learning-based, kernel-based, matrix factorization-based로 나눠서 볼 것이다.\n\t- 또한 [Ritchie et al.(2015)](https://pubmed.ncbi.nlm.nih.gov/25582081/)의 분류방법에 따라 vertical integration(논문에서는 meta-dimensinal)을 concatenation-based, transformation-based, model-based로 나눠서 볼 것이다.\n# 2. Challenges\n- 생물학 데이터는 필연적으로 오류를 동반하고 있는데, 이렇게 깨끗하지 못한 여러 omics 데이터를 종합하여 특정한 패턴을 관찰한다는 것은 쉽지 않다.\n- 또한 multi-omics 연구의 특성상 적은 샘플에서 많은 feature를 관찰해야 하는데, 이는 machine learing에서 가장 이상적이지 못한 상황으로 차원의 저주([[20230924120348|Curse of The Dimensionality]])라고 불리운다.\n\t- 이런 상황에서 ML을 통해 데이터를 훈련시키면 overfit된다.\n- 생물학 omics 데이터들이 각각 너무 성격이 다르다는 점도 하나의 문제이다.\n\t- 유전자 발현 데이터의 경우 수만개의 variable이 있을 수 있지만, 대사물질 데이터는 기껏해서 몇천일 수가 있다.\n\t- 이렇게 서로 다른 데이터들 사이의 수, 종류, 연속성, 등의 차이를 극복하고 데이터들을 잘 integration 해야한다는 어려움이 있다.\n- 데이터 간의 class imbalance 또한 ML 훈려에 악영향을 미친다.\n\t- 데이터에서 특정 class의 데이터만 기하급수적으로 많다면, 이를 바탕으로 훈련시키는 것이 부적절하다.\n\t- 왜냐하면 ML 훈련은 모든 class의 데이터 량이 엇비슷하다고 가정하고 훈련되기 때문이다.\n\t- 이를 해결하기 위한 방법으로 sampling, cost-sensitive learning과 같은 방법이 존재한다.\n- Missing data 문제의 경우 해당 샘플을 지워버린다던가 혹은 Random Forest, K-Nearest Neighbor같은 방법이 있다.\n# 3. Main integration strategies\n### Integration 5가지 소개\n- **Early integration**이란, 데이터들의 복잡도를 낮추지 않고, 단순히 둘 이상의 데이터들을 합친 뒤(wise concatenation) 그 결과물인 matrix를 ML모델에 넣는 것이다.\n\t- 현실적으로는 단순하게 합쳐진 데이터는 여전히 복잡하기 때문에 ML모델이 잘 돌아가지 않는다다\n- **Mixed integration**이란 데이터들의 복잡도를 개별적으로 낮춰서 사용하는 것이다.\n- **Intermediate integration**이란 여러 데이터들의 복잡도를 함께 낮춰서 사용하는 것이다.\n- **Late integration**이란 정반대로 데이터를 합치지 않고 각각의 omics 데이터를 개별적으로 분석하는 것이다.\n- **Hierarchical integration**이란 DNA-RNA-단백질로 이루어지는 시스템을 표현한 central dogma에 입각해서 omics 데이터 간의 조절 관계를 반영하여 합치는 것이다.\n## 3.1 Dimensionality reduction for multi-omic integration\n**Dimensionality reduction은 왜 필요할까?**\n- Multi-omics 연구에 있어서 샘플들이 너무 많은 variable을 가질 경우 machine learning 기술 적용에 차질을 빚는다. 따라서 variable의 수를 줄여서 차원의 축소와 데이터셋의 noise 감소를 유도한다.\n- 차원축소는 필수적이지는 않고, 위에서 얘기했던 5가지의 integration 방법에 모두 적용이 가능하다.\n\t- 다만 early integration와 intermediate integration의 경우 차원축소를 미리 적용하는 것이 정확도 향상에 기여한다.\n- 차원축소를 위한 방법은 크게 두 가지가 있다:\n\t- Feature selection\n\t- Redundant varaibles and feature extraction\n### 3.1.1. [[20230924120364|Feature Selection]] (FS) \n- Feature를 줄이는 작업\n- 그 결과 데이터의 복잡도, noise가 줄어들어서 데이터가 깔끔해진다.\n- 데이터가 깔끔해지니까, ML 모델이 overfitting될 위험도 적어진다.\n- 데이터가 깔끔해져서 해석하기도 편해진다.\n- 너무 많은 variable이 삭제될 경우?\n\t- 데이터셋 사이에 feautre의 수가 다르면 좋지 않다\n\t- Feature Selection은 각 omics block이 동일한 수의 feature를 가지도록 할 수 있다\n- Feature Selection은 세 가지로 나뉜다:\n\t- Filter-based Feature Selection\n\t- Wrapper-based Feature Selection\n\t- Embedded Feature Selection\n### Filter-based Feature Selection\n- ML 모델과 상관없이 적용 가능\n- Feature 수를 줄일 수 있는 방향으로 관련된 variable을 추려내는 통계학적인 분석방법임\n- **CFS, RCA** 같이 correlation 방법을 사용하거나,\n- **ReliefF** 같이 distance 방법을 사용하거나,\n- **mRMR** 같이 information gain 방법을 사용함\n### Wrapper-based Feature Selection\n- ML 모델을 활용하여 서로 다른 feature 그룹을 학습시키고 평가한다.\n- 가장 성능이 좋았던 feature 그룹만 선택한다.\n- Recursive Feature Selection이 한 가지 예시이다.\n\t- 먼저 전체 feauture를 대상으로 모델을 학습시킨다.\n\t- 그 다음 모델 퍼포먼스에 기여도가 떨어지는 feature를 차례대로 제거해간다.\n- 이 방법은 어느 ML 모델에서든 활용 가능하다는 장점이 있다.\n- 그러나 반복적으로 ML 모델을 훈련시켜야 하기 때문에, 데이터셋이 크다면 컴퓨터 성능에 따라 시간이 오래 걸린다는 단점이 있다.\n### Embedded Feature Selection\n- Classifier 모델 안에 feature selection 알고리즘이 넣어져있는 경우\n- Tree-based feature selection과 regularization 방법이 있다.\n\t- Regularization 방법은 loss function과 penalization funtion을 결합한 방법이다.\n\t\t- Loss funtion은 모델이 주어진 데이터에 얼마나 잘 맞는지 확인하는 함수다.\n\t\t- Penalization funtion은 복잡도(complexity)에 대하여 페널티를 부가하는 방식으로 feature 수를 적은 쪽으로 유도한다.\n- Embedded Feature Selection 모델은 Wrapper-based Feature Selection과 마찬가지로 ML모델과 함께 사용되지만, 컴퓨터 자원을 덜 소비하기 때문에 컴퓨터 성능에 덜 의존적이라는 장점이 있다."],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120366","target":"20230924120364","report":[]},{"id":13,"context":["Also known as third-generation sequencing\n[[20230924120379|Nanopore]]\n[[20230924120391|SMRT]]"],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120374","target":"20230924120379","report":[]},{"id":14,"context":["Also known as third-generation sequencing\n[[20230924120379|Nanopore]]\n[[20230924120391|SMRT]]"],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120374","target":"20230924120391","report":[]},{"id":15,"context":["## Example\n[[20230924120352|Deep Learning]]\n[[20230924120340|Active Learning]]"],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120375","target":"20230924120352","report":[]},{"id":16,"context":["## Example\n[[20230924120352|Deep Learning]]\n[[20230924120340|Active Learning]]"],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120375","target":"20230924120340","report":[]},{"id":17,"context":["[Johnson SS, Zaikova E, Goerlitz DS, Bai Y, Tighe SW. Real-Time DNA Sequencing in the Antarctic Dry Valleys Using the Oxford Nanopore Sequencer. J Biomol Tech. 2017 Apr;28(1):2-7. doi: 10.7171/jbt.17-2801-009. Epub 2017 Mar 22. PMID: 28337073; PMCID: PMC5362188.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5362188/)\n# Abstract\n- Oxford Nanopore Technologies에서 개발한 소형 [[20230924120379|Nanopore]] sequencer를 활용하여, 연구실이 아닌 야외에서 DNA sequencing 작업을 수행하고, 인터넷에 연결되지 않은 환경에서 소프트웨어를 활용해 분석까지 수행했다.\n- 장소는 남극의 McMurdo Dry Valleys였다.\n- 사용된 sequencer는 MK1B MinION sequencer였다.\n- 2시간 30분만에 야외환경에서 sequencing protocol을 완수할 수 있었다."],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120387","target":"20230924120379","report":[]},{"id":18,"context":["[[20230924120376|mRNA]]"],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120388","target":"20230924120376","report":[]},{"id":19,"context":["Single Cell RNA Sequencing\n[[20230924120377|Multi-omics]]\n[[20230924120392|Spatial Transcriptomics]]"],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120389","target":"20230924120377","report":[]},{"id":20,"context":["Single Cell RNA Sequencing\n[[20230924120377|Multi-omics]]\n[[20230924120392|Spatial Transcriptomics]]"],"type":"undefined","shape":{"stroke":"simple","dashInterval":null},"color":"#e1e1e1","colorHighlight":"#ff6a6a","source":"20230924120389","target":"20230924120392","report":[]}]};
        const graphProperties = {"select_origin":"directory","files_origin":"C:/Users/jhlee/OneDrive/OBSIDIAN/cosma","nodes_origin":"","links_origin":"","nodes_online":"","links_online":"","images_origin":"","export_target":"","history":true,"focus_max":2,"record_types":{"undefined":{"fill":"#858585","stroke":"#858585"}},"link_types":{"undefined":{"stroke":"simple","color":"#e1e1e1"}},"record_filters":[],"graph_background_color":"#ffffff","graph_highlight_color":"#ff6a6a","graph_highlight_on_hover":true,"graph_text_size":10,"graph_arrows":true,"node_size_method":"degree","node_size":10,"node_size_max":20,"node_size_min":2,"attraction_force":200,"attraction_distance_max":250,"attraction_vertical":0,"attraction_horizontal":0,"views":{},"chronological_record_meta":"created","record_metas":[],"title":"","author":"","description":"","keywords":[],"link_symbol":"","csl":"","bibliography":"","csl_locale":"","css_custom":"","devtools":false,"lang":"en","name":""};
        const tagList = [];
        const typeList = {"undefined":{"nodes":["20230924120339","20230924120340","20230924120341","20230924120342","20230924120343","20230924120344","20230924120345","20230924120346","20230924120347","20230924120348","20230924120349","20230924120350","20230924120351","20230924120352","20230924120353","20230924120354","20230924120355","20230924120356","20230924120357","20230924120358","20230924120359","20230924120360","20230924120361","20230924120362","20230924120363","20230924120364","20230924120365","20230924120366","20230924120367","20230924120368","20230924120369","20230924120370","20230924120371","20230924120372","20230924120373","20230924120374","20230924120375","20230924120376","20230924120377","20230924120378","20230924120379","20230924120380","20230924120381","20230924120382","20230924120383","20230924120384","20230924120385","20230924120386","20230924120387","20230924120388","20230924120389","20230924120390","20230924120391","20230924120392","20230924120393","20230924120394","20230924120395"],"active":true}};
        const timeline = {"begin":1695571425.298,"end":1695571425.361};
        const sorting = {"records":[{"alphabetical":0,"chronological":0},{"alphabetical":1,"chronological":1},{"alphabetical":2,"chronological":2},{"alphabetical":3,"chronological":3},{"alphabetical":4,"chronological":4},{"alphabetical":5,"chronological":5},{"alphabetical":6,"chronological":6},{"alphabetical":7,"chronological":7},{"alphabetical":8,"chronological":8},{"alphabetical":9,"chronological":9},{"alphabetical":10,"chronological":10},{"alphabetical":11,"chronological":11},{"alphabetical":12,"chronological":12},{"alphabetical":13,"chronological":13},{"alphabetical":14,"chronological":14},{"alphabetical":15,"chronological":15},{"alphabetical":16,"chronological":16},{"alphabetical":17,"chronological":17},{"alphabetical":18,"chronological":18},{"alphabetical":19,"chronological":19},{"alphabetical":20,"chronological":20},{"alphabetical":21,"chronological":21},{"alphabetical":22,"chronological":22},{"alphabetical":23,"chronological":23},{"alphabetical":24,"chronological":24},{"alphabetical":25,"chronological":25},{"alphabetical":26,"chronological":26},{"alphabetical":27,"chronological":27},{"alphabetical":28,"chronological":28},{"alphabetical":29,"chronological":29},{"alphabetical":30,"chronological":30},{"alphabetical":31,"chronological":31},{"alphabetical":32,"chronological":32},{"alphabetical":33,"chronological":33},{"alphabetical":34,"chronological":34},{"alphabetical":35,"chronological":35},{"alphabetical":36,"chronological":36},{"alphabetical":37,"chronological":37},{"alphabetical":38,"chronological":38},{"alphabetical":39,"chronological":39},{"alphabetical":40,"chronological":40},{"alphabetical":41,"chronological":41},{"alphabetical":42,"chronological":42},{"alphabetical":43,"chronological":43},{"alphabetical":44,"chronological":44},{"alphabetical":45,"chronological":45},{"alphabetical":46,"chronological":46},{"alphabetical":47,"chronological":47},{"alphabetical":48,"chronological":48},{"alphabetical":49,"chronological":49},{"alphabetical":50,"chronological":50},{"alphabetical":51,"chronological":51},{"alphabetical":52,"chronological":52},{"alphabetical":53,"chronological":53},{"alphabetical":54,"chronological":54},{"alphabetical":55,"chronological":55},{"alphabetical":56,"chronological":56}],"tags":[]};

        const focusIsActive = true;
    </script>

    

    
    <script>
        /*! For license information please see bundle.js.LICENSE.txt */
(()=>{var t={838:()=>{window.addEventListener("DOMContentLoaded",(function(){var t=document.getElementById("citation-references"),e=null==t?void 0:t.querySelector("button"),n=null==t?void 0:t.querySelector("code");e&&e.addEventListener("click",(function(){var t=document.createElement("a");t.setAttribute("href","data:text/plain;charset=utf-8,"+encodeURIComponent(n.textContent)),t.setAttribute("download","bibliography.json"),t.style.display="none",document.body.appendChild(t),t.click(),t.remove()}))}))},187:t=>{"use strict";var e,n="object"==typeof Reflect?Reflect:null,r=n&&"function"==typeof n.apply?n.apply:function(t,e,n){return Function.prototype.apply.call(t,e,n)};e=n&&"function"==typeof n.ownKeys?n.ownKeys:Object.getOwnPropertySymbols?function(t){return Object.getOwnPropertyNames(t).concat(Object.getOwnPropertySymbols(t))}:function(t){return Object.getOwnPropertyNames(t)};var i=Number.isNaN||function(t){return t!=t};function o(){o.init.call(this)}t.exports=o,t.exports.once=function(t,e){return new Promise((function(n,r){function i(n){t.removeListener(e,o),r(n)}function o(){"function"==typeof t.removeListener&&t.removeListener("error",i),n([].slice.call(arguments))}y(t,e,o,{once:!0}),"error"!==e&&function(t,e,n){"function"==typeof t.on&&y(t,"error",e,{once:!0})}(t,i)}))},o.EventEmitter=o,o.prototype._events=void 0,o.prototype._eventsCount=0,o.prototype._maxListeners=void 0;var a=10;function s(t){if("function"!=typeof t)throw new TypeError('The "listener" argument must be of type Function. Received type '+typeof t)}function u(t){return void 0===t._maxListeners?o.defaultMaxListeners:t._maxListeners}function c(t,e,n,r){var i,o,a,c;if(s(n),void 0===(o=t._events)?(o=t._events=Object.create(null),t._eventsCount=0):(void 0!==o.newListener&&(t.emit("newListener",e,n.listener?n.listener:n),o=t._events),a=o[e]),void 0===a)a=o[e]=n,++t._eventsCount;else if("function"==typeof a?a=o[e]=r?[n,a]:[a,n]:r?a.unshift(n):a.push(n),(i=u(t))>0&&a.length>i&&!a.warned){a.warned=!0;var h=new Error("Possible EventEmitter memory leak detected. "+a.length+" "+String(e)+" listeners added. Use emitter.setMaxListeners() to increase limit");h.name="MaxListenersExceededWarning",h.emitter=t,h.type=e,h.count=a.length,c=h,console&&console.warn&&console.warn(c)}return t}function h(){if(!this.fired)return this.target.removeListener(this.type,this.wrapFn),this.fired=!0,0===arguments.length?this.listener.call(this.target):this.listener.apply(this.target,arguments)}function f(t,e,n){var r={fired:!1,wrapFn:void 0,target:t,type:e,listener:n},i=h.bind(r);return i.listener=n,r.wrapFn=i,i}function l(t,e,n){var r=t._events;if(void 0===r)return[];var i=r[e];return void 0===i?[]:"function"==typeof i?n?[i.listener||i]:[i]:n?function(t){for(var e=new Array(t.length),n=0;n<e.length;++n)e[n]=t[n].listener||t[n];return e}(i):p(i,i.length)}function d(t){var e=this._events;if(void 0!==e){var n=e[t];if("function"==typeof n)return 1;if(void 0!==n)return n.length}return 0}function p(t,e){for(var n=new Array(e),r=0;r<e;++r)n[r]=t[r];return n}function y(t,e,n,r){if("function"==typeof t.on)r.once?t.once(e,n):t.on(e,n);else{if("function"!=typeof t.addEventListener)throw new TypeError('The "emitter" argument must be of type EventEmitter. Received type '+typeof t);t.addEventListener(e,(function i(o){r.once&&t.removeEventListener(e,i),n(o)}))}}Object.defineProperty(o,"defaultMaxListeners",{enumerable:!0,get:function(){return a},set:function(t){if("number"!=typeof t||t<0||i(t))throw new RangeError('The value of "defaultMaxListeners" is out of range. It must be a non-negative number. Received '+t+".");a=t}}),o.init=function(){void 0!==this._events&&this._events!==Object.getPrototypeOf(this)._events||(this._events=Object.create(null),this._eventsCount=0),this._maxListeners=this._maxListeners||void 0},o.prototype.setMaxListeners=function(t){if("number"!=typeof t||t<0||i(t))throw new RangeError('The value of "n" is out of range. It must be a non-negative number. Received '+t+".");return this._maxListeners=t,this},o.prototype.getMaxListeners=function(){return u(this)},o.prototype.emit=function(t){for(var e=[],n=1;n<arguments.length;n++)e.push(arguments[n]);var i="error"===t,o=this._events;if(void 0!==o)i=i&&void 0===o.error;else if(!i)return!1;if(i){var a;if(e.length>0&&(a=e[0]),a instanceof Error)throw a;var s=new Error("Unhandled error."+(a?" ("+a.message+")":""));throw s.context=a,s}var u=o[t];if(void 0===u)return!1;if("function"==typeof u)r(u,this,e);else{var c=u.length,h=p(u,c);for(n=0;n<c;++n)r(h[n],this,e)}return!0},o.prototype.addListener=function(t,e){return c(this,t,e,!1)},o.prototype.on=o.prototype.addListener,o.prototype.prependListener=function(t,e){return c(this,t,e,!0)},o.prototype.once=function(t,e){return s(e),this.on(t,f(this,t,e)),this},o.prototype.prependOnceListener=function(t,e){return s(e),this.prependListener(t,f(this,t,e)),this},o.prototype.removeListener=function(t,e){var n,r,i,o,a;if(s(e),void 0===(r=this._events))return this;if(void 0===(n=r[t]))return this;if(n===e||n.listener===e)0==--this._eventsCount?this._events=Object.create(null):(delete r[t],r.removeListener&&this.emit("removeListener",t,n.listener||e));else if("function"!=typeof n){for(i=-1,o=n.length-1;o>=0;o--)if(n[o]===e||n[o].listener===e){a=n[o].listener,i=o;break}if(i<0)return this;0===i?n.shift():function(t,e){for(;e+1<t.length;e++)t[e]=t[e+1];t.pop()}(n,i),1===n.length&&(r[t]=n[0]),void 0!==r.removeListener&&this.emit("removeListener",t,a||e)}return this},o.prototype.off=o.prototype.removeListener,o.prototype.removeAllListeners=function(t){var e,n,r;if(void 0===(n=this._events))return this;if(void 0===n.removeListener)return 0===arguments.length?(this._events=Object.create(null),this._eventsCount=0):void 0!==n[t]&&(0==--this._eventsCount?this._events=Object.create(null):delete n[t]),this;if(0===arguments.length){var i,o=Object.keys(n);for(r=0;r<o.length;++r)"removeListener"!==(i=o[r])&&this.removeAllListeners(i);return this.removeAllListeners("removeListener"),this._events=Object.create(null),this._eventsCount=0,this}if("function"==typeof(e=n[t]))this.removeListener(t,e);else if(void 0!==e)for(r=e.length-1;r>=0;r--)this.removeListener(t,e[r]);return this},o.prototype.listeners=function(t){return l(this,t,!0)},o.prototype.rawListeners=function(t){return l(this,t,!1)},o.listenerCount=function(t,e){return"function"==typeof t.listenerCount?t.listenerCount(e):d.call(t,e)},o.prototype.listenerCount=d,o.prototype.eventNames=function(){return this._eventsCount>0?e(this._events):[]}},691:function(t){t.exports=function(){"use strict";function t(t,e){var n=Object.keys(t);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(t);e&&(r=r.filter((function(e){return Object.getOwnPropertyDescriptor(t,e).enumerable}))),n.push.apply(n,r)}return n}function e(e){for(var n=1;n<arguments.length;n++){var r=null!=arguments[n]?arguments[n]:{};n%2?t(Object(r),!0).forEach((function(t){a(e,t,r[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(r)):t(Object(r)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(r,t))}))}return e}function n(t){return n="function"==typeof Symbol&&"symbol"==typeof Symbol.iterator?function(t){return typeof t}:function(t){return t&&"function"==typeof Symbol&&t.constructor===Symbol&&t!==Symbol.prototype?"symbol":typeof t},n(t)}function r(t,e){if(!(t instanceof e))throw new TypeError("Cannot call a class as a function")}function i(t,e){for(var n=0;n<e.length;n++){var r=e[n];r.enumerable=r.enumerable||!1,r.configurable=!0,"value"in r&&(r.writable=!0),Object.defineProperty(t,r.key,r)}}function o(t,e,n){return e&&i(t.prototype,e),n&&i(t,n),Object.defineProperty(t,"prototype",{writable:!1}),t}function a(t,e,n){return e in t?Object.defineProperty(t,e,{value:n,enumerable:!0,configurable:!0,writable:!0}):t[e]=n,t}function s(t){return function(t){if(Array.isArray(t))return u(t)}(t)||function(t){if("undefined"!=typeof Symbol&&null!=t[Symbol.iterator]||null!=t["@@iterator"])return Array.from(t)}(t)||function(t,e){if(t){if("string"==typeof t)return u(t,e);var n=Object.prototype.toString.call(t).slice(8,-1);return"Object"===n&&t.constructor&&(n=t.constructor.name),"Map"===n||"Set"===n?Array.from(t):"Arguments"===n||/^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)?u(t,e):void 0}}(t)||function(){throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.")}()}function u(t,e){(null==e||e>t.length)&&(e=t.length);for(var n=0,r=new Array(e);n<e;n++)r[n]=t[n];return r}function c(t){return Array.isArray?Array.isArray(t):"[object Array]"===m(t)}var h=1/0;function f(t){return null==t?"":function(t){if("string"==typeof t)return t;var e=t+"";return"0"==e&&1/t==-h?"-0":e}(t)}function l(t){return"string"==typeof t}function d(t){return"number"==typeof t}function p(t){return!0===t||!1===t||function(t){return y(t)&&null!==t}(t)&&"[object Boolean]"==m(t)}function y(t){return"object"===n(t)}function g(t){return null!=t}function v(t){return!t.trim().length}function m(t){return null==t?void 0===t?"[object Undefined]":"[object Null]":Object.prototype.toString.call(t)}var _=function(t){return"Missing ".concat(t," property in key")},b=function(t){return"Property 'weight' in key '".concat(t,"' must be a positive integer")},w=Object.prototype.hasOwnProperty,x=function(){function t(e){var n=this;r(this,t),this._keys=[],this._keyMap={};var i=0;e.forEach((function(t){var e=k(t);i+=e.weight,n._keys.push(e),n._keyMap[e.id]=e,i+=e.weight})),this._keys.forEach((function(t){t.weight/=i}))}return o(t,[{key:"get",value:function(t){return this._keyMap[t]}},{key:"keys",value:function(){return this._keys}},{key:"toJSON",value:function(){return JSON.stringify(this._keys)}}]),t}();function k(t){var e=null,n=null,r=null,i=1,o=null;if(l(t)||c(t))r=t,e=M(t),n=E(t);else{if(!w.call(t,"name"))throw new Error(_("name"));var a=t.name;if(r=a,w.call(t,"weight")&&(i=t.weight)<=0)throw new Error(b(a));e=M(a),n=E(a),o=t.getFn}return{path:e,id:n,weight:i,src:r,getFn:o}}function M(t){return c(t)?t:t.split(".")}function E(t){return c(t)?t.join("."):t}var A={useExtendedSearch:!1,getFn:function(t,e){var n=[],r=!1;return function t(e,i,o){if(g(e))if(i[o]){var a=e[i[o]];if(!g(a))return;if(o===i.length-1&&(l(a)||d(a)||p(a)))n.push(f(a));else if(c(a)){r=!0;for(var s=0,u=a.length;s<u;s+=1)t(a[s],i,o+1)}else i.length&&t(a,i,o+1)}else n.push(e)}(t,l(e)?e.split("."):e,0),r?n:n[0]},ignoreLocation:!1,ignoreFieldNorm:!1,fieldNormWeight:1},S=e(e(e(e({},{isCaseSensitive:!1,includeScore:!1,keys:[],shouldSort:!0,sortFn:function(t,e){return t.score===e.score?t.idx<e.idx?-1:1:t.score<e.score?-1:1}}),{includeMatches:!1,findAllMatches:!1,minMatchCharLength:1}),{location:0,threshold:.6,distance:100}),A),L=/[^ ]+/g;var N=function(){function t(){var e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{},n=e.getFn,i=void 0===n?S.getFn:n,o=e.fieldNormWeight,a=void 0===o?S.fieldNormWeight:o;r(this,t),this.norm=function(){var t=arguments.length>0&&void 0!==arguments[0]?arguments[0]:1,e=arguments.length>1&&void 0!==arguments[1]?arguments[1]:3,n=new Map,r=Math.pow(10,e);return{get:function(e){var i=e.match(L).length;if(n.has(i))return n.get(i);var o=1/Math.pow(i,.5*t),a=parseFloat(Math.round(o*r)/r);return n.set(i,a),a},clear:function(){n.clear()}}}(a,3),this.getFn=i,this.isCreated=!1,this.setIndexRecords()}return o(t,[{key:"setSources",value:function(){var t=arguments.length>0&&void 0!==arguments[0]?arguments[0]:[];this.docs=t}},{key:"setIndexRecords",value:function(){var t=arguments.length>0&&void 0!==arguments[0]?arguments[0]:[];this.records=t}},{key:"setKeys",value:function(){var t=this,e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:[];this.keys=e,this._keysMap={},e.forEach((function(e,n){t._keysMap[e.id]=n}))}},{key:"create",value:function(){var t=this;!this.isCreated&&this.docs.length&&(this.isCreated=!0,l(this.docs[0])?this.docs.forEach((function(e,n){t._addString(e,n)})):this.docs.forEach((function(e,n){t._addObject(e,n)})),this.norm.clear())}},{key:"add",value:function(t){var e=this.size();l(t)?this._addString(t,e):this._addObject(t,e)}},{key:"removeAt",value:function(t){this.records.splice(t,1);for(var e=t,n=this.size();e<n;e+=1)this.records[e].i-=1}},{key:"getValueForItemAtKeyId",value:function(t,e){return t[this._keysMap[e]]}},{key:"size",value:function(){return this.records.length}},{key:"_addString",value:function(t,e){if(g(t)&&!v(t)){var n={v:t,i:e,n:this.norm.get(t)};this.records.push(n)}}},{key:"_addObject",value:function(t,e){var n=this,r={i:e,$:{}};this.keys.forEach((function(e,i){var o=e.getFn?e.getFn(t):n.getFn(t,e.path);if(g(o))if(c(o))!function(){for(var t=[],e=[{nestedArrIndex:-1,value:o}];e.length;){var a=e.pop(),s=a.nestedArrIndex,u=a.value;if(g(u))if(l(u)&&!v(u)){var h={v:u,i:s,n:n.norm.get(u)};t.push(h)}else c(u)&&u.forEach((function(t,n){e.push({nestedArrIndex:n,value:t})}))}r.$[i]=t}();else if(l(o)&&!v(o)){var a={v:o,n:n.norm.get(o)};r.$[i]=a}})),this.records.push(r)}},{key:"toJSON",value:function(){return{keys:this.keys,records:this.records}}}]),t}();function $(t,e){var n=arguments.length>2&&void 0!==arguments[2]?arguments[2]:{},r=n.getFn,i=void 0===r?S.getFn:r,o=n.fieldNormWeight,a=void 0===o?S.fieldNormWeight:o,s=new N({getFn:i,fieldNormWeight:a});return s.setKeys(t.map(k)),s.setSources(e),s.create(),s}function C(t){var e=arguments.length>1&&void 0!==arguments[1]?arguments[1]:{},n=e.errors,r=void 0===n?0:n,i=e.currentLocation,o=void 0===i?0:i,a=e.expectedLocation,s=void 0===a?0:a,u=e.distance,c=void 0===u?S.distance:u,h=e.ignoreLocation,f=void 0===h?S.ignoreLocation:h,l=r/t.length;if(f)return l;var d=Math.abs(s-o);return c?l+d/c:d?1:l}var T=32;function G(t,e,n){var r=arguments.length>3&&void 0!==arguments[3]?arguments[3]:{},i=r.location,o=void 0===i?S.location:i,a=r.distance,s=void 0===a?S.distance:a,u=r.threshold,c=void 0===u?S.threshold:u,h=r.findAllMatches,f=void 0===h?S.findAllMatches:h,l=r.minMatchCharLength,d=void 0===l?S.minMatchCharLength:l,p=r.includeMatches,y=void 0===p?S.includeMatches:p,g=r.ignoreLocation,v=void 0===g?S.ignoreLocation:g;if(e.length>T)throw new Error("Pattern length exceeds max of ".concat(T,"."));for(var m,_=e.length,b=t.length,w=Math.max(0,Math.min(o,b)),x=c,k=w,M=d>1||y,E=M?Array(b):[];(m=t.indexOf(e,k))>-1;){var A=C(e,{currentLocation:m,expectedLocation:w,distance:s,ignoreLocation:v});if(x=Math.min(A,x),k=m+_,M)for(var L=0;L<_;)E[m+L]=1,L+=1}k=-1;for(var N=[],$=1,G=_+b,U=1<<_-1,D=0;D<_;D+=1){for(var O=0,j=G;O<j;)C(e,{errors:D,currentLocation:w+j,expectedLocation:w,distance:s,ignoreLocation:v})<=x?O=j:G=j,j=Math.floor((G-O)/2+O);G=j;var z=Math.max(1,w-j+1),P=f?b:Math.min(w+j,b)+_,I=Array(P+2);I[P+1]=(1<<D)-1;for(var R=P;R>=z;R-=1){var F=R-1,q=n[t.charAt(F)];if(M&&(E[F]=+!!q),I[R]=(I[R+1]<<1|1)&q,D&&(I[R]|=(N[R+1]|N[R])<<1|1|N[R+1]),I[R]&U&&($=C(e,{errors:D,currentLocation:F,expectedLocation:w,distance:s,ignoreLocation:v}))<=x){if(x=$,(k=F)<=w)break;z=Math.max(1,2*w-k)}}if(C(e,{errors:D+1,currentLocation:w,expectedLocation:w,distance:s,ignoreLocation:v})>x)break;N=I}var Y={isMatch:k>=0,score:Math.max(.001,$)};if(M){var B=function(){for(var t=arguments.length>0&&void 0!==arguments[0]?arguments[0]:[],e=arguments.length>1&&void 0!==arguments[1]?arguments[1]:S.minMatchCharLength,n=[],r=-1,i=-1,o=0,a=t.length;o<a;o+=1){var s=t[o];s&&-1===r?r=o:s||-1===r||((i=o-1)-r+1>=e&&n.push([r,i]),r=-1)}return t[o-1]&&o-r>=e&&n.push([r,o-1]),n}(E,d);B.length?y&&(Y.indices=B):Y.isMatch=!1}return Y}function U(t){for(var e={},n=0,r=t.length;n<r;n+=1){var i=t.charAt(n);e[i]=(e[i]||0)|1<<r-n-1}return e}var D=function(){function t(e){var n=this,i=arguments.length>1&&void 0!==arguments[1]?arguments[1]:{},o=i.location,a=void 0===o?S.location:o,s=i.threshold,u=void 0===s?S.threshold:s,c=i.distance,h=void 0===c?S.distance:c,f=i.includeMatches,l=void 0===f?S.includeMatches:f,d=i.findAllMatches,p=void 0===d?S.findAllMatches:d,y=i.minMatchCharLength,g=void 0===y?S.minMatchCharLength:y,v=i.isCaseSensitive,m=void 0===v?S.isCaseSensitive:v,_=i.ignoreLocation,b=void 0===_?S.ignoreLocation:_;if(r(this,t),this.options={location:a,threshold:u,distance:h,includeMatches:l,findAllMatches:p,minMatchCharLength:g,isCaseSensitive:m,ignoreLocation:b},this.pattern=m?e:e.toLowerCase(),this.chunks=[],this.pattern.length){var w=function(t,e){n.chunks.push({pattern:t,alphabet:U(t),startIndex:e})},x=this.pattern.length;if(x>T){for(var k=0,M=x%T,E=x-M;k<E;)w(this.pattern.substr(k,T),k),k+=T;if(M){var A=x-T;w(this.pattern.substr(A),A)}}else w(this.pattern,0)}}return o(t,[{key:"searchIn",value:function(t){var e=this.options,n=e.isCaseSensitive,r=e.includeMatches;if(n||(t=t.toLowerCase()),this.pattern===t){var i={isMatch:!0,score:0};return r&&(i.indices=[[0,t.length-1]]),i}var o=this.options,a=o.location,u=o.distance,c=o.threshold,h=o.findAllMatches,f=o.minMatchCharLength,l=o.ignoreLocation,d=[],p=0,y=!1;this.chunks.forEach((function(e){var n=e.pattern,i=e.alphabet,o=e.startIndex,g=G(t,n,i,{location:a+o,distance:u,threshold:c,findAllMatches:h,minMatchCharLength:f,includeMatches:r,ignoreLocation:l}),v=g.isMatch,m=g.score,_=g.indices;v&&(y=!0),p+=m,v&&_&&(d=[].concat(s(d),s(_)))}));var g={isMatch:y,score:y?p/this.chunks.length:1};return y&&r&&(g.indices=d),g}}]),t}(),O=[];function j(t,e){for(var n=0,r=O.length;n<r;n+=1){var i=O[n];if(i.condition(t,e))return new i(t,e)}return new D(t,e)}var z="$and",P="$or",I="$path",R="$val",F=function(t){return!(!t[z]&&!t[P])},q=function(t){return a({},z,Object.keys(t).map((function(e){return a({},e,t[e])})))};function Y(t,e){var n=t.matches;e.matches=[],g(n)&&n.forEach((function(t){if(g(t.indices)&&t.indices.length){var n={indices:t.indices,value:t.value};t.key&&(n.key=t.key.src),t.idx>-1&&(n.refIndex=t.idx),e.matches.push(n)}}))}function B(t,e){e.score=t.score}var W=function(){function t(n){var i=arguments.length>1&&void 0!==arguments[1]?arguments[1]:{},o=arguments.length>2?arguments[2]:void 0;if(r(this,t),this.options=e(e({},S),i),this.options.useExtendedSearch)throw new Error("Extended search is not available");this._keyStore=new x(this.options.keys),this.setCollection(n,o)}return o(t,[{key:"setCollection",value:function(t,e){if(this._docs=t,e&&!(e instanceof N))throw new Error("Incorrect 'index' type");this._myIndex=e||$(this.options.keys,this._docs,{getFn:this.options.getFn,fieldNormWeight:this.options.fieldNormWeight})}},{key:"add",value:function(t){g(t)&&(this._docs.push(t),this._myIndex.add(t))}},{key:"remove",value:function(){for(var t=arguments.length>0&&void 0!==arguments[0]?arguments[0]:function(){return!1},e=[],n=0,r=this._docs.length;n<r;n+=1){var i=this._docs[n];t(i,n)&&(this.removeAt(n),n-=1,r-=1,e.push(i))}return e}},{key:"removeAt",value:function(t){this._docs.splice(t,1),this._myIndex.removeAt(t)}},{key:"getIndex",value:function(){return this._myIndex}},{key:"search",value:function(t){var e=(arguments.length>1&&void 0!==arguments[1]?arguments[1]:{}).limit,n=void 0===e?-1:e,r=this.options,i=r.includeMatches,o=r.includeScore,a=r.shouldSort,s=r.sortFn,u=r.ignoreFieldNorm,c=l(t)?l(this._docs[0])?this._searchStringList(t):this._searchObjectList(t):this._searchLogical(t);return function(t,e){var n=e.ignoreFieldNorm,r=void 0===n?S.ignoreFieldNorm:n;t.forEach((function(t){var e=1;t.matches.forEach((function(t){var n=t.key,i=t.norm,o=t.score,a=n?n.weight:null;e*=Math.pow(0===o&&a?Number.EPSILON:o,(a||1)*(r?1:i))})),t.score=e}))}(c,{ignoreFieldNorm:u}),a&&c.sort(s),d(n)&&n>-1&&(c=c.slice(0,n)),function(t,e){var n=arguments.length>2&&void 0!==arguments[2]?arguments[2]:{},r=n.includeMatches,i=void 0===r?S.includeMatches:r,o=n.includeScore,a=void 0===o?S.includeScore:o,s=[];return i&&s.push(Y),a&&s.push(B),t.map((function(t){var n=t.idx,r={item:e[n],refIndex:n};return s.length&&s.forEach((function(e){e(t,r)})),r}))}(c,this._docs,{includeMatches:i,includeScore:o})}},{key:"_searchStringList",value:function(t){var e=j(t,this.options),n=this._myIndex.records,r=[];return n.forEach((function(t){var n=t.v,i=t.i,o=t.n;if(g(n)){var a=e.searchIn(n),s=a.isMatch,u=a.score,c=a.indices;s&&r.push({item:n,idx:i,matches:[{score:u,value:n,norm:o,indices:c}]})}})),r}},{key:"_searchLogical",value:function(t){throw new Error("Logical search is not available")}},{key:"_searchObjectList",value:function(t){var e=this,n=j(t,this.options),r=this._myIndex,i=r.keys,o=r.records,a=[];return o.forEach((function(t){var r=t.$,o=t.i;if(g(r)){var u=[];i.forEach((function(t,i){u.push.apply(u,s(e._findMatches({key:t,value:r[i],searcher:n})))})),u.length&&a.push({idx:o,item:r,matches:u})}})),a}},{key:"_findMatches",value:function(t){var e=t.key,n=t.value,r=t.searcher;if(!g(n))return[];var i=[];if(c(n))n.forEach((function(t){var n=t.v,o=t.i,a=t.n;if(g(n)){var s=r.searchIn(n),u=s.isMatch,c=s.score,h=s.indices;u&&i.push({score:c,key:e,value:n,idx:o,norm:a,indices:h})}}));else{var o=n.v,a=n.n,s=r.searchIn(o),u=s.isMatch,h=s.score,f=s.indices;u&&i.push({score:h,key:e,value:o,norm:a,indices:f})}return i}}]),t}();return W.version="6.6.2",W.createIndex=$,W.parseIndex=function(t){var e=arguments.length>1&&void 0!==arguments[1]?arguments[1]:{},n=e.getFn,r=void 0===n?S.getFn:n,i=e.fieldNormWeight,o=void 0===i?S.fieldNormWeight:i,a=t.keys,s=t.records,u=new N({getFn:r,fieldNormWeight:o});return u.setKeys(a),u.setIndexRecords(s),u},W.config=S,W.parseQuery=function(t,e){var n=(arguments.length>2&&void 0!==arguments[2]?arguments[2]:{}).auto,r=void 0===n||n;return F(t)||(t=q(t)),function t(n){var i=Object.keys(n),o=function(t){return!!t[I]}(n);if(!o&&i.length>1&&!F(n))return t(q(n));if(function(t){return!c(t)&&y(t)&&!F(t)}(n)){var a=o?n[I]:i[0],s=o?n[R]:n[a];if(!l(s))throw new Error(function(t){return"Invalid value for key ".concat(t)}(a));var u={keyId:E(a),pattern:s};return r&&(u.searcher=j(s,e)),u}var h={children:[],operator:i[0]};return i.forEach((function(e){var r=n[e];c(r)&&r.forEach((function(e){h.children.push(t(e))}))})),h}(t)},W}()},12:(t,e,n)=>{var r=n(453);function i(t){this.graph=t,this.queue=new r(Array,t.order),this.seen=new Set,this.size=0}i.prototype.hasAlreadySeenEverything=function(){return this.seen.size===this.graph.order},i.prototype.countUnseenNodes=function(){return this.graph.order-this.seen.size},i.prototype.forEachNodeYetUnseen=function(t){var e=this.seen,n=this.graph;n.someNode((function(r,i){return e.size===n.order||!e.has(r)&&!!t(r,i)}))},i.prototype.has=function(t){return this.seen.has(t)},i.prototype.push=function(t){var e=this.seen.size;return this.seen.add(t),e!==this.seen.size&&(this.queue.push(t),this.size++,!0)},i.prototype.pushWith=function(t,e){var n=this.seen.size;return this.seen.add(t),n!==this.seen.size&&(this.queue.push(e),this.size++,!0)},i.prototype.shift=function(){var t=this.queue.shift();return this.size=this.queue.size,t},t.exports=i},181:(t,e,n)=>{var r=n(186),i=n(12),o=n(480),a=o.TraversalRecord,s=o.capitalize;e.a=function(t,e,n,o){if(o=o||{},!r(t))throw new Error("graphology-traversal/bfs: expecting a graphology instance.");if("function"!=typeof n)throw new Error("graphology-traversal/bfs: given callback is not a function.");if(0!==t.order){var u,c=new i(t),h=t["forEach"+s(o.mode||"outbound")+"Neighbor"].bind(t);(null===e?c.forEachNodeYetUnseen.bind(c):function(n){n(e=""+e,t.getNodeAttributes(e))})((function(t,e){for(c.pushWith(t,new a(t,e,0));0!==c.size;)u=c.shift(),!0!==n(u.node,u.attributes,u.depth)&&h(u.node,f)}))}function f(t,e){c.pushWith(t,new a(t,e,u.depth+1))}}},480:(t,e)=>{e.TraversalRecord=function(t,e,n){this.node=t,this.attributes=e,this.depth=n},e.capitalize=function(t){return t[0].toUpperCase()+t.slice(1)}},186:t=>{t.exports=function(t){return null!==t&&"object"==typeof t&&"function"==typeof t.addUndirectedEdgeWithKey&&"function"==typeof t.dropNode&&"boolean"==typeof t.multi}},453:(t,e,n)=>{var r=n(650),i=n(993);function o(t,e){if(arguments.length<2)throw new Error("mnemonist/fixed-deque: expecting an Array class and a capacity.");if("number"!=typeof e||e<=0)throw new Error("mnemonist/fixed-deque: `capacity` should be a positive number.");this.ArrayClass=t,this.capacity=e,this.items=new t(this.capacity),this.clear()}o.prototype.clear=function(){this.start=0,this.size=0},o.prototype.push=function(t){if(this.size===this.capacity)throw new Error("mnemonist/fixed-deque.push: deque capacity ("+this.capacity+") exceeded!");var e=(this.start+this.size)%this.capacity;return this.items[e]=t,++this.size},o.prototype.unshift=function(t){if(this.size===this.capacity)throw new Error("mnemonist/fixed-deque.unshift: deque capacity ("+this.capacity+") exceeded!");var e=this.start-1;return 0===this.start&&(e=this.capacity-1),this.items[e]=t,this.start=e,++this.size},o.prototype.pop=function(){if(0===this.size)return;const t=(this.start+this.size-1)%this.capacity;return this.size--,this.items[t]},o.prototype.shift=function(){if(0!==this.size){var t=this.start;return this.size--,this.start++,this.start===this.capacity&&(this.start=0),this.items[t]}},o.prototype.peekFirst=function(){if(0!==this.size)return this.items[this.start]},o.prototype.peekLast=function(){if(0!==this.size){var t=this.start+this.size-1;return t>this.capacity&&(t-=this.capacity),this.items[t]}},o.prototype.get=function(t){if(0!==this.size)return(t=this.start+t)>this.capacity&&(t-=this.capacity),this.items[t]},o.prototype.forEach=function(t,e){e=arguments.length>1?e:this;for(var n=this.capacity,r=this.size,i=this.start,o=0;o<r;)t.call(e,this.items[i],o,this),o++,++i===n&&(i=0)},o.prototype.toArray=function(){var t=this.start+this.size;if(t<this.capacity)return this.items.slice(this.start,t);for(var e=new this.ArrayClass(this.size),n=this.capacity,r=this.size,i=this.start,o=0;o<r;)e[o]=this.items[i],o++,++i===n&&(i=0);return e},o.prototype.values=function(){var t=this.items,e=this.capacity,n=this.size,r=this.start,o=0;return new i((function(){if(o>=n)return{done:!0};var i=t[r];return r++,o++,r===e&&(r=0),{value:i,done:!1}}))},o.prototype.entries=function(){var t=this.items,e=this.capacity,n=this.size,r=this.start,o=0;return new i((function(){if(o>=n)return{done:!0};var i=t[r];return++r===e&&(r=0),{value:[o++,i],done:!1}}))},"undefined"!=typeof Symbol&&(o.prototype[Symbol.iterator]=o.prototype.values),o.prototype.inspect=function(){var t=this.toArray();return t.type=this.ArrayClass.name,t.capacity=this.capacity,Object.defineProperty(t,"constructor",{value:o,enumerable:!1}),t},"undefined"!=typeof Symbol&&(o.prototype[Symbol.for("nodejs.util.inspect.custom")]=o.prototype.inspect),o.from=function(t,e,n){if(arguments.length<3&&"number"!=typeof(n=r.guessLength(t)))throw new Error("mnemonist/fixed-deque.from: could not guess iterable length. Please provide desired capacity as last argument.");var i=new o(e,n);if(r.isArrayLike(t)){var a,s;for(a=0,s=t.length;a<s;a++)i.items[a]=t[a];return i.size=s,i}return r.forEach(t,(function(t){i.push(t)})),i},t.exports=o},650:(t,e,n)=>{var r=n(113),i=n(801);function o(t){return"number"==typeof t.length?t.length:"number"==typeof t.size?t.size:void 0}e.isArrayLike=function(t){return Array.isArray(t)||i.isTypedArray(t)},e.guessLength=o,e.toArray=function(t){var e=o(t),n="number"==typeof e?new Array(e):[],i=0;return r(t,(function(t){n[i++]=t})),n},e.toArrayWithIndices=function(t){var e=o(t),n="number"==typeof e?i.getPointerArray(e):Array,a="number"==typeof e?new Array(e):[],s="number"==typeof e?new n(e):[],u=0;return r(t,(function(t){a[u]=t,s[u]=u++})),[a,s]}},801:(t,e)=>{var n=Math.pow(2,8)-1,r=Math.pow(2,16)-1,i=Math.pow(2,32)-1,o=Math.pow(2,7)-1,a=Math.pow(2,15)-1,s=Math.pow(2,31)-1;e.getPointerArray=function(t){var e=t-1;if(e<=n)return Uint8Array;if(e<=r)return Uint16Array;if(e<=i)return Uint32Array;throw new Error("mnemonist: Pointer Array of size > 4294967295 is not supported.")},e.getSignedPointerArray=function(t){var e=t-1;return e<=o?Int8Array:e<=a?Int16Array:e<=s?Int32Array:Float64Array},e.getNumberType=function(t){return t===(0|t)?-1===Math.sign(t)?t<=127&&t>=-128?Int8Array:t<=32767&&t>=-32768?Int16Array:Int32Array:t<=255?Uint8Array:t<=65535?Uint16Array:Uint32Array:Float64Array};var u={Uint8Array:1,Int8Array:2,Uint16Array:3,Int16Array:4,Uint32Array:5,Int32Array:6,Float32Array:7,Float64Array:8};e.getMinimalRepresentation=function(t,n){var r,i,o,a,s,c=null,h=0;for(a=0,s=t.length;a<s;a++)o=n?n(t[a]):t[a],i=e.getNumberType(o),(r=u[i.name])>h&&(h=r,c=i);return c},e.isTypedArray=function(t){return"undefined"!=typeof ArrayBuffer&&ArrayBuffer.isView(t)},e.concat=function(){var t,e,n,r=0;for(t=0,n=arguments.length;t<n;t++)r+=arguments[t].length;var i=new arguments[0].constructor(r);for(t=0,e=0;t<n;t++)i.set(arguments[t],e),e+=arguments[t].length;return i},e.indices=function(t){for(var n=new(e.getPointerArray(t))(t),r=0;r<t;r++)n[r]=r;return n}},11:(t,e,n)=>{var r=n(993),i=n(346);t.exports=function(){var t=arguments,e=null,n=-1;return new r((function(){for(var r=null;;){if(null===e){if(++n>=t.length)return{done:!0};e=i(t[n])}if(!0!==(r=e.next()).done)break;e=null}return r}))}},113:(t,e,n)=>{var r=n(844),i=r.ARRAY_BUFFER_SUPPORT,o=r.SYMBOL_SUPPORT;t.exports=function(t,e){var n,r,a,s,u;if(!t)throw new Error("obliterator/forEach: invalid iterable.");if("function"!=typeof e)throw new Error("obliterator/forEach: expecting a callback.");if(Array.isArray(t)||i&&ArrayBuffer.isView(t)||"string"==typeof t||"[object Arguments]"===t.toString())for(a=0,s=t.length;a<s;a++)e(t[a],a);else if("function"!=typeof t.forEach)if(o&&Symbol.iterator in t&&"function"!=typeof t.next&&(t=t[Symbol.iterator]()),"function"!=typeof t.next)for(r in t)t.hasOwnProperty(r)&&e(t[r],r);else for(n=t,a=0;!0!==(u=n.next()).done;)e(u.value,a),a++;else t.forEach(e)}},346:(t,e,n)=>{var r=n(993),i=n(844),o=i.ARRAY_BUFFER_SUPPORT,a=i.SYMBOL_SUPPORT;t.exports=function(t){var e=function(t){return"string"==typeof t||Array.isArray(t)||o&&ArrayBuffer.isView(t)?r.fromSequence(t):"object"!=typeof t||null===t?null:a&&"function"==typeof t[Symbol.iterator]?t[Symbol.iterator]():"function"==typeof t.next?t:null}(t);if(!e)throw new Error("obliterator: target is not iterable nor a valid iterator.");return e}},993:t=>{function e(t){if("function"!=typeof t)throw new Error("obliterator/iterator: expecting a function!");this.next=t}"undefined"!=typeof Symbol&&(e.prototype[Symbol.iterator]=function(){return this}),e.of=function(){var t=arguments,n=t.length,r=0;return new e((function(){return r>=n?{done:!0}:{done:!1,value:t[r++]}}))},e.empty=function(){return new e((function(){return{done:!0}}))},e.fromSequence=function(t){var n=0,r=t.length;return new e((function(){return n>=r?{done:!0}:{done:!1,value:t[n++]}}))},e.is=function(t){return t instanceof e||"object"==typeof t&&null!==t&&"function"==typeof t.next},t.exports=e},844:(t,e)=>{e.ARRAY_BUFFER_SUPPORT="undefined"!=typeof ArrayBuffer,e.SYMBOL_SUPPORT="undefined"!=typeof Symbol},641:(t,e,n)=>{var r=n(346);t.exports=function(t,e){for(var n,i=arguments.length>1?e:1/0,o=i!==1/0?new Array(i):[],a=0,s=r(t);;){if(a===i)return o;if((n=s.next()).done)return a!==e&&(o.length=a),o;o[a++]=n.value}}}},e={};function n(r){var i=e[r];if(void 0!==i)return i.exports;var o=e[r]={exports:{}};return t[r].call(o.exports,o,o.exports,n),o.exports}n.n=t=>{var e=t&&t.__esModule?()=>t.default:()=>t;return n.d(e,{a:e}),e},n.d=(t,e)=>{for(var r in e)n.o(e,r)&&!n.o(t,r)&&Object.defineProperty(t,r,{enumerable:!0,get:e[r]})},n.o=(t,e)=>Object.prototype.hasOwnProperty.call(t,e),(()=>{"use strict";function t(e){return t="function"==typeof Symbol&&"symbol"==typeof Symbol.iterator?function(t){return typeof t}:function(t){return t&&"function"==typeof Symbol&&t.constructor===Symbol&&t!==Symbol.prototype?"symbol":typeof t},t(e)}function e(t,e){for(var n=0;n<e.length;n++){var r=e[n];r.enumerable=r.enumerable||!1,r.configurable=!0,"value"in r&&(r.writable=!0),Object.defineProperty(t,o(r.key),r)}}function r(t,n,r){return n&&e(t.prototype,n),r&&e(t,r),Object.defineProperty(t,"prototype",{writable:!1}),t}function i(t,e,n){return(e=o(e))in t?Object.defineProperty(t,e,{value:n,enumerable:!0,configurable:!0,writable:!0}):t[e]=n,t}function o(e){var n=function(e,n){if("object"!==t(e)||null===e)return e;var r=e[Symbol.toPrimitive];if(void 0!==r){var i=r.call(e,"string");if("object"!==t(i))return i;throw new TypeError("@@toPrimitive must return a primitive value.")}return String(e)}(e);return"symbol"===t(n)?n:String(n)}var a,s,u=r((function t(){!function(t,e){if(!(t instanceof e))throw new TypeError("Cannot call a class as a function")}(this,t)}));function c(t,e){return t<e?-1:t>e?1:t>=e?0:NaN}i(u,"highlightedNodes",[]),i(u,"focusMode",!1),i(u,"openedRecordId",void 0),i(u,"position",{x:0,y:0,zoom:1}),window.addEventListener("DOMContentLoaded",(function(){document.getElementById("view-save").addEventListener("click",(function(){var t=new URL(window.location),e=Array.from(document.querySelectorAll("#types-form input:checked")).map((function(t){return t.name})),n=Array.from(document.querySelectorAll("#tags-form input:checked")).map((function(t){return t.name})),r=document.getElementById("focus-input").value;e.length>0&&t.searchParams.set("filters",e.join("-")),n.length>0&&t.searchParams.set("tags",n.join("-")),document.querySelector("#focus-check:checked")&&t.searchParams.set("focus",r),window.location.replace(t)}))})),1===(a=c).length&&(s=a,a=function(t,e){return c(s(t),e)});var h=Array.prototype;h.slice,h.map,Math.sqrt(50),Math.sqrt(10),Math.sqrt(2),Array.prototype.slice;var f={value:function(){}};function l(){for(var t,e=0,n=arguments.length,r={};e<n;++e){if(!(t=arguments[e]+"")||t in r)throw new Error("illegal type: "+t);r[t]=[]}return new d(r)}function d(t){this._=t}function p(t,e){for(var n,r=0,i=t.length;r<i;++r)if((n=t[r]).name===e)return n.value}function y(t,e,n){for(var r=0,i=t.length;r<i;++r)if(t[r].name===e){t[r]=f,t=t.slice(0,r).concat(t.slice(r+1));break}return null!=n&&t.push({name:e,value:n}),t}d.prototype=l.prototype={constructor:d,on:function(t,e){var n,r,i=this._,o=(r=i,(t+"").trim().split(/^|\s+/).map((function(t){var e="",n=t.indexOf(".");if(n>=0&&(e=t.slice(n+1),t=t.slice(0,n)),t&&!r.hasOwnProperty(t))throw new Error("unknown type: "+t);return{type:t,name:e}}))),a=-1,s=o.length;if(!(arguments.length<2)){if(null!=e&&"function"!=typeof e)throw new Error("invalid callback: "+e);for(;++a<s;)if(n=(t=o[a]).type)i[n]=y(i[n],t.name,e);else if(null==e)for(n in i)i[n]=y(i[n],t.name,null);return this}for(;++a<s;)if((n=(t=o[a]).type)&&(n=p(i[n],t.name)))return n},copy:function(){var t={},e=this._;for(var n in e)t[n]=e[n].slice();return new d(t)},call:function(t,e){if((n=arguments.length-2)>0)for(var n,r,i=new Array(n),o=0;o<n;++o)i[o]=arguments[o+2];if(!this._.hasOwnProperty(t))throw new Error("unknown type: "+t);for(o=0,n=(r=this._[t]).length;o<n;++o)r[o].value.apply(e,i)},apply:function(t,e,n){if(!this._.hasOwnProperty(t))throw new Error("unknown type: "+t);for(var r=this._[t],i=0,o=r.length;i<o;++i)r[i].value.apply(e,n)}};const g=l;var v="http://www.w3.org/1999/xhtml";const m={svg:"http://www.w3.org/2000/svg",xhtml:v,xlink:"http://www.w3.org/1999/xlink",xml:"http://www.w3.org/XML/1998/namespace",xmlns:"http://www.w3.org/2000/xmlns/"};function _(t){var e=t+="",n=e.indexOf(":");return n>=0&&"xmlns"!==(e=t.slice(0,n))&&(t=t.slice(n+1)),m.hasOwnProperty(e)?{space:m[e],local:t}:t}function b(t){return function(){var e=this.ownerDocument,n=this.namespaceURI;return n===v&&e.documentElement.namespaceURI===v?e.createElement(t):e.createElementNS(n,t)}}function w(t){return function(){return this.ownerDocument.createElementNS(t.space,t.local)}}function x(t){var e=_(t);return(e.local?w:b)(e)}function k(){}function M(t){return null==t?k:function(){return this.querySelector(t)}}function E(){return[]}function A(t){return null==t?E:function(){return this.querySelectorAll(t)}}var S=function(t){return function(){return this.matches(t)}};if("undefined"!=typeof document){var L=document.documentElement;if(!L.matches){var N=L.webkitMatchesSelector||L.msMatchesSelector||L.mozMatchesSelector||L.oMatchesSelector;S=function(t){return function(){return N.call(this,t)}}}}const $=S;function C(t){return new Array(t.length)}function T(t,e){this.ownerDocument=t.ownerDocument,this.namespaceURI=t.namespaceURI,this._next=null,this._parent=t,this.__data__=e}T.prototype={constructor:T,appendChild:function(t){return this._parent.insertBefore(t,this._next)},insertBefore:function(t,e){return this._parent.insertBefore(t,e)},querySelector:function(t){return this._parent.querySelector(t)},querySelectorAll:function(t){return this._parent.querySelectorAll(t)}};function G(t,e,n,r,i,o){for(var a,s=0,u=e.length,c=o.length;s<c;++s)(a=e[s])?(a.__data__=o[s],r[s]=a):n[s]=new T(t,o[s]);for(;s<u;++s)(a=e[s])&&(i[s]=a)}function U(t,e,n,r,i,o,a){var s,u,c,h={},f=e.length,l=o.length,d=new Array(f);for(s=0;s<f;++s)(u=e[s])&&(d[s]=c="$"+a.call(u,u.__data__,s,e),c in h?i[s]=u:h[c]=u);for(s=0;s<l;++s)(u=h[c="$"+a.call(t,o[s],s,o)])?(r[s]=u,u.__data__=o[s],h[c]=null):n[s]=new T(t,o[s]);for(s=0;s<f;++s)(u=e[s])&&h[d[s]]===u&&(i[s]=u)}function D(t,e){return t<e?-1:t>e?1:t>=e?0:NaN}function O(t){return function(){this.removeAttribute(t)}}function j(t){return function(){this.removeAttributeNS(t.space,t.local)}}function z(t,e){return function(){this.setAttribute(t,e)}}function P(t,e){return function(){this.setAttributeNS(t.space,t.local,e)}}function I(t,e){return function(){var n=e.apply(this,arguments);null==n?this.removeAttribute(t):this.setAttribute(t,n)}}function R(t,e){return function(){var n=e.apply(this,arguments);null==n?this.removeAttributeNS(t.space,t.local):this.setAttributeNS(t.space,t.local,n)}}function F(t){return t.ownerDocument&&t.ownerDocument.defaultView||t.document&&t||t.defaultView}function q(t){return function(){this.style.removeProperty(t)}}function Y(t,e,n){return function(){this.style.setProperty(t,e,n)}}function B(t,e,n){return function(){var r=e.apply(this,arguments);null==r?this.style.removeProperty(t):this.style.setProperty(t,r,n)}}function W(t,e){return t.style.getPropertyValue(e)||F(t).getComputedStyle(t,null).getPropertyValue(e)}function K(t){return function(){delete this[t]}}function H(t,e){return function(){this[t]=e}}function X(t,e){return function(){var n=e.apply(this,arguments);null==n?delete this[t]:this[t]=n}}function V(t){return t.trim().split(/^|\s+/)}function Z(t){return t.classList||new J(t)}function J(t){this._node=t,this._names=V(t.getAttribute("class")||"")}function Q(t,e){for(var n=Z(t),r=-1,i=e.length;++r<i;)n.add(e[r])}function tt(t,e){for(var n=Z(t),r=-1,i=e.length;++r<i;)n.remove(e[r])}function et(t){return function(){Q(this,t)}}function nt(t){return function(){tt(this,t)}}function rt(t,e){return function(){(e.apply(this,arguments)?Q:tt)(this,t)}}function it(){this.textContent=""}function ot(t){return function(){this.textContent=t}}function at(t){return function(){var e=t.apply(this,arguments);this.textContent=null==e?"":e}}function st(){this.innerHTML=""}function ut(t){return function(){this.innerHTML=t}}function ct(t){return function(){var e=t.apply(this,arguments);this.innerHTML=null==e?"":e}}function ht(){this.nextSibling&&this.parentNode.appendChild(this)}function ft(){this.previousSibling&&this.parentNode.insertBefore(this,this.parentNode.firstChild)}function lt(){return null}function dt(){var t=this.parentNode;t&&t.removeChild(this)}function pt(){return this.parentNode.insertBefore(this.cloneNode(!1),this.nextSibling)}function yt(){return this.parentNode.insertBefore(this.cloneNode(!0),this.nextSibling)}J.prototype={add:function(t){this._names.indexOf(t)<0&&(this._names.push(t),this._node.setAttribute("class",this._names.join(" ")))},remove:function(t){var e=this._names.indexOf(t);e>=0&&(this._names.splice(e,1),this._node.setAttribute("class",this._names.join(" ")))},contains:function(t){return this._names.indexOf(t)>=0}};var gt={},vt=null;function mt(t,e,n){return t=_t(t,e,n),function(e){var n=e.relatedTarget;n&&(n===this||8&n.compareDocumentPosition(this))||t.call(this,e)}}function _t(t,e,n){return function(r){var i=vt;vt=r;try{t.call(this,this.__data__,e,n)}finally{vt=i}}}function bt(t){return function(){var e=this.__on;if(e){for(var n,r=0,i=-1,o=e.length;r<o;++r)n=e[r],t.type&&n.type!==t.type||n.name!==t.name?e[++i]=n:this.removeEventListener(n.type,n.listener,n.capture);++i?e.length=i:delete this.__on}}}function wt(t,e,n){var r=gt.hasOwnProperty(t.type)?mt:_t;return function(i,o,a){var s,u=this.__on,c=r(e,o,a);if(u)for(var h=0,f=u.length;h<f;++h)if((s=u[h]).type===t.type&&s.name===t.name)return this.removeEventListener(s.type,s.listener,s.capture),this.addEventListener(s.type,s.listener=c,s.capture=n),void(s.value=e);this.addEventListener(t.type,c,n),s={type:t.type,name:t.name,value:e,listener:c,capture:n},u?u.push(s):this.__on=[s]}}function xt(t,e,n,r){var i=vt;t.sourceEvent=vt,vt=t;try{return e.apply(n,r)}finally{vt=i}}function kt(t,e,n){var r=F(t),i=r.CustomEvent;"function"==typeof i?i=new i(e,n):(i=r.document.createEvent("Event"),n?(i.initEvent(e,n.bubbles,n.cancelable),i.detail=n.detail):i.initEvent(e,!1,!1)),t.dispatchEvent(i)}function Mt(t,e){return function(){return kt(this,t,e)}}function Et(t,e){return function(){return kt(this,t,e.apply(this,arguments))}}"undefined"!=typeof document&&("onmouseenter"in document.documentElement||(gt={mouseenter:"mouseover",mouseleave:"mouseout"}));var At=[null];function St(t,e){this._groups=t,this._parents=e}function Lt(){return new St([[document.documentElement]],At)}St.prototype=Lt.prototype={constructor:St,select:function(t){"function"!=typeof t&&(t=M(t));for(var e=this._groups,n=e.length,r=new Array(n),i=0;i<n;++i)for(var o,a,s=e[i],u=s.length,c=r[i]=new Array(u),h=0;h<u;++h)(o=s[h])&&(a=t.call(o,o.__data__,h,s))&&("__data__"in o&&(a.__data__=o.__data__),c[h]=a);return new St(r,this._parents)},selectAll:function(t){"function"!=typeof t&&(t=A(t));for(var e=this._groups,n=e.length,r=[],i=[],o=0;o<n;++o)for(var a,s=e[o],u=s.length,c=0;c<u;++c)(a=s[c])&&(r.push(t.call(a,a.__data__,c,s)),i.push(a));return new St(r,i)},filter:function(t){"function"!=typeof t&&(t=$(t));for(var e=this._groups,n=e.length,r=new Array(n),i=0;i<n;++i)for(var o,a=e[i],s=a.length,u=r[i]=[],c=0;c<s;++c)(o=a[c])&&t.call(o,o.__data__,c,a)&&u.push(o);return new St(r,this._parents)},data:function(t,e){if(!t)return p=new Array(this.size()),h=-1,this.each((function(t){p[++h]=t})),p;var n,r=e?U:G,i=this._parents,o=this._groups;"function"!=typeof t&&(n=t,t=function(){return n});for(var a=o.length,s=new Array(a),u=new Array(a),c=new Array(a),h=0;h<a;++h){var f=i[h],l=o[h],d=l.length,p=t.call(f,f&&f.__data__,h,i),y=p.length,g=u[h]=new Array(y),v=s[h]=new Array(y);r(f,l,g,v,c[h]=new Array(d),p,e);for(var m,_,b=0,w=0;b<y;++b)if(m=g[b]){for(b>=w&&(w=b+1);!(_=v[w])&&++w<y;);m._next=_||null}}return(s=new St(s,i))._enter=u,s._exit=c,s},enter:function(){return new St(this._enter||this._groups.map(C),this._parents)},exit:function(){return new St(this._exit||this._groups.map(C),this._parents)},merge:function(t){for(var e=this._groups,n=t._groups,r=e.length,i=n.length,o=Math.min(r,i),a=new Array(r),s=0;s<o;++s)for(var u,c=e[s],h=n[s],f=c.length,l=a[s]=new Array(f),d=0;d<f;++d)(u=c[d]||h[d])&&(l[d]=u);for(;s<r;++s)a[s]=e[s];return new St(a,this._parents)},order:function(){for(var t=this._groups,e=-1,n=t.length;++e<n;)for(var r,i=t[e],o=i.length-1,a=i[o];--o>=0;)(r=i[o])&&(a&&a!==r.nextSibling&&a.parentNode.insertBefore(r,a),a=r);return this},sort:function(t){function e(e,n){return e&&n?t(e.__data__,n.__data__):!e-!n}t||(t=D);for(var n=this._groups,r=n.length,i=new Array(r),o=0;o<r;++o){for(var a,s=n[o],u=s.length,c=i[o]=new Array(u),h=0;h<u;++h)(a=s[h])&&(c[h]=a);c.sort(e)}return new St(i,this._parents).order()},call:function(){var t=arguments[0];return arguments[0]=this,t.apply(null,arguments),this},nodes:function(){var t=new Array(this.size()),e=-1;return this.each((function(){t[++e]=this})),t},node:function(){for(var t=this._groups,e=0,n=t.length;e<n;++e)for(var r=t[e],i=0,o=r.length;i<o;++i){var a=r[i];if(a)return a}return null},size:function(){var t=0;return this.each((function(){++t})),t},empty:function(){return!this.node()},each:function(t){for(var e=this._groups,n=0,r=e.length;n<r;++n)for(var i,o=e[n],a=0,s=o.length;a<s;++a)(i=o[a])&&t.call(i,i.__data__,a,o);return this},attr:function(t,e){var n=_(t);if(arguments.length<2){var r=this.node();return n.local?r.getAttributeNS(n.space,n.local):r.getAttribute(n)}return this.each((null==e?n.local?j:O:"function"==typeof e?n.local?R:I:n.local?P:z)(n,e))},style:function(t,e,n){return arguments.length>1?this.each((null==e?q:"function"==typeof e?B:Y)(t,e,null==n?"":n)):W(this.node(),t)},property:function(t,e){return arguments.length>1?this.each((null==e?K:"function"==typeof e?X:H)(t,e)):this.node()[t]},classed:function(t,e){var n=V(t+"");if(arguments.length<2){for(var r=Z(this.node()),i=-1,o=n.length;++i<o;)if(!r.contains(n[i]))return!1;return!0}return this.each(("function"==typeof e?rt:e?et:nt)(n,e))},text:function(t){return arguments.length?this.each(null==t?it:("function"==typeof t?at:ot)(t)):this.node().textContent},html:function(t){return arguments.length?this.each(null==t?st:("function"==typeof t?ct:ut)(t)):this.node().innerHTML},raise:function(){return this.each(ht)},lower:function(){return this.each(ft)},append:function(t){var e="function"==typeof t?t:x(t);return this.select((function(){return this.appendChild(e.apply(this,arguments))}))},insert:function(t,e){var n="function"==typeof t?t:x(t),r=null==e?lt:"function"==typeof e?e:M(e);return this.select((function(){return this.insertBefore(n.apply(this,arguments),r.apply(this,arguments)||null)}))},remove:function(){return this.each(dt)},clone:function(t){return this.select(t?yt:pt)},datum:function(t){return arguments.length?this.property("__data__",t):this.node().__data__},on:function(t,e,n){var r,i,o=function(t){return t.trim().split(/^|\s+/).map((function(t){var e="",n=t.indexOf(".");return n>=0&&(e=t.slice(n+1),t=t.slice(0,n)),{type:t,name:e}}))}(t+""),a=o.length;if(!(arguments.length<2)){for(s=e?wt:bt,null==n&&(n=!1),r=0;r<a;++r)this.each(s(o[r],e,n));return this}var s=this.node().__on;if(s)for(var u,c=0,h=s.length;c<h;++c)for(r=0,u=s[c];r<a;++r)if((i=o[r]).type===u.type&&i.name===u.name)return u.value},dispatch:function(t,e){return this.each(("function"==typeof e?Et:Mt)(t,e))}};const Nt=Lt;function $t(t){return"string"==typeof t?new St([[document.querySelector(t)]],[document.documentElement]):new St([[t]],At)}var Ct=0;function Tt(){this._="@"+(++Ct).toString(36)}function Gt(){for(var t,e=vt;t=e.sourceEvent;)e=t;return e}function Ut(t,e){var n=t.ownerSVGElement||t;if(n.createSVGPoint){var r=n.createSVGPoint();return r.x=e.clientX,r.y=e.clientY,[(r=r.matrixTransform(t.getScreenCTM().inverse())).x,r.y]}var i=t.getBoundingClientRect();return[e.clientX-i.left-t.clientLeft,e.clientY-i.top-t.clientTop]}function Dt(t){var e=Gt();return e.changedTouches&&(e=e.changedTouches[0]),Ut(t,e)}function Ot(t,e,n){arguments.length<3&&(n=e,e=Gt().changedTouches);for(var r,i=0,o=e?e.length:0;i<o;++i)if((r=e[i]).identifier===n)return Ut(t,r);return null}function jt(){vt.stopImmediatePropagation()}function zt(){vt.preventDefault(),vt.stopImmediatePropagation()}function Pt(t){var e=t.document.documentElement,n=$t(t).on("dragstart.drag",zt,!0);"onselectstart"in e?n.on("selectstart.drag",zt,!0):(e.__noselect=e.style.MozUserSelect,e.style.MozUserSelect="none")}function It(t,e){var n=t.document.documentElement,r=$t(t).on("dragstart.drag",null);e&&(r.on("click.drag",zt,!0),setTimeout((function(){r.on("click.drag",null)}),0)),"onselectstart"in n?r.on("selectstart.drag",null):(n.style.MozUserSelect=n.__noselect,delete n.__noselect)}function Rt(t){return function(){return t}}function Ft(t,e,n,r,i,o,a,s,u,c){this.target=t,this.type=e,this.subject=n,this.identifier=r,this.active=i,this.x=o,this.y=a,this.dx=s,this.dy=u,this._=c}function qt(){return!vt.button}function Yt(){return this.parentNode}function Bt(t){return null==t?{x:vt.x,y:vt.y}:t}function Wt(){return"ontouchstart"in this}function Kt(t,e,n){t.prototype=e.prototype=n,n.constructor=t}function Ht(t,e){var n=Object.create(t.prototype);for(var r in e)n[r]=e[r];return n}function Xt(){}Tt.prototype=function(){return new Tt}.prototype={constructor:Tt,get:function(t){for(var e=this._;!(e in t);)if(!(t=t.parentNode))return;return t[e]},set:function(t,e){return t[this._]=e},remove:function(t){return this._ in t&&delete t[this._]},toString:function(){return this._}},Ft.prototype.on=function(){var t=this._.on.apply(this._,arguments);return t===this._?this:t};var Vt=.7,Zt=1/Vt,Jt="\\s*([+-]?\\d+)\\s*",Qt="\\s*([+-]?\\d*\\.?\\d+(?:[eE][+-]?\\d+)?)\\s*",te="\\s*([+-]?\\d*\\.?\\d+(?:[eE][+-]?\\d+)?)%\\s*",ee=/^#([0-9a-f]{3})$/,ne=/^#([0-9a-f]{6})$/,re=new RegExp("^rgb\\("+[Jt,Jt,Jt]+"\\)$"),ie=new RegExp("^rgb\\("+[te,te,te]+"\\)$"),oe=new RegExp("^rgba\\("+[Jt,Jt,Jt,Qt]+"\\)$"),ae=new RegExp("^rgba\\("+[te,te,te,Qt]+"\\)$"),se=new RegExp("^hsl\\("+[Qt,te,te]+"\\)$"),ue=new RegExp("^hsla\\("+[Qt,te,te,Qt]+"\\)$"),ce={aliceblue:15792383,antiquewhite:16444375,aqua:65535,aquamarine:8388564,azure:15794175,beige:16119260,bisque:16770244,black:0,blanchedalmond:16772045,blue:255,blueviolet:9055202,brown:10824234,burlywood:14596231,cadetblue:6266528,chartreuse:8388352,chocolate:13789470,coral:16744272,cornflowerblue:6591981,cornsilk:16775388,crimson:14423100,cyan:65535,darkblue:139,darkcyan:35723,darkgoldenrod:12092939,darkgray:11119017,darkgreen:25600,darkgrey:11119017,darkkhaki:12433259,darkmagenta:9109643,darkolivegreen:5597999,darkorange:16747520,darkorchid:10040012,darkred:9109504,darksalmon:15308410,darkseagreen:9419919,darkslateblue:4734347,darkslategray:3100495,darkslategrey:3100495,darkturquoise:52945,darkviolet:9699539,deeppink:16716947,deepskyblue:49151,dimgray:6908265,dimgrey:6908265,dodgerblue:2003199,firebrick:11674146,floralwhite:16775920,forestgreen:2263842,fuchsia:16711935,gainsboro:14474460,ghostwhite:16316671,gold:16766720,goldenrod:14329120,gray:8421504,green:32768,greenyellow:11403055,grey:8421504,honeydew:15794160,hotpink:16738740,indianred:13458524,indigo:4915330,ivory:16777200,khaki:15787660,lavender:15132410,lavenderblush:16773365,lawngreen:8190976,lemonchiffon:16775885,lightblue:11393254,lightcoral:15761536,lightcyan:14745599,lightgoldenrodyellow:16448210,lightgray:13882323,lightgreen:9498256,lightgrey:13882323,lightpink:16758465,lightsalmon:16752762,lightseagreen:2142890,lightskyblue:8900346,lightslategray:7833753,lightslategrey:7833753,lightsteelblue:11584734,lightyellow:16777184,lime:65280,limegreen:3329330,linen:16445670,magenta:16711935,maroon:8388608,mediumaquamarine:6737322,mediumblue:205,mediumorchid:12211667,mediumpurple:9662683,mediumseagreen:3978097,mediumslateblue:8087790,mediumspringgreen:64154,mediumturquoise:4772300,mediumvioletred:13047173,midnightblue:1644912,mintcream:16121850,mistyrose:16770273,moccasin:16770229,navajowhite:16768685,navy:128,oldlace:16643558,olive:8421376,olivedrab:7048739,orange:16753920,orangered:16729344,orchid:14315734,palegoldenrod:15657130,palegreen:10025880,paleturquoise:11529966,palevioletred:14381203,papayawhip:16773077,peachpuff:16767673,peru:13468991,pink:16761035,plum:14524637,powderblue:11591910,purple:8388736,rebeccapurple:6697881,red:16711680,rosybrown:12357519,royalblue:4286945,saddlebrown:9127187,salmon:16416882,sandybrown:16032864,seagreen:3050327,seashell:16774638,sienna:10506797,silver:12632256,skyblue:8900331,slateblue:6970061,slategray:7372944,slategrey:7372944,snow:16775930,springgreen:65407,steelblue:4620980,tan:13808780,teal:32896,thistle:14204888,tomato:16737095,turquoise:4251856,violet:15631086,wheat:16113331,white:16777215,whitesmoke:16119285,yellow:16776960,yellowgreen:10145074};function he(t){var e;return t=(t+"").trim().toLowerCase(),(e=ee.exec(t))?new ye((e=parseInt(e[1],16))>>8&15|e>>4&240,e>>4&15|240&e,(15&e)<<4|15&e,1):(e=ne.exec(t))?fe(parseInt(e[1],16)):(e=re.exec(t))?new ye(e[1],e[2],e[3],1):(e=ie.exec(t))?new ye(255*e[1]/100,255*e[2]/100,255*e[3]/100,1):(e=oe.exec(t))?le(e[1],e[2],e[3],e[4]):(e=ae.exec(t))?le(255*e[1]/100,255*e[2]/100,255*e[3]/100,e[4]):(e=se.exec(t))?ge(e[1],e[2]/100,e[3]/100,1):(e=ue.exec(t))?ge(e[1],e[2]/100,e[3]/100,e[4]):ce.hasOwnProperty(t)?fe(ce[t]):"transparent"===t?new ye(NaN,NaN,NaN,0):null}function fe(t){return new ye(t>>16&255,t>>8&255,255&t,1)}function le(t,e,n,r){return r<=0&&(t=e=n=NaN),new ye(t,e,n,r)}function de(t){return t instanceof Xt||(t=he(t)),t?new ye((t=t.rgb()).r,t.g,t.b,t.opacity):new ye}function pe(t,e,n,r){return 1===arguments.length?de(t):new ye(t,e,n,null==r?1:r)}function ye(t,e,n,r){this.r=+t,this.g=+e,this.b=+n,this.opacity=+r}function ge(t,e,n,r){return r<=0?t=e=n=NaN:n<=0||n>=1?t=e=NaN:e<=0&&(t=NaN),new me(t,e,n,r)}function ve(t,e,n,r){return 1===arguments.length?function(t){if(t instanceof me)return new me(t.h,t.s,t.l,t.opacity);if(t instanceof Xt||(t=he(t)),!t)return new me;if(t instanceof me)return t;var e=(t=t.rgb()).r/255,n=t.g/255,r=t.b/255,i=Math.min(e,n,r),o=Math.max(e,n,r),a=NaN,s=o-i,u=(o+i)/2;return s?(a=e===o?(n-r)/s+6*(n<r):n===o?(r-e)/s+2:(e-n)/s+4,s/=u<.5?o+i:2-o-i,a*=60):s=u>0&&u<1?0:a,new me(a,s,u,t.opacity)}(t):new me(t,e,n,null==r?1:r)}function me(t,e,n,r){this.h=+t,this.s=+e,this.l=+n,this.opacity=+r}function _e(t,e,n){return 255*(t<60?e+(n-e)*t/60:t<180?n:t<240?e+(n-e)*(240-t)/60:e)}Kt(Xt,he,{displayable:function(){return this.rgb().displayable()},toString:function(){return this.rgb()+""}}),Kt(ye,pe,Ht(Xt,{brighter:function(t){return t=null==t?Zt:Math.pow(Zt,t),new ye(this.r*t,this.g*t,this.b*t,this.opacity)},darker:function(t){return t=null==t?Vt:Math.pow(Vt,t),new ye(this.r*t,this.g*t,this.b*t,this.opacity)},rgb:function(){return this},displayable:function(){return 0<=this.r&&this.r<=255&&0<=this.g&&this.g<=255&&0<=this.b&&this.b<=255&&0<=this.opacity&&this.opacity<=1},toString:function(){var t=this.opacity;return(1===(t=isNaN(t)?1:Math.max(0,Math.min(1,t)))?"rgb(":"rgba(")+Math.max(0,Math.min(255,Math.round(this.r)||0))+", "+Math.max(0,Math.min(255,Math.round(this.g)||0))+", "+Math.max(0,Math.min(255,Math.round(this.b)||0))+(1===t?")":", "+t+")")}})),Kt(me,ve,Ht(Xt,{brighter:function(t){return t=null==t?Zt:Math.pow(Zt,t),new me(this.h,this.s,this.l*t,this.opacity)},darker:function(t){return t=null==t?Vt:Math.pow(Vt,t),new me(this.h,this.s,this.l*t,this.opacity)},rgb:function(){var t=this.h%360+360*(this.h<0),e=isNaN(t)||isNaN(this.s)?0:this.s,n=this.l,r=n+(n<.5?n:1-n)*e,i=2*n-r;return new ye(_e(t>=240?t-240:t+120,i,r),_e(t,i,r),_e(t<120?t+240:t-120,i,r),this.opacity)},displayable:function(){return(0<=this.s&&this.s<=1||isNaN(this.s))&&0<=this.l&&this.l<=1&&0<=this.opacity&&this.opacity<=1}}));var be=Math.PI/180,we=180/Math.PI,xe=.95047,ke=1,Me=1.08883,Ee=4/29,Ae=6/29,Se=3*Ae*Ae,Le=Ae*Ae*Ae;function Ne(t){if(t instanceof $e)return new $e(t.l,t.a,t.b,t.opacity);if(t instanceof Oe){var e=t.h*be;return new $e(t.l,Math.cos(e)*t.c,Math.sin(e)*t.c,t.opacity)}t instanceof ye||(t=de(t));var n=Ue(t.r),r=Ue(t.g),i=Ue(t.b),o=Ce((.4124564*n+.3575761*r+.1804375*i)/xe),a=Ce((.2126729*n+.7151522*r+.072175*i)/ke);return new $e(116*a-16,500*(o-a),200*(a-Ce((.0193339*n+.119192*r+.9503041*i)/Me)),t.opacity)}function $e(t,e,n,r){this.l=+t,this.a=+e,this.b=+n,this.opacity=+r}function Ce(t){return t>Le?Math.pow(t,1/3):t/Se+Ee}function Te(t){return t>Ae?t*t*t:Se*(t-Ee)}function Ge(t){return 255*(t<=.0031308?12.92*t:1.055*Math.pow(t,1/2.4)-.055)}function Ue(t){return(t/=255)<=.04045?t/12.92:Math.pow((t+.055)/1.055,2.4)}function De(t,e,n,r){return 1===arguments.length?function(t){if(t instanceof Oe)return new Oe(t.h,t.c,t.l,t.opacity);t instanceof $e||(t=Ne(t));var e=Math.atan2(t.b,t.a)*we;return new Oe(e<0?e+360:e,Math.sqrt(t.a*t.a+t.b*t.b),t.l,t.opacity)}(t):new Oe(t,e,n,null==r?1:r)}function Oe(t,e,n,r){this.h=+t,this.c=+e,this.l=+n,this.opacity=+r}Kt($e,(function(t,e,n,r){return 1===arguments.length?Ne(t):new $e(t,e,n,null==r?1:r)}),Ht(Xt,{brighter:function(t){return new $e(this.l+18*(null==t?1:t),this.a,this.b,this.opacity)},darker:function(t){return new $e(this.l-18*(null==t?1:t),this.a,this.b,this.opacity)},rgb:function(){var t=(this.l+16)/116,e=isNaN(this.a)?t:t+this.a/500,n=isNaN(this.b)?t:t-this.b/200;return t=ke*Te(t),new ye(Ge(3.2404542*(e=xe*Te(e))-1.5371385*t-.4985314*(n=Me*Te(n))),Ge(-.969266*e+1.8760108*t+.041556*n),Ge(.0556434*e-.2040259*t+1.0572252*n),this.opacity)}})),Kt(Oe,De,Ht(Xt,{brighter:function(t){return new Oe(this.h,this.c,this.l+18*(null==t?1:t),this.opacity)},darker:function(t){return new Oe(this.h,this.c,this.l-18*(null==t?1:t),this.opacity)},rgb:function(){return Ne(this).rgb()}}));var je=-.14861,ze=1.78277,Pe=-.29227,Ie=-.90649,Re=1.97294,Fe=Re*Ie,qe=Re*ze,Ye=ze*Pe-Ie*je;function Be(t,e,n,r){return 1===arguments.length?function(t){if(t instanceof We)return new We(t.h,t.s,t.l,t.opacity);t instanceof ye||(t=de(t));var e=t.r/255,n=t.g/255,r=t.b/255,i=(Ye*r+Fe*e-qe*n)/(Ye+Fe-qe),o=r-i,a=(Re*(n-i)-Pe*o)/Ie,s=Math.sqrt(a*a+o*o)/(Re*i*(1-i)),u=s?Math.atan2(a,o)*we-120:NaN;return new We(u<0?u+360:u,s,i,t.opacity)}(t):new We(t,e,n,null==r?1:r)}function We(t,e,n,r){this.h=+t,this.s=+e,this.l=+n,this.opacity=+r}function Ke(t,e,n,r,i){var o=t*t,a=o*t;return((1-3*t+3*o-a)*e+(4-6*o+3*a)*n+(1+3*t+3*o-3*a)*r+a*i)/6}function He(t){return function(){return t}}function Xe(t,e){return function(n){return t+n*e}}function Ve(t,e){var n=e-t;return n?Xe(t,n>180||n<-180?n-360*Math.round(n/360):n):He(isNaN(t)?e:t)}function Ze(t,e){var n=e-t;return n?Xe(t,n):He(isNaN(t)?e:t)}Kt(We,Be,Ht(Xt,{brighter:function(t){return t=null==t?Zt:Math.pow(Zt,t),new We(this.h,this.s,this.l*t,this.opacity)},darker:function(t){return t=null==t?Vt:Math.pow(Vt,t),new We(this.h,this.s,this.l*t,this.opacity)},rgb:function(){var t=isNaN(this.h)?0:(this.h+120)*be,e=+this.l,n=isNaN(this.s)?0:this.s*e*(1-e),r=Math.cos(t),i=Math.sin(t);return new ye(255*(e+n*(je*r+ze*i)),255*(e+n*(Pe*r+Ie*i)),255*(e+n*(Re*r)),this.opacity)}}));const Je=function t(e){var n=function(t){return 1==(t=+t)?Ze:function(e,n){return n-e?function(t,e,n){return t=Math.pow(t,n),e=Math.pow(e,n)-t,n=1/n,function(r){return Math.pow(t+r*e,n)}}(e,n,t):He(isNaN(e)?n:e)}}(e);function r(t,e){var r=n((t=pe(t)).r,(e=pe(e)).r),i=n(t.g,e.g),o=n(t.b,e.b),a=Ze(t.opacity,e.opacity);return function(e){return t.r=r(e),t.g=i(e),t.b=o(e),t.opacity=a(e),t+""}}return r.gamma=t,r}(1);function Qe(t){return function(e){var n,r,i=e.length,o=new Array(i),a=new Array(i),s=new Array(i);for(n=0;n<i;++n)r=pe(e[n]),o[n]=r.r||0,a[n]=r.g||0,s[n]=r.b||0;return o=t(o),a=t(a),s=t(s),r.opacity=1,function(t){return r.r=o(t),r.g=a(t),r.b=s(t),r+""}}}function tn(t,e){return e-=t=+t,function(n){return t+e*n}}Qe((function(t){var e=t.length-1;return function(n){var r=n<=0?n=0:n>=1?(n=1,e-1):Math.floor(n*e),i=t[r],o=t[r+1],a=r>0?t[r-1]:2*i-o,s=r<e-1?t[r+2]:2*o-i;return Ke((n-r/e)*e,a,i,o,s)}})),Qe((function(t){var e=t.length;return function(n){var r=Math.floor(((n%=1)<0?++n:n)*e),i=t[(r+e-1)%e],o=t[r%e],a=t[(r+1)%e],s=t[(r+2)%e];return Ke((n-r/e)*e,i,o,a,s)}}));var en=/[-+]?(?:\d+\.?\d*|\.?\d+)(?:[eE][-+]?\d+)?/g,nn=new RegExp(en.source,"g");function rn(t,e){var n,r,i,o=en.lastIndex=nn.lastIndex=0,a=-1,s=[],u=[];for(t+="",e+="";(n=en.exec(t))&&(r=nn.exec(e));)(i=r.index)>o&&(i=e.slice(o,i),s[a]?s[a]+=i:s[++a]=i),(n=n[0])===(r=r[0])?s[a]?s[a]+=r:s[++a]=r:(s[++a]=null,u.push({i:a,x:tn(n,r)})),o=nn.lastIndex;return o<e.length&&(i=e.slice(o),s[a]?s[a]+=i:s[++a]=i),s.length<2?u[0]?function(t){return function(e){return t(e)+""}}(u[0].x):function(t){return function(){return t}}(e):(e=u.length,function(t){for(var n,r=0;r<e;++r)s[(n=u[r]).i]=n.x(t);return s.join("")})}var on,an,sn,un,cn=180/Math.PI,hn={translateX:0,translateY:0,rotate:0,skewX:0,scaleX:1,scaleY:1};function fn(t,e,n,r,i,o){var a,s,u;return(a=Math.sqrt(t*t+e*e))&&(t/=a,e/=a),(u=t*n+e*r)&&(n-=t*u,r-=e*u),(s=Math.sqrt(n*n+r*r))&&(n/=s,r/=s,u/=s),t*r<e*n&&(t=-t,e=-e,u=-u,a=-a),{translateX:i,translateY:o,rotate:Math.atan2(e,t)*cn,skewX:Math.atan(u)*cn,scaleX:a,scaleY:s}}function ln(t,e,n,r){function i(t){return t.length?t.pop()+" ":""}return function(o,a){var s=[],u=[];return o=t(o),a=t(a),function(t,r,i,o,a,s){if(t!==i||r!==o){var u=a.push("translate(",null,e,null,n);s.push({i:u-4,x:tn(t,i)},{i:u-2,x:tn(r,o)})}else(i||o)&&a.push("translate("+i+e+o+n)}(o.translateX,o.translateY,a.translateX,a.translateY,s,u),function(t,e,n,o){t!==e?(t-e>180?e+=360:e-t>180&&(t+=360),o.push({i:n.push(i(n)+"rotate(",null,r)-2,x:tn(t,e)})):e&&n.push(i(n)+"rotate("+e+r)}(o.rotate,a.rotate,s,u),function(t,e,n,o){t!==e?o.push({i:n.push(i(n)+"skewX(",null,r)-2,x:tn(t,e)}):e&&n.push(i(n)+"skewX("+e+r)}(o.skewX,a.skewX,s,u),function(t,e,n,r,o,a){if(t!==n||e!==r){var s=o.push(i(o)+"scale(",null,",",null,")");a.push({i:s-4,x:tn(t,n)},{i:s-2,x:tn(e,r)})}else 1===n&&1===r||o.push(i(o)+"scale("+n+","+r+")")}(o.scaleX,o.scaleY,a.scaleX,a.scaleY,s,u),o=a=null,function(t){for(var e,n=-1,r=u.length;++n<r;)s[(e=u[n]).i]=e.x(t);return s.join("")}}}var dn=ln((function(t){return"none"===t?hn:(on||(on=document.createElement("DIV"),an=document.documentElement,sn=document.defaultView),on.style.transform=t,t=sn.getComputedStyle(an.appendChild(on),null).getPropertyValue("transform"),an.removeChild(on),fn(+(t=t.slice(7,-1).split(","))[0],+t[1],+t[2],+t[3],+t[4],+t[5]))}),"px, ","px)","deg)"),pn=ln((function(t){return null==t?hn:(un||(un=document.createElementNS("http://www.w3.org/2000/svg","g")),un.setAttribute("transform",t),(t=un.transform.baseVal.consolidate())?fn((t=t.matrix).a,t.b,t.c,t.d,t.e,t.f):hn)}),", ",")",")"),yn=Math.SQRT2;function gn(t){return((t=Math.exp(t))+1/t)/2}function vn(t,e){var n,r,i=t[0],o=t[1],a=t[2],s=e[0],u=e[1],c=e[2],h=s-i,f=u-o,l=h*h+f*f;if(l<1e-12)r=Math.log(c/a)/yn,n=function(t){return[i+t*h,o+t*f,a*Math.exp(yn*t*r)]};else{var d=Math.sqrt(l),p=(c*c-a*a+4*l)/(2*a*2*d),y=(c*c-a*a-4*l)/(2*c*2*d),g=Math.log(Math.sqrt(p*p+1)-p),v=Math.log(Math.sqrt(y*y+1)-y);r=(v-g)/yn,n=function(t){var e,n=t*r,s=gn(g),u=a/(2*d)*(s*(e=yn*n+g,((e=Math.exp(2*e))-1)/(e+1))-function(t){return((t=Math.exp(t))-1/t)/2}(g));return[i+u*h,o+u*f,a*s/gn(yn*n+g)]}}return n.duration=1e3*r,n}function mn(t){return function(e,n){var r=t((e=ve(e)).h,(n=ve(n)).h),i=Ze(e.s,n.s),o=Ze(e.l,n.l),a=Ze(e.opacity,n.opacity);return function(t){return e.h=r(t),e.s=i(t),e.l=o(t),e.opacity=a(t),e+""}}}function _n(t){return function(e,n){var r=t((e=De(e)).h,(n=De(n)).h),i=Ze(e.c,n.c),o=Ze(e.l,n.l),a=Ze(e.opacity,n.opacity);return function(t){return e.h=r(t),e.c=i(t),e.l=o(t),e.opacity=a(t),e+""}}}function bn(t){return function e(n){function r(e,r){var i=t((e=Be(e)).h,(r=Be(r)).h),o=Ze(e.s,r.s),a=Ze(e.l,r.l),s=Ze(e.opacity,r.opacity);return function(t){return e.h=i(t),e.s=o(t),e.l=a(Math.pow(t,n)),e.opacity=s(t),e+""}}return n=+n,r.gamma=e,r}(1)}mn(Ve),mn(Ze),_n(Ve),_n(Ze),bn(Ve);var wn,xn,kn=bn(Ze),Mn=0,En=0,An=0,Sn=1e3,Ln=0,Nn=0,$n=0,Cn="object"==typeof performance&&performance.now?performance:Date,Tn="object"==typeof window&&window.requestAnimationFrame?window.requestAnimationFrame.bind(window):function(t){setTimeout(t,17)};function Gn(){return Nn||(Tn(Un),Nn=Cn.now()+$n)}function Un(){Nn=0}function Dn(){this._call=this._time=this._next=null}function On(t,e,n){var r=new Dn;return r.restart(t,e,n),r}function jn(){Nn=(Ln=Cn.now())+$n,Mn=En=0;try{!function(){Gn(),++Mn;for(var t,e=wn;e;)(t=Nn-e._time)>=0&&e._call.call(null,t),e=e._next;--Mn}()}finally{Mn=0,function(){for(var t,e,n=wn,r=1/0;n;)n._call?(r>n._time&&(r=n._time),t=n,n=n._next):(e=n._next,n._next=null,n=t?t._next=e:wn=e);xn=t,Pn(r)}(),Nn=0}}function zn(){var t=Cn.now(),e=t-Ln;e>Sn&&($n-=e,Ln=t)}function Pn(t){Mn||(En&&(En=clearTimeout(En)),t-Nn>24?(t<1/0&&(En=setTimeout(jn,t-Cn.now()-$n)),An&&(An=clearInterval(An))):(An||(Ln=Cn.now(),An=setInterval(zn,Sn)),Mn=1,Tn(jn)))}function In(t,e,n){var r=new Dn;return e=null==e?0:+e,r.restart((function(n){r.stop(),t(n+e)}),e,n),r}Dn.prototype=On.prototype={constructor:Dn,restart:function(t,e,n){if("function"!=typeof t)throw new TypeError("callback is not a function");n=(null==n?Gn():+n)+(null==e?0:+e),this._next||xn===this||(xn?xn._next=this:wn=this,xn=this),this._call=t,this._time=n,Pn()},stop:function(){this._call&&(this._call=null,this._time=1/0,Pn())}};var Rn=g("start","end","interrupt"),Fn=[],qn=0,Yn=2;function Bn(t,e,n,r,i,o){var a=t.__transition;if(a){if(n in a)return}else t.__transition={};!function(t,e,n){var r,i=t.__transition;function o(u){var c,h,f,l;if(1!==n.state)return s();for(c in i)if((l=i[c]).name===n.name){if(3===l.state)return In(o);4===l.state?(l.state=6,l.timer.stop(),l.on.call("interrupt",t,t.__data__,l.index,l.group),delete i[c]):+c<e&&(l.state=6,l.timer.stop(),delete i[c])}if(In((function(){3===n.state&&(n.state=4,n.timer.restart(a,n.delay,n.time),a(u))})),n.state=Yn,n.on.call("start",t,t.__data__,n.index,n.group),n.state===Yn){for(n.state=3,r=new Array(f=n.tween.length),c=0,h=-1;c<f;++c)(l=n.tween[c].value.call(t,t.__data__,n.index,n.group))&&(r[++h]=l);r.length=h+1}}function a(e){for(var i=e<n.duration?n.ease.call(null,e/n.duration):(n.timer.restart(s),n.state=5,1),o=-1,a=r.length;++o<a;)r[o].call(null,i);5===n.state&&(n.on.call("end",t,t.__data__,n.index,n.group),s())}function s(){for(var r in n.state=6,n.timer.stop(),delete i[e],i)return;delete t.__transition}i[e]=n,n.timer=On((function(t){n.state=1,n.timer.restart(o,n.delay,n.time),n.delay<=t&&o(t-n.delay)}),0,n.time)}(t,n,{name:e,index:r,group:i,on:Rn,tween:Fn,time:o.time,delay:o.delay,duration:o.duration,ease:o.ease,timer:null,state:qn})}function Wn(t,e){var n=Hn(t,e);if(n.state>qn)throw new Error("too late; already scheduled");return n}function Kn(t,e){var n=Hn(t,e);if(n.state>Yn)throw new Error("too late; already started");return n}function Hn(t,e){var n=t.__transition;if(!n||!(n=n[e]))throw new Error("transition not found");return n}function Xn(t,e){var n,r,i,o=t.__transition,a=!0;if(o){for(i in e=null==e?null:e+"",o)(n=o[i]).name===e?(r=n.state>Yn&&n.state<5,n.state=6,n.timer.stop(),r&&n.on.call("interrupt",t,t.__data__,n.index,n.group),delete o[i]):a=!1;a&&delete t.__transition}}function Vn(t,e){var n,r;return function(){var i=Kn(this,t),o=i.tween;if(o!==n)for(var a=0,s=(r=n=o).length;a<s;++a)if(r[a].name===e){(r=r.slice()).splice(a,1);break}i.tween=r}}function Zn(t,e,n){var r,i;if("function"!=typeof n)throw new Error;return function(){var o=Kn(this,t),a=o.tween;if(a!==r){i=(r=a).slice();for(var s={name:e,value:n},u=0,c=i.length;u<c;++u)if(i[u].name===e){i[u]=s;break}u===c&&i.push(s)}o.tween=i}}function Jn(t,e,n){var r=t._id;return t.each((function(){var t=Kn(this,r);(t.value||(t.value={}))[e]=n.apply(this,arguments)})),function(t){return Hn(t,r).value[e]}}function Qn(t,e){var n;return("number"==typeof e?tn:e instanceof he?Je:(n=he(e))?(e=n,Je):rn)(t,e)}function tr(t){return function(){this.removeAttribute(t)}}function er(t){return function(){this.removeAttributeNS(t.space,t.local)}}function nr(t,e,n){var r,i;return function(){var o=this.getAttribute(t);return o===n?null:o===r?i:i=e(r=o,n)}}function rr(t,e,n){var r,i;return function(){var o=this.getAttributeNS(t.space,t.local);return o===n?null:o===r?i:i=e(r=o,n)}}function ir(t,e,n){var r,i,o;return function(){var a,s=n(this);if(null!=s)return(a=this.getAttribute(t))===s?null:a===r&&s===i?o:o=e(r=a,i=s);this.removeAttribute(t)}}function or(t,e,n){var r,i,o;return function(){var a,s=n(this);if(null!=s)return(a=this.getAttributeNS(t.space,t.local))===s?null:a===r&&s===i?o:o=e(r=a,i=s);this.removeAttributeNS(t.space,t.local)}}function ar(t,e){function n(){var n=this,r=e.apply(n,arguments);return r&&function(e){n.setAttributeNS(t.space,t.local,r(e))}}return n._value=e,n}function sr(t,e){function n(){var n=this,r=e.apply(n,arguments);return r&&function(e){n.setAttribute(t,r(e))}}return n._value=e,n}function ur(t,e){return function(){Wn(this,t).delay=+e.apply(this,arguments)}}function cr(t,e){return e=+e,function(){Wn(this,t).delay=e}}function hr(t,e){return function(){Kn(this,t).duration=+e.apply(this,arguments)}}function fr(t,e){return e=+e,function(){Kn(this,t).duration=e}}var lr=Nt.prototype.constructor,dr=0;function pr(t,e,n,r){this._groups=t,this._parents=e,this._name=n,this._id=r}function yr(){return++dr}var gr=Nt.prototype;pr.prototype=function(t){return Nt().transition(t)}.prototype={constructor:pr,select:function(t){var e=this._name,n=this._id;"function"!=typeof t&&(t=M(t));for(var r=this._groups,i=r.length,o=new Array(i),a=0;a<i;++a)for(var s,u,c=r[a],h=c.length,f=o[a]=new Array(h),l=0;l<h;++l)(s=c[l])&&(u=t.call(s,s.__data__,l,c))&&("__data__"in s&&(u.__data__=s.__data__),f[l]=u,Bn(f[l],e,n,l,f,Hn(s,n)));return new pr(o,this._parents,e,n)},selectAll:function(t){var e=this._name,n=this._id;"function"!=typeof t&&(t=A(t));for(var r=this._groups,i=r.length,o=[],a=[],s=0;s<i;++s)for(var u,c=r[s],h=c.length,f=0;f<h;++f)if(u=c[f]){for(var l,d=t.call(u,u.__data__,f,c),p=Hn(u,n),y=0,g=d.length;y<g;++y)(l=d[y])&&Bn(l,e,n,y,d,p);o.push(d),a.push(u)}return new pr(o,a,e,n)},filter:function(t){"function"!=typeof t&&(t=$(t));for(var e=this._groups,n=e.length,r=new Array(n),i=0;i<n;++i)for(var o,a=e[i],s=a.length,u=r[i]=[],c=0;c<s;++c)(o=a[c])&&t.call(o,o.__data__,c,a)&&u.push(o);return new pr(r,this._parents,this._name,this._id)},merge:function(t){if(t._id!==this._id)throw new Error;for(var e=this._groups,n=t._groups,r=e.length,i=n.length,o=Math.min(r,i),a=new Array(r),s=0;s<o;++s)for(var u,c=e[s],h=n[s],f=c.length,l=a[s]=new Array(f),d=0;d<f;++d)(u=c[d]||h[d])&&(l[d]=u);for(;s<r;++s)a[s]=e[s];return new pr(a,this._parents,this._name,this._id)},selection:function(){return new lr(this._groups,this._parents)},transition:function(){for(var t=this._name,e=this._id,n=yr(),r=this._groups,i=r.length,o=0;o<i;++o)for(var a,s=r[o],u=s.length,c=0;c<u;++c)if(a=s[c]){var h=Hn(a,e);Bn(a,t,n,c,s,{time:h.time+h.delay+h.duration,delay:0,duration:h.duration,ease:h.ease})}return new pr(r,this._parents,t,n)},call:gr.call,nodes:gr.nodes,node:gr.node,size:gr.size,empty:gr.empty,each:gr.each,on:function(t,e){var n=this._id;return arguments.length<2?Hn(this.node(),n).on.on(t):this.each(function(t,e,n){var r,i,o=function(t){return(t+"").trim().split(/^|\s+/).every((function(t){var e=t.indexOf(".");return e>=0&&(t=t.slice(0,e)),!t||"start"===t}))}(e)?Wn:Kn;return function(){var a=o(this,t),s=a.on;s!==r&&(i=(r=s).copy()).on(e,n),a.on=i}}(n,t,e))},attr:function(t,e){var n=_(t),r="transform"===n?pn:Qn;return this.attrTween(t,"function"==typeof e?(n.local?or:ir)(n,r,Jn(this,"attr."+t,e)):null==e?(n.local?er:tr)(n):(n.local?rr:nr)(n,r,e+""))},attrTween:function(t,e){var n="attr."+t;if(arguments.length<2)return(n=this.tween(n))&&n._value;if(null==e)return this.tween(n,null);if("function"!=typeof e)throw new Error;var r=_(t);return this.tween(n,(r.local?ar:sr)(r,e))},style:function(t,e,n){var r="transform"==(t+="")?dn:Qn;return null==e?this.styleTween(t,function(t,e){var n,r,i;return function(){var o=W(this,t),a=(this.style.removeProperty(t),W(this,t));return o===a?null:o===n&&a===r?i:i=e(n=o,r=a)}}(t,r)).on("end.style."+t,function(t){return function(){this.style.removeProperty(t)}}(t)):this.styleTween(t,"function"==typeof e?function(t,e,n){var r,i,o;return function(){var a=W(this,t),s=n(this);return null==s&&(this.style.removeProperty(t),s=W(this,t)),a===s?null:a===r&&s===i?o:o=e(r=a,i=s)}}(t,r,Jn(this,"style."+t,e)):function(t,e,n){var r,i;return function(){var o=W(this,t);return o===n?null:o===r?i:i=e(r=o,n)}}(t,r,e+""),n)},styleTween:function(t,e,n){var r="style."+(t+="");if(arguments.length<2)return(r=this.tween(r))&&r._value;if(null==e)return this.tween(r,null);if("function"!=typeof e)throw new Error;return this.tween(r,function(t,e,n){function r(){var r=this,i=e.apply(r,arguments);return i&&function(e){r.style.setProperty(t,i(e),n)}}return r._value=e,r}(t,e,null==n?"":n))},text:function(t){return this.tween("text","function"==typeof t?function(t){return function(){var e=t(this);this.textContent=null==e?"":e}}(Jn(this,"text",t)):function(t){return function(){this.textContent=t}}(null==t?"":t+""))},remove:function(){return this.on("end.remove",function(t){return function(){var e=this.parentNode;for(var n in this.__transition)if(+n!==t)return;e&&e.removeChild(this)}}(this._id))},tween:function(t,e){var n=this._id;if(t+="",arguments.length<2){for(var r,i=Hn(this.node(),n).tween,o=0,a=i.length;o<a;++o)if((r=i[o]).name===t)return r.value;return null}return this.each((null==e?Vn:Zn)(n,t,e))},delay:function(t){var e=this._id;return arguments.length?this.each(("function"==typeof t?ur:cr)(e,t)):Hn(this.node(),e).delay},duration:function(t){var e=this._id;return arguments.length?this.each(("function"==typeof t?hr:fr)(e,t)):Hn(this.node(),e).duration},ease:function(t){var e=this._id;return arguments.length?this.each(function(t,e){if("function"!=typeof e)throw new Error;return function(){Kn(this,t).ease=e}}(e,t)):Hn(this.node(),e).ease}},function t(e){function n(t){return Math.pow(t,e)}return e=+e,n.exponent=t,n}(3),function t(e){function n(t){return 1-Math.pow(1-t,e)}return e=+e,n.exponent=t,n}(3),function t(e){function n(t){return((t*=2)<=1?Math.pow(t,e):2-Math.pow(2-t,e))/2}return e=+e,n.exponent=t,n}(3),Math.PI;var vr=1.70158,mr=(function t(e){function n(t){return t*t*((e+1)*t-e)}return e=+e,n.overshoot=t,n}(vr),function t(e){function n(t){return--t*t*((e+1)*t+e)+1}return e=+e,n.overshoot=t,n}(vr),function t(e){function n(t){return((t*=2)<1?t*t*((e+1)*t-e):(t-=2)*t*((e+1)*t+e)+2)/2}return e=+e,n.overshoot=t,n}(vr),2*Math.PI),_r=(function t(e,n){var r=Math.asin(1/(e=Math.max(1,e)))*(n/=mr);function i(t){return e*Math.pow(2,10*--t)*Math.sin((r-t)/n)}return i.amplitude=function(e){return t(e,n*mr)},i.period=function(n){return t(e,n)},i}(1,.3),function t(e,n){var r=Math.asin(1/(e=Math.max(1,e)))*(n/=mr);function i(t){return 1-e*Math.pow(2,-10*(t=+t))*Math.sin((t+r)/n)}return i.amplitude=function(e){return t(e,n*mr)},i.period=function(n){return t(e,n)},i}(1,.3),function t(e,n){var r=Math.asin(1/(e=Math.max(1,e)))*(n/=mr);function i(t){return((t=2*t-1)<0?e*Math.pow(2,10*t)*Math.sin((r-t)/n):2-e*Math.pow(2,-10*t)*Math.sin((r+t)/n))/2}return i.amplitude=function(e){return t(e,n*mr)},i.period=function(n){return t(e,n)},i}(1,.3),{time:null,delay:0,duration:250,ease:function(t){return((t*=2)<=1?t*t*t:(t-=2)*t*t+2)/2}});function br(t,e){for(var n;!(n=t.__transition)||!(n=n[e]);)if(!(t=t.parentNode))return _r.time=Gn(),_r;return n}function wr(t){return{type:t}}function xr(t,e){return t<e?-1:t>e?1:t>=e?0:NaN}Nt.prototype.interrupt=function(t){return this.each((function(){Xn(this,t)}))},Nt.prototype.transition=function(t){var e,n;t instanceof pr?(e=t._id,t=t._name):(e=yr(),(n=_r).time=Gn(),t=null==t?null:t+"");for(var r=this._groups,i=r.length,o=0;o<i;++o)for(var a,s=r[o],u=s.length,c=0;c<u;++c)(a=s[c])&&Bn(a,t,e,c,s,n||br(a,e));return new pr(r,this._parents,t,e)},["e","w"].map(wr),["n","s"].map(wr),["n","e","s","w","nw","ne","se","sw"].map(wr);var kr=function(t){return 1===t.length&&(t=function(t){return function(e,n){return xr(t(e),n)}}(t)),{left:function(e,n,r,i){for(null==r&&(r=0),null==i&&(i=e.length);r<i;){var o=r+i>>>1;t(e[o],n)<0?r=o+1:i=o}return r},right:function(e,n,r,i){for(null==r&&(r=0),null==i&&(i=e.length);r<i;){var o=r+i>>>1;t(e[o],n)>0?i=o:r=o+1}return r}}}(xr);kr.right,kr.left;var Mr=Array.prototype;Mr.slice,Mr.map,Math.sqrt(50),Math.sqrt(10),Math.sqrt(2),Math.cos,Math.sin,Math.PI,Math.max,Array.prototype.slice;var Er=Math.PI,Ar=2*Er,Sr=1e-6,Lr=Ar-Sr;function Nr(){this._x0=this._y0=this._x1=this._y1=null,this._=""}Nr.prototype=function(){return new Nr}.prototype={constructor:Nr,moveTo:function(t,e){this._+="M"+(this._x0=this._x1=+t)+","+(this._y0=this._y1=+e)},closePath:function(){null!==this._x1&&(this._x1=this._x0,this._y1=this._y0,this._+="Z")},lineTo:function(t,e){this._+="L"+(this._x1=+t)+","+(this._y1=+e)},quadraticCurveTo:function(t,e,n,r){this._+="Q"+ +t+","+ +e+","+(this._x1=+n)+","+(this._y1=+r)},bezierCurveTo:function(t,e,n,r,i,o){this._+="C"+ +t+","+ +e+","+ +n+","+ +r+","+(this._x1=+i)+","+(this._y1=+o)},arcTo:function(t,e,n,r,i){t=+t,e=+e,n=+n,r=+r,i=+i;var o=this._x1,a=this._y1,s=n-t,u=r-e,c=o-t,h=a-e,f=c*c+h*h;if(i<0)throw new Error("negative radius: "+i);if(null===this._x1)this._+="M"+(this._x1=t)+","+(this._y1=e);else if(f>Sr)if(Math.abs(h*s-u*c)>Sr&&i){var l=n-o,d=r-a,p=s*s+u*u,y=l*l+d*d,g=Math.sqrt(p),v=Math.sqrt(f),m=i*Math.tan((Er-Math.acos((p+f-y)/(2*g*v)))/2),_=m/v,b=m/g;Math.abs(_-1)>Sr&&(this._+="L"+(t+_*c)+","+(e+_*h)),this._+="A"+i+","+i+",0,0,"+ +(h*l>c*d)+","+(this._x1=t+b*s)+","+(this._y1=e+b*u)}else this._+="L"+(this._x1=t)+","+(this._y1=e)},arc:function(t,e,n,r,i,o){t=+t,e=+e;var a=(n=+n)*Math.cos(r),s=n*Math.sin(r),u=t+a,c=e+s,h=1^o,f=o?r-i:i-r;if(n<0)throw new Error("negative radius: "+n);null===this._x1?this._+="M"+u+","+c:(Math.abs(this._x1-u)>Sr||Math.abs(this._y1-c)>Sr)&&(this._+="L"+u+","+c),n&&(f<0&&(f=f%Ar+Ar),f>Lr?this._+="A"+n+","+n+",0,1,"+h+","+(t-a)+","+(e-s)+"A"+n+","+n+",0,1,"+h+","+(this._x1=u)+","+(this._y1=c):f>Sr&&(this._+="A"+n+","+n+",0,"+ +(f>=Er)+","+h+","+(this._x1=t+n*Math.cos(i))+","+(this._y1=e+n*Math.sin(i))))},rect:function(t,e,n,r){this._+="M"+(this._x0=this._x1=+t)+","+(this._y0=this._y1=+e)+"h"+ +n+"v"+ +r+"h"+-n+"Z"},toString:function(){return this._}};var $r="$";function Cr(){}function Tr(t,e){var n=new Cr;if(t instanceof Cr)t.each((function(t,e){n.set(e,t)}));else if(Array.isArray(t)){var r,i=-1,o=t.length;if(null==e)for(;++i<o;)n.set(i,t[i]);else for(;++i<o;)n.set(e(r=t[i],i,t),r)}else if(t)for(var a in t)n.set(a,t[a]);return n}Cr.prototype=Tr.prototype={constructor:Cr,has:function(t){return $r+t in this},get:function(t){return this[$r+t]},set:function(t,e){return this[$r+t]=e,this},remove:function(t){var e=$r+t;return e in this&&delete this[e]},clear:function(){for(var t in this)t[0]===$r&&delete this[t]},keys:function(){var t=[];for(var e in this)e[0]===$r&&t.push(e.slice(1));return t},values:function(){var t=[];for(var e in this)e[0]===$r&&t.push(this[e]);return t},entries:function(){var t=[];for(var e in this)e[0]===$r&&t.push({key:e.slice(1),value:this[e]});return t},size:function(){var t=0;for(var e in this)e[0]===$r&&++t;return t},empty:function(){for(var t in this)if(t[0]===$r)return!1;return!0},each:function(t){for(var e in this)e[0]===$r&&t(this[e],e.slice(1),this)}};const Gr=Tr;function Ur(){}var Dr=Gr.prototype;Ur.prototype=function(t,e){var n=new Ur;if(t instanceof Ur)t.each((function(t){n.add(t)}));else if(t){var r=-1,i=t.length;if(null==e)for(;++r<i;)n.add(t[r]);else for(;++r<i;)n.add(e(t[r],r,t))}return n}.prototype={constructor:Ur,has:Dr.has,add:function(t){return this[$r+(t+="")]=t,this},remove:Dr.remove,clear:Dr.clear,values:Dr.keys,size:Dr.size,empty:Dr.empty,each:Dr.each};var Or={},jr={};function zr(t){return new Function("d","return {"+t.map((function(t,e){return JSON.stringify(t)+": d["+e+"]"})).join(",")+"}")}function Pr(t){var e=new RegExp('["'+t+"\n\r]"),n=t.charCodeAt(0);function r(t,e){var r,i=[],o=t.length,a=0,s=0,u=o<=0,c=!1;function h(){if(u)return jr;if(c)return c=!1,Or;var e,r,i=a;if(34===t.charCodeAt(i)){for(;a++<o&&34!==t.charCodeAt(a)||34===t.charCodeAt(++a););return(e=a)>=o?u=!0:10===(r=t.charCodeAt(a++))?c=!0:13===r&&(c=!0,10===t.charCodeAt(a)&&++a),t.slice(i+1,e-1).replace(/""/g,'"')}for(;a<o;){if(10===(r=t.charCodeAt(e=a++)))c=!0;else if(13===r)c=!0,10===t.charCodeAt(a)&&++a;else if(r!==n)continue;return t.slice(i,e)}return u=!0,t.slice(i,o)}for(10===t.charCodeAt(o-1)&&--o,13===t.charCodeAt(o-1)&&--o;(r=h())!==jr;){for(var f=[];r!==Or&&r!==jr;)f.push(r),r=h();e&&null==(f=e(f,s++))||i.push(f)}return i}function i(e){return e.map(o).join(t)}function o(t){return null==t?"":e.test(t+="")?'"'+t.replace(/"/g,'""')+'"':t}return{parse:function(t,e){var n,i,o=r(t,(function(t,r){if(n)return n(t,r-1);i=t,n=e?function(t,e){var n=zr(t);return function(r,i){return e(n(r),i,t)}}(t,e):zr(t)}));return o.columns=i||[],o},parseRows:r,format:function(e,n){return null==n&&(n=function(t){var e=Object.create(null),n=[];return t.forEach((function(t){for(var r in t)r in e||n.push(e[r]=r)})),n}(e)),[n.map(o).join(t)].concat(e.map((function(e){return n.map((function(t){return o(e[t])})).join(t)}))).join("\n")},formatRows:function(t){return t.map(i).join("\n")}}}var Ir=Pr(","),Rr=Ir.parse,Fr=(Ir.parseRows,Ir.format,Ir.formatRows,Pr("\t")),qr=Fr.parse;function Yr(t,e,n,r){if(isNaN(e)||isNaN(n))return t;var i,o,a,s,u,c,h,f,l,d=t._root,p={data:r},y=t._x0,g=t._y0,v=t._x1,m=t._y1;if(!d)return t._root=p,t;for(;d.length;)if((c=e>=(o=(y+v)/2))?y=o:v=o,(h=n>=(a=(g+m)/2))?g=a:m=a,i=d,!(d=d[f=h<<1|c]))return i[f]=p,t;if(s=+t._x.call(null,d.data),u=+t._y.call(null,d.data),e===s&&n===u)return p.next=d,i?i[f]=p:t._root=p,t;do{i=i?i[f]=new Array(4):t._root=new Array(4),(c=e>=(o=(y+v)/2))?y=o:v=o,(h=n>=(a=(g+m)/2))?g=a:m=a}while((f=h<<1|c)==(l=(u>=a)<<1|s>=o));return i[l]=d,i[f]=p,t}function Br(t,e,n,r,i){this.node=t,this.x0=e,this.y0=n,this.x1=r,this.y1=i}function Wr(t){return t[0]}function Kr(t){return t[1]}function Hr(t,e,n){var r=new Xr(null==e?Wr:e,null==n?Kr:n,NaN,NaN,NaN,NaN);return null==t?r:r.addAll(t)}function Xr(t,e,n,r,i,o){this._x=t,this._y=e,this._x0=n,this._y0=r,this._x1=i,this._y1=o,this._root=void 0}function Vr(t){for(var e={data:t.data},n=e;t=t.next;)n=n.next={data:t.data};return e}Fr.parseRows,Fr.format,Fr.formatRows;var Zr=Hr.prototype=Xr.prototype;function Jr(t){return function(){return t}}function Qr(){return 1e-6*(Math.random()-.5)}function ti(t){return t.index}function ei(t,e){var n=t.get(e);if(!n)throw new Error("missing: "+e);return n}function ni(t){return t.x}function ri(t){return t.y}Zr.copy=function(){var t,e,n=new Xr(this._x,this._y,this._x0,this._y0,this._x1,this._y1),r=this._root;if(!r)return n;if(!r.length)return n._root=Vr(r),n;for(t=[{source:r,target:n._root=new Array(4)}];r=t.pop();)for(var i=0;i<4;++i)(e=r.source[i])&&(e.length?t.push({source:e,target:r.target[i]=new Array(4)}):r.target[i]=Vr(e));return n},Zr.add=function(t){var e=+this._x.call(null,t),n=+this._y.call(null,t);return Yr(this.cover(e,n),e,n,t)},Zr.addAll=function(t){var e,n,r,i,o=t.length,a=new Array(o),s=new Array(o),u=1/0,c=1/0,h=-1/0,f=-1/0;for(n=0;n<o;++n)isNaN(r=+this._x.call(null,e=t[n]))||isNaN(i=+this._y.call(null,e))||(a[n]=r,s[n]=i,r<u&&(u=r),r>h&&(h=r),i<c&&(c=i),i>f&&(f=i));for(h<u&&(u=this._x0,h=this._x1),f<c&&(c=this._y0,f=this._y1),this.cover(u,c).cover(h,f),n=0;n<o;++n)Yr(this,a[n],s[n],t[n]);return this},Zr.cover=function(t,e){if(isNaN(t=+t)||isNaN(e=+e))return this;var n=this._x0,r=this._y0,i=this._x1,o=this._y1;if(isNaN(n))i=(n=Math.floor(t))+1,o=(r=Math.floor(e))+1;else{if(!(n>t||t>i||r>e||e>o))return this;var a,s,u=i-n,c=this._root;switch(s=(e<(r+o)/2)<<1|t<(n+i)/2){case 0:do{(a=new Array(4))[s]=c,c=a}while(o=r+(u*=2),t>(i=n+u)||e>o);break;case 1:do{(a=new Array(4))[s]=c,c=a}while(o=r+(u*=2),(n=i-u)>t||e>o);break;case 2:do{(a=new Array(4))[s]=c,c=a}while(r=o-(u*=2),t>(i=n+u)||r>e);break;case 3:do{(a=new Array(4))[s]=c,c=a}while(r=o-(u*=2),(n=i-u)>t||r>e)}this._root&&this._root.length&&(this._root=c)}return this._x0=n,this._y0=r,this._x1=i,this._y1=o,this},Zr.data=function(){var t=[];return this.visit((function(e){if(!e.length)do{t.push(e.data)}while(e=e.next)})),t},Zr.extent=function(t){return arguments.length?this.cover(+t[0][0],+t[0][1]).cover(+t[1][0],+t[1][1]):isNaN(this._x0)?void 0:[[this._x0,this._y0],[this._x1,this._y1]]},Zr.find=function(t,e,n){var r,i,o,a,s,u,c,h=this._x0,f=this._y0,l=this._x1,d=this._y1,p=[],y=this._root;for(y&&p.push(new Br(y,h,f,l,d)),null==n?n=1/0:(h=t-n,f=e-n,l=t+n,d=e+n,n*=n);u=p.pop();)if(!(!(y=u.node)||(i=u.x0)>l||(o=u.y0)>d||(a=u.x1)<h||(s=u.y1)<f))if(y.length){var g=(i+a)/2,v=(o+s)/2;p.push(new Br(y[3],g,v,a,s),new Br(y[2],i,v,g,s),new Br(y[1],g,o,a,v),new Br(y[0],i,o,g,v)),(c=(e>=v)<<1|t>=g)&&(u=p[p.length-1],p[p.length-1]=p[p.length-1-c],p[p.length-1-c]=u)}else{var m=t-+this._x.call(null,y.data),_=e-+this._y.call(null,y.data),b=m*m+_*_;if(b<n){var w=Math.sqrt(n=b);h=t-w,f=e-w,l=t+w,d=e+w,r=y.data}}return r},Zr.remove=function(t){if(isNaN(o=+this._x.call(null,t))||isNaN(a=+this._y.call(null,t)))return this;var e,n,r,i,o,a,s,u,c,h,f,l,d=this._root,p=this._x0,y=this._y0,g=this._x1,v=this._y1;if(!d)return this;if(d.length)for(;;){if((c=o>=(s=(p+g)/2))?p=s:g=s,(h=a>=(u=(y+v)/2))?y=u:v=u,e=d,!(d=d[f=h<<1|c]))return this;if(!d.length)break;(e[f+1&3]||e[f+2&3]||e[f+3&3])&&(n=e,l=f)}for(;d.data!==t;)if(r=d,!(d=d.next))return this;return(i=d.next)&&delete d.next,r?(i?r.next=i:delete r.next,this):e?(i?e[f]=i:delete e[f],(d=e[0]||e[1]||e[2]||e[3])&&d===(e[3]||e[2]||e[1]||e[0])&&!d.length&&(n?n[l]=d:this._root=d),this):(this._root=i,this)},Zr.removeAll=function(t){for(var e=0,n=t.length;e<n;++e)this.remove(t[e]);return this},Zr.root=function(){return this._root},Zr.size=function(){var t=0;return this.visit((function(e){if(!e.length)do{++t}while(e=e.next)})),t},Zr.visit=function(t){var e,n,r,i,o,a,s=[],u=this._root;for(u&&s.push(new Br(u,this._x0,this._y0,this._x1,this._y1));e=s.pop();)if(!t(u=e.node,r=e.x0,i=e.y0,o=e.x1,a=e.y1)&&u.length){var c=(r+o)/2,h=(i+a)/2;(n=u[3])&&s.push(new Br(n,c,h,o,a)),(n=u[2])&&s.push(new Br(n,r,h,c,a)),(n=u[1])&&s.push(new Br(n,c,i,o,h)),(n=u[0])&&s.push(new Br(n,r,i,c,h))}return this},Zr.visitAfter=function(t){var e,n=[],r=[];for(this._root&&n.push(new Br(this._root,this._x0,this._y0,this._x1,this._y1));e=n.pop();){var i=e.node;if(i.length){var o,a=e.x0,s=e.y0,u=e.x1,c=e.y1,h=(a+u)/2,f=(s+c)/2;(o=i[0])&&n.push(new Br(o,a,s,h,f)),(o=i[1])&&n.push(new Br(o,h,s,u,f)),(o=i[2])&&n.push(new Br(o,a,f,h,c)),(o=i[3])&&n.push(new Br(o,h,f,u,c))}r.push(e)}for(;e=r.pop();)t(e.node,e.x0,e.y0,e.x1,e.y1);return this},Zr.x=function(t){return arguments.length?(this._x=t,this):this._x},Zr.y=function(t){return arguments.length?(this._y=t,this):this._y};var ii,oi=Math.PI*(3-Math.sqrt(5));function ai(t,e){if((n=(t=e?t.toExponential(e-1):t.toExponential()).indexOf("e"))<0)return null;var n,r=t.slice(0,n);return[r.length>1?r[0]+r.slice(2):r,+t.slice(n+1)]}function si(t,e){var n=ai(t,e);if(!n)return t+"";var r=n[0],i=n[1];return i<0?"0."+new Array(-i).join("0")+r:r.length>i+1?r.slice(0,i+1)+"."+r.slice(i+1):r+new Array(i-r.length+2).join("0")}const ui={"":function(t,e){t:for(var n,r=(t=t.toPrecision(e)).length,i=1,o=-1;i<r;++i)switch(t[i]){case".":o=n=i;break;case"0":0===o&&(o=i),n=i;break;case"e":break t;default:o>0&&(o=0)}return o>0?t.slice(0,o)+t.slice(n+1):t},"%":function(t,e){return(100*t).toFixed(e)},b:function(t){return Math.round(t).toString(2)},c:function(t){return t+""},d:function(t){return Math.round(t).toString(10)},e:function(t,e){return t.toExponential(e)},f:function(t,e){return t.toFixed(e)},g:function(t,e){return t.toPrecision(e)},o:function(t){return Math.round(t).toString(8)},p:function(t,e){return si(100*t,e)},r:si,s:function(t,e){var n=ai(t,e);if(!n)return t+"";var r=n[0],i=n[1],o=i-(ii=3*Math.max(-8,Math.min(8,Math.floor(i/3))))+1,a=r.length;return o===a?r:o>a?r+new Array(o-a+1).join("0"):o>0?r.slice(0,o)+"."+r.slice(o):"0."+new Array(1-o).join("0")+ai(t,Math.max(0,e+o-1))[0]},X:function(t){return Math.round(t).toString(16).toUpperCase()},x:function(t){return Math.round(t).toString(16)}};var ci=/^(?:(.)?([<>=^]))?([+\-\( ])?([$#])?(0)?(\d+)?(,)?(\.\d+)?([a-z%])?$/i;function hi(t){return new fi(t)}function fi(t){if(!(e=ci.exec(t)))throw new Error("invalid format: "+t);var e,n=e[1]||" ",r=e[2]||">",i=e[3]||"-",o=e[4]||"",a=!!e[5],s=e[6]&&+e[6],u=!!e[7],c=e[8]&&+e[8].slice(1),h=e[9]||"";"n"===h?(u=!0,h="g"):ui[h]||(h=""),(a||"0"===n&&"="===r)&&(a=!0,n="0",r="="),this.fill=n,this.align=r,this.sign=i,this.symbol=o,this.zero=a,this.width=s,this.comma=u,this.precision=c,this.type=h}function li(t){return t}hi.prototype=fi.prototype,fi.prototype.toString=function(){return this.fill+this.align+this.sign+this.symbol+(this.zero?"0":"")+(null==this.width?"":Math.max(1,0|this.width))+(this.comma?",":"")+(null==this.precision?"":"."+Math.max(0,0|this.precision))+this.type};var di,pi=["y","z","a","f","p","n","µ","m","","k","M","G","T","P","E","Z","Y"];function yi(){return new gi}function gi(){this.reset()}di=function(t){var e,n,r=t.grouping&&t.thousands?(e=t.grouping,n=t.thousands,function(t,r){for(var i=t.length,o=[],a=0,s=e[0],u=0;i>0&&s>0&&(u+s+1>r&&(s=Math.max(1,r-u)),o.push(t.substring(i-=s,i+s)),!((u+=s+1)>r));)s=e[a=(a+1)%e.length];return o.reverse().join(n)}):li,i=t.currency,o=t.decimal,a=t.numerals?function(t){return function(e){return e.replace(/[0-9]/g,(function(e){return t[+e]}))}}(t.numerals):li,s=t.percent||"%";function u(t){var e=(t=hi(t)).fill,n=t.align,u=t.sign,c=t.symbol,h=t.zero,f=t.width,l=t.comma,d=t.precision,p=t.type,y="$"===c?i[0]:"#"===c&&/[boxX]/.test(p)?"0"+p.toLowerCase():"",g="$"===c?i[1]:/[%p]/.test(p)?s:"",v=ui[p],m=!p||/[defgprs%]/.test(p);function _(t){var i,s,c,_=y,b=g;if("c"===p)b=v(t)+b,t="";else{var w=(t=+t)<0;if(t=v(Math.abs(t),d),w&&0==+t&&(w=!1),_=(w?"("===u?u:"-":"-"===u||"("===u?"":u)+_,b=("s"===p?pi[8+ii/3]:"")+b+(w&&"("===u?")":""),m)for(i=-1,s=t.length;++i<s;)if(48>(c=t.charCodeAt(i))||c>57){b=(46===c?o+t.slice(i+1):t.slice(i))+b,t=t.slice(0,i);break}}l&&!h&&(t=r(t,1/0));var x=_.length+t.length+b.length,k=x<f?new Array(f-x+1).join(e):"";switch(l&&h&&(t=r(k+t,k.length?f-b.length:1/0),k=""),n){case"<":t=_+t+b+k;break;case"=":t=_+k+t+b;break;case"^":t=k.slice(0,x=k.length>>1)+_+t+b+k.slice(x);break;default:t=k+_+t+b}return a(t)}return d=null==d?p?6:12:/[gprs]/.test(p)?Math.max(1,Math.min(21,d)):Math.max(0,Math.min(20,d)),_.toString=function(){return t+""},_}return{format:u,formatPrefix:function(t,e){var n,r=u(((t=hi(t)).type="f",t)),i=3*Math.max(-8,Math.min(8,Math.floor((n=e,((n=ai(Math.abs(n)))?n[1]:NaN)/3)))),o=Math.pow(10,-i),a=pi[8+i/3];return function(t){return r(o*t)+a}}}}({decimal:".",thousands:",",grouping:[3],currency:["$",""]}),di.format,di.formatPrefix,gi.prototype={constructor:gi,reset:function(){this.s=this.t=0},add:function(t){mi(vi,t,this.t),mi(this,vi.s,this.s),this.s?this.t+=vi.t:this.s=vi.t},valueOf:function(){return this.s}};var vi=new gi;function mi(t,e,n){var r=t.s=e+n,i=r-e,o=r-i;t.t=e-o+(n-i)}var _i=Math.PI,bi=_i/2,wi=2*_i,xi=_i/180,ki=(Math.abs,Math.atan),Mi=Math.atan2,Ei=Math.cos,Ai=(Math.ceil,Math.exp,Math.floor,Math.log,Math.pow,Math.sin),Si=(Math.sign,Math.sqrt);Math.tan;function Li(t){return t>1?bi:t<-1?-bi:Math.asin(t)}function Ni(t,e){return[t>_i?t-wi:t<-_i?t+wi:t,e]}yi(),yi(),yi(),Ni.invert=Ni;yi();function $i(t,e){return t<e?-1:t>e?1:t>=e?0:NaN}var Ci=function(t){return 1===t.length&&(t=function(t){return function(e,n){return $i(t(e),n)}}(t)),{left:function(e,n,r,i){for(null==r&&(r=0),null==i&&(i=e.length);r<i;){var o=r+i>>>1;t(e[o],n)<0?r=o+1:i=o}return r},right:function(e,n,r,i){for(null==r&&(r=0),null==i&&(i=e.length);r<i;){var o=r+i>>>1;t(e[o],n)>0?i=o:r=o+1}return r}}}($i);Ci.right,Ci.left;var Ti=Array.prototype;function Gi(){}function Ui(t){return function(e,n){var r=Ei(e),i=Ei(n),o=t(r*i);return[o*i*Ai(e),o*Ai(n)]}}function Di(t){return function(e,n){var r=Si(e*e+n*n),i=t(r),o=Ai(i),a=Ei(i);return[Mi(e*o,r*a),Li(r&&n*o/r)]}}function Oi(t,e){return[t,e]}function ji(t){var e=0,n=t.children,r=n&&n.length;if(r)for(;--r>=0;)e+=n[r].value;else e=1;t.value=e}function zi(t,e){var n,r,i,o,a,s=new Fi(t),u=+t.value&&(s.value=t.value),c=[s];for(null==e&&(e=Pi);n=c.pop();)if(u&&(n.value=+n.data.value),(i=e(n.data))&&(a=i.length))for(n.children=new Array(a),o=a-1;o>=0;--o)c.push(r=n.children[o]=new Fi(i[o])),r.parent=n,r.depth=n.depth+1;return s.eachBefore(Ri)}function Pi(t){return t.children}function Ii(t){t.data=t.data.data}function Ri(t){var e=0;do{t.height=e}while((t=t.parent)&&t.height<++e)}function Fi(t){this.data=t,this.depth=this.height=0,this.parent=null}function qi(t,e,n,r,i){for(var o,a=t.children,s=-1,u=a.length,c=t.value&&(r-e)/t.value;++s<u;)(o=a[s]).y0=n,o.y1=i,o.x0=e,o.x1=e+=o.value*c}function Yi(t,e,n,r,i){for(var o,a=t.children,s=-1,u=a.length,c=t.value&&(i-n)/t.value;++s<u;)(o=a[s]).x0=e,o.x1=r,o.y0=n,o.y1=n+=o.value*c}Ti.slice,Ti.map,Math.sqrt(50),Math.sqrt(10),Math.sqrt(2),yi(),yi(),yi(),yi(),Gi.prototype={constructor:Gi,point:function(t,e){this.stream.point(t,e)},sphere:function(){this.stream.sphere()},lineStart:function(){this.stream.lineStart()},lineEnd:function(){this.stream.lineEnd()},polygonStart:function(){this.stream.polygonStart()},polygonEnd:function(){this.stream.polygonEnd()}},Ei(30*xi),Ui((function(t){return Si(2/(1+t))})).invert=Di((function(t){return 2*Li(t/2)})),Ui((function(t){return(t=(e=t)>1?0:e<-1?_i:Math.acos(e))&&t/Ai(t);var e})).invert=Di((function(t){return t})),Oi.invert=Oi,Di(ki),Di(Li),Di((function(t){return 2*ki(t)})),Fi.prototype=zi.prototype={constructor:Fi,count:function(){return this.eachAfter(ji)},each:function(t){var e,n,r,i,o=this,a=[o];do{for(e=a.reverse(),a=[];o=e.pop();)if(t(o),n=o.children)for(r=0,i=n.length;r<i;++r)a.push(n[r])}while(a.length);return this},eachAfter:function(t){for(var e,n,r,i=this,o=[i],a=[];i=o.pop();)if(a.push(i),e=i.children)for(n=0,r=e.length;n<r;++n)o.push(e[n]);for(;i=a.pop();)t(i);return this},eachBefore:function(t){for(var e,n,r=this,i=[r];r=i.pop();)if(t(r),e=r.children)for(n=e.length-1;n>=0;--n)i.push(e[n]);return this},sum:function(t){return this.eachAfter((function(e){for(var n=+t(e.data)||0,r=e.children,i=r&&r.length;--i>=0;)n+=r[i].value;e.value=n}))},sort:function(t){return this.eachBefore((function(e){e.children&&e.children.sort(t)}))},path:function(t){for(var e=this,n=function(t,e){if(t===e)return t;var n=t.ancestors(),r=e.ancestors(),i=null;for(t=n.pop(),e=r.pop();t===e;)i=t,t=n.pop(),e=r.pop();return i}(e,t),r=[e];e!==n;)e=e.parent,r.push(e);for(var i=r.length;t!==n;)r.splice(i,0,t),t=t.parent;return r},ancestors:function(){for(var t=this,e=[t];t=t.parent;)e.push(t);return e},descendants:function(){var t=[];return this.each((function(e){t.push(e)})),t},leaves:function(){var t=[];return this.eachBefore((function(e){e.children||t.push(e)})),t},links:function(){var t=this,e=[];return t.each((function(n){n!==t&&e.push({source:n.parent,target:n})})),e},copy:function(){return zi(this).eachBefore(Ii)}},Array.prototype.slice,Object.create(Fi.prototype);var Bi=(1+Math.sqrt(5))/2;function Wi(t,e,n,r,i,o){for(var a,s,u,c,h,f,l,d,p,y,g,v=[],m=e.children,_=0,b=0,w=m.length,x=e.value;_<w;){u=i-n,c=o-r;do{h=m[b++].value}while(!h&&b<w);for(f=l=h,g=h*h*(y=Math.max(c/u,u/c)/(x*t)),p=Math.max(l/g,g/f);b<w;++b){if(h+=s=m[b].value,s<f&&(f=s),s>l&&(l=s),g=h*h*y,(d=Math.max(l/g,g/f))>p){h-=s;break}p=d}v.push(a={value:h,dice:u<c,children:m.slice(_,b)}),a.dice?qi(a,n,r,i,x?r+=c*h/x:o):Yi(a,n,r,x?n+=u*h/x:i,o),x-=h,_=b}return v}!function t(e){function n(t,n,r,i,o){Wi(e,t,n,r,i,o)}return n.ratio=function(e){return t((e=+e)>1?e:1)},n}(Bi),function t(e){function n(t,n,r,i,o){if((a=t._squarify)&&a.ratio===e)for(var a,s,u,c,h,f=-1,l=a.length,d=t.value;++f<l;){for(u=(s=a[f]).children,c=s.value=0,h=u.length;c<h;++c)s.value+=u[c].value;s.dice?qi(s,n,r,i,r+=(o-r)*s.value/d):Yi(s,n,r,n+=(i-n)*s.value/d,o),d-=s.value}else t._squarify=a=Wi(e,t,n,r,i,o),a.ratio=e}return n.ratio=function(e){return t((e=+e)>1?e:1)},n}(Bi);var Ki=[].slice,Hi={};function Xi(t){this._size=t,this._call=this._error=null,this._tasks=[],this._data=[],this._waiting=this._active=this._ended=this._start=0}function Vi(t){if(!t._start)try{!function(t){for(;t._start=t._waiting&&t._active<t._size;){var e=t._ended+t._active,n=t._tasks[e],r=n.length-1,i=n[r];n[r]=Zi(t,e),--t._waiting,++t._active,n=i.apply(null,n),t._tasks[e]&&(t._tasks[e]=n||Hi)}}(t)}catch(e){if(t._tasks[t._ended+t._active-1])Ji(t,e);else if(!t._data)throw e}}function Zi(t,e){return function(n,r){t._tasks[e]&&(--t._active,++t._ended,t._tasks[e]=null,null==t._error&&(null!=n?Ji(t,n):(t._data[e]=r,t._waiting?Vi(t):Qi(t))))}}function Ji(t,e){var n,r=t._tasks.length;for(t._error=e,t._data=void 0,t._waiting=NaN;--r>=0;)if((n=t._tasks[r])&&(t._tasks[r]=null,n.abort))try{n.abort()}catch(e){}t._active=NaN,Qi(t)}function Qi(t){if(!t._active&&t._call){var e=t._data;t._data=void 0,t._call(t._error,e)}}function to(){return Math.random()}Xi.prototype=function(t){if(null==t)t=1/0;else if(!((t=+t)>=1))throw new Error("invalid concurrency");return new Xi(t)}.prototype={constructor:Xi,defer:function(t){if("function"!=typeof t)throw new Error("invalid callback");if(this._call)throw new Error("defer after await");if(null!=this._error)return this;var e=Ki.call(arguments,1);return e.push(t),++this._waiting,this._tasks.push(e),Vi(this),this},abort:function(){return null==this._error&&Ji(this,new Error("abort")),this},await:function(t){if("function"!=typeof t)throw new Error("invalid callback");if(this._call)throw new Error("multiple await");return this._call=function(e,n){t.apply(null,[e].concat(n))},Qi(this),this},awaitAll:function(t){if("function"!=typeof t)throw new Error("invalid callback");if(this._call)throw new Error("multiple await");return this._call=t,Qi(this),this}},function t(e){function n(t,n){return t=null==t?0:+t,n=null==n?1:+n,1===arguments.length?(n=t,t=0):n-=t,function(){return e()*n+t}}return n.source=t,n}(to);const eo=function t(e){function n(t,n){var r,i;return t=null==t?0:+t,n=null==n?1:+n,function(){var o;if(null!=r)o=r,r=null;else do{r=2*e()-1,o=2*e()-1,i=r*r+o*o}while(!i||i>1);return t+n*o*Math.sqrt(-2*Math.log(i)/i)}}return n.source=t,n}(to),no=(function t(e){function n(){var t=eo.source(e).apply(this,arguments);return function(){return Math.exp(t())}}return n.source=t,n}(to),function t(e){function n(t){return function(){for(var n=0,r=0;r<t;++r)n+=e();return n}}return n.source=t,n}(to));function ro(t,e){var n,r,i,o,a=g("beforesend","progress","load","error"),s=Gr(),u=new XMLHttpRequest,c=null,h=null,f=0;function l(t){var e,r=u.status;if(!r&&function(t){var e=t.responseType;return e&&"text"!==e?t.response:t.responseText}(u)||r>=200&&r<300||304===r){if(i)try{e=i.call(n,u)}catch(t){return void a.call("error",n,t)}else e=u;a.call("load",n,e)}else a.call("error",n,t)}if("undefined"!=typeof XDomainRequest&&!("withCredentials"in u)&&/^(http(s)?:)?\/\//.test(t)&&(u=new XDomainRequest),"onload"in u?u.onload=u.onerror=u.ontimeout=l:u.onreadystatechange=function(t){u.readyState>3&&l(t)},u.onprogress=function(t){a.call("progress",n,t)},n={header:function(t,e){return t=(t+"").toLowerCase(),arguments.length<2?s.get(t):(null==e?s.remove(t):s.set(t,e+""),n)},mimeType:function(t){return arguments.length?(r=null==t?null:t+"",n):r},responseType:function(t){return arguments.length?(o=t,n):o},timeout:function(t){return arguments.length?(f=+t,n):f},user:function(t){return arguments.length<1?c:(c=null==t?null:t+"",n)},password:function(t){return arguments.length<1?h:(h=null==t?null:t+"",n)},response:function(t){return i=t,n},get:function(t,e){return n.send("GET",t,e)},post:function(t,e){return n.send("POST",t,e)},send:function(e,i,l){return u.open(e,t,!0,c,h),null==r||s.has("accept")||s.set("accept",r+",*/*"),u.setRequestHeader&&s.each((function(t,e){u.setRequestHeader(e,t)})),null!=r&&u.overrideMimeType&&u.overrideMimeType(r),null!=o&&(u.responseType=o),f>0&&(u.timeout=f),null==l&&"function"==typeof i&&(l=i,i=null),null!=l&&1===l.length&&(l=function(t){return function(e,n){t(null==e?n:null)}}(l)),null!=l&&n.on("error",l).on("load",(function(t){l(null,t)})),a.call("beforesend",n,u),u.send(null==i?null:i),n},abort:function(){return u.abort(),n},on:function(){var t=a.on.apply(a,arguments);return t===a?n:t}},null!=e){if("function"!=typeof e)throw new Error("invalid callback: "+e);return n.get(e)}return n}function io(t,e){return function(n,r){var i=ro(n).mimeType(t).response(e);if(null!=r){if("function"!=typeof r)throw new Error("invalid callback: "+r);return i.get(r)}return i}}function oo(t,e){return function(n,r,i){arguments.length<3&&(i=r,r=null);var o=ro(n).mimeType(t);return o.row=function(t){return arguments.length?o.response(function(t,e){return function(n){return t(n.responseText,e)}}(e,r=t)):r},o.row(r),i?o.get(i):o}}(function t(e){function n(t){var n=no.source(e)(t);return function(){return n()/t}}return n.source=t,n})(to),function t(e){function n(t){return function(){return-Math.log(1-e())/t}}return n.source=t,n}(to),io("text/html",(function(t){return document.createRange().createContextualFragment(t.responseText)})),io("application/json",(function(t){return JSON.parse(t.responseText)})),io("text/plain",(function(t){return t.responseText})),io("application/xml",(function(t){var e=t.responseXML;if(!e)throw new Error("parse error");return e})),oo("text/csv",Rr),oo("text/tab-separated-values",qr);var ao=Array.prototype;ao.map,ao.slice;var so=new Date,uo=new Date;function co(t,e,n,r){function i(e){return t(e=new Date(+e)),e}return i.floor=i,i.ceil=function(n){return t(n=new Date(n-1)),e(n,1),t(n),n},i.round=function(t){var e=i(t),n=i.ceil(t);return t-e<n-t?e:n},i.offset=function(t,n){return e(t=new Date(+t),null==n?1:Math.floor(n)),t},i.range=function(n,r,o){var a,s=[];if(n=i.ceil(n),o=null==o?1:Math.floor(o),!(n<r&&o>0))return s;do{s.push(a=new Date(+n)),e(n,o),t(n)}while(a<n&&n<r);return s},i.filter=function(n){return co((function(e){if(e>=e)for(;t(e),!n(e);)e.setTime(e-1)}),(function(t,r){if(t>=t)if(r<0)for(;++r<=0;)for(;e(t,-1),!n(t););else for(;--r>=0;)for(;e(t,1),!n(t););}))},n&&(i.count=function(e,r){return so.setTime(+e),uo.setTime(+r),t(so),t(uo),Math.floor(n(so,uo))},i.every=function(t){return t=Math.floor(t),isFinite(t)&&t>0?t>1?i.filter(r?function(e){return r(e)%t==0}:function(e){return i.count(0,e)%t==0}):i:null}),i}var ho=co((function(){}),(function(t,e){t.setTime(+t+e)}),(function(t,e){return e-t}));ho.every=function(t){return t=Math.floor(t),isFinite(t)&&t>0?t>1?co((function(e){e.setTime(Math.floor(e/t)*t)}),(function(e,n){e.setTime(+e+n*t)}),(function(e,n){return(n-e)/t})):ho:null},ho.range;var fo=1e3,lo=6e4,po=36e5,yo=864e5,go=6048e5,vo=co((function(t){t.setTime(Math.floor(t/fo)*fo)}),(function(t,e){t.setTime(+t+e*fo)}),(function(t,e){return(e-t)/fo}),(function(t){return t.getUTCSeconds()}));vo.range;var mo=co((function(t){t.setTime(Math.floor(t/lo)*lo)}),(function(t,e){t.setTime(+t+e*lo)}),(function(t,e){return(e-t)/lo}),(function(t){return t.getMinutes()}));mo.range;var _o=co((function(t){var e=t.getTimezoneOffset()*lo%po;e<0&&(e+=po),t.setTime(Math.floor((+t-e)/po)*po+e)}),(function(t,e){t.setTime(+t+e*po)}),(function(t,e){return(e-t)/po}),(function(t){return t.getHours()}));_o.range;var bo=co((function(t){t.setHours(0,0,0,0)}),(function(t,e){t.setDate(t.getDate()+e)}),(function(t,e){return(e-t-(e.getTimezoneOffset()-t.getTimezoneOffset())*lo)/yo}),(function(t){return t.getDate()-1}));const wo=bo;function xo(t){return co((function(e){e.setDate(e.getDate()-(e.getDay()+7-t)%7),e.setHours(0,0,0,0)}),(function(t,e){t.setDate(t.getDate()+7*e)}),(function(t,e){return(e-t-(e.getTimezoneOffset()-t.getTimezoneOffset())*lo)/go}))}bo.range;var ko=xo(0),Mo=xo(1),Eo=xo(2),Ao=xo(3),So=xo(4),Lo=xo(5),No=xo(6),$o=(ko.range,Mo.range,Eo.range,Ao.range,So.range,Lo.range,No.range,co((function(t){t.setDate(1),t.setHours(0,0,0,0)}),(function(t,e){t.setMonth(t.getMonth()+e)}),(function(t,e){return e.getMonth()-t.getMonth()+12*(e.getFullYear()-t.getFullYear())}),(function(t){return t.getMonth()})));$o.range;var Co=co((function(t){t.setMonth(0,1),t.setHours(0,0,0,0)}),(function(t,e){t.setFullYear(t.getFullYear()+e)}),(function(t,e){return e.getFullYear()-t.getFullYear()}),(function(t){return t.getFullYear()}));Co.every=function(t){return isFinite(t=Math.floor(t))&&t>0?co((function(e){e.setFullYear(Math.floor(e.getFullYear()/t)*t),e.setMonth(0,1),e.setHours(0,0,0,0)}),(function(e,n){e.setFullYear(e.getFullYear()+n*t)})):null};const To=Co;Co.range;var Go=co((function(t){t.setUTCSeconds(0,0)}),(function(t,e){t.setTime(+t+e*lo)}),(function(t,e){return(e-t)/lo}),(function(t){return t.getUTCMinutes()}));Go.range;var Uo=co((function(t){t.setUTCMinutes(0,0,0)}),(function(t,e){t.setTime(+t+e*po)}),(function(t,e){return(e-t)/po}),(function(t){return t.getUTCHours()}));Uo.range;var Do=co((function(t){t.setUTCHours(0,0,0,0)}),(function(t,e){t.setUTCDate(t.getUTCDate()+e)}),(function(t,e){return(e-t)/yo}),(function(t){return t.getUTCDate()-1}));const Oo=Do;function jo(t){return co((function(e){e.setUTCDate(e.getUTCDate()-(e.getUTCDay()+7-t)%7),e.setUTCHours(0,0,0,0)}),(function(t,e){t.setUTCDate(t.getUTCDate()+7*e)}),(function(t,e){return(e-t)/go}))}Do.range;var zo=jo(0),Po=jo(1),Io=jo(2),Ro=jo(3),Fo=jo(4),qo=jo(5),Yo=jo(6),Bo=(zo.range,Po.range,Io.range,Ro.range,Fo.range,qo.range,Yo.range,co((function(t){t.setUTCDate(1),t.setUTCHours(0,0,0,0)}),(function(t,e){t.setUTCMonth(t.getUTCMonth()+e)}),(function(t,e){return e.getUTCMonth()-t.getUTCMonth()+12*(e.getUTCFullYear()-t.getUTCFullYear())}),(function(t){return t.getUTCMonth()})));Bo.range;var Wo=co((function(t){t.setUTCMonth(0,1),t.setUTCHours(0,0,0,0)}),(function(t,e){t.setUTCFullYear(t.getUTCFullYear()+e)}),(function(t,e){return e.getUTCFullYear()-t.getUTCFullYear()}),(function(t){return t.getUTCFullYear()}));Wo.every=function(t){return isFinite(t=Math.floor(t))&&t>0?co((function(e){e.setUTCFullYear(Math.floor(e.getUTCFullYear()/t)*t),e.setUTCMonth(0,1),e.setUTCHours(0,0,0,0)}),(function(e,n){e.setUTCFullYear(e.getUTCFullYear()+n*t)})):null};const Ko=Wo;function Ho(t){if(0<=t.y&&t.y<100){var e=new Date(-1,t.m,t.d,t.H,t.M,t.S,t.L);return e.setFullYear(t.y),e}return new Date(t.y,t.m,t.d,t.H,t.M,t.S,t.L)}function Xo(t){if(0<=t.y&&t.y<100){var e=new Date(Date.UTC(-1,t.m,t.d,t.H,t.M,t.S,t.L));return e.setUTCFullYear(t.y),e}return new Date(Date.UTC(t.y,t.m,t.d,t.H,t.M,t.S,t.L))}function Vo(t){return{y:t,m:0,d:1,H:0,M:0,S:0,L:0}}Wo.range;var Zo,Jo,Qo,ta={"-":"",_:" ",0:"0"},ea=/^\s*\d+/,na=/^%/,ra=/[\\^$*+?|[\]().{}]/g;function ia(t,e,n){var r=t<0?"-":"",i=(r?-t:t)+"",o=i.length;return r+(o<n?new Array(n-o+1).join(e)+i:i)}function oa(t){return t.replace(ra,"\\$&")}function aa(t){return new RegExp("^(?:"+t.map(oa).join("|")+")","i")}function sa(t){for(var e={},n=-1,r=t.length;++n<r;)e[t[n].toLowerCase()]=n;return e}function ua(t,e,n){var r=ea.exec(e.slice(n,n+1));return r?(t.w=+r[0],n+r[0].length):-1}function ca(t,e,n){var r=ea.exec(e.slice(n,n+1));return r?(t.u=+r[0],n+r[0].length):-1}function ha(t,e,n){var r=ea.exec(e.slice(n,n+2));return r?(t.U=+r[0],n+r[0].length):-1}function fa(t,e,n){var r=ea.exec(e.slice(n,n+2));return r?(t.V=+r[0],n+r[0].length):-1}function la(t,e,n){var r=ea.exec(e.slice(n,n+2));return r?(t.W=+r[0],n+r[0].length):-1}function da(t,e,n){var r=ea.exec(e.slice(n,n+4));return r?(t.y=+r[0],n+r[0].length):-1}function pa(t,e,n){var r=ea.exec(e.slice(n,n+2));return r?(t.y=+r[0]+(+r[0]>68?1900:2e3),n+r[0].length):-1}function ya(t,e,n){var r=/^(Z)|([+-]\d\d)(?::?(\d\d))?/.exec(e.slice(n,n+6));return r?(t.Z=r[1]?0:-(r[2]+(r[3]||"00")),n+r[0].length):-1}function ga(t,e,n){var r=ea.exec(e.slice(n,n+2));return r?(t.m=r[0]-1,n+r[0].length):-1}function va(t,e,n){var r=ea.exec(e.slice(n,n+2));return r?(t.d=+r[0],n+r[0].length):-1}function ma(t,e,n){var r=ea.exec(e.slice(n,n+3));return r?(t.m=0,t.d=+r[0],n+r[0].length):-1}function _a(t,e,n){var r=ea.exec(e.slice(n,n+2));return r?(t.H=+r[0],n+r[0].length):-1}function ba(t,e,n){var r=ea.exec(e.slice(n,n+2));return r?(t.M=+r[0],n+r[0].length):-1}function wa(t,e,n){var r=ea.exec(e.slice(n,n+2));return r?(t.S=+r[0],n+r[0].length):-1}function xa(t,e,n){var r=ea.exec(e.slice(n,n+3));return r?(t.L=+r[0],n+r[0].length):-1}function ka(t,e,n){var r=ea.exec(e.slice(n,n+6));return r?(t.L=Math.floor(r[0]/1e3),n+r[0].length):-1}function Ma(t,e,n){var r=na.exec(e.slice(n,n+1));return r?n+r[0].length:-1}function Ea(t,e,n){var r=ea.exec(e.slice(n));return r?(t.Q=+r[0],n+r[0].length):-1}function Aa(t,e,n){var r=ea.exec(e.slice(n));return r?(t.Q=1e3*+r[0],n+r[0].length):-1}function Sa(t,e){return ia(t.getDate(),e,2)}function La(t,e){return ia(t.getHours(),e,2)}function Na(t,e){return ia(t.getHours()%12||12,e,2)}function $a(t,e){return ia(1+wo.count(To(t),t),e,3)}function Ca(t,e){return ia(t.getMilliseconds(),e,3)}function Ta(t,e){return Ca(t,e)+"000"}function Ga(t,e){return ia(t.getMonth()+1,e,2)}function Ua(t,e){return ia(t.getMinutes(),e,2)}function Da(t,e){return ia(t.getSeconds(),e,2)}function Oa(t){var e=t.getDay();return 0===e?7:e}function ja(t,e){return ia(ko.count(To(t),t),e,2)}function za(t,e){var n=t.getDay();return t=n>=4||0===n?So(t):So.ceil(t),ia(So.count(To(t),t)+(4===To(t).getDay()),e,2)}function Pa(t){return t.getDay()}function Ia(t,e){return ia(Mo.count(To(t),t),e,2)}function Ra(t,e){return ia(t.getFullYear()%100,e,2)}function Fa(t,e){return ia(t.getFullYear()%1e4,e,4)}function qa(t){var e=t.getTimezoneOffset();return(e>0?"-":(e*=-1,"+"))+ia(e/60|0,"0",2)+ia(e%60,"0",2)}function Ya(t,e){return ia(t.getUTCDate(),e,2)}function Ba(t,e){return ia(t.getUTCHours(),e,2)}function Wa(t,e){return ia(t.getUTCHours()%12||12,e,2)}function Ka(t,e){return ia(1+Oo.count(Ko(t),t),e,3)}function Ha(t,e){return ia(t.getUTCMilliseconds(),e,3)}function Xa(t,e){return Ha(t,e)+"000"}function Va(t,e){return ia(t.getUTCMonth()+1,e,2)}function Za(t,e){return ia(t.getUTCMinutes(),e,2)}function Ja(t,e){return ia(t.getUTCSeconds(),e,2)}function Qa(t){var e=t.getUTCDay();return 0===e?7:e}function ts(t,e){return ia(zo.count(Ko(t),t),e,2)}function es(t,e){var n=t.getUTCDay();return t=n>=4||0===n?Fo(t):Fo.ceil(t),ia(Fo.count(Ko(t),t)+(4===Ko(t).getUTCDay()),e,2)}function ns(t){return t.getUTCDay()}function rs(t,e){return ia(Po.count(Ko(t),t),e,2)}function is(t,e){return ia(t.getUTCFullYear()%100,e,2)}function os(t,e){return ia(t.getUTCFullYear()%1e4,e,4)}function as(){return"+0000"}function ss(){return"%"}function us(t){return+t}function cs(t){return Math.floor(+t/1e3)}Zo=function(t){var e=t.dateTime,n=t.date,r=t.time,i=t.periods,o=t.days,a=t.shortDays,s=t.months,u=t.shortMonths,c=aa(i),h=sa(i),f=aa(o),l=sa(o),d=aa(a),p=sa(a),y=aa(s),g=sa(s),v=aa(u),m=sa(u),_={a:function(t){return a[t.getDay()]},A:function(t){return o[t.getDay()]},b:function(t){return u[t.getMonth()]},B:function(t){return s[t.getMonth()]},c:null,d:Sa,e:Sa,f:Ta,H:La,I:Na,j:$a,L:Ca,m:Ga,M:Ua,p:function(t){return i[+(t.getHours()>=12)]},Q:us,s:cs,S:Da,u:Oa,U:ja,V:za,w:Pa,W:Ia,x:null,X:null,y:Ra,Y:Fa,Z:qa,"%":ss},b={a:function(t){return a[t.getUTCDay()]},A:function(t){return o[t.getUTCDay()]},b:function(t){return u[t.getUTCMonth()]},B:function(t){return s[t.getUTCMonth()]},c:null,d:Ya,e:Ya,f:Xa,H:Ba,I:Wa,j:Ka,L:Ha,m:Va,M:Za,p:function(t){return i[+(t.getUTCHours()>=12)]},Q:us,s:cs,S:Ja,u:Qa,U:ts,V:es,w:ns,W:rs,x:null,X:null,y:is,Y:os,Z:as,"%":ss},w={a:function(t,e,n){var r=d.exec(e.slice(n));return r?(t.w=p[r[0].toLowerCase()],n+r[0].length):-1},A:function(t,e,n){var r=f.exec(e.slice(n));return r?(t.w=l[r[0].toLowerCase()],n+r[0].length):-1},b:function(t,e,n){var r=v.exec(e.slice(n));return r?(t.m=m[r[0].toLowerCase()],n+r[0].length):-1},B:function(t,e,n){var r=y.exec(e.slice(n));return r?(t.m=g[r[0].toLowerCase()],n+r[0].length):-1},c:function(t,n,r){return M(t,e,n,r)},d:va,e:va,f:ka,H:_a,I:_a,j:ma,L:xa,m:ga,M:ba,p:function(t,e,n){var r=c.exec(e.slice(n));return r?(t.p=h[r[0].toLowerCase()],n+r[0].length):-1},Q:Ea,s:Aa,S:wa,u:ca,U:ha,V:fa,w:ua,W:la,x:function(t,e,r){return M(t,n,e,r)},X:function(t,e,n){return M(t,r,e,n)},y:pa,Y:da,Z:ya,"%":Ma};function x(t,e){return function(n){var r,i,o,a=[],s=-1,u=0,c=t.length;for(n instanceof Date||(n=new Date(+n));++s<c;)37===t.charCodeAt(s)&&(a.push(t.slice(u,s)),null!=(i=ta[r=t.charAt(++s)])?r=t.charAt(++s):i="e"===r?" ":"0",(o=e[r])&&(r=o(n,i)),a.push(r),u=s+1);return a.push(t.slice(u,s)),a.join("")}}function k(t,e){return function(n){var r,i,o=Vo(1900);if(M(o,t,n+="",0)!=n.length)return null;if("Q"in o)return new Date(o.Q);if("p"in o&&(o.H=o.H%12+12*o.p),"V"in o){if(o.V<1||o.V>53)return null;"w"in o||(o.w=1),"Z"in o?(i=(r=Xo(Vo(o.y))).getUTCDay(),r=i>4||0===i?Po.ceil(r):Po(r),r=Oo.offset(r,7*(o.V-1)),o.y=r.getUTCFullYear(),o.m=r.getUTCMonth(),o.d=r.getUTCDate()+(o.w+6)%7):(i=(r=e(Vo(o.y))).getDay(),r=i>4||0===i?Mo.ceil(r):Mo(r),r=wo.offset(r,7*(o.V-1)),o.y=r.getFullYear(),o.m=r.getMonth(),o.d=r.getDate()+(o.w+6)%7)}else("W"in o||"U"in o)&&("w"in o||(o.w="u"in o?o.u%7:"W"in o?1:0),i="Z"in o?Xo(Vo(o.y)).getUTCDay():e(Vo(o.y)).getDay(),o.m=0,o.d="W"in o?(o.w+6)%7+7*o.W-(i+5)%7:o.w+7*o.U-(i+6)%7);return"Z"in o?(o.H+=o.Z/100|0,o.M+=o.Z%100,Xo(o)):e(o)}}function M(t,e,n,r){for(var i,o,a=0,s=e.length,u=n.length;a<s;){if(r>=u)return-1;if(37===(i=e.charCodeAt(a++))){if(i=e.charAt(a++),!(o=w[i in ta?e.charAt(a++):i])||(r=o(t,n,r))<0)return-1}else if(i!=n.charCodeAt(r++))return-1}return r}return _.x=x(n,_),_.X=x(r,_),_.c=x(e,_),b.x=x(n,b),b.X=x(r,b),b.c=x(e,b),{format:function(t){var e=x(t+="",_);return e.toString=function(){return t},e},parse:function(t){var e=k(t+="",Ho);return e.toString=function(){return t},e},utcFormat:function(t){var e=x(t+="",b);return e.toString=function(){return t},e},utcParse:function(t){var e=k(t,Xo);return e.toString=function(){return t},e}}}({dateTime:"%x, %X",date:"%-m/%-d/%Y",time:"%-I:%M:%S %p",periods:["AM","PM"],days:["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"],shortDays:["Sun","Mon","Tue","Wed","Thu","Fri","Sat"],months:["January","February","March","April","May","June","July","August","September","October","November","December"],shortMonths:["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"]}),Zo.format,Zo.parse,Jo=Zo.utcFormat,Qo=Zo.utcParse;var hs="%Y-%m-%dT%H:%M:%S.%LZ";function fs(t){return t.match(/.{6}/g).map((function(t){return"#"+t}))}function ls(t){var e=t.length;return function(n){return t[Math.max(0,Math.min(e-1,Math.floor(n*e)))]}}Date.prototype.toISOString||Jo(hs),+new Date("2000-01-01T00:00:00.000Z")||Qo(hs),fs("1f77b4ff7f0e2ca02cd627289467bd8c564be377c27f7f7fbcbd2217becf"),fs("393b795254a36b6ecf9c9ede6379398ca252b5cf6bcedb9c8c6d31bd9e39e7ba52e7cb94843c39ad494ad6616be7969c7b4173a55194ce6dbdde9ed6"),fs("3182bd6baed69ecae1c6dbefe6550dfd8d3cfdae6bfdd0a231a35474c476a1d99bc7e9c0756bb19e9ac8bcbddcdadaeb636363969696bdbdbdd9d9d9"),fs("1f77b4aec7e8ff7f0effbb782ca02c98df8ad62728ff98969467bdc5b0d58c564bc49c94e377c2f7b6d27f7f7fc7c7c7bcbd22dbdb8d17becf9edae5"),kn(Be(300,.5,0),Be(-240,.5,1)),kn(Be(-100,.75,.35),Be(80,1.5,.8)),kn(Be(260,.75,.35),Be(80,1.5,.8)),Be(),ls(fs("44015444025645045745055946075a46085c460a5d460b5e470d60470e6147106347116447136548146748166848176948186a481a6c481b6d481c6e481d6f481f70482071482173482374482475482576482677482878482979472a7a472c7a472d7b472e7c472f7d46307e46327e46337f463480453581453781453882443983443a83443b84433d84433e85423f854240864241864142874144874045884046883f47883f48893e49893e4a893e4c8a3d4d8a3d4e8a3c4f8a3c508b3b518b3b528b3a538b3a548c39558c39568c38588c38598c375a8c375b8d365c8d365d8d355e8d355f8d34608d34618d33628d33638d32648e32658e31668e31678e31688e30698e306a8e2f6b8e2f6c8e2e6d8e2e6e8e2e6f8e2d708e2d718e2c718e2c728e2c738e2b748e2b758e2a768e2a778e2a788e29798e297a8e297b8e287c8e287d8e277e8e277f8e27808e26818e26828e26828e25838e25848e25858e24868e24878e23888e23898e238a8d228b8d228c8d228d8d218e8d218f8d21908d21918c20928c20928c20938c1f948c1f958b1f968b1f978b1f988b1f998a1f9a8a1e9b8a1e9c891e9d891f9e891f9f881fa0881fa1881fa1871fa28720a38620a48621a58521a68522a78522a88423a98324aa8325ab8225ac8226ad8127ad8128ae8029af7f2ab07f2cb17e2db27d2eb37c2fb47c31b57b32b67a34b67935b77937b87838b9773aba763bbb753dbc743fbc7340bd7242be7144bf7046c06f48c16e4ac16d4cc26c4ec36b50c46a52c56954c56856c66758c7655ac8645cc8635ec96260ca6063cb5f65cb5e67cc5c69cd5b6ccd5a6ece5870cf5773d05675d05477d1537ad1517cd2507fd34e81d34d84d44b86d54989d5488bd6468ed64590d74393d74195d84098d83e9bd93c9dd93ba0da39a2da37a5db36a8db34aadc32addc30b0dd2fb2dd2db5de2bb8de29bade28bddf26c0df25c2df23c5e021c8e020cae11fcde11dd0e11cd2e21bd5e21ad8e219dae319dde318dfe318e2e418e5e419e7e419eae51aece51befe51cf1e51df4e61ef6e620f8e621fbe723fde725")),ls(fs("00000401000501010601010802010902020b02020d03030f03031204041405041606051806051a07061c08071e0907200a08220b09240c09260d0a290e0b2b100b2d110c2f120d31130d34140e36150e38160f3b180f3d19103f1a10421c10441d11471e114920114b21114e22115024125325125527125829115a2a115c2c115f2d11612f116331116533106734106936106b38106c390f6e3b0f703d0f713f0f72400f74420f75440f764510774710784910784a10794c117a4e117b4f127b51127c52137c54137d56147d57157e59157e5a167e5c167f5d177f5f187f601880621980641a80651a80671b80681c816a1c816b1d816d1d816e1e81701f81721f817320817521817621817822817922827b23827c23827e24828025828125818326818426818627818827818928818b29818c29818e2a81902a81912b81932b80942c80962c80982d80992d809b2e7f9c2e7f9e2f7fa02f7fa1307ea3307ea5317ea6317da8327daa337dab337cad347cae347bb0357bb2357bb3367ab5367ab73779b83779ba3878bc3978bd3977bf3a77c03a76c23b75c43c75c53c74c73d73c83e73ca3e72cc3f71cd4071cf4070d0416fd2426fd3436ed5446dd6456cd8456cd9466bdb476adc4869de4968df4a68e04c67e24d66e34e65e44f64e55064e75263e85362e95462ea5661eb5760ec5860ed5a5fee5b5eef5d5ef05f5ef1605df2625df2645cf3655cf4675cf4695cf56b5cf66c5cf66e5cf7705cf7725cf8745cf8765cf9785df9795df97b5dfa7d5efa7f5efa815ffb835ffb8560fb8761fc8961fc8a62fc8c63fc8e64fc9065fd9266fd9467fd9668fd9869fd9a6afd9b6bfe9d6cfe9f6dfea16efea36ffea571fea772fea973feaa74feac76feae77feb078feb27afeb47bfeb67cfeb77efeb97ffebb81febd82febf84fec185fec287fec488fec68afec88cfeca8dfecc8ffecd90fecf92fed194fed395fed597fed799fed89afdda9cfddc9efddea0fde0a1fde2a3fde3a5fde5a7fde7a9fde9aafdebacfcecaefceeb0fcf0b2fcf2b4fcf4b6fcf6b8fcf7b9fcf9bbfcfbbdfcfdbf")),ls(fs("00000401000501010601010802010a02020c02020e03021004031204031405041706041907051b08051d09061f0a07220b07240c08260d08290e092b10092d110a30120a32140b34150b37160b39180c3c190c3e1b0c411c0c431e0c451f0c48210c4a230c4c240c4f260c51280b53290b552b0b572d0b592f0a5b310a5c320a5e340a5f3609613809623909633b09643d09653e0966400a67420a68440a68450a69470b6a490b6a4a0c6b4c0c6b4d0d6c4f0d6c510e6c520e6d540f6d550f6d57106e59106e5a116e5c126e5d126e5f136e61136e62146e64156e65156e67166e69166e6a176e6c186e6d186e6f196e71196e721a6e741a6e751b6e771c6d781c6d7a1d6d7c1d6d7d1e6d7f1e6c801f6c82206c84206b85216b87216b88226a8a226a8c23698d23698f24699025689225689326679526679727669827669a28659b29649d29649f2a63a02a63a22b62a32c61a52c60a62d60a82e5fa92e5eab2f5ead305dae305cb0315bb1325ab3325ab43359b63458b73557b93556ba3655bc3754bd3853bf3952c03a51c13a50c33b4fc43c4ec63d4dc73e4cc83f4bca404acb4149cc4248ce4347cf4446d04545d24644d34743d44842d54a41d74b3fd84c3ed94d3dda4e3cdb503bdd513ade5238df5337e05536e15635e25734e35933e45a31e55c30e65d2fe75e2ee8602de9612bea632aeb6429eb6628ec6726ed6925ee6a24ef6c23ef6e21f06f20f1711ff1731df2741cf3761bf37819f47918f57b17f57d15f67e14f68013f78212f78410f8850ff8870ef8890cf98b0bf98c0af98e09fa9008fa9207fa9407fb9606fb9706fb9906fb9b06fb9d07fc9f07fca108fca309fca50afca60cfca80dfcaa0ffcac11fcae12fcb014fcb216fcb418fbb61afbb81dfbba1ffbbc21fbbe23fac026fac228fac42afac62df9c72ff9c932f9cb35f8cd37f8cf3af7d13df7d340f6d543f6d746f5d949f5db4cf4dd4ff4df53f4e156f3e35af3e55df2e661f2e865f2ea69f1ec6df1ed71f1ef75f1f179f2f27df2f482f3f586f3f68af4f88ef5f992f6fa96f8fb9af9fc9dfafda1fcffa4")),ls(fs("0d088710078813078916078a19068c1b068d1d068e20068f2206902406912605912805922a05932c05942e05952f059631059733059735049837049938049a3a049a3c049b3e049c3f049c41049d43039e44039e46039f48039f4903a04b03a14c02a14e02a25002a25102a35302a35502a45601a45801a45901a55b01a55c01a65e01a66001a66100a76300a76400a76600a76700a86900a86a00a86c00a86e00a86f00a87100a87201a87401a87501a87701a87801a87a02a87b02a87d03a87e03a88004a88104a78305a78405a78606a68707a68808a68a09a58b0aa58d0ba58e0ca48f0da4910ea3920fa39410a29511a19613a19814a099159f9a169f9c179e9d189d9e199da01a9ca11b9ba21d9aa31e9aa51f99a62098a72197a82296aa2395ab2494ac2694ad2793ae2892b02991b12a90b22b8fb32c8eb42e8db52f8cb6308bb7318ab83289ba3388bb3488bc3587bd3786be3885bf3984c03a83c13b82c23c81c33d80c43e7fc5407ec6417dc7427cc8437bc9447aca457acb4679cc4778cc4977cd4a76ce4b75cf4c74d04d73d14e72d24f71d35171d45270d5536fd5546ed6556dd7566cd8576bd9586ada5a6ada5b69db5c68dc5d67dd5e66de5f65de6164df6263e06363e16462e26561e26660e3685fe4695ee56a5de56b5de66c5ce76e5be76f5ae87059e97158e97257ea7457eb7556eb7655ec7754ed7953ed7a52ee7b51ef7c51ef7e50f07f4ff0804ef1814df1834cf2844bf3854bf3874af48849f48948f58b47f58c46f68d45f68f44f79044f79143f79342f89441f89540f9973ff9983ef99a3efa9b3dfa9c3cfa9e3bfb9f3afba139fba238fca338fca537fca636fca835fca934fdab33fdac33fdae32fdaf31fdb130fdb22ffdb42ffdb52efeb72dfeb82cfeba2cfebb2bfebd2afebe2afec029fdc229fdc328fdc527fdc627fdc827fdca26fdcb26fccd25fcce25fcd025fcd225fbd324fbd524fbd724fad824fada24f9dc24f9dd25f8df25f8e125f7e225f7e425f6e626f6e826f5e926f5eb27f4ed27f3ee27f3f027f2f227f1f426f1f525f0f724f0f921")),Math.abs,Math.atan2,Math.cos,Math.max,Math.min,Math.sin,Math.sqrt;var ds=Math.PI,ps=2*ds;function ys(t){this._context=t}function gs(t){this._curve=t}ys.prototype={areaStart:function(){this._line=0},areaEnd:function(){this._line=NaN},lineStart:function(){this._point=0},lineEnd:function(){(this._line||0!==this._line&&1===this._point)&&this._context.closePath(),this._line=1-this._line},point:function(t,e){switch(t=+t,e=+e,this._point){case 0:this._point=1,this._line?this._context.lineTo(t,e):this._context.moveTo(t,e);break;case 1:this._point=2;default:this._context.lineTo(t,e)}}},gs.prototype={areaStart:function(){this._curve.areaStart()},areaEnd:function(){this._curve.areaEnd()},lineStart:function(){this._curve.lineStart()},lineEnd:function(){this._curve.lineEnd()},point:function(t,e){this._curve.point(e*Math.sin(t),e*-Math.cos(t))}},Array.prototype.slice,Math.sqrt(1/3);Math.sin(ds/10),Math.sin(7*ds/10);function vs(){}function ms(t,e,n){t._context.bezierCurveTo((2*t._x0+t._x1)/3,(2*t._y0+t._y1)/3,(t._x0+2*t._x1)/3,(t._y0+2*t._y1)/3,(t._x0+4*t._x1+e)/6,(t._y0+4*t._y1+n)/6)}function _s(t){this._context=t}function bs(t,e){this._basis=new _s(t),this._beta=e}function ws(t,e,n){t._context.bezierCurveTo(t._x1+t._k*(t._x2-t._x0),t._y1+t._k*(t._y2-t._y0),t._x2+t._k*(t._x1-e),t._y2+t._k*(t._y1-n),t._x2,t._y2)}function xs(t,e){this._context=t,this._k=(1-e)/6}function ks(t,e){this._context=t,this._k=(1-e)/6}function Ms(t,e){this._context=t,this._k=(1-e)/6}function Es(t,e,n){var r=t._x1,i=t._y1,o=t._x2,a=t._y2;if(t._l01_a>1e-12){var s=2*t._l01_2a+3*t._l01_a*t._l12_a+t._l12_2a,u=3*t._l01_a*(t._l01_a+t._l12_a);r=(r*s-t._x0*t._l12_2a+t._x2*t._l01_2a)/u,i=(i*s-t._y0*t._l12_2a+t._y2*t._l01_2a)/u}if(t._l23_a>1e-12){var c=2*t._l23_2a+3*t._l23_a*t._l12_a+t._l12_2a,h=3*t._l23_a*(t._l23_a+t._l12_a);o=(o*c+t._x1*t._l23_2a-e*t._l12_2a)/h,a=(a*c+t._y1*t._l23_2a-n*t._l12_2a)/h}t._context.bezierCurveTo(r,i,o,a,t._x2,t._y2)}function As(t,e){this._context=t,this._alpha=e}function Ss(t,e){this._context=t,this._alpha=e}function Ls(t,e){this._context=t,this._alpha=e}function Ns(t){return t<0?-1:1}function $s(t,e,n){var r=t._x1-t._x0,i=e-t._x1,o=(t._y1-t._y0)/(r||i<0&&-0),a=(n-t._y1)/(i||r<0&&-0),s=(o*i+a*r)/(r+i);return(Ns(o)+Ns(a))*Math.min(Math.abs(o),Math.abs(a),.5*Math.abs(s))||0}function Cs(t,e){var n=t._x1-t._x0;return n?(3*(t._y1-t._y0)/n-e)/2:e}function Ts(t,e,n){var r=t._x0,i=t._y0,o=t._x1,a=t._y1,s=(o-r)/3;t._context.bezierCurveTo(r+s,i+s*e,o-s,a-s*n,o,a)}function Gs(t){this._context=t}function Us(t){this._context=t}function Ds(){this._=null}function Os(t){t.U=t.C=t.L=t.R=t.P=t.N=null}function js(t,e){var n=e,r=e.R,i=n.U;i?i.L===n?i.L=r:i.R=r:t._=r,r.U=i,n.U=r,n.R=r.L,n.R&&(n.R.U=n),r.L=n}function zs(t,e){var n=e,r=e.L,i=n.U;i?i.L===n?i.L=r:i.R=r:t._=r,r.U=i,n.U=r,n.L=r.R,n.L&&(n.L.U=n),r.R=n}function Ps(t){for(;t.L;)t=t.L;return t}Math.sin(ps/10),Math.cos(ps/10),Math.sqrt(3),Math.sqrt(3),Math.sqrt(12),_s.prototype={areaStart:function(){this._line=0},areaEnd:function(){this._line=NaN},lineStart:function(){this._x0=this._x1=this._y0=this._y1=NaN,this._point=0},lineEnd:function(){switch(this._point){case 3:ms(this,this._x1,this._y1);case 2:this._context.lineTo(this._x1,this._y1)}(this._line||0!==this._line&&1===this._point)&&this._context.closePath(),this._line=1-this._line},point:function(t,e){switch(t=+t,e=+e,this._point){case 0:this._point=1,this._line?this._context.lineTo(t,e):this._context.moveTo(t,e);break;case 1:this._point=2;break;case 2:this._point=3,this._context.lineTo((5*this._x0+this._x1)/6,(5*this._y0+this._y1)/6);default:ms(this,t,e)}this._x0=this._x1,this._x1=t,this._y0=this._y1,this._y1=e}},bs.prototype={lineStart:function(){this._x=[],this._y=[],this._basis.lineStart()},lineEnd:function(){var t=this._x,e=this._y,n=t.length-1;if(n>0)for(var r,i=t[0],o=e[0],a=t[n]-i,s=e[n]-o,u=-1;++u<=n;)r=u/n,this._basis.point(this._beta*t[u]+(1-this._beta)*(i+r*a),this._beta*e[u]+(1-this._beta)*(o+r*s));this._x=this._y=null,this._basis.lineEnd()},point:function(t,e){this._x.push(+t),this._y.push(+e)}},function t(e){function n(t){return 1===e?new _s(t):new bs(t,e)}return n.beta=function(e){return t(+e)},n}(.85),xs.prototype={areaStart:function(){this._line=0},areaEnd:function(){this._line=NaN},lineStart:function(){this._x0=this._x1=this._x2=this._y0=this._y1=this._y2=NaN,this._point=0},lineEnd:function(){switch(this._point){case 2:this._context.lineTo(this._x2,this._y2);break;case 3:ws(this,this._x1,this._y1)}(this._line||0!==this._line&&1===this._point)&&this._context.closePath(),this._line=1-this._line},point:function(t,e){switch(t=+t,e=+e,this._point){case 0:this._point=1,this._line?this._context.lineTo(t,e):this._context.moveTo(t,e);break;case 1:this._point=2,this._x1=t,this._y1=e;break;case 2:this._point=3;default:ws(this,t,e)}this._x0=this._x1,this._x1=this._x2,this._x2=t,this._y0=this._y1,this._y1=this._y2,this._y2=e}},function t(e){function n(t){return new xs(t,e)}return n.tension=function(e){return t(+e)},n}(0),ks.prototype={areaStart:vs,areaEnd:vs,lineStart:function(){this._x0=this._x1=this._x2=this._x3=this._x4=this._x5=this._y0=this._y1=this._y2=this._y3=this._y4=this._y5=NaN,this._point=0},lineEnd:function(){switch(this._point){case 1:this._context.moveTo(this._x3,this._y3),this._context.closePath();break;case 2:this._context.lineTo(this._x3,this._y3),this._context.closePath();break;case 3:this.point(this._x3,this._y3),this.point(this._x4,this._y4),this.point(this._x5,this._y5)}},point:function(t,e){switch(t=+t,e=+e,this._point){case 0:this._point=1,this._x3=t,this._y3=e;break;case 1:this._point=2,this._context.moveTo(this._x4=t,this._y4=e);break;case 2:this._point=3,this._x5=t,this._y5=e;break;default:ws(this,t,e)}this._x0=this._x1,this._x1=this._x2,this._x2=t,this._y0=this._y1,this._y1=this._y2,this._y2=e}},function t(e){function n(t){return new ks(t,e)}return n.tension=function(e){return t(+e)},n}(0),Ms.prototype={areaStart:function(){this._line=0},areaEnd:function(){this._line=NaN},lineStart:function(){this._x0=this._x1=this._x2=this._y0=this._y1=this._y2=NaN,this._point=0},lineEnd:function(){(this._line||0!==this._line&&3===this._point)&&this._context.closePath(),this._line=1-this._line},point:function(t,e){switch(t=+t,e=+e,this._point){case 0:this._point=1;break;case 1:this._point=2;break;case 2:this._point=3,this._line?this._context.lineTo(this._x2,this._y2):this._context.moveTo(this._x2,this._y2);break;case 3:this._point=4;default:ws(this,t,e)}this._x0=this._x1,this._x1=this._x2,this._x2=t,this._y0=this._y1,this._y1=this._y2,this._y2=e}},function t(e){function n(t){return new Ms(t,e)}return n.tension=function(e){return t(+e)},n}(0),As.prototype={areaStart:function(){this._line=0},areaEnd:function(){this._line=NaN},lineStart:function(){this._x0=this._x1=this._x2=this._y0=this._y1=this._y2=NaN,this._l01_a=this._l12_a=this._l23_a=this._l01_2a=this._l12_2a=this._l23_2a=this._point=0},lineEnd:function(){switch(this._point){case 2:this._context.lineTo(this._x2,this._y2);break;case 3:this.point(this._x2,this._y2)}(this._line||0!==this._line&&1===this._point)&&this._context.closePath(),this._line=1-this._line},point:function(t,e){if(t=+t,e=+e,this._point){var n=this._x2-t,r=this._y2-e;this._l23_a=Math.sqrt(this._l23_2a=Math.pow(n*n+r*r,this._alpha))}switch(this._point){case 0:this._point=1,this._line?this._context.lineTo(t,e):this._context.moveTo(t,e);break;case 1:this._point=2;break;case 2:this._point=3;default:Es(this,t,e)}this._l01_a=this._l12_a,this._l12_a=this._l23_a,this._l01_2a=this._l12_2a,this._l12_2a=this._l23_2a,this._x0=this._x1,this._x1=this._x2,this._x2=t,this._y0=this._y1,this._y1=this._y2,this._y2=e}},function t(e){function n(t){return e?new As(t,e):new xs(t,0)}return n.alpha=function(e){return t(+e)},n}(.5),Ss.prototype={areaStart:vs,areaEnd:vs,lineStart:function(){this._x0=this._x1=this._x2=this._x3=this._x4=this._x5=this._y0=this._y1=this._y2=this._y3=this._y4=this._y5=NaN,this._l01_a=this._l12_a=this._l23_a=this._l01_2a=this._l12_2a=this._l23_2a=this._point=0},lineEnd:function(){switch(this._point){case 1:this._context.moveTo(this._x3,this._y3),this._context.closePath();break;case 2:this._context.lineTo(this._x3,this._y3),this._context.closePath();break;case 3:this.point(this._x3,this._y3),this.point(this._x4,this._y4),this.point(this._x5,this._y5)}},point:function(t,e){if(t=+t,e=+e,this._point){var n=this._x2-t,r=this._y2-e;this._l23_a=Math.sqrt(this._l23_2a=Math.pow(n*n+r*r,this._alpha))}switch(this._point){case 0:this._point=1,this._x3=t,this._y3=e;break;case 1:this._point=2,this._context.moveTo(this._x4=t,this._y4=e);break;case 2:this._point=3,this._x5=t,this._y5=e;break;default:Es(this,t,e)}this._l01_a=this._l12_a,this._l12_a=this._l23_a,this._l01_2a=this._l12_2a,this._l12_2a=this._l23_2a,this._x0=this._x1,this._x1=this._x2,this._x2=t,this._y0=this._y1,this._y1=this._y2,this._y2=e}},function t(e){function n(t){return e?new Ss(t,e):new ks(t,0)}return n.alpha=function(e){return t(+e)},n}(.5),Ls.prototype={areaStart:function(){this._line=0},areaEnd:function(){this._line=NaN},lineStart:function(){this._x0=this._x1=this._x2=this._y0=this._y1=this._y2=NaN,this._l01_a=this._l12_a=this._l23_a=this._l01_2a=this._l12_2a=this._l23_2a=this._point=0},lineEnd:function(){(this._line||0!==this._line&&3===this._point)&&this._context.closePath(),this._line=1-this._line},point:function(t,e){if(t=+t,e=+e,this._point){var n=this._x2-t,r=this._y2-e;this._l23_a=Math.sqrt(this._l23_2a=Math.pow(n*n+r*r,this._alpha))}switch(this._point){case 0:this._point=1;break;case 1:this._point=2;break;case 2:this._point=3,this._line?this._context.lineTo(this._x2,this._y2):this._context.moveTo(this._x2,this._y2);break;case 3:this._point=4;default:Es(this,t,e)}this._l01_a=this._l12_a,this._l12_a=this._l23_a,this._l01_2a=this._l12_2a,this._l12_2a=this._l23_2a,this._x0=this._x1,this._x1=this._x2,this._x2=t,this._y0=this._y1,this._y1=this._y2,this._y2=e}},function t(e){function n(t){return e?new Ls(t,e):new Ms(t,0)}return n.alpha=function(e){return t(+e)},n}(.5),Gs.prototype={areaStart:function(){this._line=0},areaEnd:function(){this._line=NaN},lineStart:function(){this._x0=this._x1=this._y0=this._y1=this._t0=NaN,this._point=0},lineEnd:function(){switch(this._point){case 2:this._context.lineTo(this._x1,this._y1);break;case 3:Ts(this,this._t0,Cs(this,this._t0))}(this._line||0!==this._line&&1===this._point)&&this._context.closePath(),this._line=1-this._line},point:function(t,e){var n=NaN;if(e=+e,(t=+t)!==this._x1||e!==this._y1){switch(this._point){case 0:this._point=1,this._line?this._context.lineTo(t,e):this._context.moveTo(t,e);break;case 1:this._point=2;break;case 2:this._point=3,Ts(this,Cs(this,n=$s(this,t,e)),n);break;default:Ts(this,this._t0,n=$s(this,t,e))}this._x0=this._x1,this._x1=t,this._y0=this._y1,this._y1=e,this._t0=n}}},(function(t){this._context=new Us(t)}.prototype=Object.create(Gs.prototype)).point=function(t,e){Gs.prototype.point.call(this,e,t)},Us.prototype={moveTo:function(t,e){this._context.moveTo(e,t)},closePath:function(){this._context.closePath()},lineTo:function(t,e){this._context.lineTo(e,t)},bezierCurveTo:function(t,e,n,r,i,o){this._context.bezierCurveTo(e,t,r,n,o,i)}},Ds.prototype={constructor:Ds,insert:function(t,e){var n,r,i;if(t){if(e.P=t,e.N=t.N,t.N&&(t.N.P=e),t.N=e,t.R){for(t=t.R;t.L;)t=t.L;t.L=e}else t.R=e;n=t}else this._?(t=Ps(this._),e.P=null,e.N=t,t.P=t.L=e,n=t):(e.P=e.N=null,this._=e,n=null);for(e.L=e.R=null,e.U=n,e.C=!0,t=e;n&&n.C;)n===(r=n.U).L?(i=r.R)&&i.C?(n.C=i.C=!1,r.C=!0,t=r):(t===n.R&&(js(this,n),n=(t=n).U),n.C=!1,r.C=!0,zs(this,r)):(i=r.L)&&i.C?(n.C=i.C=!1,r.C=!0,t=r):(t===n.L&&(zs(this,n),n=(t=n).U),n.C=!1,r.C=!0,js(this,r)),n=t.U;this._.C=!1},remove:function(t){t.N&&(t.N.P=t.P),t.P&&(t.P.N=t.N),t.N=t.P=null;var e,n,r,i=t.U,o=t.L,a=t.R;if(n=o?a?Ps(a):o:a,i?i.L===t?i.L=n:i.R=n:this._=n,o&&a?(r=n.C,n.C=t.C,n.L=o,o.U=n,n!==a?(i=n.U,n.U=t.U,t=n.R,i.L=t,n.R=a,a.U=n):(n.U=i,i=n,t=n.R)):(r=t.C,t=n),t&&(t.U=i),!r)if(t&&t.C)t.C=!1;else{do{if(t===this._)break;if(t===i.L){if((e=i.R).C&&(e.C=!1,i.C=!0,js(this,i),e=i.R),e.L&&e.L.C||e.R&&e.R.C){e.R&&e.R.C||(e.L.C=!1,e.C=!0,zs(this,e),e=i.R),e.C=i.C,i.C=e.R.C=!1,js(this,i),t=this._;break}}else if((e=i.L).C&&(e.C=!1,i.C=!0,zs(this,i),e=i.L),e.L&&e.L.C||e.R&&e.R.C){e.L&&e.L.C||(e.R.C=!1,e.C=!0,js(this,e),e=i.L),e.C=i.C,i.C=e.L.C=!1,zs(this,i),t=this._;break}e.C=!0,t=i,i=i.U}while(!t.C);t&&(t.C=!1)}}};const Is=Ds;function Rs(t,e,n,r){var i=[null,null],o=fu.push(i)-1;return i.left=t,i.right=e,n&&qs(i,t,e,n),r&&qs(i,e,t,r),cu[t.index].halfedges.push(o),cu[e.index].halfedges.push(o),i}function Fs(t,e,n){var r=[e,n];return r.left=t,r}function qs(t,e,n,r){t[0]||t[1]?t.left===n?t[1]=r:t[0]=r:(t[0]=r,t.left=e,t.right=n)}function Ys(t,e,n,r,i){var o,a=t[0],s=t[1],u=a[0],c=a[1],h=0,f=1,l=s[0]-u,d=s[1]-c;if(o=e-u,l||!(o>0)){if(o/=l,l<0){if(o<h)return;o<f&&(f=o)}else if(l>0){if(o>f)return;o>h&&(h=o)}if(o=r-u,l||!(o<0)){if(o/=l,l<0){if(o>f)return;o>h&&(h=o)}else if(l>0){if(o<h)return;o<f&&(f=o)}if(o=n-c,d||!(o>0)){if(o/=d,d<0){if(o<h)return;o<f&&(f=o)}else if(d>0){if(o>f)return;o>h&&(h=o)}if(o=i-c,d||!(o<0)){if(o/=d,d<0){if(o>f)return;o>h&&(h=o)}else if(d>0){if(o<h)return;o<f&&(f=o)}return!(h>0||f<1)||(h>0&&(t[0]=[u+h*l,c+h*d]),f<1&&(t[1]=[u+f*l,c+f*d]),!0)}}}}}function Bs(t,e,n,r,i){var o=t[1];if(o)return!0;var a,s,u=t[0],c=t.left,h=t.right,f=c[0],l=c[1],d=h[0],p=h[1],y=(f+d)/2,g=(l+p)/2;if(p===l){if(y<e||y>=r)return;if(f>d){if(u){if(u[1]>=i)return}else u=[y,n];o=[y,i]}else{if(u){if(u[1]<n)return}else u=[y,i];o=[y,n]}}else if(s=g-(a=(f-d)/(p-l))*y,a<-1||a>1)if(f>d){if(u){if(u[1]>=i)return}else u=[(n-s)/a,n];o=[(i-s)/a,i]}else{if(u){if(u[1]<n)return}else u=[(i-s)/a,i];o=[(n-s)/a,n]}else if(l<p){if(u){if(u[0]>=r)return}else u=[e,a*e+s];o=[r,a*r+s]}else{if(u){if(u[0]<e)return}else u=[r,a*r+s];o=[e,a*e+s]}return t[0]=u,t[1]=o,!0}function Ws(t,e){var n=t.site,r=e.left,i=e.right;return n===i&&(i=r,r=n),i?Math.atan2(i[1]-r[1],i[0]-r[0]):(n===r?(r=e[1],i=e[0]):(r=e[0],i=e[1]),Math.atan2(r[0]-i[0],i[1]-r[1]))}function Ks(t,e){return e[+(e.left!==t.site)]}function Hs(t,e){return e[+(e.left===t.site)]}var Xs,Vs=[];function Zs(){Os(this),this.x=this.y=this.arc=this.site=this.cy=null}function Js(t){var e=t.P,n=t.N;if(e&&n){var r=e.site,i=t.site,o=n.site;if(r!==o){var a=i[0],s=i[1],u=r[0]-a,c=r[1]-s,h=o[0]-a,f=o[1]-s,l=2*(u*f-c*h);if(!(l>=-du)){var d=u*u+c*c,p=h*h+f*f,y=(f*d-c*p)/l,g=(u*p-h*d)/l,v=Vs.pop()||new Zs;v.arc=t,v.site=i,v.x=y+a,v.y=(v.cy=g+s)+Math.sqrt(y*y+g*g),t.circle=v;for(var m=null,_=hu._;_;)if(v.y<_.y||v.y===_.y&&v.x<=_.x){if(!_.L){m=_.P;break}_=_.L}else{if(!_.R){m=_;break}_=_.R}hu.insert(m,v),m||(Xs=v)}}}}function Qs(t){var e=t.circle;e&&(e.P||(Xs=e.N),hu.remove(e),Vs.push(e),Os(e),t.circle=null)}var tu=[];function eu(){Os(this),this.edge=this.site=this.circle=null}function nu(t){var e=tu.pop()||new eu;return e.site=t,e}function ru(t){Qs(t),uu.remove(t),tu.push(t),Os(t)}function iu(t){var e=t.circle,n=e.x,r=e.cy,i=[n,r],o=t.P,a=t.N,s=[t];ru(t);for(var u=o;u.circle&&Math.abs(n-u.circle.x)<lu&&Math.abs(r-u.circle.cy)<lu;)o=u.P,s.unshift(u),ru(u),u=o;s.unshift(u),Qs(u);for(var c=a;c.circle&&Math.abs(n-c.circle.x)<lu&&Math.abs(r-c.circle.cy)<lu;)a=c.N,s.push(c),ru(c),c=a;s.push(c),Qs(c);var h,f=s.length;for(h=1;h<f;++h)c=s[h],u=s[h-1],qs(c.edge,u.site,c.site,i);u=s[0],(c=s[f-1]).edge=Rs(u.site,c.site,null,i),Js(u),Js(c)}function ou(t){for(var e,n,r,i,o=t[0],a=t[1],s=uu._;s;)if((r=au(s,a)-o)>lu)s=s.L;else{if(!((i=o-su(s,a))>lu)){r>-lu?(e=s.P,n=s):i>-lu?(e=s,n=s.N):e=n=s;break}if(!s.R){e=s;break}s=s.R}!function(t){cu[t.index]={site:t,halfedges:[]}}(t);var u=nu(t);if(uu.insert(e,u),e||n){if(e===n)return Qs(e),n=nu(e.site),uu.insert(u,n),u.edge=n.edge=Rs(e.site,u.site),Js(e),void Js(n);if(n){Qs(e),Qs(n);var c=e.site,h=c[0],f=c[1],l=t[0]-h,d=t[1]-f,p=n.site,y=p[0]-h,g=p[1]-f,v=2*(l*g-d*y),m=l*l+d*d,_=y*y+g*g,b=[(g*m-d*_)/v+h,(l*_-y*m)/v+f];qs(n.edge,c,p,b),u.edge=Rs(c,t,null,b),n.edge=Rs(t,p,null,b),Js(e),Js(n)}else u.edge=Rs(e.site,u.site)}}function au(t,e){var n=t.site,r=n[0],i=n[1],o=i-e;if(!o)return r;var a=t.P;if(!a)return-1/0;var s=(n=a.site)[0],u=n[1],c=u-e;if(!c)return s;var h=s-r,f=1/o-1/c,l=h/c;return f?(-l+Math.sqrt(l*l-2*f*(h*h/(-2*c)-u+c/2+i-o/2)))/f+r:(r+s)/2}function su(t,e){var n=t.N;if(n)return au(n,e);var r=t.site;return r[1]===e?r[0]:1/0}var uu,cu,hu,fu,lu=1e-6,du=1e-12;function pu(t,e){return e[1]-t[1]||e[0]-t[0]}function yu(t,e){var n,r,i,o=t.sort(pu).pop();for(fu=[],cu=new Array(t.length),uu=new Is,hu=new Is;;)if(i=Xs,o&&(!i||o[1]<i.y||o[1]===i.y&&o[0]<i.x))o[0]===n&&o[1]===r||(ou(o),n=o[0],r=o[1]),o=t.pop();else{if(!i)break;iu(i.arc)}if(function(){for(var t,e,n,r,i=0,o=cu.length;i<o;++i)if((t=cu[i])&&(r=(e=t.halfedges).length)){var a=new Array(r),s=new Array(r);for(n=0;n<r;++n)a[n]=n,s[n]=Ws(t,fu[e[n]]);for(a.sort((function(t,e){return s[e]-s[t]})),n=0;n<r;++n)s[n]=e[a[n]];for(n=0;n<r;++n)e[n]=s[n]}}(),e){var a=+e[0][0],s=+e[0][1],u=+e[1][0],c=+e[1][1];!function(t,e,n,r){for(var i,o=fu.length;o--;)Bs(i=fu[o],t,e,n,r)&&Ys(i,t,e,n,r)&&(Math.abs(i[0][0]-i[1][0])>lu||Math.abs(i[0][1]-i[1][1])>lu)||delete fu[o]}(a,s,u,c),function(t,e,n,r){var i,o,a,s,u,c,h,f,l,d,p,y,g=cu.length,v=!0;for(i=0;i<g;++i)if(o=cu[i]){for(a=o.site,s=(u=o.halfedges).length;s--;)fu[u[s]]||u.splice(s,1);for(s=0,c=u.length;s<c;)p=(d=Hs(o,fu[u[s]]))[0],y=d[1],f=(h=Ks(o,fu[u[++s%c]]))[0],l=h[1],(Math.abs(p-f)>lu||Math.abs(y-l)>lu)&&(u.splice(s,0,fu.push(Fs(a,d,Math.abs(p-t)<lu&&r-y>lu?[t,Math.abs(f-t)<lu?l:r]:Math.abs(y-r)<lu&&n-p>lu?[Math.abs(l-r)<lu?f:n,r]:Math.abs(p-n)<lu&&y-e>lu?[n,Math.abs(f-n)<lu?l:e]:Math.abs(y-e)<lu&&p-t>lu?[Math.abs(l-e)<lu?f:t,e]:null))-1),++c);c&&(v=!1)}if(v){var m,_,b,w=1/0;for(i=0,v=null;i<g;++i)(o=cu[i])&&(b=(m=(a=o.site)[0]-t)*m+(_=a[1]-e)*_)<w&&(w=b,v=o);if(v){var x=[t,e],k=[t,r],M=[n,r],E=[n,e];v.halfedges.push(fu.push(Fs(a=v.site,x,k))-1,fu.push(Fs(a,k,M))-1,fu.push(Fs(a,M,E))-1,fu.push(Fs(a,E,x))-1)}}for(i=0;i<g;++i)(o=cu[i])&&(o.halfedges.length||delete cu[i])}(a,s,u,c)}this.edges=fu,this.cells=cu,uu=hu=fu=cu=null}function gu(t){return function(){return t}}function vu(t,e,n){this.target=t,this.type=e,this.transform=n}function mu(t,e,n){this.k=t,this.x=e,this.y=n}yu.prototype={constructor:yu,polygons:function(){var t=this.edges;return this.cells.map((function(e){var n=e.halfedges.map((function(n){return Ks(e,t[n])}));return n.data=e.site.data,n}))},triangles:function(){var t=[],e=this.edges;return this.cells.forEach((function(n,r){if(o=(i=n.halfedges).length)for(var i,o,a,s=n.site,u=-1,c=e[i[o-1]],h=c.left===s?c.right:c.left;++u<o;)a=h,h=(c=e[i[u]]).left===s?c.right:c.left,a&&h&&r<a.index&&r<h.index&&(l=a,d=h,((f=s)[0]-d[0])*(l[1]-f[1])-(f[0]-l[0])*(d[1]-f[1])<0)&&t.push([s.data,a.data,h.data]);var f,l,d})),t},links:function(){return this.edges.filter((function(t){return t.right})).map((function(t){return{source:t.left.data,target:t.right.data}}))},find:function(t,e,n){for(var r,i,o=this,a=o._found||0,s=o.cells.length;!(i=o.cells[a]);)if(++a>=s)return null;var u=t-i.site[0],c=e-i.site[1],h=u*u+c*c;do{i=o.cells[r=a],a=null,i.halfedges.forEach((function(n){var r=o.edges[n],s=r.left;if(s!==i.site&&s||(s=r.right)){var u=t-s[0],c=e-s[1],f=u*u+c*c;f<h&&(h=f,a=s.index)}}))}while(null!==a);return o._found=r,null==n||h<=n*n?i.site:null}},mu.prototype={constructor:mu,scale:function(t){return 1===t?this:new mu(this.k*t,this.x,this.y)},translate:function(t,e){return 0===t&0===e?this:new mu(this.k,this.x+this.k*t,this.y+this.k*e)},apply:function(t){return[t[0]*this.k+this.x,t[1]*this.k+this.y]},applyX:function(t){return t*this.k+this.x},applyY:function(t){return t*this.k+this.y},invert:function(t){return[(t[0]-this.x)/this.k,(t[1]-this.y)/this.k]},invertX:function(t){return(t-this.x)/this.k},invertY:function(t){return(t-this.y)/this.k},rescaleX:function(t){return t.copy().domain(t.range().map(this.invertX,this).map(t.invert,t))},rescaleY:function(t){return t.copy().domain(t.range().map(this.invertY,this).map(t.invert,t))},toString:function(){return"translate("+this.x+","+this.y+") scale("+this.k+")"}};var _u=new mu(1,0,0);function bu(){vt.stopImmediatePropagation()}function wu(){vt.preventDefault(),vt.stopImmediatePropagation()}function xu(){return!vt.button}function ku(){var t,e,n=this;return n instanceof SVGElement?(t=(n=n.ownerSVGElement||n).width.baseVal.value,e=n.height.baseVal.value):(t=n.clientWidth,e=n.clientHeight),[[0,0],[t,e]]}function Mu(){return this.__zoom||_u}function Eu(){return-vt.deltaY*(vt.deltaMode?120:1)/500}function Au(){return"ontouchstart"in this}function Su(t,e,n){var r=t.invertX(e[0][0])-n[0][0],i=t.invertX(e[1][0])-n[1][0],o=t.invertY(e[0][1])-n[0][1],a=t.invertY(e[1][1])-n[1][1];return t.translate(i>r?(r+i)/2:Math.min(0,r)||Math.max(0,i),a>o?(o+a)/2:Math.min(0,o)||Math.max(0,a))}mu.prototype;var Lu="undefined"!=typeof navigator&&navigator.userAgent.toLowerCase().indexOf("firefox")>0;function Nu(t,e,n,r){t.addEventListener?t.addEventListener(e,n,r):t.attachEvent&&t.attachEvent("on".concat(e),(function(){n(window.event)}))}function $u(t,e){for(var n=e.slice(0,e.length-1),r=0;r<n.length;r++)n[r]=t[n[r].toLowerCase()];return n}function Cu(t){"string"!=typeof t&&(t="");for(var e=(t=t.replace(/\s/g,"")).split(","),n=e.lastIndexOf("");n>=0;)e[n-1]+=",",e.splice(n,1),n=e.lastIndexOf("");return e}for(var Tu={backspace:8,"⌫":8,tab:9,clear:12,enter:13,"↩":13,return:13,esc:27,escape:27,space:32,left:37,up:38,right:39,down:40,del:46,delete:46,ins:45,insert:45,home:36,end:35,pageup:33,pagedown:34,capslock:20,num_0:96,num_1:97,num_2:98,num_3:99,num_4:100,num_5:101,num_6:102,num_7:103,num_8:104,num_9:105,num_multiply:106,num_add:107,num_enter:108,num_subtract:109,num_decimal:110,num_divide:111,"⇪":20,",":188,".":190,"/":191,"`":192,"-":Lu?173:189,"=":Lu?61:187,";":Lu?59:186,"'":222,"[":219,"]":221,"\\":220},Gu={"⇧":16,shift:16,"⌥":18,alt:18,option:18,"⌃":17,ctrl:17,control:17,"⌘":91,cmd:91,command:91},Uu={16:"shiftKey",18:"altKey",17:"ctrlKey",91:"metaKey",shiftKey:16,ctrlKey:17,altKey:18,metaKey:91},Du={16:!1,18:!1,17:!1,91:!1},Ou={},ju=1;ju<20;ju++)Tu["f".concat(ju)]=111+ju;var zu=[],Pu=!1,Iu="all",Ru=[],Fu=function(t){return Tu[t.toLowerCase()]||Gu[t.toLowerCase()]||t.toUpperCase().charCodeAt(0)};function qu(t){Iu=t||"all"}function Yu(){return Iu||"all"}var Bu=function(t){var e=t.key,n=t.scope,r=t.method,i=t.splitKey,o=void 0===i?"+":i;Cu(e).forEach((function(t){var e=t.split(o),i=e.length,a=e[i-1],s="*"===a?"*":Fu(a);if(Ou[s]){n||(n=Yu());var u=i>1?$u(Gu,e):[];Ou[s]=Ou[s].filter((function(t){return!((!r||t.method===r)&&t.scope===n&&function(t,e){for(var n=t.length>=e.length?t:e,r=t.length>=e.length?e:t,i=!0,o=0;o<n.length;o++)-1===r.indexOf(n[o])&&(i=!1);return i}(t.mods,u))}))}}))};function Wu(t,e,n,r){var i;if(e.element===r&&(e.scope===n||"all"===e.scope)){for(var o in i=e.mods.length>0,Du)Object.prototype.hasOwnProperty.call(Du,o)&&(!Du[o]&&e.mods.indexOf(+o)>-1||Du[o]&&-1===e.mods.indexOf(+o))&&(i=!1);(0!==e.mods.length||Du[16]||Du[18]||Du[17]||Du[91])&&!i&&"*"!==e.shortcut||(e.keys=[],e.keys=e.keys.concat(zu),!1===e.method(t,e)&&(t.preventDefault?t.preventDefault():t.returnValue=!1,t.stopPropagation&&t.stopPropagation(),t.cancelBubble&&(t.cancelBubble=!0)))}}function Ku(t,e){var n=Ou["*"],r=t.keyCode||t.which||t.charCode;if(Hu.filter.call(this,t)){if(93!==r&&224!==r||(r=91),-1===zu.indexOf(r)&&229!==r&&zu.push(r),["ctrlKey","altKey","shiftKey","metaKey"].forEach((function(e){var n=Uu[e];t[e]&&-1===zu.indexOf(n)?zu.push(n):!t[e]&&zu.indexOf(n)>-1?zu.splice(zu.indexOf(n),1):"metaKey"===e&&t[e]&&3===zu.length&&(t.ctrlKey||t.shiftKey||t.altKey||(zu=zu.slice(zu.indexOf(n))))})),r in Du){for(var i in Du[r]=!0,Gu)Gu[i]===r&&(Hu[i]=!0);if(!n)return}for(var o in Du)Object.prototype.hasOwnProperty.call(Du,o)&&(Du[o]=t[Uu[o]]);t.getModifierState&&(!t.altKey||t.ctrlKey)&&t.getModifierState("AltGraph")&&(-1===zu.indexOf(17)&&zu.push(17),-1===zu.indexOf(18)&&zu.push(18),Du[17]=!0,Du[18]=!0);var a=Yu();if(n)for(var s=0;s<n.length;s++)n[s].scope===a&&("keydown"===t.type&&n[s].keydown||"keyup"===t.type&&n[s].keyup)&&Wu(t,n[s],a,e);if(r in Ou)for(var u=0;u<Ou[r].length;u++)if(("keydown"===t.type&&Ou[r][u].keydown||"keyup"===t.type&&Ou[r][u].keyup)&&Ou[r][u].key){for(var c=Ou[r][u],h=c.splitKey,f=c.key.split(h),l=[],d=0;d<f.length;d++)l.push(Fu(f[d]));l.sort().join("")===zu.sort().join("")&&Wu(t,c,a,e)}}}function Hu(t,e,n){zu=[];var r=Cu(t),i=[],o="all",a=document,s=0,u=!1,c=!0,h="+",f=!1;for(void 0===n&&"function"==typeof e&&(n=e),"[object Object]"===Object.prototype.toString.call(e)&&(e.scope&&(o=e.scope),e.element&&(a=e.element),e.keyup&&(u=e.keyup),void 0!==e.keydown&&(c=e.keydown),void 0!==e.capture&&(f=e.capture),"string"==typeof e.splitKey&&(h=e.splitKey)),"string"==typeof e&&(o=e);s<r.length;s++)i=[],(t=r[s].split(h)).length>1&&(i=$u(Gu,t)),(t="*"===(t=t[t.length-1])?"*":Fu(t))in Ou||(Ou[t]=[]),Ou[t].push({keyup:u,keydown:c,scope:o,mods:i,shortcut:r[s],method:n,key:r[s],splitKey:h,element:a});void 0!==a&&!function(t){return Ru.indexOf(t)>-1}(a)&&window&&(Ru.push(a),Nu(a,"keydown",(function(t){Ku(t,a)}),f),Pu||(Pu=!0,Nu(window,"focus",(function(){zu=[]}),f)),Nu(a,"keyup",(function(t){Ku(t,a),function(t){var e=t.keyCode||t.which||t.charCode,n=zu.indexOf(e);if(n>=0&&zu.splice(n,1),t.key&&"meta"===t.key.toLowerCase()&&zu.splice(0,zu.length),93!==e&&224!==e||(e=91),e in Du)for(var r in Du[e]=!1,Gu)Gu[r]===e&&(Hu[r]=!1)}(t)}),f))}var Xu={getPressedKeyString:function(){return zu.map((function(t){return e=t,Object.keys(Tu).find((function(t){return Tu[t]===e}))||function(t){return Object.keys(Gu).find((function(e){return Gu[e]===t}))}(t)||String.fromCharCode(t);var e}))},setScope:qu,getScope:Yu,deleteScope:function(t,e){var n,r;for(var i in t||(t=Yu()),Ou)if(Object.prototype.hasOwnProperty.call(Ou,i))for(n=Ou[i],r=0;r<n.length;)n[r].scope===t?n.splice(r,1):r++;Yu()===t&&qu(e||"all")},getPressedKeyCodes:function(){return zu.slice(0)},isPressed:function(t){return"string"==typeof t&&(t=Fu(t)),-1!==zu.indexOf(t)},filter:function(t){var e=t.target||t.srcElement,n=e.tagName,r=!0;return!e.isContentEditable&&("INPUT"!==n&&"TEXTAREA"!==n&&"SELECT"!==n||e.readOnly)||(r=!1),r},trigger:function(t){var e=arguments.length>1&&void 0!==arguments[1]?arguments[1]:"all";Object.keys(Ou).forEach((function(n){Ou[n].filter((function(n){return n.scope===e&&n.shortcut===t})).forEach((function(t){t&&t.method&&t.method()}))}))},unbind:function(t){if(void 0===t)Object.keys(Ou).forEach((function(t){return delete Ou[t]}));else if(Array.isArray(t))t.forEach((function(t){t.key&&Bu(t)}));else if("object"==typeof t)t.key&&Bu(t);else if("string"==typeof t){for(var e=arguments.length,n=new Array(e>1?e-1:0),r=1;r<e;r++)n[r-1]=arguments[r];var i=n[0],o=n[1];"function"==typeof i&&(o=i,i=""),Bu({key:t,scope:i,method:o,splitKey:"+"})}},keyMap:Tu,modifier:Gu,modifierMap:Uu};for(var Vu in Xu)Object.prototype.hasOwnProperty.call(Xu,Vu)&&(Hu[Vu]=Xu[Vu]);if("undefined"!=typeof window){var Zu=window.hotkeys;Hu.noConflict=function(t){return t&&window.hotkeys===Hu&&(window.hotkeys=Zu),Hu},window.hotkeys=Hu}const Ju={notFiltered:0,filteredByTimeline:1,filteredByFocus:2,filteredByTag:3,filteredByType:4};var Qu=[];data.nodes=data.nodes.map((function(t){return Qu.push(t.id),t.hidden=Ju.notFiltered,t.isolated=!1,t.highlighted=!1,t}));var tc=$t("#graph-canvas"),ec=tc.append("svg"),nc=tc.node().getBoundingClientRect(),rc=nc.width,ic=nc.height;ec.attr("viewBox",[0,0,rc,ic]).attr("preserveAspectRatio","xMinYMin meet");var oc=function(t){var e,n=1,r=.001,i=1-Math.pow(r,1/300),o=0,a=.6,s=Gr(),u=On(h),c=g("tick","end");function h(){f(),c.call("tick",e),n<r&&(u.stop(),c.call("end",e))}function f(){var e,r,u=t.length;for(n+=(o-n)*i,s.each((function(t){t(n)})),e=0;e<u;++e)null==(r=t[e]).fx?r.x+=r.vx*=a:(r.x=r.fx,r.vx=0),null==r.fy?r.y+=r.vy*=a:(r.y=r.fy,r.vy=0)}function l(){for(var e,n=0,r=t.length;n<r;++n){if((e=t[n]).index=n,isNaN(e.x)||isNaN(e.y)){var i=10*Math.sqrt(n),o=n*oi;e.x=i*Math.cos(o),e.y=i*Math.sin(o)}(isNaN(e.vx)||isNaN(e.vy))&&(e.vx=e.vy=0)}}function d(e){return e.initialize&&e.initialize(t),e}return null==t&&(t=[]),l(),e={tick:f,restart:function(){return u.restart(h),e},stop:function(){return u.stop(),e},nodes:function(n){return arguments.length?(t=n,l(),s.each(d),e):t},alpha:function(t){return arguments.length?(n=+t,e):n},alphaMin:function(t){return arguments.length?(r=+t,e):r},alphaDecay:function(t){return arguments.length?(i=+t,e):+i},alphaTarget:function(t){return arguments.length?(o=+t,e):o},velocityDecay:function(t){return arguments.length?(a=1-t,e):1-a},force:function(t,n){return arguments.length>1?(null==n?s.remove(t):s.set(t,d(n)),e):s.get(t)},find:function(e,n,r){var i,o,a,s,u,c=0,h=t.length;for(null==r?r=1/0:r*=r,c=0;c<h;++c)(a=(i=e-(s=t[c]).x)*i+(o=n-s.y)*o)<r&&(u=s,r=a);return u},on:function(t,n){return arguments.length>1?(c.on(t,n),e):c.on(t)}}}(data.nodes).force("link",function(t){var e,n,r,i,o,a=ti,s=function(t){return 1/Math.min(i[t.source.index],i[t.target.index])},u=Jr(30),c=1;function h(r){for(var i=0,a=t.length;i<c;++i)for(var s,u,h,f,l,d,p,y=0;y<a;++y)u=(s=t[y]).source,f=(h=s.target).x+h.vx-u.x-u.vx||Qr(),l=h.y+h.vy-u.y-u.vy||Qr(),f*=d=((d=Math.sqrt(f*f+l*l))-n[y])/d*r*e[y],l*=d,h.vx-=f*(p=o[y]),h.vy-=l*p,u.vx+=f*(p=1-p),u.vy+=l*p}function f(){if(r){var s,u,c=r.length,h=t.length,f=Gr(r,a);for(s=0,i=new Array(c);s<h;++s)(u=t[s]).index=s,"object"!=typeof u.source&&(u.source=ei(f,u.source)),"object"!=typeof u.target&&(u.target=ei(f,u.target)),i[u.source.index]=(i[u.source.index]||0)+1,i[u.target.index]=(i[u.target.index]||0)+1;for(s=0,o=new Array(h);s<h;++s)u=t[s],o[s]=i[u.source.index]/(i[u.source.index]+i[u.target.index]);e=new Array(h),l(),n=new Array(h),d()}}function l(){if(r)for(var n=0,i=t.length;n<i;++n)e[n]=+s(t[n],n,t)}function d(){if(r)for(var e=0,i=t.length;e<i;++e)n[e]=+u(t[e],e,t)}return null==t&&(t=[]),h.initialize=function(t){r=t,f()},h.links=function(e){return arguments.length?(t=e,f(),h):t},h.id=function(t){return arguments.length?(a=t,h):a},h.iterations=function(t){return arguments.length?(c=+t,h):c},h.strength=function(t){return arguments.length?(s="function"==typeof t?t:Jr(+t),l(),h):s},h.distance=function(t){return arguments.length?(u="function"==typeof t?t:Jr(+t),d(),h):u},h}(data.links).id((function(t){return t.id}))).force("charge",function(){var t,e,n,r,i=Jr(-30),o=1,a=1/0,s=.81;function u(r){var i,o=t.length,a=Hr(t,ni,ri).visitAfter(h);for(n=r,i=0;i<o;++i)e=t[i],a.visit(f)}function c(){if(t){var e,n,o=t.length;for(r=new Array(o),e=0;e<o;++e)n=t[e],r[n.index]=+i(n,e,t)}}function h(t){var e,n,i,o,a,s=0,u=0;if(t.length){for(i=o=a=0;a<4;++a)(e=t[a])&&(n=Math.abs(e.value))&&(s+=e.value,u+=n,i+=n*e.x,o+=n*e.y);t.x=i/u,t.y=o/u}else{(e=t).x=e.data.x,e.y=e.data.y;do{s+=r[e.data.index]}while(e=e.next)}t.value=s}function f(t,i,u,c){if(!t.value)return!0;var h=t.x-e.x,f=t.y-e.y,l=c-i,d=h*h+f*f;if(l*l/s<d)return d<a&&(0===h&&(d+=(h=Qr())*h),0===f&&(d+=(f=Qr())*f),d<o&&(d=Math.sqrt(o*d)),e.vx+=h*t.value*n/d,e.vy+=f*t.value*n/d),!0;if(!(t.length||d>=a)){(t.data!==e||t.next)&&(0===h&&(d+=(h=Qr())*h),0===f&&(d+=(f=Qr())*f),d<o&&(d=Math.sqrt(o*d)));do{t.data!==e&&(l=r[t.data.index]*n/d,e.vx+=h*l,e.vy+=f*l)}while(t=t.next)}}return u.initialize=function(e){t=e,c()},u.strength=function(t){return arguments.length?(i="function"==typeof t?t:Jr(+t),c(),u):i},u.distanceMin=function(t){return arguments.length?(o=t*t,u):Math.sqrt(o)},u.distanceMax=function(t){return arguments.length?(a=t*t,u):Math.sqrt(a)},u.theta=function(t){return arguments.length?(s=t*t,u):Math.sqrt(s)},u}()).force("center",function(t,e){var n;function r(){var r,i,o=n.length,a=0,s=0;for(r=0;r<o;++r)a+=(i=n[r]).x,s+=i.y;for(a=a/o-t,s=s/o-e,r=0;r<o;++r)(i=n[r]).x-=a,i.y-=s}return null==t&&(t=0),null==e&&(e=0),r.initialize=function(t){n=t},r.x=function(e){return arguments.length?(t=+e,r):t},r.y=function(t){return arguments.length?(e=+t,r):e},r}()).force("forceX",function(t){var e,n,r,i=Jr(.1);function o(t){for(var i,o=0,a=e.length;o<a;++o)(i=e[o]).vx+=(r[o]-i.x)*n[o]*t}function a(){if(e){var o,a=e.length;for(n=new Array(a),r=new Array(a),o=0;o<a;++o)n[o]=isNaN(r[o]=+t(e[o],o,e))?0:+i(e[o],o,e)}}return"function"!=typeof t&&(t=Jr(null==t?0:+t)),o.initialize=function(t){e=t,a()},o.strength=function(t){return arguments.length?(i="function"==typeof t?t:Jr(+t),a(),o):i},o.x=function(e){return arguments.length?(t="function"==typeof e?e:Jr(+e),a(),o):t},o}()).force("forceY",function(t){var e,n,r,i=Jr(.1);function o(t){for(var i,o=0,a=e.length;o<a;++o)(i=e[o]).vy+=(r[o]-i.y)*n[o]*t}function a(){if(e){var o,a=e.length;for(n=new Array(a),r=new Array(a),o=0;o<a;++o)n[o]=isNaN(r[o]=+t(e[o],o,e))?0:+i(e[o],o,e)}}return"function"!=typeof t&&(t=Jr(null==t?0:+t)),o.initialize=function(t){e=t,a()},o.strength=function(t){return arguments.length?(i="function"==typeof t?t:Jr(+t),a(),o):i},o.y=function(e){return arguments.length?(t="function"==typeof e?e:Jr(+e),a(),o):t},o}());oc.force("center").x(.5*rc).y(.5*ic),window.updateForces=function(){oc.force("charge").strength(-Math.abs(graphProperties.attraction_force)).distanceMax(graphProperties.attraction_distance_max),oc.force("forceX").strength(graphProperties.attraction_vertical),oc.force("forceY").strength(graphProperties.attraction_horizontal),oc.alpha(1).restart()},updateForces(),Hu("space",(function(t){t.preventDefault(),updateForces()})),oc.on("tick",(function(){ac.links.attr("x1",(function(t){return t.source.x})).attr("y1",(function(t){return t.source.y})).attr("x2",(function(t){return t.target.x})).attr("y2",(function(t){return t.target.y})),ac.nodes.attr("transform",(function(t){return"translate("+t.x+","+t.y+")"})),$t("#load-bar-value").style("flex-basis",100*oc.alpha()+"%")}));var ac={};function sc(t){var e=ac.nodes.filter((function(t){return t.hidden===Ju.notFiltered})).data().map((function(t){return t.id}));return{nodes:ac.nodes.filter((function(e){return t.includes(e.id)})),links:ac.links.filter((function(n){return!(!t.includes(n.source.id)&&!t.includes(n.target.id)||!e.includes(n.source.id)||!e.includes(n.target.id))}))}}function uc(t,e){var n=[],r=[];Qu.forEach((function(e){t.includes(e)?r.push(e):n.push(e)})),function(t,e){if(hideNodeNetwork(t),function(t){var e,n=pc(t);try{for(n.s();!(e=n.n()).done;){var r=e.value;mc.querySelectorAll('[data-index="'+r+'"]').forEach((function(t){t.style.display="none"}))}}catch(t){n.e(t)}finally{n.f()}}(t),void 0===e)throw new Error("Need priority");ac.nodes.data().map((function(n){var r=n.id,i=n.hidden;return t.includes(r)&&i<=e&&(n.hidden=e),n}))}(n,e),cc(r,e)}function cc(t){var e=arguments.length>1&&void 0!==arguments[1]?arguments[1]:Ju.notFiltered,n=[];ac.nodes.data().map((function(r){var i=r.id,o=r.hidden;return t.includes(i)&&o<=e&&(n.push(i),r.hidden=Ju.notFiltered),r})),displayNodeNetwork(n),function(t){var e,n=pc(t=t.filter((function(t){if(dc.find((function(e){return e.id===t})).hidden===Ju.notFiltered)return!0})));try{for(n.s();!(e=n.n()).done;){var r=e.value;mc.querySelectorAll('[data-index="'+r+'"]').forEach((function(t){t.style.display=null}))}}catch(t){n.e(t)}finally{n.f()}}(n)}function hc(){var t=arguments.length>0&&void 0!==arguments[0]?arguments[0]:Ju.notFiltered;cc(Qu,t)}function fc(){if(0!==u.highlightedNodes.length){var t=sc(u.highlightedNodes),e=t.nodes,n=t.links;e.selectAll("circle").style("stroke",null),n.style("stroke",null),u.highlightedNodes=[]}}function lc(){var t=u.position,e=t.x,n=t.y,r=t.zoom,i=[-e,-n,rc/r,ic/r].join(" ");ec.attr("viewBox",i)}ac.links=ec.append("g").selectAll("line").data(data.links).enter().append("line").attr("stroke",(function(t){return t.color})).attr("title",(function(t){return t.title})).attr("data-source",(function(t){return t.source.id})).attr("data-target",(function(t){return t.target.id})).attr("stroke-dasharray",(function(t){return("dash"===t.shape.stroke||"dotted"===t.shape.stroke)&&t.shape.dashInterval})).attr("filter",(function(t){return"double"===t.shape.stroke&&"url(#double)"})),!0===graphProperties.graph_arrows&&ac.links.attr("marker-end","url(#arrow)"),ac.nodes=ec.append("g").selectAll("g").data(data.nodes).enter().append("g").attr("data-node",(function(t){return t.id})).append("a").attr("href",(function(t){return"#"+t.id})),ac.circles=ac.nodes.append("circle").attr("r",(function(t){return t.size})).attr("fill",(function(t){return t.fill})).attr("stroke",(function(t){return t.colorStroke})).attr("stroke-width",(function(t){return t.strokeWidth})).call(function(){var t,e,n,r,i=qt,o=Yt,a=Bt,s=Wt,u={},c=g("start","drag","end"),h=0,f=0;function l(t){t.on("mousedown.drag",d).filter(s).on("touchstart.drag",v).on("touchmove.drag",m).on("touchend.drag touchcancel.drag",_).style("touch-action","none").style("-webkit-tap-highlight-color","rgba(0,0,0,0)")}function d(){if(!r&&i.apply(this,arguments)){var a=b("mouse",o.apply(this,arguments),Dt,this,arguments);a&&($t(vt.view).on("mousemove.drag",p,!0).on("mouseup.drag",y,!0),Pt(vt.view),jt(),n=!1,t=vt.clientX,e=vt.clientY,a("start"))}}function p(){if(zt(),!n){var r=vt.clientX-t,i=vt.clientY-e;n=r*r+i*i>f}u.mouse("drag")}function y(){$t(vt.view).on("mousemove.drag mouseup.drag",null),It(vt.view,n),zt(),u.mouse("end")}function v(){if(i.apply(this,arguments)){var t,e,n=vt.changedTouches,r=o.apply(this,arguments),a=n.length;for(t=0;t<a;++t)(e=b(n[t].identifier,r,Ot,this,arguments))&&(jt(),e("start"))}}function m(){var t,e,n=vt.changedTouches,r=n.length;for(t=0;t<r;++t)(e=u[n[t].identifier])&&(zt(),e("drag"))}function _(){var t,e,n=vt.changedTouches,i=n.length;for(r&&clearTimeout(r),r=setTimeout((function(){r=null}),500),t=0;t<i;++t)(e=u[n[t].identifier])&&(jt(),e("end"))}function b(t,e,n,r,i){var o,s,f,d=n(e,t),p=c.copy();if(xt(new Ft(l,"beforestart",o,t,h,d[0],d[1],0,0,p),(function(){return null!=(vt.subject=o=a.apply(r,i))&&(s=o.x-d[0]||0,f=o.y-d[1]||0,!0)})))return function a(c){var y,g=d;switch(c){case"start":u[t]=a,y=h++;break;case"end":delete u[t],--h;case"drag":d=n(e,t),y=h}xt(new Ft(l,c,o,t,y,d[0]+s,d[1]+f,d[0]-g[0],d[1]-g[1],p),p.apply,p,[c,r,i])}}return l.filter=function(t){return arguments.length?(i="function"==typeof t?t:Rt(!!t),l):i},l.container=function(t){return arguments.length?(o="function"==typeof t?t:Rt(t),l):o},l.subject=function(t){return arguments.length?(a="function"==typeof t?t:Rt(t),l):a},l.touchable=function(t){return arguments.length?(s="function"==typeof t?t:Rt(!!t),l):s},l.on=function(){var t=c.on.apply(c,arguments);return t===c?l:t},l.clickDistance=function(t){return arguments.length?(f=(t=+t)*t,l):Math.sqrt(f)},l}().on("start",(function(t){vt.active||oc.alphaTarget(.3).restart(),t.fx=t.x,t.fy=t.y})).on("drag",(function(t){t.fx=vt.x,t.fy=vt.y})).on("end",(function(t){vt.active||oc.alphaTarget(1e-4),t.fx=null,t.fy=null}))).on("mouseenter",(function(t){if(graphProperties.graph_highlight_on_hover){var e=[t.id],n=ac.links.filter((function(n){return n.source.id!==t.id&&n.target.id!==t.id||(e.push(n.source.id,n.target.id),!1)})),r=ac.nodes.filter((function(t){return!e.includes(t.id)})),i=ac.links.filter((function(e){return e.source.id===t.id||e.target.id===t.id}));ac.nodes.filter((function(t){return!!e.includes(t.id)})).selectAll("circle").attr("stroke",(function(t){return t.highlight})),i.attr("stroke",(function(t){return t.colorHighlight})),r.attr("opacity",.5),n.attr("stroke-opacity",.5)}})).on("mouseout",(function(){graphProperties.graph_highlight_on_hover&&(ac.nodes.selectAll("circle").attr("stroke",(function(t){return t.colorStroke})),ac.links.attr("stroke",(function(t){return t.color})),ac.nodes.attr("opacity",1),ac.links.attr("stroke-opacity",1))})),ac.labels=ac.nodes.append("text").each((function(t){for(var e=t.label.split(" "),n=$t(this),r="",i=0;i<e.length;i++)(r+=e[i]+" ").length<25&&i!==e.length-1||(n.append("tspan").attr("x",0).attr("dy","1.2em").text(r.slice(0,-1)),r="")})).attr("font-size",graphProperties.graph_text_size).attr("x",0).attr("y",(function(t){return t.size})).attr("dominant-baseline","middle").attr("text-anchor","middle"),window.zoomToNode=function(t){var e=ac.nodes.filter((function(e){return e.id===t})).datum(),n=e.x,r=e.y;n=-Math.abs(n-200),r=-Math.abs(r-300),u.position={x:n,y:r,zoom:2},lc()},Hu("c",(function(t){t.preventDefault(),u.openedRecordId&&zoomToNode(Number(u.openedRecordId))})),window.hideNodeNetwork=function(t){var e=sc(t),n=e.nodes,r=e.links;n.style("display","none"),r.style("display","none")},window.displayNodeNetwork=function(t){var e=sc(t),n=e.nodes,r=e.links;n.style("display",null),r.style("display",null)},window.linksDisplayToggle=function(t){t?ac.links.style("display",null):ac.links.style("display","none")},window.labelDisplayToggle=function(t){t?ac.labels.style("display",null):ac.labels.style("display","none")},window.labelHighlight=function(t){ac.nodes.data().map((function(e){return t.includes(e.id)&&(e.highlighted=!0),e})),ac.nodes.filter((function(e){return t.includes(e.id)})).select("text").style("fill","var(--highlight)")},window.updateFontsize=function(){ac.labels.attr("font-size",graphProperties.text_size)},window.labelUnlight=function(t){ac.nodes.data().map((function(e){return t.includes(e.id)&&(e.highlighted=!1),e})),ac.nodes.filter((function(e){return t.includes(e.id)})).select("text").style("fill",null)},window.labelUnlightAll=function(){data.nodes=data.nodes.map((function(t){return t.highlighted=!1,t})),ac.labels.style("fill",null)};var dc=ac.nodes.data();function pc(t,e){var n="undefined"!=typeof Symbol&&t[Symbol.iterator]||t["@@iterator"];if(!n){if(Array.isArray(t)||(n=yc(t))||e&&t&&"number"==typeof t.length){n&&(t=n);var r=0,i=function(){};return{s:i,n:function(){return r>=t.length?{done:!0}:{done:!1,value:t[r++]}},e:function(t){throw t},f:i}}throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.")}var o,a=!0,s=!1;return{s:function(){n=n.call(t)},n:function(){var t=n.next();return a=t.done,t},e:function(t){s=!0,o=t},f:function(){try{a||null==n.return||n.return()}finally{if(s)throw o}}}}function yc(t,e){if(t){if("string"==typeof t)return gc(t,e);var n=Object.prototype.toString.call(t).slice(8,-1);return"Object"===n&&t.constructor&&(n=t.constructor.name),"Map"===n||"Set"===n?Array.from(t):"Arguments"===n||/^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)?gc(t,e):void 0}}function gc(t,e){(null==e||e>t.length)&&(e=t.length);for(var n=0,r=new Array(e);n<e;n++)r[n]=t[n];return r}ac.links.data();var vc=document.getElementById("record-container");window.openRecord=function(t){var e,n,r,i,o=document.getElementById(t);if(o){void 0!==u.openedRecordId&&document.getElementById(u.openedRecordId).classList.remove("active"),vc.classList.add("active"),vc.scrollTo({top:0}),o.classList.add("active"),u.openedRecordId=t,fc(),n=sc(e=[t]),r=n.nodes,i=n.links,r.selectAll("circle").style("stroke","var(--highlight)"),i.style("stroke","var(--highlight)"),u.highlightedNodes=u.highlightedNodes.concat(e);var a=o.querySelector("h1").textContent;document.title=a}},window.closeRecord=function(){vc.classList.remove("active"),document.getElementById(u.openedRecordId).classList.remove("active"),u.openedRecordId=void 0,fc()},Hu("escape",(function(t){t.preventDefault(),closeRecord()})),window.hashRecord=function(){var t=new URL(window.location).hash;if(t){var e=decodeURI(t.substring(1));openRecord(e)}},window.addEventListener("hashchange",window.hashRecord),window.addEventListener("DOMContentLoaded",window.hashRecord);var mc=document.getElementById("index");window.addEventListener("DOMContentLoaded",(function(){var t=mc.querySelector(".sorting-select"),e=sorting.records;function n(){for(var n=mc.querySelectorAll("li"),r=function(t,e){return function(t){if(Array.isArray(t))return t}(t)||function(t,e){var n=null==t?null:"undefined"!=typeof Symbol&&t[Symbol.iterator]||t["@@iterator"];if(null!=n){var r,i,o,a,s=[],u=!0,c=!1;try{if(o=(n=n.call(t)).next,0===e){if(Object(n)!==n)return;u=!1}else for(;!(u=(r=o.call(n)).done)&&(s.push(r.value),s.length!==e);u=!0);}catch(t){c=!0,i=t}finally{try{if(!u&&null!=n.return&&(a=n.return(),Object(a)!==a))return}finally{if(c)throw i}}return s}}(t,e)||yc(t,e)||function(){throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.")}()}(t.value.split(":"),2),i=r[0],o=r[1],a=0;a<n.length;a++){var s=e[a][i],u=n[a];"reverse"===o&&(s=n.length-1-s),u.style.order=s}}n(),t.addEventListener("change",n)}));var _c=n(691),bc=new(n.n(_c)())([],{includeScore:!1,keys:["label"]});window.addEventListener("DOMContentLoaded",(function(){var t=[],e=0,n=document.getElementById("search"),r=document.getElementById("search-result-list");function i(){r.childNodes[e].classList.add("outline")}function o(){r.childNodes[e].classList.remove("outline")}n.addEventListener("focus",(function(){bc.setCollection(dc.filter((function(t){return 0===t.hidden}))),console.log(bc),n.addEventListener("input",(function(){if(r.innerHTML="",e=0,t=[],""!==n.value){t=bc.search(n.value);for(var i=0;i<Math.min(5,t.length);i++){var o=t[i].item,a=o.id,s=o.label,u=o.type,c=document.createElement("li");c.classList.add("search-result-item"),c.innerHTML='\n        <a href="#'.concat(a,'">\n          <span\n            class="record-type-point"\n            style="color:var(--n_').concat(u,')"\n          >⬤</span>\n          <span>').concat(s,"</span>\n        </a>"),r.appendChild(c)}}}))})),Hu("s",(function(t){t.preventDefault(),n.focus()})),document.addEventListener("keydown",(function(r){if(0!==t.length)switch(r.key){case"ArrowUp":if(r.preventDefault(),0===e)return void n.focus();o(),e--,i();break;case"ArrowDown":if(r.preventDefault(),4===e||e===t.length-1)return;o(),e++,i(),1===e&&n.blur();break;case"Enter":r.preventDefault(),openRecord(t[e].item.id),n.blur()}}))}));var wc=tc.node().getBoundingClientRect(),xc=wc.width,kc=wc.height,Mc=function(){var t,e,n=xu,r=ku,i=Su,o=Eu,a=Au,s=[0,1/0],u=[[-1/0,-1/0],[1/0,1/0]],c=250,h=vn,f=[],l=g("start","zoom","end"),d=500,p=150,y=0;function v(t){t.property("__zoom",Mu).on("wheel.zoom",M).on("mousedown.zoom",E).on("dblclick.zoom",A).filter(a).on("touchstart.zoom",S).on("touchmove.zoom",L).on("touchend.zoom touchcancel.zoom",N).style("touch-action","none").style("-webkit-tap-highlight-color","rgba(0,0,0,0)")}function m(t,e){return(e=Math.max(s[0],Math.min(s[1],e)))===t.k?t:new mu(e,t.x,t.y)}function _(t,e,n){var r=e[0]-n[0]*t.k,i=e[1]-n[1]*t.k;return r===t.x&&i===t.y?t:new mu(t.k,r,i)}function b(t){return[(+t[0][0]+ +t[1][0])/2,(+t[0][1]+ +t[1][1])/2]}function w(t,e,n){t.on("start.zoom",(function(){x(this,arguments).start()})).on("interrupt.zoom end.zoom",(function(){x(this,arguments).end()})).tween("zoom",(function(){var t=this,i=arguments,o=x(t,i),a=r.apply(t,i),s=n||b(a),u=Math.max(a[1][0]-a[0][0],a[1][1]-a[0][1]),c=t.__zoom,f="function"==typeof e?e.apply(t,i):e,l=h(c.invert(s).concat(u/c.k),f.invert(s).concat(u/f.k));return function(t){if(1===t)t=f;else{var e=l(t),n=u/e[2];t=new mu(n,s[0]-e[0]*n,s[1]-e[1]*n)}o.zoom(null,t)}}))}function x(t,e){for(var n,r=0,i=f.length;r<i;++r)if((n=f[r]).that===t)return n;return new k(t,e)}function k(t,e){this.that=t,this.args=e,this.index=-1,this.active=0,this.extent=r.apply(t,e)}function M(){if(n.apply(this,arguments)){var t=x(this,arguments),e=this.__zoom,r=Math.max(s[0],Math.min(s[1],e.k*Math.pow(2,o.apply(this,arguments)))),a=Dt(this);if(t.wheel)t.mouse[0][0]===a[0]&&t.mouse[0][1]===a[1]||(t.mouse[1]=e.invert(t.mouse[0]=a)),clearTimeout(t.wheel);else{if(e.k===r)return;t.mouse=[a,e.invert(a)],Xn(this),t.start()}wu(),t.wheel=setTimeout((function(){t.wheel=null,t.end()}),p),t.zoom("mouse",i(_(m(e,r),t.mouse[0],t.mouse[1]),t.extent,u))}}function E(){if(!e&&n.apply(this,arguments)){var t=x(this,arguments),r=$t(vt.view).on("mousemove.zoom",(function(){if(wu(),!t.moved){var e=vt.clientX-a,n=vt.clientY-s;t.moved=e*e+n*n>y}t.zoom("mouse",i(_(t.that.__zoom,t.mouse[0]=Dt(t.that),t.mouse[1]),t.extent,u))}),!0).on("mouseup.zoom",(function(){r.on("mousemove.zoom mouseup.zoom",null),It(vt.view,t.moved),wu(),t.end()}),!0),o=Dt(this),a=vt.clientX,s=vt.clientY;Pt(vt.view),bu(),t.mouse=[o,this.__zoom.invert(o)],Xn(this),t.start()}}function A(){if(n.apply(this,arguments)){var t=this.__zoom,e=Dt(this),o=t.invert(e),a=t.k*(vt.shiftKey?.5:2),s=i(_(m(t,a),e,o),r.apply(this,arguments),u);wu(),c>0?$t(this).transition().duration(c).call(w,s,e):$t(this).call(v.transform,s)}}function S(){if(n.apply(this,arguments)){var e,r,i,o,a=x(this,arguments),s=vt.changedTouches,u=s.length;for(bu(),r=0;r<u;++r)o=[o=Ot(this,s,(i=s[r]).identifier),this.__zoom.invert(o),i.identifier],a.touch0?a.touch1||(a.touch1=o):(a.touch0=o,e=!0);if(t&&(t=clearTimeout(t),!a.touch1))return a.end(),void((o=$t(this).on("dblclick.zoom"))&&o.apply(this,arguments));e&&(t=setTimeout((function(){t=null}),d),Xn(this),a.start())}}function L(){var e,n,r,o,a=x(this,arguments),s=vt.changedTouches,c=s.length;for(wu(),t&&(t=clearTimeout(t)),e=0;e<c;++e)r=Ot(this,s,(n=s[e]).identifier),a.touch0&&a.touch0[2]===n.identifier?a.touch0[0]=r:a.touch1&&a.touch1[2]===n.identifier&&(a.touch1[0]=r);if(n=a.that.__zoom,a.touch1){var h=a.touch0[0],f=a.touch0[1],l=a.touch1[0],d=a.touch1[1],p=(p=l[0]-h[0])*p+(p=l[1]-h[1])*p,y=(y=d[0]-f[0])*y+(y=d[1]-f[1])*y;n=m(n,Math.sqrt(p/y)),r=[(h[0]+l[0])/2,(h[1]+l[1])/2],o=[(f[0]+d[0])/2,(f[1]+d[1])/2]}else{if(!a.touch0)return;r=a.touch0[0],o=a.touch0[1]}a.zoom("touch",i(_(n,r,o),a.extent,u))}function N(){var t,n,r=x(this,arguments),i=vt.changedTouches,o=i.length;for(bu(),e&&clearTimeout(e),e=setTimeout((function(){e=null}),d),t=0;t<o;++t)n=i[t],r.touch0&&r.touch0[2]===n.identifier?delete r.touch0:r.touch1&&r.touch1[2]===n.identifier&&delete r.touch1;r.touch1&&!r.touch0&&(r.touch0=r.touch1,delete r.touch1),r.touch0?r.touch0[1]=this.__zoom.invert(r.touch0[0]):r.end()}return v.transform=function(t,e){var n=t.selection?t.selection():t;n.property("__zoom",Mu),t!==n?w(t,e):n.interrupt().each((function(){x(this,arguments).start().zoom(null,"function"==typeof e?e.apply(this,arguments):e).end()}))},v.scaleBy=function(t,e){v.scaleTo(t,(function(){return this.__zoom.k*("function"==typeof e?e.apply(this,arguments):e)}))},v.scaleTo=function(t,e){v.transform(t,(function(){var t=r.apply(this,arguments),n=this.__zoom,o=b(t),a=n.invert(o),s="function"==typeof e?e.apply(this,arguments):e;return i(_(m(n,s),o,a),t,u)}))},v.translateBy=function(t,e,n){v.transform(t,(function(){return i(this.__zoom.translate("function"==typeof e?e.apply(this,arguments):e,"function"==typeof n?n.apply(this,arguments):n),r.apply(this,arguments),u)}))},v.translateTo=function(t,e,n){v.transform(t,(function(){var t=r.apply(this,arguments),o=this.__zoom,a=b(t);return i(_u.translate(a[0],a[1]).scale(o.k).translate("function"==typeof e?-e.apply(this,arguments):-e,"function"==typeof n?-n.apply(this,arguments):-n),t,u)}))},k.prototype={start:function(){return 1==++this.active&&(this.index=f.push(this)-1,this.emit("start")),this},zoom:function(t,e){return this.mouse&&"mouse"!==t&&(this.mouse[1]=e.invert(this.mouse[0])),this.touch0&&"touch"!==t&&(this.touch0[1]=e.invert(this.touch0[0])),this.touch1&&"touch"!==t&&(this.touch1[1]=e.invert(this.touch1[0])),this.that.__zoom=e,this.emit("zoom"),this},end:function(){return 0==--this.active&&(f.splice(this.index,1),this.index=-1,this.emit("end")),this},emit:function(t){xt(new vu(v,t,this.that.__zoom),l.apply,l,[t,this.that,this.args])}},v.wheelDelta=function(t){return arguments.length?(o="function"==typeof t?t:gu(+t),v):o},v.filter=function(t){return arguments.length?(n="function"==typeof t?t:gu(!!t),v):n},v.touchable=function(t){return arguments.length?(a="function"==typeof t?t:gu(!!t),v):a},v.extent=function(t){return arguments.length?(r="function"==typeof t?t:gu([[+t[0][0],+t[0][1]],[+t[1][0],+t[1][1]]]),v):r},v.scaleExtent=function(t){return arguments.length?(s[0]=+t[0],s[1]=+t[1],v):[s[0],s[1]]},v.translateExtent=function(t){return arguments.length?(u[0][0]=+t[0][0],u[1][0]=+t[1][0],u[0][1]=+t[0][1],u[1][1]=+t[1][1],v):[[u[0][0],u[0][1]],[u[1][0],u[1][1]]]},v.constrain=function(t){return arguments.length?(i=t,v):i},v.duration=function(t){return arguments.length?(c=+t,v):c},v.interpolate=function(t){return arguments.length?(h=t,v):h},v.on=function(){var t=l.on.apply(l,arguments);return t===l?v:t},v.clickDistance=function(t){return arguments.length?(y=(t=+t)*t,v):Math.sqrt(y)},v}().extent([[0,0],[xc,kc]]).scaleExtent([1,10]).on("zoom",(function(){var t=vt.transform,e=t.x,n=t.y,r=t.k;u.position.x=e/r||0,u.position.y=n/r||0,u.position.zoom=r||1,lc()}));function Ec(){u.position.zoom=1,u.position.x=0,u.position.y=0,tc.call(Mc.transform,_u.translate(u.position.y,u.position.x).scale(u.position.zoom)),lc()}function Ac(t,e){var n="undefined"!=typeof Symbol&&t[Symbol.iterator]||t["@@iterator"];if(!n){if(Array.isArray(t)||(n=Sc(t))||e&&t&&"number"==typeof t.length){n&&(t=n);var r=0,i=function(){};return{s:i,n:function(){return r>=t.length?{done:!0}:{done:!1,value:t[r++]}},e:function(t){throw t},f:i}}throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.")}var o,a=!0,s=!1;return{s:function(){n=n.call(t)},n:function(){var t=n.next();return a=t.done,t},e:function(t){s=!0,o=t},f:function(){try{a||null==n.return||n.return()}finally{if(s)throw o}}}}function Sc(t,e){if(t){if("string"==typeof t)return Lc(t,e);var n=Object.prototype.toString.call(t).slice(8,-1);return"Object"===n&&t.constructor&&(n=t.constructor.name),"Map"===n||"Set"===n?Array.from(t):"Arguments"===n||/^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)?Lc(t,e):void 0}}function Lc(t,e){(null==e||e>t.length)&&(e=t.length);for(var n=0,r=new Array(e);n<e;n++)r[n]=t[n];return r}function Nc(){var t,e=document.getElementById("index-counter"),n=document.querySelectorAll(".menu-types-list .badge"),r=0,i={},o=Ac(data.nodes);try{for(o.s();!(t=o.n()).done;){var a=t.value,s=a.hidden,u=a.type;!0!==s&&(r+=1,i[u]?i[u]+=1:i[u]=1)}}catch(t){o.e(t)}finally{o.f()}var c,h=Ac(n);try{for(h.s();!(c=h.n()).done;){var f=c.value;$c(f,i[f.previousElementSibling.name]||0)}}catch(t){h.e(t)}finally{h.f()}$c(e,r)}function $c(t,e){var n,r,i=(n=t.textContent.split("/",2),r=2,function(t){if(Array.isArray(t))return t}(n)||function(t,e){var n=null==t?null:"undefined"!=typeof Symbol&&t[Symbol.iterator]||t["@@iterator"];if(null!=n){var r,i,o,a,s=[],u=!0,c=!1;try{if(o=(n=n.call(t)).next,0===e){if(Object(n)!==n)return;u=!1}else for(;!(u=(r=o.call(n)).done)&&(s.push(r.value),s.length!==e);u=!0);}catch(t){c=!0,i=t}finally{try{if(!u&&null!=n.return&&(a=n.return(),Object(a)!==a))return}finally{if(c)throw i}}return s}}(n,r)||Sc(n,r)||function(){throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.")}()),o=i[0],a=i[1];o=Number(o),a=Number(a)||o,t.textContent=e===a?a:"".concat(e,"/").concat(a)}function Cc(t,e){(null==e||e>t.length)&&(e=t.length);for(var n=0,r=new Array(e);n<e;n++)r[n]=t[n];return r}tc.call(Mc),Hu("e,alt+r",(function(t){t.preventDefault(),Ec()})),window.zoomMore=function(){Mc.scaleTo(tc,u.position.zoom+.2)},window.zoomLess=function(){Mc.scaleTo(tc,u.position.zoom-.2)},window.zoomReset=Ec,n(838);var Tc=timeline,Gc=Tc.begin,Uc=Tc.end;function Dc(t){return new Date(1e3*t).toLocaleDateString(graphProperties.lang)}function Oc(t,e){return function(t){if(Array.isArray(t))return t}(t)||function(t,e){var n=null==t?null:"undefined"!=typeof Symbol&&t[Symbol.iterator]||t["@@iterator"];if(null!=n){var r,i,o,a,s=[],u=!0,c=!1;try{if(o=(n=n.call(t)).next,0===e){if(Object(n)!==n)return;u=!1}else for(;!(u=(r=o.call(n)).done)&&(s.push(r.value),s.length!==e);u=!0);}catch(t){c=!0,i=t}finally{try{if(!u&&null!=n.return&&(a=n.return(),Object(a)!==a))return}finally{if(c)throw i}}return s}}(t,e)||jc(t,e)||function(){throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.")}()}function jc(t,e){if(t){if("string"==typeof t)return zc(t,e);var n=Object.prototype.toString.call(t).slice(8,-1);return"Object"===n&&t.constructor&&(n=t.constructor.name),"Map"===n||"Set"===n?Array.from(t):"Arguments"===n||/^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)?zc(t,e):void 0}}function zc(t,e){(null==e||e>t.length)&&(e=t.length);for(var n=0,r=new Array(e);n<e;n++)r[n]=t[n];return r}function Pc(t,e){if(t){if("string"==typeof t)return Ic(t,e);var n=Object.prototype.toString.call(t).slice(8,-1);return"Object"===n&&t.constructor&&(n=t.constructor.name),"Map"===n||"Set"===n?Array.from(t):"Arguments"===n||/^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)?Ic(t,e):void 0}}function Ic(t,e){(null==e||e>t.length)&&(e=t.length);for(var n=0,r=new Array(e);n<e;n++)r[n]=t[n];return r}window.addEventListener("DOMContentLoaded",(function(){var t=document.getElementById("timeline-checkbox"),e=document.getElementById("timeline-form"),n=document.getElementById("timeline-ticks");if(e){var r=e.querySelector("output"),i=e.querySelector('input[type="range"]');t.checked=!1,t.addEventListener("change",(function(){t.checked?(e.classList.add("active"),a(),o(),window.addEventListener("resize",a),i.addEventListener("input",o)):(e.classList.remove("active"),window.removeEventListener("resize",a),i.removeEventListener("input",o),hc(Ju.filteredByTimeline),Nc())}))}function o(){var t=i.value;r.value=Dc(t);var e,n=[],o=function(t,e){var n="undefined"!=typeof Symbol&&t[Symbol.iterator]||t["@@iterator"];if(!n){if(Array.isArray(t)||(n=function(t,e){if(t){if("string"==typeof t)return Cc(t,e);var n=Object.prototype.toString.call(t).slice(8,-1);return"Object"===n&&t.constructor&&(n=t.constructor.name),"Map"===n||"Set"===n?Array.from(t):"Arguments"===n||/^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)?Cc(t,e):void 0}}(t))||e&&t&&"number"==typeof t.length){n&&(t=n);var r=0,i=function(){};return{s:i,n:function(){return r>=t.length?{done:!0}:{done:!1,value:t[r++]}},e:function(t){throw t},f:i}}throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.")}var o,a=!0,s=!1;return{s:function(){n=n.call(t)},n:function(){var t=n.next();return a=t.done,t},e:function(t){s=!0,o=t},f:function(){try{a||null==n.return||n.return()}finally{if(s)throw o}}}}(data.nodes);try{for(o.s();!(e=o.n()).done;){var a=e.value,s=a.begin,u=a.end,c=a.id;void 0===u&&(u=Uc),void 0===s&&(s=Gc),t>=s&&t<=u&&n.push(c)}}catch(t){o.e(t)}finally{o.f()}uc(n,Ju.filteredByTimeline),Nc()}function a(){var t=Math.round(e.offsetWidth/100),r=Math.round((Uc-Gc)/t);n.setAttribute("style","--list-length: ".concat(t+1,";")),n.innerHTML="",n.insertAdjacentHTML("beforeend",'<option value="'.concat(Gc,'">').concat(Dc(Gc),"</option>"));for(var i=1;i<=t;i++)i!==t?n.insertAdjacentHTML("beforeend",'<option value="'.concat(Gc+r*i,'">').concat(Dc(Gc+r*i),"</option>")):n.insertAdjacentHTML("beforeend",'<option value="'.concat(Uc,'">').concat(Dc(Uc),"</option>"))}})),window.addEventListener("DOMContentLoaded",(function(){for(var t,e=document.getElementById("types-form"),n=e.querySelectorAll("input"),r=0,i=Object.entries(typeList);r<i.length;r++){var o=Oc(i[r],2),a=o[0],s=o[1].active;e.querySelector('[name="'.concat(a,'"]')).checked=s}d();var u,c=null===(t=new URL(window.location).searchParams.get("filters"))||void 0===t?void 0:t.split("-");if(null!=c&&c.length){for(var h=0,f=Object.entries(typeList);h<f.length;h++){var l=Oc(f[h],1)[0];e.querySelector('[name="'.concat(l,'"]')).checked=c.includes(l)}d()}function d(){var t=new FormData(e);t=Object.fromEntries(t);var n=new Set;t=Object.entries(typeList).filter((function(e){var n=Oc(e,1)[0];return!!t[n]})).forEach((function(t){Oc(t,2)[1].nodes.forEach((function(t){return n.add(t)}))})),uc(Array.from(n),Ju.filteredByType)}e.addEventListener("change",d);var p,y=function(t,e){var n="undefined"!=typeof Symbol&&t[Symbol.iterator]||t["@@iterator"];if(!n){if(Array.isArray(t)||(n=jc(t))){n&&(t=n);var r=0,i=function(){};return{s:i,n:function(){return r>=t.length?{done:!0}:{done:!1,value:t[r++]}},e:function(t){throw t},f:i}}throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.")}var o,a=!0,s=!1;return{s:function(){n=n.call(t)},n:function(){var t=n.next();return a=t.done,t},e:function(t){s=!0,o=t},f:function(){try{a||null==n.return||n.return()}finally{if(s)throw o}}}}(n);try{var g=function(){var t=p.value,r=t.name;t.checked,t.parentElement.addEventListener("click",(function(t){t.altKey&&(t.stopPropagation(),t.preventDefault(),u===r?(v(),u=void 0):(function(t){n.forEach((function(e){return e.checked=t===e.name})),e.dispatchEvent(new Event("change"))}(r),u=r))}))};for(y.s();!(p=y.n()).done;)g()}catch(t){y.e(t)}finally{y.f()}function v(){e.querySelectorAll("input:not(:checked)").forEach((function(t){return t.checked=!0})),e.dispatchEvent(new Event("change"))}hotkeys("alt+r",(function(t){t.preventDefault(),v()}))})),window.addEventListener("DOMContentLoaded",(function(){var t,e=document.getElementById("tags-form"),n=document.querySelector(".menu-tags .sorting-select"),r=sorting.tags;u(),c(),n.addEventListener("change",u),e.addEventListener("change",c);var i=null===(t=new URL(window.location).searchParams.get("tags"))||void 0===t?void 0:t.split("-");if(null!=i&&i.length){var o,a=function(t,e){var n="undefined"!=typeof Symbol&&t[Symbol.iterator]||t["@@iterator"];if(!n){if(Array.isArray(t)||(n=Pc(t))){n&&(t=n);var r=0,i=function(){};return{s:i,n:function(){return r>=t.length?{done:!0}:{done:!1,value:t[r++]}},e:function(t){throw t},f:i}}throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.")}var o,a=!0,s=!1;return{s:function(){n=n.call(t)},n:function(){var t=n.next();return a=t.done,t},e:function(t){s=!0,o=t},f:function(){try{a||null==n.return||n.return()}finally{if(s)throw o}}}}(i);try{for(a.s();!(o=a.n()).done;){var s=o.value;e.querySelector('input[name="'.concat(s,'"]')).checked=!0}}catch(t){a.e(t)}finally{a.f()}e.dispatchEvent(new Event("change"))}function u(){for(var t=e.querySelectorAll("label"),i=function(t,e){return function(t){if(Array.isArray(t))return t}(t)||function(t,e){var n=null==t?null:"undefined"!=typeof Symbol&&t[Symbol.iterator]||t["@@iterator"];if(null!=n){var r,i,o,a,s=[],u=!0,c=!1;try{if(o=(n=n.call(t)).next,0===e){if(Object(n)!==n)return;u=!1}else for(;!(u=(r=o.call(n)).done)&&(s.push(r.value),s.length!==e);u=!0);}catch(t){c=!0,i=t}finally{try{if(!u&&null!=n.return&&(a=n.return(),Object(a)!==a))return}finally{if(c)throw i}}return s}}(t,e)||Pc(t,e)||function(){throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.")}()}(n.value.split(":"),2),o=i[0],a=i[1],s=0;s<t.length;s++){var u=r[s][o],c=t[s];"reverse"===a&&(u=t.length-1-u),c.style.order=u}}function c(){var t=new FormData(e);t=Object.fromEntries(t);var n=new Set;tagList.filter((function(e){var n=e.name;return!!t[n]})).forEach((function(t){t.nodes.forEach((function(t){return n.add(t)}))})),0!==n.size?uc(Array.from(n),Ju.filteredByTag):hc(Ju.filteredByTag)}Hu("alt+r",(function(t){t.preventDefault(),e.querySelectorAll("input:checked").forEach((function(t){return t.checked=!1})),e.dispatchEvent(new Event("change"))})),window.activeTag=function(t){var n=e.querySelector('input[name="'.concat(t,'"]'));n.checked=!n.checked,e.dispatchEvent(new Event("change"))}}));var Rc=n(187),Fc=n(993),qc=n(641),Yc=n(11);let Bc=function(){const t=arguments[0];for(let e=1,n=arguments.length;e<n;e++)if(arguments[e])for(const n in arguments[e])t[n]=arguments[e][n];return t};function Wc(t,e,n,r){const i=t._nodes.get(e);let o=null;return i?(o="mixed"===r?i.out&&i.out[n]||i.undirected&&i.undirected[n]:"directed"===r?i.out&&i.out[n]:i.undirected&&i.undirected[n],o):o}function Kc(t){return"object"==typeof t&&null!==t}function Hc(t){let e;for(e in t)return!1;return!0}function Xc(t,e,n){Object.defineProperty(t,e,{enumerable:!1,configurable:!1,writable:!0,value:n})}function Vc(t,e,n){const r={enumerable:!0,configurable:!0};"function"==typeof n?r.get=n:(r.value=n,r.writable=!1),Object.defineProperty(t,e,r)}function Zc(t){return!(!Kc(t)||t.attributes&&!Array.isArray(t.attributes))}"function"==typeof Object.assign&&(Bc=Object.assign);class Jc extends Error{constructor(t){super(),this.name="GraphError",this.message=t}}class Qc extends Jc{constructor(t){super(t),this.name="InvalidArgumentsGraphError","function"==typeof Error.captureStackTrace&&Error.captureStackTrace(this,Qc.prototype.constructor)}}class th extends Jc{constructor(t){super(t),this.name="NotFoundGraphError","function"==typeof Error.captureStackTrace&&Error.captureStackTrace(this,th.prototype.constructor)}}class eh extends Jc{constructor(t){super(t),this.name="UsageGraphError","function"==typeof Error.captureStackTrace&&Error.captureStackTrace(this,eh.prototype.constructor)}}function nh(t,e){this.key=t,this.attributes=e,this.clear()}function rh(t,e){this.key=t,this.attributes=e,this.clear()}function ih(t,e){this.key=t,this.attributes=e,this.clear()}function oh(t,e,n,r,i){this.key=e,this.attributes=i,this.undirected=t,this.source=n,this.target=r}nh.prototype.clear=function(){this.inDegree=0,this.outDegree=0,this.undirectedDegree=0,this.undirectedLoops=0,this.directedLoops=0,this.in={},this.out={},this.undirected={}},rh.prototype.clear=function(){this.inDegree=0,this.outDegree=0,this.directedLoops=0,this.in={},this.out={}},ih.prototype.clear=function(){this.undirectedDegree=0,this.undirectedLoops=0,this.undirected={}},oh.prototype.attach=function(){let t="out",e="in";this.undirected&&(t=e="undirected");const n=this.source.key,r=this.target.key;this.source[t][r]=this,this.undirected&&n===r||(this.target[e][n]=this)},oh.prototype.attachMulti=function(){let t="out",e="in";const n=this.source.key,r=this.target.key;this.undirected&&(t=e="undirected");const i=this.source[t],o=i[r];if(void 0===o)return i[r]=this,void(this.undirected&&n===r||(this.target[e][n]=this));o.previous=this,this.next=o,i[r]=this,this.target[e][n]=this},oh.prototype.detach=function(){const t=this.source.key,e=this.target.key;let n="out",r="in";this.undirected&&(n=r="undirected"),delete this.source[n][e],delete this.target[r][t]},oh.prototype.detachMulti=function(){const t=this.source.key,e=this.target.key;let n="out",r="in";this.undirected&&(n=r="undirected"),void 0===this.previous?void 0===this.next?(delete this.source[n][e],delete this.target[r][t]):(this.next.previous=void 0,this.source[n][e]=this.next,this.target[r][t]=this.next):(this.previous.next=this.next,void 0!==this.next&&(this.next.previous=this.previous))};function ah(t,e,n,r,i,o,a){let s,u,c,h;if(r=""+r,0===n){if(s=t._nodes.get(r),!s)throw new th(`Graph.${e}: could not find the "${r}" node in the graph.`);c=i,h=o}else if(3===n){if(i=""+i,u=t._edges.get(i),!u)throw new th(`Graph.${e}: could not find the "${i}" edge in the graph.`);const n=u.source.key,f=u.target.key;if(r===n)s=u.target;else{if(r!==f)throw new th(`Graph.${e}: the "${r}" node is not attached to the "${i}" edge (${n}, ${f}).`);s=u.source}c=o,h=a}else{if(u=t._edges.get(r),!u)throw new th(`Graph.${e}: could not find the "${r}" edge in the graph.`);s=1===n?u.source:u.target,c=i,h=o}return[s,c,h]}const sh=[{name:t=>`get${t}Attribute`,attacher:function(t,e,n){t.prototype[e]=function(t,r,i){const[o,a]=ah(this,e,n,t,r,i);return o.attributes[a]}}},{name:t=>`get${t}Attributes`,attacher:function(t,e,n){t.prototype[e]=function(t,r){const[i]=ah(this,e,n,t,r);return i.attributes}}},{name:t=>`has${t}Attribute`,attacher:function(t,e,n){t.prototype[e]=function(t,r,i){const[o,a]=ah(this,e,n,t,r,i);return o.attributes.hasOwnProperty(a)}}},{name:t=>`set${t}Attribute`,attacher:function(t,e,n){t.prototype[e]=function(t,r,i,o){const[a,s,u]=ah(this,e,n,t,r,i,o);return a.attributes[s]=u,this.emit("nodeAttributesUpdated",{key:a.key,type:"set",attributes:a.attributes,name:s}),this}}},{name:t=>`update${t}Attribute`,attacher:function(t,e,n){t.prototype[e]=function(t,r,i,o){const[a,s,u]=ah(this,e,n,t,r,i,o);if("function"!=typeof u)throw new Qc(`Graph.${e}: updater should be a function.`);const c=a.attributes,h=u(c[s]);return c[s]=h,this.emit("nodeAttributesUpdated",{key:a.key,type:"set",attributes:a.attributes,name:s}),this}}},{name:t=>`remove${t}Attribute`,attacher:function(t,e,n){t.prototype[e]=function(t,r,i){const[o,a]=ah(this,e,n,t,r,i);return delete o.attributes[a],this.emit("nodeAttributesUpdated",{key:o.key,type:"remove",attributes:o.attributes,name:a}),this}}},{name:t=>`replace${t}Attributes`,attacher:function(t,e,n){t.prototype[e]=function(t,r,i){const[o,a]=ah(this,e,n,t,r,i);if(!Kc(a))throw new Qc(`Graph.${e}: provided attributes are not a plain object.`);return o.attributes=a,this.emit("nodeAttributesUpdated",{key:o.key,type:"replace",attributes:o.attributes}),this}}},{name:t=>`merge${t}Attributes`,attacher:function(t,e,n){t.prototype[e]=function(t,r,i){const[o,a]=ah(this,e,n,t,r,i);if(!Kc(a))throw new Qc(`Graph.${e}: provided attributes are not a plain object.`);return Bc(o.attributes,a),this.emit("nodeAttributesUpdated",{key:o.key,type:"merge",attributes:o.attributes,data:a}),this}}},{name:t=>`update${t}Attributes`,attacher:function(t,e,n){t.prototype[e]=function(t,r,i){const[o,a]=ah(this,e,n,t,r,i);if("function"!=typeof a)throw new Qc(`Graph.${e}: provided updater is not a function.`);return o.attributes=a(o.attributes),this.emit("nodeAttributesUpdated",{key:o.key,type:"update",attributes:o.attributes}),this}}}],uh=[{name:t=>`get${t}Attribute`,attacher:function(t,e,n){t.prototype[e]=function(t,r){let i;if("mixed"!==this.type&&"mixed"!==n&&n!==this.type)throw new eh(`Graph.${e}: cannot find this type of edges in your ${this.type} graph.`);if(arguments.length>2){if(this.multi)throw new eh(`Graph.${e}: cannot use a {source,target} combo when asking about an edge's attributes in a MultiGraph since we cannot infer the one you want information about.`);const o=""+t,a=""+r;if(r=arguments[2],i=Wc(this,o,a,n),!i)throw new th(`Graph.${e}: could not find an edge for the given path ("${o}" - "${a}").`)}else{if("mixed"!==n)throw new eh(`Graph.${e}: calling this method with only a key (vs. a source and target) does not make sense since an edge with this key could have the other type.`);if(t=""+t,i=this._edges.get(t),!i)throw new th(`Graph.${e}: could not find the "${t}" edge in the graph.`)}return i.attributes[r]}}},{name:t=>`get${t}Attributes`,attacher:function(t,e,n){t.prototype[e]=function(t){let r;if("mixed"!==this.type&&"mixed"!==n&&n!==this.type)throw new eh(`Graph.${e}: cannot find this type of edges in your ${this.type} graph.`);if(arguments.length>1){if(this.multi)throw new eh(`Graph.${e}: cannot use a {source,target} combo when asking about an edge's attributes in a MultiGraph since we cannot infer the one you want information about.`);const i=""+t,o=""+arguments[1];if(r=Wc(this,i,o,n),!r)throw new th(`Graph.${e}: could not find an edge for the given path ("${i}" - "${o}").`)}else{if("mixed"!==n)throw new eh(`Graph.${e}: calling this method with only a key (vs. a source and target) does not make sense since an edge with this key could have the other type.`);if(t=""+t,r=this._edges.get(t),!r)throw new th(`Graph.${e}: could not find the "${t}" edge in the graph.`)}return r.attributes}}},{name:t=>`has${t}Attribute`,attacher:function(t,e,n){t.prototype[e]=function(t,r){let i;if("mixed"!==this.type&&"mixed"!==n&&n!==this.type)throw new eh(`Graph.${e}: cannot find this type of edges in your ${this.type} graph.`);if(arguments.length>2){if(this.multi)throw new eh(`Graph.${e}: cannot use a {source,target} combo when asking about an edge's attributes in a MultiGraph since we cannot infer the one you want information about.`);const o=""+t,a=""+r;if(r=arguments[2],i=Wc(this,o,a,n),!i)throw new th(`Graph.${e}: could not find an edge for the given path ("${o}" - "${a}").`)}else{if("mixed"!==n)throw new eh(`Graph.${e}: calling this method with only a key (vs. a source and target) does not make sense since an edge with this key could have the other type.`);if(t=""+t,i=this._edges.get(t),!i)throw new th(`Graph.${e}: could not find the "${t}" edge in the graph.`)}return i.attributes.hasOwnProperty(r)}}},{name:t=>`set${t}Attribute`,attacher:function(t,e,n){t.prototype[e]=function(t,r,i){let o;if("mixed"!==this.type&&"mixed"!==n&&n!==this.type)throw new eh(`Graph.${e}: cannot find this type of edges in your ${this.type} graph.`);if(arguments.length>3){if(this.multi)throw new eh(`Graph.${e}: cannot use a {source,target} combo when asking about an edge's attributes in a MultiGraph since we cannot infer the one you want information about.`);const a=""+t,s=""+r;if(r=arguments[2],i=arguments[3],o=Wc(this,a,s,n),!o)throw new th(`Graph.${e}: could not find an edge for the given path ("${a}" - "${s}").`)}else{if("mixed"!==n)throw new eh(`Graph.${e}: calling this method with only a key (vs. a source and target) does not make sense since an edge with this key could have the other type.`);if(t=""+t,o=this._edges.get(t),!o)throw new th(`Graph.${e}: could not find the "${t}" edge in the graph.`)}return o.attributes[r]=i,this.emit("edgeAttributesUpdated",{key:o.key,type:"set",attributes:o.attributes,name:r}),this}}},{name:t=>`update${t}Attribute`,attacher:function(t,e,n){t.prototype[e]=function(t,r,i){let o;if("mixed"!==this.type&&"mixed"!==n&&n!==this.type)throw new eh(`Graph.${e}: cannot find this type of edges in your ${this.type} graph.`);if(arguments.length>3){if(this.multi)throw new eh(`Graph.${e}: cannot use a {source,target} combo when asking about an edge's attributes in a MultiGraph since we cannot infer the one you want information about.`);const a=""+t,s=""+r;if(r=arguments[2],i=arguments[3],o=Wc(this,a,s,n),!o)throw new th(`Graph.${e}: could not find an edge for the given path ("${a}" - "${s}").`)}else{if("mixed"!==n)throw new eh(`Graph.${e}: calling this method with only a key (vs. a source and target) does not make sense since an edge with this key could have the other type.`);if(t=""+t,o=this._edges.get(t),!o)throw new th(`Graph.${e}: could not find the "${t}" edge in the graph.`)}if("function"!=typeof i)throw new Qc(`Graph.${e}: updater should be a function.`);return o.attributes[r]=i(o.attributes[r]),this.emit("edgeAttributesUpdated",{key:o.key,type:"set",attributes:o.attributes,name:r}),this}}},{name:t=>`remove${t}Attribute`,attacher:function(t,e,n){t.prototype[e]=function(t,r){let i;if("mixed"!==this.type&&"mixed"!==n&&n!==this.type)throw new eh(`Graph.${e}: cannot find this type of edges in your ${this.type} graph.`);if(arguments.length>2){if(this.multi)throw new eh(`Graph.${e}: cannot use a {source,target} combo when asking about an edge's attributes in a MultiGraph since we cannot infer the one you want information about.`);const o=""+t,a=""+r;if(r=arguments[2],i=Wc(this,o,a,n),!i)throw new th(`Graph.${e}: could not find an edge for the given path ("${o}" - "${a}").`)}else{if("mixed"!==n)throw new eh(`Graph.${e}: calling this method with only a key (vs. a source and target) does not make sense since an edge with this key could have the other type.`);if(t=""+t,i=this._edges.get(t),!i)throw new th(`Graph.${e}: could not find the "${t}" edge in the graph.`)}return delete i.attributes[r],this.emit("edgeAttributesUpdated",{key:i.key,type:"remove",attributes:i.attributes,name:r}),this}}},{name:t=>`replace${t}Attributes`,attacher:function(t,e,n){t.prototype[e]=function(t,r){let i;if("mixed"!==this.type&&"mixed"!==n&&n!==this.type)throw new eh(`Graph.${e}: cannot find this type of edges in your ${this.type} graph.`);if(arguments.length>2){if(this.multi)throw new eh(`Graph.${e}: cannot use a {source,target} combo when asking about an edge's attributes in a MultiGraph since we cannot infer the one you want information about.`);const o=""+t,a=""+r;if(r=arguments[2],i=Wc(this,o,a,n),!i)throw new th(`Graph.${e}: could not find an edge for the given path ("${o}" - "${a}").`)}else{if("mixed"!==n)throw new eh(`Graph.${e}: calling this method with only a key (vs. a source and target) does not make sense since an edge with this key could have the other type.`);if(t=""+t,i=this._edges.get(t),!i)throw new th(`Graph.${e}: could not find the "${t}" edge in the graph.`)}if(!Kc(r))throw new Qc(`Graph.${e}: provided attributes are not a plain object.`);return i.attributes=r,this.emit("edgeAttributesUpdated",{key:i.key,type:"replace",attributes:i.attributes}),this}}},{name:t=>`merge${t}Attributes`,attacher:function(t,e,n){t.prototype[e]=function(t,r){let i;if("mixed"!==this.type&&"mixed"!==n&&n!==this.type)throw new eh(`Graph.${e}: cannot find this type of edges in your ${this.type} graph.`);if(arguments.length>2){if(this.multi)throw new eh(`Graph.${e}: cannot use a {source,target} combo when asking about an edge's attributes in a MultiGraph since we cannot infer the one you want information about.`);const o=""+t,a=""+r;if(r=arguments[2],i=Wc(this,o,a,n),!i)throw new th(`Graph.${e}: could not find an edge for the given path ("${o}" - "${a}").`)}else{if("mixed"!==n)throw new eh(`Graph.${e}: calling this method with only a key (vs. a source and target) does not make sense since an edge with this key could have the other type.`);if(t=""+t,i=this._edges.get(t),!i)throw new th(`Graph.${e}: could not find the "${t}" edge in the graph.`)}if(!Kc(r))throw new Qc(`Graph.${e}: provided attributes are not a plain object.`);return Bc(i.attributes,r),this.emit("edgeAttributesUpdated",{key:i.key,type:"merge",attributes:i.attributes,data:r}),this}}},{name:t=>`update${t}Attributes`,attacher:function(t,e,n){t.prototype[e]=function(t,r){let i;if("mixed"!==this.type&&"mixed"!==n&&n!==this.type)throw new eh(`Graph.${e}: cannot find this type of edges in your ${this.type} graph.`);if(arguments.length>2){if(this.multi)throw new eh(`Graph.${e}: cannot use a {source,target} combo when asking about an edge's attributes in a MultiGraph since we cannot infer the one you want information about.`);const o=""+t,a=""+r;if(r=arguments[2],i=Wc(this,o,a,n),!i)throw new th(`Graph.${e}: could not find an edge for the given path ("${o}" - "${a}").`)}else{if("mixed"!==n)throw new eh(`Graph.${e}: calling this method with only a key (vs. a source and target) does not make sense since an edge with this key could have the other type.`);if(t=""+t,i=this._edges.get(t),!i)throw new th(`Graph.${e}: could not find the "${t}" edge in the graph.`)}if("function"!=typeof r)throw new Qc(`Graph.${e}: provided updater is not a function.`);return i.attributes=r(i.attributes),this.emit("edgeAttributesUpdated",{key:i.key,type:"update",attributes:i.attributes}),this}}}],ch=[{name:"edges",type:"mixed"},{name:"inEdges",type:"directed",direction:"in"},{name:"outEdges",type:"directed",direction:"out"},{name:"inboundEdges",type:"mixed",direction:"in"},{name:"outboundEdges",type:"mixed",direction:"out"},{name:"directedEdges",type:"directed"},{name:"undirectedEdges",type:"undirected"}];function hh(t,e,n,r){let i=!1;for(const o in e){if(o===r)continue;const a=e[o];if(i=n(a.key,a.attributes,a.source.key,a.target.key,a.source.attributes,a.target.attributes,a.undirected),t&&i)return a.key}}function fh(t,e,n,r){let i,o,a,s=!1;for(const u in e)if(u!==r){i=e[u];do{if(o=i.source,a=i.target,s=n(i.key,i.attributes,o.key,a.key,o.attributes,a.attributes,i.undirected),t&&s)return i.key;i=i.next}while(void 0!==i)}}function lh(t,e){const n=Object.keys(t),r=n.length;let i,o=0;return new Fc((function(){do{if(i)i=i.next;else{if(o>=r)return{done:!0};const a=n[o++];if(a===e){i=void 0;continue}i=t[a]}}while(!i);return{done:!1,value:{edge:i.key,attributes:i.attributes,source:i.source.key,target:i.target.key,sourceAttributes:i.source.attributes,targetAttributes:i.target.attributes,undirected:i.undirected}}}))}function dh(t,e,n,r){const i=e[n];if(!i)return;const o=i.source,a=i.target;return r(i.key,i.attributes,o.key,a.key,o.attributes,a.attributes,i.undirected)&&t?i.key:void 0}function ph(t,e,n,r){let i=e[n];if(!i)return;let o=!1;do{if(o=r(i.key,i.attributes,i.source.key,i.target.key,i.source.attributes,i.target.attributes,i.undirected),t&&o)return i.key;i=i.next}while(void 0!==i)}function yh(t,e){let n=t[e];return void 0!==n.next?new Fc((function(){if(!n)return{done:!0};const t={edge:n.key,attributes:n.attributes,source:n.source.key,target:n.target.key,sourceAttributes:n.source.attributes,targetAttributes:n.target.attributes,undirected:n.undirected};return n=n.next,{done:!1,value:t}})):Fc.of({edge:n.key,attributes:n.attributes,source:n.source.key,target:n.target.key,sourceAttributes:n.source.attributes,targetAttributes:n.target.attributes,undirected:n.undirected})}function gh(t,e,n,r){if(0===e.size)return;const i="mixed"!==n&&n!==e.type,o="undirected"===n;let a,s,u=!1;const c=e._edges.values();for(;a=c.next(),!0!==a.done;){if(s=a.value,i&&s.undirected!==o)continue;const{key:e,attributes:n,source:c,target:h}=s;if(u=r(e,n,c.key,h.key,c.attributes,h.attributes,s.undirected),t&&u)return e}}function vh(t,e,n,r,i,o){const a=e?fh:hh;let s;if("undirected"!==n){if("out"!==r&&(s=a(t,i.in,o),t&&s))return s;if("in"!==r&&(s=a(t,i.out,o,r?void 0:i.key),t&&s))return s}if("directed"!==n&&(s=a(t,i.undirected,o),t&&s))return s}function mh(t,e,n,r,i,o,a){const s=n?ph:dh;let u;if("undirected"!==e){if(void 0!==i.in&&"out"!==r&&(u=s(t,i.in,o,a),t&&u))return u;if(void 0!==i.out&&"in"!==r&&(r||i.key!==o)&&(u=s(t,i.out,o,a),t&&u))return u}if("directed"!==e&&void 0!==i.undirected&&(u=s(t,i.undirected,o,a),t&&u))return u}const _h=[{name:"neighbors",type:"mixed"},{name:"inNeighbors",type:"directed",direction:"in"},{name:"outNeighbors",type:"directed",direction:"out"},{name:"inboundNeighbors",type:"mixed",direction:"in"},{name:"outboundNeighbors",type:"mixed",direction:"out"},{name:"directedNeighbors",type:"directed"},{name:"undirectedNeighbors",type:"undirected"}];function bh(){this.A=null,this.B=null}function wh(t,e,n,r,i){for(const o in r){const a=r[o],s=a.source,u=a.target,c=s===n?u:s;if(e&&e.has(c.key))continue;const h=i(c.key,c.attributes);if(t&&h)return c.key}}function xh(t,e,n,r,i){if("mixed"!==e){if("undirected"===e)return wh(t,null,r,r.undirected,i);if("string"==typeof n)return wh(t,null,r,r[n],i)}const o=new bh;let a;if("undirected"!==e){if("out"!==n){if(a=wh(t,null,r,r.in,i),t&&a)return a;o.wrap(r.in)}if("in"!==n){if(a=wh(t,o,r,r.out,i),t&&a)return a;o.wrap(r.out)}}if("directed"!==e&&(a=wh(t,o,r,r.undirected,i),t&&a))return a}function kh(t,e,n){const r=Object.keys(n),i=r.length;let o=0;return new Fc((function(){let a=null;do{if(o>=i)return t&&t.wrap(n),{done:!0};const s=n[r[o++]],u=s.source,c=s.target;a=u===e?c:u,t&&t.has(a.key)&&(a=null)}while(null===a);return{done:!1,value:{neighbor:a.key,attributes:a.attributes}}}))}function Mh(t,e,n,r,i){const o=r._nodes.values(),a=r.type;let s,u,c,h,f,l,d;for(;s=o.next(),!0!==s.done;){let r=!1;if(u=s.value,"undirected"!==a)for(c in h=u.out,h){f=h[c];do{if(l=f.target,r=!0,d=i(u.key,l.key,u.attributes,l.attributes,f.key,f.attributes,f.undirected),t&&d)return f;f=f.next}while(f)}if("directed"!==a)for(c in h=u.undirected,h)if(!(e&&u.key>c)){f=h[c];do{if(l=f.target,l.key!==c&&(l=f.source),r=!0,d=i(u.key,l.key,u.attributes,l.attributes,f.key,f.attributes,f.undirected),t&&d)return f;f=f.next}while(f)}if(n&&!r&&(d=i(u.key,null,u.attributes,null,null,null,null),t&&d))return null}}function Eh(t){if(!Kc(t))throw new Qc('Graph.import: invalid serialized node. A serialized node should be a plain object with at least a "key" property.');if(!("key"in t))throw new Qc("Graph.import: serialized node is missing its key.");if("attributes"in t&&(!Kc(t.attributes)||null===t.attributes))throw new Qc("Graph.import: invalid attributes. Attributes should be a plain object, null or omitted.")}function Ah(t){if(!Kc(t))throw new Qc('Graph.import: invalid serialized edge. A serialized edge should be a plain object with at least a "source" & "target" property.');if(!("source"in t))throw new Qc("Graph.import: serialized edge is missing its source.");if(!("target"in t))throw new Qc("Graph.import: serialized edge is missing its target.");if("attributes"in t&&(!Kc(t.attributes)||null===t.attributes))throw new Qc("Graph.import: invalid attributes. Attributes should be a plain object, null or omitted.");if("undirected"in t&&"boolean"!=typeof t.undirected)throw new Qc("Graph.import: invalid undirectedness information. Undirected should be boolean or omitted.")}bh.prototype.wrap=function(t){null===this.A?this.A=t:null===this.B&&(this.B=t)},bh.prototype.has=function(t){return null!==this.A&&t in this.A||null!==this.B&&t in this.B};const Sh=function(){let t=255&Math.floor(256*Math.random());return()=>t++}(),Lh=new Set(["directed","undirected","mixed"]),Nh=new Set(["domain","_events","_eventsCount","_maxListeners"]),$h={allowSelfLoops:!0,multi:!1,type:"mixed"};function Ch(t,e,n){const r=new t.NodeDataClass(e,n);return t._nodes.set(e,r),t.emit("nodeAdded",{key:e,attributes:n}),r}function Th(t,e,n,r,i,o,a,s){if(!r&&"undirected"===t.type)throw new eh(`Graph.${e}: you cannot add a directed edge to an undirected graph. Use the #.addEdge or #.addUndirectedEdge instead.`);if(r&&"directed"===t.type)throw new eh(`Graph.${e}: you cannot add an undirected edge to a directed graph. Use the #.addEdge or #.addDirectedEdge instead.`);if(s&&!Kc(s))throw new Qc(`Graph.${e}: invalid attributes. Expecting an object but got "${s}"`);if(o=""+o,a=""+a,s=s||{},!t.allowSelfLoops&&o===a)throw new eh(`Graph.${e}: source & target are the same ("${o}"), thus creating a loop explicitly forbidden by this graph 'allowSelfLoops' option set to false.`);const u=t._nodes.get(o),c=t._nodes.get(a);if(!u)throw new th(`Graph.${e}: source node "${o}" not found.`);if(!c)throw new th(`Graph.${e}: target node "${a}" not found.`);const h={key:null,undirected:r,source:o,target:a,attributes:s};if(n)i=t._edgeKeyGenerator();else if(i=""+i,t._edges.has(i))throw new eh(`Graph.${e}: the "${i}" edge already exists in the graph.`);if(!t.multi&&(r?void 0!==u.undirected[a]:void 0!==u.out[a]))throw new eh(`Graph.${e}: an edge linking "${o}" to "${a}" already exists. If you really want to add multiple edges linking those nodes, you should create a multi graph by using the 'multi' option.`);const f=new oh(r,i,u,c,s);t._edges.set(i,f);const l=o===a;return r?(u.undirectedDegree++,c.undirectedDegree++,l&&(u.undirectedLoops++,t._undirectedSelfLoopCount++)):(u.outDegree++,c.inDegree++,l&&(u.directedLoops++,t._directedSelfLoopCount++)),t.multi?f.attachMulti():f.attach(),r?t._undirectedSize++:t._directedSize++,h.key=i,t.emit("edgeAdded",h),i}function Gh(t,e,n,r,i,o,a,s,u){if(!r&&"undirected"===t.type)throw new eh(`Graph.${e}: you cannot merge/update a directed edge to an undirected graph. Use the #.mergeEdge/#.updateEdge or #.addUndirectedEdge instead.`);if(r&&"directed"===t.type)throw new eh(`Graph.${e}: you cannot merge/update an undirected edge to a directed graph. Use the #.mergeEdge/#.updateEdge or #.addDirectedEdge instead.`);if(s)if(u){if("function"!=typeof s)throw new Qc(`Graph.${e}: invalid updater function. Expecting a function but got "${s}"`)}else if(!Kc(s))throw new Qc(`Graph.${e}: invalid attributes. Expecting an object but got "${s}"`);let c;if(o=""+o,a=""+a,u&&(c=s,s=void 0),!t.allowSelfLoops&&o===a)throw new eh(`Graph.${e}: source & target are the same ("${o}"), thus creating a loop explicitly forbidden by this graph 'allowSelfLoops' option set to false.`);let h,f,l=t._nodes.get(o),d=t._nodes.get(a);if(!n&&(h=t._edges.get(i),h)){if(!(h.source.key===o&&h.target.key===a||r&&h.source.key===a&&h.target.key===o))throw new eh(`Graph.${e}: inconsistency detected when attempting to merge the "${i}" edge with "${o}" source & "${a}" target vs. ("${h.source.key}", "${h.target.key}").`);f=h}if(f||t.multi||!l||(f=r?l.undirected[a]:l.out[a]),f){const e=[f.key,!1,!1,!1];if(u?!c:!s)return e;if(u){const e=f.attributes;f.attributes=c(e),t.emit("edgeAttributesUpdated",{type:"replace",key:f.key,attributes:f.attributes})}else Bc(f.attributes,s),t.emit("edgeAttributesUpdated",{type:"merge",key:f.key,attributes:f.attributes,data:s});return e}s=s||{},u&&c&&(s=c(s));const p={key:null,undirected:r,source:o,target:a,attributes:s};if(n)i=t._edgeKeyGenerator();else if(i=""+i,t._edges.has(i))throw new eh(`Graph.${e}: the "${i}" edge already exists in the graph.`);let y=!1,g=!1;l||(l=Ch(t,o,{}),y=!0,o===a&&(d=l,g=!0)),d||(d=Ch(t,a,{}),g=!0),h=new oh(r,i,l,d,s),t._edges.set(i,h);const v=o===a;return r?(l.undirectedDegree++,d.undirectedDegree++,v&&(l.undirectedLoops++,t._undirectedSelfLoopCount++)):(l.outDegree++,d.inDegree++,v&&(l.directedLoops++,t._directedSelfLoopCount++)),t.multi?h.attachMulti():h.attach(),r?t._undirectedSize++:t._directedSize++,p.key=i,t.emit("edgeAdded",p),[i,!0,y,g]}function Uh(t,e){t._edges.delete(e.key);const{source:n,target:r,attributes:i}=e,o=e.undirected,a=n===r;o?(n.undirectedDegree--,r.undirectedDegree--,a&&(n.undirectedLoops--,t._undirectedSelfLoopCount--)):(n.outDegree--,r.inDegree--,a&&(n.directedLoops--,t._directedSelfLoopCount--)),t.multi?e.detachMulti():e.detach(),o?t._undirectedSize--:t._directedSize--,t.emit("edgeDropped",{key:e.key,attributes:i,source:n.key,target:r.key,undirected:o})}class Dh extends Rc.EventEmitter{constructor(t){if(super(),"boolean"!=typeof(t=Bc({},$h,t)).multi)throw new Qc(`Graph.constructor: invalid 'multi' option. Expecting a boolean but got "${t.multi}".`);if(!Lh.has(t.type))throw new Qc(`Graph.constructor: invalid 'type' option. Should be one of "mixed", "directed" or "undirected" but got "${t.type}".`);if("boolean"!=typeof t.allowSelfLoops)throw new Qc(`Graph.constructor: invalid 'allowSelfLoops' option. Expecting a boolean but got "${t.allowSelfLoops}".`);Xc(this,"NodeDataClass","mixed"===t.type?nh:"directed"===t.type?rh:ih);const e="geid_"+Sh()+"_";let n=0;Xc(this,"_attributes",{}),Xc(this,"_nodes",new Map),Xc(this,"_edges",new Map),Xc(this,"_directedSize",0),Xc(this,"_undirectedSize",0),Xc(this,"_directedSelfLoopCount",0),Xc(this,"_undirectedSelfLoopCount",0),Xc(this,"_edgeKeyGenerator",(()=>{let t;do{t=e+n++}while(this._edges.has(t));return t})),Xc(this,"_options",t),Nh.forEach((t=>Xc(this,t,this[t]))),Vc(this,"order",(()=>this._nodes.size)),Vc(this,"size",(()=>this._edges.size)),Vc(this,"directedSize",(()=>this._directedSize)),Vc(this,"undirectedSize",(()=>this._undirectedSize)),Vc(this,"selfLoopCount",(()=>this._directedSelfLoopCount+this._undirectedSelfLoopCount)),Vc(this,"directedSelfLoopCount",(()=>this._directedSelfLoopCount)),Vc(this,"undirectedSelfLoopCount",(()=>this._undirectedSelfLoopCount)),Vc(this,"multi",this._options.multi),Vc(this,"type",this._options.type),Vc(this,"allowSelfLoops",this._options.allowSelfLoops),Vc(this,"implementation",(()=>"graphology"))}_resetInstanceCounters(){this._directedSize=0,this._undirectedSize=0,this._directedSelfLoopCount=0,this._undirectedSelfLoopCount=0}hasNode(t){return this._nodes.has(""+t)}hasDirectedEdge(t,e){if("undirected"===this.type)return!1;if(1===arguments.length){const e=""+t,n=this._edges.get(e);return!!n&&!n.undirected}if(2===arguments.length){t=""+t,e=""+e;const n=this._nodes.get(t);return!!n&&n.out.hasOwnProperty(e)}throw new Qc(`Graph.hasDirectedEdge: invalid arity (${arguments.length}, instead of 1 or 2). You can either ask for an edge id or for the existence of an edge between a source & a target.`)}hasUndirectedEdge(t,e){if("directed"===this.type)return!1;if(1===arguments.length){const e=""+t,n=this._edges.get(e);return!!n&&n.undirected}if(2===arguments.length){t=""+t,e=""+e;const n=this._nodes.get(t);return!!n&&n.undirected.hasOwnProperty(e)}throw new Qc(`Graph.hasDirectedEdge: invalid arity (${arguments.length}, instead of 1 or 2). You can either ask for an edge id or for the existence of an edge between a source & a target.`)}hasEdge(t,e){if(1===arguments.length){const e=""+t;return this._edges.has(e)}if(2===arguments.length){t=""+t,e=""+e;const n=this._nodes.get(t);return!!n&&(void 0!==n.out&&n.out.hasOwnProperty(e)||void 0!==n.undirected&&n.undirected.hasOwnProperty(e))}throw new Qc(`Graph.hasEdge: invalid arity (${arguments.length}, instead of 1 or 2). You can either ask for an edge id or for the existence of an edge between a source & a target.`)}directedEdge(t,e){if("undirected"===this.type)return;if(t=""+t,e=""+e,this.multi)throw new eh("Graph.directedEdge: this method is irrelevant with multigraphs since there might be multiple edges between source & target. See #.directedEdges instead.");const n=this._nodes.get(t);if(!n)throw new th(`Graph.directedEdge: could not find the "${t}" source node in the graph.`);if(!this._nodes.has(e))throw new th(`Graph.directedEdge: could not find the "${e}" target node in the graph.`);const r=n.out&&n.out[e]||void 0;return r?r.key:void 0}undirectedEdge(t,e){if("directed"===this.type)return;if(t=""+t,e=""+e,this.multi)throw new eh("Graph.undirectedEdge: this method is irrelevant with multigraphs since there might be multiple edges between source & target. See #.undirectedEdges instead.");const n=this._nodes.get(t);if(!n)throw new th(`Graph.undirectedEdge: could not find the "${t}" source node in the graph.`);if(!this._nodes.has(e))throw new th(`Graph.undirectedEdge: could not find the "${e}" target node in the graph.`);const r=n.undirected&&n.undirected[e]||void 0;return r?r.key:void 0}edge(t,e){if(this.multi)throw new eh("Graph.edge: this method is irrelevant with multigraphs since there might be multiple edges between source & target. See #.edges instead.");t=""+t,e=""+e;const n=this._nodes.get(t);if(!n)throw new th(`Graph.edge: could not find the "${t}" source node in the graph.`);if(!this._nodes.has(e))throw new th(`Graph.edge: could not find the "${e}" target node in the graph.`);const r=n.out&&n.out[e]||n.undirected&&n.undirected[e]||void 0;if(r)return r.key}areDirectedNeighbors(t,e){t=""+t,e=""+e;const n=this._nodes.get(t);if(!n)throw new th(`Graph.areDirectedNeighbors: could not find the "${t}" node in the graph.`);return"undirected"!==this.type&&(e in n.in||e in n.out)}areOutNeighbors(t,e){t=""+t,e=""+e;const n=this._nodes.get(t);if(!n)throw new th(`Graph.areOutNeighbors: could not find the "${t}" node in the graph.`);return"undirected"!==this.type&&e in n.out}areInNeighbors(t,e){t=""+t,e=""+e;const n=this._nodes.get(t);if(!n)throw new th(`Graph.areInNeighbors: could not find the "${t}" node in the graph.`);return"undirected"!==this.type&&e in n.in}areUndirectedNeighbors(t,e){t=""+t,e=""+e;const n=this._nodes.get(t);if(!n)throw new th(`Graph.areUndirectedNeighbors: could not find the "${t}" node in the graph.`);return"directed"!==this.type&&e in n.undirected}areNeighbors(t,e){t=""+t,e=""+e;const n=this._nodes.get(t);if(!n)throw new th(`Graph.areNeighbors: could not find the "${t}" node in the graph.`);return"undirected"!==this.type&&(e in n.in||e in n.out)||"directed"!==this.type&&e in n.undirected}areInboundNeighbors(t,e){t=""+t,e=""+e;const n=this._nodes.get(t);if(!n)throw new th(`Graph.areInboundNeighbors: could not find the "${t}" node in the graph.`);return"undirected"!==this.type&&e in n.in||"directed"!==this.type&&e in n.undirected}areOutboundNeighbors(t,e){t=""+t,e=""+e;const n=this._nodes.get(t);if(!n)throw new th(`Graph.areOutboundNeighbors: could not find the "${t}" node in the graph.`);return"undirected"!==this.type&&e in n.out||"directed"!==this.type&&e in n.undirected}inDegree(t){t=""+t;const e=this._nodes.get(t);if(!e)throw new th(`Graph.inDegree: could not find the "${t}" node in the graph.`);return"undirected"===this.type?0:e.inDegree}outDegree(t){t=""+t;const e=this._nodes.get(t);if(!e)throw new th(`Graph.outDegree: could not find the "${t}" node in the graph.`);return"undirected"===this.type?0:e.outDegree}directedDegree(t){t=""+t;const e=this._nodes.get(t);if(!e)throw new th(`Graph.directedDegree: could not find the "${t}" node in the graph.`);return"undirected"===this.type?0:e.inDegree+e.outDegree}undirectedDegree(t){t=""+t;const e=this._nodes.get(t);if(!e)throw new th(`Graph.undirectedDegree: could not find the "${t}" node in the graph.`);return"directed"===this.type?0:e.undirectedDegree}inboundDegree(t){t=""+t;const e=this._nodes.get(t);if(!e)throw new th(`Graph.inboundDegree: could not find the "${t}" node in the graph.`);let n=0;return"directed"!==this.type&&(n+=e.undirectedDegree),"undirected"!==this.type&&(n+=e.inDegree),n}outboundDegree(t){t=""+t;const e=this._nodes.get(t);if(!e)throw new th(`Graph.outboundDegree: could not find the "${t}" node in the graph.`);let n=0;return"directed"!==this.type&&(n+=e.undirectedDegree),"undirected"!==this.type&&(n+=e.outDegree),n}degree(t){t=""+t;const e=this._nodes.get(t);if(!e)throw new th(`Graph.degree: could not find the "${t}" node in the graph.`);let n=0;return"directed"!==this.type&&(n+=e.undirectedDegree),"undirected"!==this.type&&(n+=e.inDegree+e.outDegree),n}inDegreeWithoutSelfLoops(t){t=""+t;const e=this._nodes.get(t);if(!e)throw new th(`Graph.inDegreeWithoutSelfLoops: could not find the "${t}" node in the graph.`);return"undirected"===this.type?0:e.inDegree-e.directedLoops}outDegreeWithoutSelfLoops(t){t=""+t;const e=this._nodes.get(t);if(!e)throw new th(`Graph.outDegreeWithoutSelfLoops: could not find the "${t}" node in the graph.`);return"undirected"===this.type?0:e.outDegree-e.directedLoops}directedDegreeWithoutSelfLoops(t){t=""+t;const e=this._nodes.get(t);if(!e)throw new th(`Graph.directedDegreeWithoutSelfLoops: could not find the "${t}" node in the graph.`);return"undirected"===this.type?0:e.inDegree+e.outDegree-2*e.directedLoops}undirectedDegreeWithoutSelfLoops(t){t=""+t;const e=this._nodes.get(t);if(!e)throw new th(`Graph.undirectedDegreeWithoutSelfLoops: could not find the "${t}" node in the graph.`);return"directed"===this.type?0:e.undirectedDegree-2*e.undirectedLoops}inboundDegreeWithoutSelfLoops(t){t=""+t;const e=this._nodes.get(t);if(!e)throw new th(`Graph.inboundDegreeWithoutSelfLoops: could not find the "${t}" node in the graph.`);let n=0,r=0;return"directed"!==this.type&&(n+=e.undirectedDegree,r+=2*e.undirectedLoops),"undirected"!==this.type&&(n+=e.inDegree,r+=e.directedLoops),n-r}outboundDegreeWithoutSelfLoops(t){t=""+t;const e=this._nodes.get(t);if(!e)throw new th(`Graph.outboundDegreeWithoutSelfLoops: could not find the "${t}" node in the graph.`);let n=0,r=0;return"directed"!==this.type&&(n+=e.undirectedDegree,r+=2*e.undirectedLoops),"undirected"!==this.type&&(n+=e.outDegree,r+=e.directedLoops),n-r}degreeWithoutSelfLoops(t){t=""+t;const e=this._nodes.get(t);if(!e)throw new th(`Graph.degreeWithoutSelfLoops: could not find the "${t}" node in the graph.`);let n=0,r=0;return"directed"!==this.type&&(n+=e.undirectedDegree,r+=2*e.undirectedLoops),"undirected"!==this.type&&(n+=e.inDegree+e.outDegree,r+=2*e.directedLoops),n-r}source(t){t=""+t;const e=this._edges.get(t);if(!e)throw new th(`Graph.source: could not find the "${t}" edge in the graph.`);return e.source.key}target(t){t=""+t;const e=this._edges.get(t);if(!e)throw new th(`Graph.target: could not find the "${t}" edge in the graph.`);return e.target.key}extremities(t){t=""+t;const e=this._edges.get(t);if(!e)throw new th(`Graph.extremities: could not find the "${t}" edge in the graph.`);return[e.source.key,e.target.key]}opposite(t,e){t=""+t,e=""+e;const n=this._edges.get(e);if(!n)throw new th(`Graph.opposite: could not find the "${e}" edge in the graph.`);const r=n.source.key,i=n.target.key;if(t===r)return i;if(t===i)return r;throw new th(`Graph.opposite: the "${t}" node is not attached to the "${e}" edge (${r}, ${i}).`)}hasExtremity(t,e){t=""+t,e=""+e;const n=this._edges.get(t);if(!n)throw new th(`Graph.hasExtremity: could not find the "${t}" edge in the graph.`);return n.source.key===e||n.target.key===e}isUndirected(t){t=""+t;const e=this._edges.get(t);if(!e)throw new th(`Graph.isUndirected: could not find the "${t}" edge in the graph.`);return e.undirected}isDirected(t){t=""+t;const e=this._edges.get(t);if(!e)throw new th(`Graph.isDirected: could not find the "${t}" edge in the graph.`);return!e.undirected}isSelfLoop(t){t=""+t;const e=this._edges.get(t);if(!e)throw new th(`Graph.isSelfLoop: could not find the "${t}" edge in the graph.`);return e.source===e.target}addNode(t,e){const n=function(t,e,n){if(n&&!Kc(n))throw new Qc(`Graph.addNode: invalid attributes. Expecting an object but got "${n}"`);if(e=""+e,n=n||{},t._nodes.has(e))throw new eh(`Graph.addNode: the "${e}" node already exist in the graph.`);const r=new t.NodeDataClass(e,n);return t._nodes.set(e,r),t.emit("nodeAdded",{key:e,attributes:n}),r}(this,t,e);return n.key}mergeNode(t,e){if(e&&!Kc(e))throw new Qc(`Graph.mergeNode: invalid attributes. Expecting an object but got "${e}"`);t=""+t,e=e||{};let n=this._nodes.get(t);return n?(e&&(Bc(n.attributes,e),this.emit("nodeAttributesUpdated",{type:"merge",key:t,attributes:n.attributes,data:e})),[t,!1]):(n=new this.NodeDataClass(t,e),this._nodes.set(t,n),this.emit("nodeAdded",{key:t,attributes:e}),[t,!0])}updateNode(t,e){if(e&&"function"!=typeof e)throw new Qc(`Graph.updateNode: invalid updater function. Expecting a function but got "${e}"`);t=""+t;let n=this._nodes.get(t);if(n){if(e){const r=n.attributes;n.attributes=e(r),this.emit("nodeAttributesUpdated",{type:"replace",key:t,attributes:n.attributes})}return[t,!1]}const r=e?e({}):{};return n=new this.NodeDataClass(t,r),this._nodes.set(t,n),this.emit("nodeAdded",{key:t,attributes:r}),[t,!0]}dropNode(t){t=""+t;const e=this._nodes.get(t);if(!e)throw new th(`Graph.dropNode: could not find the "${t}" node in the graph.`);let n;if("undirected"!==this.type){for(const t in e.out){n=e.out[t];do{Uh(this,n),n=n.next}while(n)}for(const t in e.in){n=e.in[t];do{Uh(this,n),n=n.next}while(n)}}if("directed"!==this.type)for(const t in e.undirected){n=e.undirected[t];do{Uh(this,n),n=n.next}while(n)}this._nodes.delete(t),this.emit("nodeDropped",{key:t,attributes:e.attributes})}dropEdge(t){let e;if(arguments.length>1){const t=""+arguments[0],n=""+arguments[1];if(e=Wc(this,t,n,this.type),!e)throw new th(`Graph.dropEdge: could not find the "${t}" -> "${n}" edge in the graph.`)}else if(t=""+t,e=this._edges.get(t),!e)throw new th(`Graph.dropEdge: could not find the "${t}" edge in the graph.`);return Uh(this,e),this}dropDirectedEdge(t,e){if(arguments.length<2)throw new eh("Graph.dropDirectedEdge: it does not make sense to try and drop a directed edge by key. What if the edge with this key is undirected? Use #.dropEdge for this purpose instead.");if(this.multi)throw new eh("Graph.dropDirectedEdge: cannot use a {source,target} combo when dropping an edge in a MultiGraph since we cannot infer the one you want to delete as there could be multiple ones.");const n=Wc(this,t=""+t,e=""+e,"directed");if(!n)throw new th(`Graph.dropDirectedEdge: could not find a "${t}" -> "${e}" edge in the graph.`);return Uh(this,n),this}dropUndirectedEdge(t,e){if(arguments.length<2)throw new eh("Graph.dropUndirectedEdge: it does not make sense to drop a directed edge by key. What if the edge with this key is undirected? Use #.dropEdge for this purpose instead.");if(this.multi)throw new eh("Graph.dropUndirectedEdge: cannot use a {source,target} combo when dropping an edge in a MultiGraph since we cannot infer the one you want to delete as there could be multiple ones.");const n=Wc(this,t,e,"undirected");if(!n)throw new th(`Graph.dropUndirectedEdge: could not find a "${t}" -> "${e}" edge in the graph.`);return Uh(this,n),this}clear(){this._edges.clear(),this._nodes.clear(),this._resetInstanceCounters(),this.emit("cleared")}clearEdges(){const t=this._nodes.values();let e;for(;e=t.next(),!0!==e.done;)e.value.clear();this._edges.clear(),this._resetInstanceCounters(),this.emit("edgesCleared")}getAttribute(t){return this._attributes[t]}getAttributes(){return this._attributes}hasAttribute(t){return this._attributes.hasOwnProperty(t)}setAttribute(t,e){return this._attributes[t]=e,this.emit("attributesUpdated",{type:"set",attributes:this._attributes,name:t}),this}updateAttribute(t,e){if("function"!=typeof e)throw new Qc("Graph.updateAttribute: updater should be a function.");const n=this._attributes[t];return this._attributes[t]=e(n),this.emit("attributesUpdated",{type:"set",attributes:this._attributes,name:t}),this}removeAttribute(t){return delete this._attributes[t],this.emit("attributesUpdated",{type:"remove",attributes:this._attributes,name:t}),this}replaceAttributes(t){if(!Kc(t))throw new Qc("Graph.replaceAttributes: provided attributes are not a plain object.");return this._attributes=t,this.emit("attributesUpdated",{type:"replace",attributes:this._attributes}),this}mergeAttributes(t){if(!Kc(t))throw new Qc("Graph.mergeAttributes: provided attributes are not a plain object.");return Bc(this._attributes,t),this.emit("attributesUpdated",{type:"merge",attributes:this._attributes,data:t}),this}updateAttributes(t){if("function"!=typeof t)throw new Qc("Graph.updateAttributes: provided updater is not a function.");return this._attributes=t(this._attributes),this.emit("attributesUpdated",{type:"update",attributes:this._attributes}),this}updateEachNodeAttributes(t,e){if("function"!=typeof t)throw new Qc("Graph.updateEachNodeAttributes: expecting an updater function.");if(e&&!Zc(e))throw new Qc("Graph.updateEachNodeAttributes: invalid hints. Expecting an object having the following shape: {attributes?: [string]}");const n=this._nodes.values();let r,i;for(;r=n.next(),!0!==r.done;)i=r.value,i.attributes=t(i.key,i.attributes);this.emit("eachNodeAttributesUpdated",{hints:e||null})}updateEachEdgeAttributes(t,e){if("function"!=typeof t)throw new Qc("Graph.updateEachEdgeAttributes: expecting an updater function.");if(e&&!Zc(e))throw new Qc("Graph.updateEachEdgeAttributes: invalid hints. Expecting an object having the following shape: {attributes?: [string]}");const n=this._edges.values();let r,i,o,a;for(;r=n.next(),!0!==r.done;)i=r.value,o=i.source,a=i.target,i.attributes=t(i.key,i.attributes,o.key,a.key,o.attributes,a.attributes,i.undirected);this.emit("eachEdgeAttributesUpdated",{hints:e||null})}forEachAdjacencyEntry(t){if("function"!=typeof t)throw new Qc("Graph.forEachAdjacencyEntry: expecting a callback.");Mh(!1,!1,!1,this,t)}forEachAdjacencyEntryWithOrphans(t){if("function"!=typeof t)throw new Qc("Graph.forEachAdjacencyEntryWithOrphans: expecting a callback.");Mh(!1,!1,!0,this,t)}forEachAssymetricAdjacencyEntry(t){if("function"!=typeof t)throw new Qc("Graph.forEachAssymetricAdjacencyEntry: expecting a callback.");Mh(!1,!0,!1,this,t)}forEachAssymetricAdjacencyEntryWithOrphans(t){if("function"!=typeof t)throw new Qc("Graph.forEachAssymetricAdjacencyEntryWithOrphans: expecting a callback.");Mh(!1,!0,!0,this,t)}nodes(){return"function"==typeof Array.from?Array.from(this._nodes.keys()):qc(this._nodes.keys(),this._nodes.size)}forEachNode(t){if("function"!=typeof t)throw new Qc("Graph.forEachNode: expecting a callback.");const e=this._nodes.values();let n,r;for(;n=e.next(),!0!==n.done;)r=n.value,t(r.key,r.attributes)}findNode(t){if("function"!=typeof t)throw new Qc("Graph.findNode: expecting a callback.");const e=this._nodes.values();let n,r;for(;n=e.next(),!0!==n.done;)if(r=n.value,t(r.key,r.attributes))return r.key}mapNodes(t){if("function"!=typeof t)throw new Qc("Graph.mapNode: expecting a callback.");const e=this._nodes.values();let n,r;const i=new Array(this.order);let o=0;for(;n=e.next(),!0!==n.done;)r=n.value,i[o++]=t(r.key,r.attributes);return i}someNode(t){if("function"!=typeof t)throw new Qc("Graph.someNode: expecting a callback.");const e=this._nodes.values();let n,r;for(;n=e.next(),!0!==n.done;)if(r=n.value,t(r.key,r.attributes))return!0;return!1}everyNode(t){if("function"!=typeof t)throw new Qc("Graph.everyNode: expecting a callback.");const e=this._nodes.values();let n,r;for(;n=e.next(),!0!==n.done;)if(r=n.value,!t(r.key,r.attributes))return!1;return!0}filterNodes(t){if("function"!=typeof t)throw new Qc("Graph.filterNodes: expecting a callback.");const e=this._nodes.values();let n,r;const i=[];for(;n=e.next(),!0!==n.done;)r=n.value,t(r.key,r.attributes)&&i.push(r.key);return i}reduceNodes(t,e){if("function"!=typeof t)throw new Qc("Graph.reduceNodes: expecting a callback.");if(arguments.length<2)throw new Qc("Graph.reduceNodes: missing initial value. You must provide it because the callback takes more than one argument and we cannot infer the initial value from the first iteration, as you could with a simple array.");let n=e;const r=this._nodes.values();let i,o;for(;i=r.next(),!0!==i.done;)o=i.value,n=t(n,o.key,o.attributes);return n}nodeEntries(){const t=this._nodes.values();return new Fc((()=>{const e=t.next();if(e.done)return e;const n=e.value;return{value:{node:n.key,attributes:n.attributes},done:!1}}))}export(){const t=new Array(this._nodes.size);let e=0;this._nodes.forEach(((n,r)=>{t[e++]=function(t,e){const n={key:t};return Hc(e.attributes)||(n.attributes=Bc({},e.attributes)),n}(r,n)}));const n=new Array(this._edges.size);return e=0,this._edges.forEach(((t,r)=>{n[e++]=function(t,e,n){const r={key:e,source:n.source.key,target:n.target.key};return Hc(n.attributes)||(r.attributes=Bc({},n.attributes)),"mixed"===t&&n.undirected&&(r.undirected=!0),r}(this.type,r,t)})),{options:{type:this.type,multi:this.multi,allowSelfLoops:this.allowSelfLoops},attributes:this.getAttributes(),nodes:t,edges:n}}import(t,e=!1){if(t instanceof Dh)return t.forEachNode(((t,n)=>{e?this.mergeNode(t,n):this.addNode(t,n)})),t.forEachEdge(((t,n,r,i,o,a,s)=>{e?s?this.mergeUndirectedEdgeWithKey(t,r,i,n):this.mergeDirectedEdgeWithKey(t,r,i,n):s?this.addUndirectedEdgeWithKey(t,r,i,n):this.addDirectedEdgeWithKey(t,r,i,n)})),this;if(!Kc(t))throw new Qc("Graph.import: invalid argument. Expecting a serialized graph or, alternatively, a Graph instance.");if(t.attributes){if(!Kc(t.attributes))throw new Qc("Graph.import: invalid attributes. Expecting a plain object.");e?this.mergeAttributes(t.attributes):this.replaceAttributes(t.attributes)}let n,r,i,o,a;if(t.nodes){if(i=t.nodes,!Array.isArray(i))throw new Qc("Graph.import: invalid nodes. Expecting an array.");for(n=0,r=i.length;n<r;n++){o=i[n],Eh(o);const{key:t,attributes:r}=o;e?this.mergeNode(t,r):this.addNode(t,r)}}if(t.edges){let o=!1;if("undirected"===this.type&&(o=!0),i=t.edges,!Array.isArray(i))throw new Qc("Graph.import: invalid edges. Expecting an array.");for(n=0,r=i.length;n<r;n++){a=i[n],Ah(a);const{source:t,target:r,attributes:s,undirected:u=o}=a;let c;"key"in a?(c=e?u?this.mergeUndirectedEdgeWithKey:this.mergeDirectedEdgeWithKey:u?this.addUndirectedEdgeWithKey:this.addDirectedEdgeWithKey,c.call(this,a.key,t,r,s)):(c=e?u?this.mergeUndirectedEdge:this.mergeDirectedEdge:u?this.addUndirectedEdge:this.addDirectedEdge,c.call(this,t,r,s))}}return this}nullCopy(t){const e=new Dh(Bc({},this._options,t));return e.replaceAttributes(Bc({},this.getAttributes())),e}emptyCopy(t){const e=this.nullCopy(t);return this._nodes.forEach(((t,n)=>{const r=Bc({},t.attributes);t=new e.NodeDataClass(n,r),e._nodes.set(n,t)})),e}copy(t){if("string"==typeof(t=t||{}).type&&t.type!==this.type&&"mixed"!==t.type)throw new eh(`Graph.copy: cannot create an incompatible copy from "${this.type}" type to "${t.type}" because this would mean losing information about the current graph.`);if("boolean"==typeof t.multi&&t.multi!==this.multi&&!0!==t.multi)throw new eh("Graph.copy: cannot create an incompatible copy by downgrading a multi graph to a simple one because this would mean losing information about the current graph.");if("boolean"==typeof t.allowSelfLoops&&t.allowSelfLoops!==this.allowSelfLoops&&!0!==t.allowSelfLoops)throw new eh("Graph.copy: cannot create an incompatible copy from a graph allowing self loops to one that does not because this would mean losing information about the current graph.");const e=this.emptyCopy(t),n=this._edges.values();let r,i;for(;r=n.next(),!0!==r.done;)i=r.value,Th(e,"copy",!1,i.undirected,i.key,i.source.key,i.target.key,Bc({},i.attributes));return e}toJSON(){return this.export()}toString(){return"[object Graph]"}inspect(){const t={};this._nodes.forEach(((e,n)=>{t[n]=e.attributes}));const e={},n={};this._edges.forEach(((t,r)=>{const i=t.undirected?"--":"->";let o,a="",s=t.source.key,u=t.target.key;t.undirected&&s>u&&(o=s,s=u,u=o);const c=`(${s})${i}(${u})`;r.startsWith("geid_")?this.multi&&(void 0===n[c]?n[c]=0:n[c]++,a+=`${n[c]}. `):a+=`[${r}]: `,a+=c,e[a]=t.attributes}));const r={};for(const t in this)this.hasOwnProperty(t)&&!Nh.has(t)&&"function"!=typeof this[t]&&"symbol"!=typeof t&&(r[t]=this[t]);return r.attributes=this._attributes,r.nodes=t,r.edges=e,Xc(r,"constructor",this.constructor),r}}"undefined"!=typeof Symbol&&(Dh.prototype[Symbol.for("nodejs.util.inspect.custom")]=Dh.prototype.inspect),[{name:t=>`${t}Edge`,generateKey:!0},{name:t=>`${t}DirectedEdge`,generateKey:!0,type:"directed"},{name:t=>`${t}UndirectedEdge`,generateKey:!0,type:"undirected"},{name:t=>`${t}EdgeWithKey`},{name:t=>`${t}DirectedEdgeWithKey`,type:"directed"},{name:t=>`${t}UndirectedEdgeWithKey`,type:"undirected"}].forEach((t=>{["add","merge","update"].forEach((e=>{const n=t.name(e),r="add"===e?Th:Gh;t.generateKey?Dh.prototype[n]=function(i,o,a){return r(this,n,!0,"undirected"===(t.type||this.type),null,i,o,a,"update"===e)}:Dh.prototype[n]=function(i,o,a,s){return r(this,n,!1,"undirected"===(t.type||this.type),i,o,a,s,"update"===e)}}))})),function(t){sh.forEach((function({name:e,attacher:n}){n(t,e("Node"),0),n(t,e("Source"),1),n(t,e("Target"),2),n(t,e("Opposite"),3)}))}(Dh),function(t){uh.forEach((function({name:e,attacher:n}){n(t,e("Edge"),"mixed"),n(t,e("DirectedEdge"),"directed"),n(t,e("UndirectedEdge"),"undirected")}))}(Dh),function(t){ch.forEach((e=>{(function(t,e){const{name:n,type:r,direction:i}=e;t.prototype[n]=function(t,e){if("mixed"!==r&&"mixed"!==this.type&&r!==this.type)return[];if(!arguments.length)return function(t,e){if(0===t.size)return[];if("mixed"===e||e===t.type)return"function"==typeof Array.from?Array.from(t._edges.keys()):qc(t._edges.keys(),t._edges.size);const n="undirected"===e?t.undirectedSize:t.directedSize,r=new Array(n),i="undirected"===e,o=t._edges.values();let a,s,u=0;for(;a=o.next(),!0!==a.done;)s=a.value,s.undirected===i&&(r[u++]=s.key);return r}(this,r);if(1===arguments.length){t=""+t;const e=this._nodes.get(t);if(void 0===e)throw new th(`Graph.${n}: could not find the "${t}" node in the graph.`);return function(t,e,n,r){const i=[];return vh(!1,t,e,n,r,(function(t){i.push(t)})),i}(this.multi,"mixed"===r?this.type:r,i,e)}if(2===arguments.length){t=""+t,e=""+e;const o=this._nodes.get(t);if(!o)throw new th(`Graph.${n}:  could not find the "${t}" source node in the graph.`);if(!this._nodes.has(e))throw new th(`Graph.${n}:  could not find the "${e}" target node in the graph.`);return function(t,e,n,r,i){const o=[];return mh(!1,t,e,n,r,i,(function(t){o.push(t)})),o}(r,this.multi,i,o,e)}throw new Qc(`Graph.${n}: too many arguments (expecting 0, 1 or 2 and got ${arguments.length}).`)}})(t,e),function(t,e){const{name:n,type:r,direction:i}=e,o="forEach"+n[0].toUpperCase()+n.slice(1,-1);t.prototype[o]=function(t,e,n){if("mixed"===r||"mixed"===this.type||r===this.type){if(1===arguments.length)return gh(!1,this,r,n=t);if(2===arguments.length){t=""+t,n=e;const a=this._nodes.get(t);if(void 0===a)throw new th(`Graph.${o}: could not find the "${t}" node in the graph.`);return vh(!1,this.multi,"mixed"===r?this.type:r,i,a,n)}if(3===arguments.length){t=""+t,e=""+e;const a=this._nodes.get(t);if(!a)throw new th(`Graph.${o}:  could not find the "${t}" source node in the graph.`);if(!this._nodes.has(e))throw new th(`Graph.${o}:  could not find the "${e}" target node in the graph.`);return mh(!1,r,this.multi,i,a,e,n)}throw new Qc(`Graph.${o}: too many arguments (expecting 1, 2 or 3 and got ${arguments.length}).`)}};const a="map"+n[0].toUpperCase()+n.slice(1);t.prototype[a]=function(){const t=Array.prototype.slice.call(arguments),e=t.pop();let n;if(0===t.length){let i=0;"directed"!==r&&(i+=this.undirectedSize),"undirected"!==r&&(i+=this.directedSize),n=new Array(i);let o=0;t.push(((t,r,i,a,s,u,c)=>{n[o++]=e(t,r,i,a,s,u,c)}))}else n=[],t.push(((t,r,i,o,a,s,u)=>{n.push(e(t,r,i,o,a,s,u))}));return this[o].apply(this,t),n};const s="filter"+n[0].toUpperCase()+n.slice(1);t.prototype[s]=function(){const t=Array.prototype.slice.call(arguments),e=t.pop(),n=[];return t.push(((t,r,i,o,a,s,u)=>{e(t,r,i,o,a,s,u)&&n.push(t)})),this[o].apply(this,t),n};const u="reduce"+n[0].toUpperCase()+n.slice(1);t.prototype[u]=function(){let t,e,n=Array.prototype.slice.call(arguments);if(n.length<2||n.length>4)throw new Qc(`Graph.${u}: invalid number of arguments (expecting 2, 3 or 4 and got ${n.length}).`);if("function"==typeof n[n.length-1]&&"function"!=typeof n[n.length-2])throw new Qc(`Graph.${u}: missing initial value. You must provide it because the callback takes more than one argument and we cannot infer the initial value from the first iteration, as you could with a simple array.`);2===n.length?(t=n[0],e=n[1],n=[]):3===n.length?(t=n[1],e=n[2],n=[n[0]]):4===n.length&&(t=n[2],e=n[3],n=[n[0],n[1]]);let r=e;return n.push(((e,n,i,o,a,s,u)=>{r=t(r,e,n,i,o,a,s,u)})),this[o].apply(this,n),r}}(t,e),function(t,e){const{name:n,type:r,direction:i}=e,o="find"+n[0].toUpperCase()+n.slice(1,-1);t.prototype[o]=function(t,e,n){if("mixed"!==r&&"mixed"!==this.type&&r!==this.type)return!1;if(1===arguments.length)return gh(!0,this,r,n=t);if(2===arguments.length){t=""+t,n=e;const a=this._nodes.get(t);if(void 0===a)throw new th(`Graph.${o}: could not find the "${t}" node in the graph.`);return vh(!0,this.multi,"mixed"===r?this.type:r,i,a,n)}if(3===arguments.length){t=""+t,e=""+e;const a=this._nodes.get(t);if(!a)throw new th(`Graph.${o}:  could not find the "${t}" source node in the graph.`);if(!this._nodes.has(e))throw new th(`Graph.${o}:  could not find the "${e}" target node in the graph.`);return mh(!0,r,this.multi,i,a,e,n)}throw new Qc(`Graph.${o}: too many arguments (expecting 1, 2 or 3 and got ${arguments.length}).`)};const a="some"+n[0].toUpperCase()+n.slice(1,-1);t.prototype[a]=function(){const t=Array.prototype.slice.call(arguments),e=t.pop();return t.push(((t,n,r,i,o,a,s)=>e(t,n,r,i,o,a,s))),!!this[o].apply(this,t)};const s="every"+n[0].toUpperCase()+n.slice(1,-1);t.prototype[s]=function(){const t=Array.prototype.slice.call(arguments),e=t.pop();return t.push(((t,n,r,i,o,a,s)=>!e(t,n,r,i,o,a,s))),!this[o].apply(this,t)}}(t,e),function(t,e){const{name:n,type:r,direction:i}=e,o=n.slice(0,-1)+"Entries";t.prototype[o]=function(t,e){if("mixed"!==r&&"mixed"!==this.type&&r!==this.type)return Fc.empty();if(!arguments.length)return function(t,e){if(0===t.size)return Fc.empty();const n="mixed"!==e&&e!==t.type,r="undirected"===e,i=t._edges.values();return new Fc((function(){let t,e;for(;;){if(t=i.next(),t.done)return t;if(e=t.value,!n||e.undirected===r)break}return{value:{edge:e.key,attributes:e.attributes,source:e.source.key,target:e.target.key,sourceAttributes:e.source.attributes,targetAttributes:e.target.attributes,undirected:e.undirected},done:!1}}))}(this,r);if(1===arguments.length){t=""+t;const e=this._nodes.get(t);if(!e)throw new th(`Graph.${o}: could not find the "${t}" node in the graph.`);return function(t,e,n){let r=Fc.empty();return"undirected"!==t&&("out"!==e&&void 0!==n.in&&(r=Yc(r,lh(n.in))),"in"!==e&&void 0!==n.out&&(r=Yc(r,lh(n.out,e?void 0:n.key)))),"directed"!==t&&void 0!==n.undirected&&(r=Yc(r,lh(n.undirected))),r}(r,i,e)}if(2===arguments.length){t=""+t,e=""+e;const n=this._nodes.get(t);if(!n)throw new th(`Graph.${o}:  could not find the "${t}" source node in the graph.`);if(!this._nodes.has(e))throw new th(`Graph.${o}:  could not find the "${e}" target node in the graph.`);return function(t,e,n,r){let i=Fc.empty();return"undirected"!==t&&(void 0!==n.in&&"out"!==e&&r in n.in&&(i=Yc(i,yh(n.in,r))),void 0!==n.out&&"in"!==e&&r in n.out&&(e||n.key!==r)&&(i=Yc(i,yh(n.out,r)))),"directed"!==t&&void 0!==n.undirected&&r in n.undirected&&(i=Yc(i,yh(n.undirected,r))),i}(r,i,n,e)}throw new Qc(`Graph.${o}: too many arguments (expecting 0, 1 or 2 and got ${arguments.length}).`)}}(t,e)}))}(Dh),function(t){_h.forEach((e=>{(function(t,e){const{name:n,type:r,direction:i}=e;t.prototype[n]=function(t){if("mixed"!==r&&"mixed"!==this.type&&r!==this.type)return[];t=""+t;const e=this._nodes.get(t);if(void 0===e)throw new th(`Graph.${n}: could not find the "${t}" node in the graph.`);return function(t,e,n){if("mixed"!==t){if("undirected"===t)return Object.keys(n.undirected);if("string"==typeof e)return Object.keys(n[e])}const r=[];return xh(!1,t,e,n,(function(t){r.push(t)})),r}("mixed"===r?this.type:r,i,e)}})(t,e),function(t,e){const{name:n,type:r,direction:i}=e,o="forEach"+n[0].toUpperCase()+n.slice(1,-1);t.prototype[o]=function(t,e){if("mixed"!==r&&"mixed"!==this.type&&r!==this.type)return;t=""+t;const n=this._nodes.get(t);if(void 0===n)throw new th(`Graph.${o}: could not find the "${t}" node in the graph.`);xh(!1,"mixed"===r?this.type:r,i,n,e)};const a="map"+n[0].toUpperCase()+n.slice(1);t.prototype[a]=function(t,e){const n=[];return this[o](t,((t,r)=>{n.push(e(t,r))})),n};const s="filter"+n[0].toUpperCase()+n.slice(1);t.prototype[s]=function(t,e){const n=[];return this[o](t,((t,r)=>{e(t,r)&&n.push(t)})),n};const u="reduce"+n[0].toUpperCase()+n.slice(1);t.prototype[u]=function(t,e,n){if(arguments.length<3)throw new Qc(`Graph.${u}: missing initial value. You must provide it because the callback takes more than one argument and we cannot infer the initial value from the first iteration, as you could with a simple array.`);let r=n;return this[o](t,((t,n)=>{r=e(r,t,n)})),r}}(t,e),function(t,e){const{name:n,type:r,direction:i}=e,o=n[0].toUpperCase()+n.slice(1,-1),a="find"+o;t.prototype[a]=function(t,e){if("mixed"!==r&&"mixed"!==this.type&&r!==this.type)return;t=""+t;const n=this._nodes.get(t);if(void 0===n)throw new th(`Graph.${a}: could not find the "${t}" node in the graph.`);return xh(!0,"mixed"===r?this.type:r,i,n,e)};const s="some"+o;t.prototype[s]=function(t,e){return!!this[a](t,e)};const u="every"+o;t.prototype[u]=function(t,e){return!this[a](t,((t,n)=>!e(t,n)))}}(t,e),function(t,e){const{name:n,type:r,direction:i}=e,o=n.slice(0,-1)+"Entries";t.prototype[o]=function(t){if("mixed"!==r&&"mixed"!==this.type&&r!==this.type)return Fc.empty();t=""+t;const e=this._nodes.get(t);if(void 0===e)throw new th(`Graph.${o}: could not find the "${t}" node in the graph.`);return function(t,e,n){if("mixed"!==t){if("undirected"===t)return kh(null,n,n.undirected);if("string"==typeof e)return kh(null,n,n[e])}let r=Fc.empty();const i=new bh;return"undirected"!==t&&("out"!==e&&(r=Yc(r,kh(i,n,n.in))),"in"!==e&&(r=Yc(r,kh(i,n,n.out)))),"directed"!==t&&(r=Yc(r,kh(i,n,n.undirected))),r}("mixed"===r?this.type:r,i,e)}}(t,e)}))}(Dh);class Oh extends Dh{constructor(t){const e=Bc({type:"directed"},t);if("multi"in e&&!1!==e.multi)throw new Qc("DirectedGraph.from: inconsistent indication that the graph should be multi in given options!");if("directed"!==e.type)throw new Qc('DirectedGraph.from: inconsistent "'+e.type+'" type in given options!');super(e)}}class jh extends Dh{constructor(t){const e=Bc({type:"undirected"},t);if("multi"in e&&!1!==e.multi)throw new Qc("UndirectedGraph.from: inconsistent indication that the graph should be multi in given options!");if("undirected"!==e.type)throw new Qc('UndirectedGraph.from: inconsistent "'+e.type+'" type in given options!');super(e)}}class zh extends Dh{constructor(t){const e=Bc({multi:!0},t);if("multi"in e&&!0!==e.multi)throw new Qc("MultiGraph.from: inconsistent indication that the graph should be simple in given options!");super(e)}}class Ph extends Dh{constructor(t){const e=Bc({type:"directed",multi:!0},t);if("multi"in e&&!0!==e.multi)throw new Qc("MultiDirectedGraph.from: inconsistent indication that the graph should be simple in given options!");if("directed"!==e.type)throw new Qc('MultiDirectedGraph.from: inconsistent "'+e.type+'" type in given options!');super(e)}}class Ih extends Dh{constructor(t){const e=Bc({type:"undirected",multi:!0},t);if("multi"in e&&!0!==e.multi)throw new Qc("MultiUndirectedGraph.from: inconsistent indication that the graph should be simple in given options!");if("undirected"!==e.type)throw new Qc('MultiUndirectedGraph.from: inconsistent "'+e.type+'" type in given options!');super(e)}}function Rh(t){t.from=function(e,n){const r=Bc({},e.options,n),i=new t(r);return i.import(e),i}}Rh(Dh),Rh(Oh),Rh(jh),Rh(zh),Rh(Ph),Rh(Ih),Dh.Graph=Dh,Dh.DirectedGraph=Oh,Dh.UndirectedGraph=jh,Dh.MultiGraph=zh,Dh.MultiDirectedGraph=Ph,Dh.MultiUndirectedGraph=Ih,Dh.InvalidArgumentsGraphError=Qc,Dh.NotFoundGraphError=th,Dh.UsageGraphError=eh;var Fh=n(181);function qh(t,e){var n="undefined"!=typeof Symbol&&t[Symbol.iterator]||t["@@iterator"];if(!n){if(Array.isArray(t)||(n=function(t,e){if(t){if("string"==typeof t)return Yh(t,e);var n=Object.prototype.toString.call(t).slice(8,-1);return"Object"===n&&t.constructor&&(n=t.constructor.name),"Map"===n||"Set"===n?Array.from(t):"Arguments"===n||/^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)?Yh(t,e):void 0}}(t))||e&&t&&"number"==typeof t.length){n&&(t=n);var r=0,i=function(){};return{s:i,n:function(){return r>=t.length?{done:!0}:{done:!1,value:t[r++]}},e:function(t){throw t},f:i}}throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.")}var o,a=!0,s=!1;return{s:function(){n=n.call(t)},n:function(){var t=n.next();return a=t.done,t},e:function(t){s=!0,o=t},f:function(){try{a||null==n.return||n.return()}finally{if(s)throw o}}}}function Yh(t,e){(null==e||e>t.length)&&(e=t.length);for(var n=0,r=new Array(e);n<e;n++)r[n]=t[n];return r}var Bh=function(){var t,e=new Dh,n=qh(data.nodes);try{for(n.s();!(t=n.n()).done;){var r=t.value,i=r.id,o=r.label;e.hasNode(i)||e.addNode(i,{label:o})}}catch(t){n.e(t)}finally{n.f()}var a,s=qh(data.links);try{for(s.s();!(a=s.n()).done;){var u=a.value,c=u.source,h=u.target;e.hasEdge(c.id,h.id)||e.addEdge(c.id,h.id)}}catch(t){s.e(t)}finally{s.f()}return e}();window.addEventListener("DOMContentLoaded",(function(){var t=document.getElementById("focus-check"),e=document.getElementById("focus-mode-select"),n=document.getElementById("focus-input");if(!1!==focusIsActive){var r;t.checked=!1;var i=new URL(window.location).searchParams,o=Number(i.get("focus"));!1===isNaN(o)&&o>0&&(n.value=o,a()),Hu("f",(function(e){e.preventDefault(),t.checked=!t.checked,t.dispatchEvent(new Event("change"))})),Hu("alt+r",(function(e){e.preventDefault(),t.checked=!1,t.dispatchEvent(new Event("change"))})),t.addEventListener("change",(function(){t.checked?a():(n.classList.remove("active"),n.removeEventListener("input",s),e.removeEventListener("change",c),hc(Ju.filteredByFocus))}))}function a(){void 0!==u.openedRecordId?(c(),e.addEventListener("change",c),s(),n.classList.add("active"),n.addEventListener("input",s),n.focus()):t.checked=!1}function s(){var t=u.openedRecordId,e=[];(0,Fh.a)(Bh,t,(function(t,r,i){return e.push(t),i>=n.valueAsNumber}),{mode:r}),uc(e,Ju.filteredByFocus)}function c(){r=e.value,s()}}))})()})();
    </script>
    

</body>

</html>